{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: index.html\n",
    "title: glycowork\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from IPython.display import HTML\n",
    "import pandas as pd\n",
    "\n",
    "from glycowork import GlycoDraw as OrigGlycoDraw\n",
    "import re\n",
    "def GlycoDraw(*args, **kwargs):\n",
    "  result = OrigGlycoDraw(*args, **kwargs)\n",
    "  if hasattr(result, 'as_svg'):\n",
    "    svg = result.as_svg()\n",
    "    result._repr_svg_ = lambda: re.sub(r'(<svg[^>]*>)', r'\\1<style>svg{background:white}</style>', svg)\n",
    "  return result\n",
    "import glycowork\n",
    "glycowork.GlycoDraw = GlycoDraw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![CI](https://github.com/BojarLab/glycowork/actions/workflows/test.yaml/badge.svg) ![PyPI - Downloads](https://img.shields.io/pypi/dm/glycowork?color=brightgreen)[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/BojarLab/glycowork/issues)[![DOI](https://zenodo.org/badge/327716604.svg)](https://zenodo.org/doi/10.5281/zenodo.10039202)[![codecov](https://codecov.io/gh/BojarLab/glycowork/graph/badge.svg?token=3AHNZ9QRZV)](https://codecov.io/gh/BojarLab/glycowork)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<img src=\"./glycowork_badge_wo_bg.jpg\" width=\"200\" alt=\"glycowork logo\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Glycans are fundamental biological sequences that are as crucial as DNA, RNA, and proteins. As complex carbohydrates forming branched structures, glycans are ubiquitous yet often overlooked in biological research.\n",
    "\n",
    "## Why Glycans are Important\n",
    "- Ubiquitous in biology\n",
    "- Integral to protein and lipid function\n",
    "- Relevant to human diseases\n",
    "\n",
    "## Challenges in Glycan Analysis\n",
    "Analyzing glycans is complicated due to their non-linear structures and enormous diversity. But that's where `glycowork` comes in.\n",
    "\n",
    "## Introducing glycowork: Your Solution for Glycan-Focused Data Science\n",
    "Glycowork is a Python package specifically designed to simplify glycan sequence processing and analysis. It offers:\n",
    "\n",
    "- Functions for glycan analysis\n",
    "- Datasets for model training\n",
    "- Full support for IUPAC-condensed string representation. Broad support for IUPAC-extended, LinearCode, Oxford, GlycoCT, WURCS, GLYCAM, CSDB-linear, GlycoWorkBench, GlyTouCan IDs, and more.\n",
    "- Powerful graph-based architecture for in-depth analysis\n",
    "\n",
    "**Documentation:** <https://bojarlab.github.io/glycowork/>\n",
    "\n",
    "**Contribute:** Interested in contributing? Read our [Contribution Guidelines](https://github.com/BojarLab/glycowork/blob/master/CONTRIBUTING.md)\n",
    "\n",
    "**Citation:** If `glycowork` adds value to your project, please cite [Thomes et al., 2021](https://academic.oup.com/glycob/advance-article/doi/10.1093/glycob/cwab067/6311240)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<u>Not familiar with Python?</u> Try our no-code, graphical user interface (`glycoworkGUI.exe`, can be downloaded at the bottom of the latest [Release](https://github.com/BojarLab/glycowork/releases) page) for accessing some of the most useful `glycowork` functions\n",
    "\n",
    "Or try our [web interface](https://canonicalize.streamlit.app/) for sequence format conversion/cleaning and drawing glycan SNFG structures!\n",
    "\n",
    "<u>via pip:</u> <br>\n",
    "`pip install glycowork` <br>\n",
    "`import glycowork`\n",
    "\n",
    "<u>alternative:</u> <br>\n",
    "`pip install git+https://github.com/BojarLab/glycowork.git` <br>\n",
    "`import glycowork`\n",
    "\n",
    "<u>Note that we have optional extra installs for specialized use (even further instructions can be found in the `Examples` tab; on Mac you might need to use `\"glycowork[ml]\"`), such as:</u> <br>\n",
    "_deep learning_ <br>\n",
    "`pip install glycowork[ml]` <br>\n",
    "_analyzing atomic/chemical properties of glycans_ <br>\n",
    "`pip install glycowork[chem]` <br>\n",
    "_everything_ <br>\n",
    "`pip install glycowork[all]` <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data & Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Glycowork` currently contains the following main datasets that are freely available to everyone:\n",
    "\n",
    "- **`df_glycan`**\n",
    "    - contains ~50,500 unique glycan sequences, including labels such as ~39,500 species associations, ~20,000 tissue associations, and ~1,000 disease associations\n",
    "- **`glycan_binding`**\n",
    "    - contains >790,000 protein-glycan binding interactions, from >2,000 unique glycan-binding proteins\n",
    "    \n",
    "Additionally, we store these trained deep learning models for easy usage, which can be retrieved with the `prep_model` function:\n",
    "\n",
    "- **`LectinOracle`**\n",
    "    - can be used to predict glycan-binding specificity of a protein, given its ESMC representation; from [Lundstrom et al., 2021](https://onlinelibrary.wiley.com/doi/10.1002/advs.202103807)\n",
    "- **`LectinOracle_flex`**\n",
    "    - operates the same as LectinOracle but can directly use the raw protein sequence as input (no ESMC representation required)\n",
    "- **`SweetNet`**\n",
    "    - a graph convolutional neural network trained to predict species from glycan, can be used to generate learned glycan representations; from [Burkholz et al., 2021](https://pubmed.ncbi.nlm.nih.gov/34133929/)\n",
    "- **`NSequonPred`**\n",
    "    - given the ESM-1b representation of an N-sequon (+/- 20 AA), this model can predict whether the sequon will be glycosylated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to use"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Glycowork` currently contains four main modules:\n",
    "\n",
    " - **`glycan_data`**\n",
    "     - stores several glycan datasets and contains helper functions\n",
    " - **`ml`**\n",
    "     - here are all the functions for training and using machine learning models, including train-test-split, getting glycan representations, etc.\n",
    " - **`motif`**\n",
    "     - contains functions for processing & drawing glycan sequences, identifying motifs and features, and analyzing them\n",
    " - **`network`**\n",
    "     - contains functions for constructing and analyzing glycan networks (e.g., biosynthetic networks)\n",
    "     \n",
    "Below are some examples of what you can do with `glycowork`; be sure to check out the other `examples` in the full documentation for everything that's there.  [--> Learn more](./05_examples.ipynb)\n",
    "A non-exhaustive list includes:\n",
    "\n",
    "- using trained AI models for prediction [--> Learn more](./02_ml.ipynb#inference)\n",
    "- training your own AI models [--> Learn more](./02_ml.ipynb#model_training)\n",
    "- motif enrichment analyses [--> Learn more](./03_motif.ipynb#analysis)\n",
    "- differential glycomics expression analysis [--> Learn more](./03_motif.ipynb#analysis)\n",
    "- annotating motifs in glycans [--> Learn more](./03_motif.ipynb#annotate)\n",
    "- drawing publication-quality glycan figures [--> Learn more](./03_motif.ipynb#draw)\n",
    "- finding out whether & where glycans are describing the same sequence [--> Learn more](./03_motif.ipynb#graph)\n",
    "- _m/z_ to composition to structure to motif mappings [--> Learn more](./03_motif.ipynb#tokenization)\n",
    "- mass calculation [--> Learn more](./03_motif.ipynb#tokenization)\n",
    "- visualizing motif distribution / glycan similarities / sequence properties [--> Learn more](./03_motif.ipynb#analysis)\n",
    "- constructing and analyzing biosynthetic networks [--> Learn more](./04_network.ipynb#biosynthesis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n",
       "     width=\"700\" height=\"510.0\" viewBox=\"-650.0 -205.0 700 510.0\"><style>svg{background:white}</style>\n",
       "<defs>\n",
       "<path d=\"M-100,0 L0,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d0\" />\n",
       "<path d=\"M-200,0 L-100,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d1\" />\n",
       "<path d=\"M-300,100.0 L-200,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d2\" />\n",
       "<path d=\"M-400,100.0 L-300,100.0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d3\" />\n",
       "<path d=\"M-500,100.0 L-400,100.0\" stroke-width=\"4.0\" stroke=\"#000000\" id=\"d4\" />\n",
       "<path d=\"M-600,100.0 L-500,100.0\" stroke-width=\"4.0\" stroke=\"#000000\" id=\"d5\" />\n",
       "<path d=\"M-400,-100.0 L-300,-100.0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d6\" />\n",
       "<path d=\"M-500,-100.0 L-400,-100.0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d7\" />\n",
       "<path d=\"M-600,-100.0 L-500,-100.0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d8\" />\n",
       "<path d=\"M-400,200.0 L-400,100.0\" stroke-width=\"4.0\" stroke=\"#000000\" id=\"d9\" />\n",
       "<path d=\"M-300,-100.0 L-200,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d10\" />\n",
       "<path d=\"M-300,0 L-200,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d11\" />\n",
       "<path d=\"M0,-100 L0,0\" stroke-width=\"4.0\" stroke=\"#D9D9D9\" id=\"d12\" />\n",
       "<path d=\"M-50,25.0 L50,25.0\" stroke-width=\"0\" id=\"d13\" />\n",
       "<path d=\"M-150,25.0 L-50,25.0\" stroke-width=\"0\" id=\"d14\" />\n",
       "<path d=\"M-250,25.0 L-150,25.0\" stroke-width=\"0\" id=\"d15\" />\n",
       "<path d=\"M-350,125.0 L-250,125.0\" stroke-width=\"0\" id=\"d16\" />\n",
       "<path d=\"M-450,125.0 L-350,125.0\" stroke-width=\"0\" id=\"d17\" />\n",
       "<path d=\"M-550,125.0 L-450,125.0\" stroke-width=\"0\" id=\"d18\" />\n",
       "<path d=\"M-650,125.0 L-550,125.0\" stroke-width=\"0\" id=\"d19\" />\n",
       "<path d=\"M-450,225.0 L-350,225.0\" stroke-width=\"0\" id=\"d20\" />\n",
       "<path d=\"M-350,-75.0 L-250,-75.0\" stroke-width=\"0\" id=\"d21\" />\n",
       "<path d=\"M-450,-75.0 L-350,-75.0\" stroke-width=\"0\" id=\"d22\" />\n",
       "<path d=\"M-550,-75.0 L-450,-75.0\" stroke-width=\"0\" id=\"d23\" />\n",
       "<path d=\"M-650,-75.0 L-550,-75.0\" stroke-width=\"0\" id=\"d24\" />\n",
       "<path d=\"M-350,25.0 L-250,25.0\" stroke-width=\"0\" id=\"d25\" />\n",
       "<path d=\"M-50,-75.0 L50,-75.0\" stroke-width=\"0\" id=\"d26\" />\n",
       "</defs>\n",
       "<g transform=\"rotate(0 -300.0 50.0)\">\n",
       "<use xlink:href=\"#d0\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d0\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 4</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d1\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d1\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 4</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d2\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d2\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 3</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d3\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d3\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 2</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d4\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#000000\" valign=\"middle\"><textPath xlink:href=\"#d4\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 4</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d5\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#000000\" valign=\"middle\"><textPath xlink:href=\"#d5\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 3</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d6\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d6\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 2</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d7\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d7\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 4</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d8\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d8\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 6</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d9\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#000000\" valign=\"middle\"><textPath xlink:href=\"#d9\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 3</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d10\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d10\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 6</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d11\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d11\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">β 4</tspan>\n",
       "</textPath></text>\n",
       "<use xlink:href=\"#d12\" />\n",
       "<text font-size=\"20.0\" text-anchor=\"middle\" fill=\"#D9D9D9\" valign=\"middle\"><textPath xlink:href=\"#d12\" startOffset=\"50%\">\n",
       "<tspan dy=\"-0.5em\">α 6</tspan>\n",
       "</textPath></text>\n",
       "<rect x=\"-25.0\" y=\"-25.0\" width=\"50\" height=\"50\" fill=\"#CDE7EF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d13\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d13\" startOffset=\"50%\"></textPath></text>\n",
       "<rect x=\"-125.0\" y=\"-25.0\" width=\"50\" height=\"50\" fill=\"#CDE7EF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d14\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d14\" startOffset=\"50%\"></textPath></text>\n",
       "<circle cx=\"-200\" cy=\"0\" r=\"25.0\" fill=\"#CDE9DF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d15\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d15\" startOffset=\"50%\"></textPath></text>\n",
       "<circle cx=\"-300\" cy=\"100.0\" r=\"25.0\" fill=\"#CDE9DF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d16\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d16\" startOffset=\"50%\"></textPath></text>\n",
       "<rect x=\"-425.0\" y=\"75.0\" width=\"50\" height=\"50\" fill=\"#0385AE\" stroke-width=\"2.0\" stroke=\"#000000\" />\n",
       "<use xlink:href=\"#d17\" />\n",
       "<text font-size=\"17.5\" fill=\"#000000\" text-anchor=\"middle\"><textPath xlink:href=\"#d17\" startOffset=\"50%\"></textPath></text>\n",
       "<circle cx=\"-500\" cy=\"100.0\" r=\"25.0\" fill=\"#FCC326\" stroke-width=\"2.0\" stroke=\"#000000\" />\n",
       "<use xlink:href=\"#d18\" />\n",
       "<text font-size=\"17.5\" fill=\"#000000\" text-anchor=\"middle\"><textPath xlink:href=\"#d18\" startOffset=\"50%\"></textPath></text>\n",
       "<path d=\"M-600,125.0 L-575.0,100.0 L-600,75.0 L-625.0,100.0 Z\" fill=\"#A15989\" stroke=\"#000000\" stroke-width=\"2.0\" />\n",
       "<use xlink:href=\"#d19\" />\n",
       "<text font-size=\"17.5\" fill=\"#000000\" text-anchor=\"middle\"><textPath xlink:href=\"#d19\" startOffset=\"50%\"></textPath></text>\n",
       "<path d=\"M-425.0,221.65063509461098 L-400,178.34936490538902 L-375.0,221.65063509461098 Z\" fill=\"#C23537\" stroke=\"#000000\" stroke-width=\"2.0\" />\n",
       "<use xlink:href=\"#d20\" />\n",
       "<text font-size=\"17.5\" fill=\"#000000\" text-anchor=\"middle\"><textPath xlink:href=\"#d20\" startOffset=\"50%\"></textPath></text>\n",
       "<circle cx=\"-300\" cy=\"-100.0\" r=\"25.0\" fill=\"#CDE9DF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d21\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d21\" startOffset=\"50%\"></textPath></text>\n",
       "<rect x=\"-425.0\" y=\"-125.0\" width=\"50\" height=\"50\" fill=\"#CDE7EF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d22\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d22\" startOffset=\"50%\"></textPath></text>\n",
       "<circle cx=\"-500\" cy=\"-100.0\" r=\"25.0\" fill=\"#FFF6DE\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d23\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d23\" startOffset=\"50%\"></textPath></text>\n",
       "<path d=\"M-600,-75.0 L-575.0,-100.0 L-600,-125.0 L-625.0,-100.0 Z\" fill=\"#EEF8FB\" stroke=\"#D9D9D9\" stroke-width=\"2.0\" />\n",
       "<use xlink:href=\"#d24\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d24\" startOffset=\"50%\"></textPath></text>\n",
       "<rect x=\"-325.0\" y=\"-25.0\" width=\"50\" height=\"50\" fill=\"#CDE7EF\" stroke-width=\"2.0\" stroke=\"#D9D9D9\" />\n",
       "<use xlink:href=\"#d25\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d25\" startOffset=\"50%\"></textPath></text>\n",
       "<path d=\"M-25.0,-78.34936490538904 L0,-121.65063509461096 L25.0,-78.34936490538904 Z\" fill=\"#F7E0E0\" stroke=\"#D9D9D9\" stroke-width=\"2.0\" />\n",
       "<use xlink:href=\"#d26\" />\n",
       "<text font-size=\"17.5\" fill=\"#D9D9D9\" text-anchor=\"middle\"><textPath xlink:href=\"#d26\" startOffset=\"50%\"></textPath></text>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<drawsvg.drawing.Drawing>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drawing publication-quality glycan figures\n",
    "from glycowork import GlycoDraw\n",
    "GlycoDraw(\"Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc\", highlight_motif = \"Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get motifs, graph features, and sequence features of a set of glycan sequences to train models or analyze glycan properties\n",
    "glycans = [\"Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc\",\n",
    "           \"Ma3(Ma6)Mb4GNb4GN;N\",\n",
    "           \"α-D-Manp-(1→3)[α-D-Manp-(1→6)]-β-D-Manp-(1→4)-β-D-GlcpNAc-(1→4)-β-D-GlcpNAc-(1→\",\n",
    "           \"F(3)XA2\",\n",
    "           \"WURCS=2.0/5,11,10/[a2122h-1b_1-5_2*NCC/3=O][a1122h-1b_1-5][a1122h-1a_1-5][a2112h-1b_1-5][a1221m-1a_1-5]/1-1-2-3-1-4-3-1-4-5-5/a4-b1_a6-k1_b4-c1_c3-d1_c6-g1_d2-e1_e4-f1_g2-h1_h4-i1_i2-j1\",\n",
    "           \"\"\"RES\n",
    "1b:b-dglc-HEX-1:5\n",
    "2s:n-acetyl\n",
    "3b:b-dglc-HEX-1:5\n",
    "4s:n-acetyl\n",
    "5b:b-dman-HEX-1:5\n",
    "6b:a-dman-HEX-1:5\n",
    "7b:b-dglc-HEX-1:5\n",
    "8s:n-acetyl\n",
    "9b:b-dgal-HEX-1:5\n",
    "10s:sulfate\n",
    "11s:n-acetyl\n",
    "12b:a-dman-HEX-1:5\n",
    "13b:b-dglc-HEX-1:5\n",
    "14s:n-acetyl\n",
    "15b:b-dgal-HEX-1:5\n",
    "16s:n-acetyl\n",
    "LIN\n",
    "1:1d(2+1)2n\n",
    "2:1o(4+1)3d\n",
    "3:3d(2+1)4n\n",
    "4:3o(4+1)5d\n",
    "5:5o(3+1)6d\n",
    "6:6o(2+1)7d\n",
    "7:7d(2+1)8n\n",
    "8:7o(4+1)9d\n",
    "9:9o(-1+1)10n\n",
    "10:9d(2+1)11n\n",
    "11:5o(6+1)12d\n",
    "12:12o(2+1)13d\n",
    "13:13d(2+1)14n\n",
    "14:13o(4+1)15d\n",
    "15:15d(2+1)16n\"\"\"]\n",
    "from glycowork.motif.annotate import annotate_dataset\n",
    "out = annotate_dataset(glycans, feature_set = ['known', 'terminal', 'exhaustive'], condense=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Internal_LewisX</th>\n",
       "      <th>Internal_LewisA</th>\n",
       "      <th>H_antigen_type2</th>\n",
       "      <th>Chitobiose</th>\n",
       "      <th>Trimannosylcore</th>\n",
       "      <th>Terminal_LacNAc_type1</th>\n",
       "      <th>Internal_LacNAc_type2</th>\n",
       "      <th>Terminal_LacNAc_type2</th>\n",
       "      <th>Terminal_LacdiNAc_type2</th>\n",
       "      <th>core_fucose</th>\n",
       "      <th>core_fucose(a1-3)</th>\n",
       "      <th>Fuc</th>\n",
       "      <th>Gal</th>\n",
       "      <th>GalNAc</th>\n",
       "      <th>GalNAcOS</th>\n",
       "      <th>GlcNAc</th>\n",
       "      <th>Man</th>\n",
       "      <th>Neu5Ac</th>\n",
       "      <th>Xyl</th>\n",
       "      <th>Fuc(a1-2)Gal</th>\n",
       "      <th>Fuc(a1-3)GlcNAc</th>\n",
       "      <th>Fuc(a1-4)GlcNAc</th>\n",
       "      <th>Fuc(a1-6)GlcNAc</th>\n",
       "      <th>Fuc(a1-?)GlcNAc</th>\n",
       "      <th>Gal(b1-3)GlcNAc</th>\n",
       "      <th>Gal(b1-4)GlcNAc</th>\n",
       "      <th>Gal(b1-?)GlcNAc</th>\n",
       "      <th>GalNAc(b1-4)GlcNAc</th>\n",
       "      <th>GalNAcOS(b1-4)GlcNAc</th>\n",
       "      <th>GlcNAc(b1-2)Man</th>\n",
       "      <th>GlcNAc(b1-4)GlcNAc</th>\n",
       "      <th>Man(a1-3)Man</th>\n",
       "      <th>Man(a1-6)Man</th>\n",
       "      <th>Man(a1-?)Man</th>\n",
       "      <th>Man(b1-4)GlcNAc</th>\n",
       "      <th>Neu5Ac(a2-3)Gal</th>\n",
       "      <th>Xyl(b1-2)Man</th>\n",
       "      <th>Terminal_GalNAcOS(b1-4)</th>\n",
       "      <th>Terminal_Fuc(a1-3)</th>\n",
       "      <th>Terminal_Gal(b1-4)</th>\n",
       "      <th>Terminal_Fuc(a1-4)</th>\n",
       "      <th>Terminal_Fuc(a1-6)</th>\n",
       "      <th>Terminal_Gal(b1-3)</th>\n",
       "      <th>Terminal_Neu5Ac(a2-3)</th>\n",
       "      <th>Terminal_GlcNAc(b1-2)</th>\n",
       "      <th>Terminal_Fuc(a1-2)</th>\n",
       "      <th>Terminal_Man(a1-6)</th>\n",
       "      <th>Terminal_Xyl(b1-2)</th>\n",
       "      <th>Terminal_Man(a1-3)</th>\n",
       "      <th>Terminal_GalNAc(b1-4)</th>\n",
       "      <th>Terminal_Fuc(a1-?)</th>\n",
       "      <th>Terminal_Gal(b1-?)</th>\n",
       "      <th>Terminal_Man(a1-?)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GalNAcOS(b1-4)GlcNAc(b1-2)Man(a1-3)[GalNAc(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "HTML(out.to_html())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "False\n",
      "\n",
      "Terminal structures:\n",
      "['Man(a1-3)', 'Man(a1-6)', 'Fuc(a1-6)']\n"
     ]
    }
   ],
   "source": [
    "#using graphs, you can easily check whether a glycan contains a specific motif; how about internal Lewis A/X motifs?\n",
    "from glycowork.motif.graph import subgraph_isomorphism\n",
    "print(subgraph_isomorphism('Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
    "                     'Fuc(a1-?)[Gal(b1-?)]GlcNAc', termini_list = ['terminal', 'internal', 'flexible']))\n",
    "print(subgraph_isomorphism('Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
    "                     'Fuc(a1-3/4)[Gal(b1-3/4)]GlcNAc', termini_list = ['t', 'i', 'f']))\n",
    "print(subgraph_isomorphism('Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
    "                     'dHex(a1-?)[Hex(b1-?)]GlcNAc', termini_list = ['t', 'i', 'f']))\n",
    "\n",
    "#or you could find the terminal epitopes of a glycan\n",
    "from glycowork.motif.annotate import get_terminal_structures\n",
    "print(\"\\nTerminal structures:\")\n",
    "print(get_terminal_structures('Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 compositions could not be matched. Run with verbose = True to see which compositions.\n",
      "                                               glycan  abundance\n",
      "0   GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Ma...          0\n",
      "1   GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Man(a1-6)]...          0\n",
      "2   GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[Man(a1-6)]...          0\n",
      "3   GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Ma...          0\n",
      "4   GalNAc(b1-3/4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]...          0\n",
      "5   GlcNAc(b1-2)Man(a1-6)[Man(a1-3)][GlcNAc(b1-4)]...          0\n",
      "6   GlcNAc(b1-2)Man(a1-3/6)[GlcNAc(b1-4)][Man(a1-3...          0\n",
      "7   Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]...          0\n",
      "8   GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-6)Man(a1-6)]Ma...          0\n",
      "9   GlcNAc(b1-4)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Ma...          0\n",
      "10  GlcNAc(b1-4)Man(a1-3)[GlcNAc(b1-4)Man(a1-6)]Ma...          0\n",
      "11  GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)[GlcNAc(b1-4...          0\n",
      "12  GlcNAc(b1-4)Man(a1-3)[GlcNAc(b1-6)Man(a1-6)]Ma...          0\n",
      "13  Man(a1-3)[GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)]...          0\n",
      "14  GalNAc(b1-4)GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Ma...          0\n",
      "\n",
      "Mass of the composition Hex3HexNAc4\n",
      "1316.4865545999999\n",
      "1316.4865545999999\n",
      "1316.4865545999999\n"
     ]
    }
   ],
   "source": [
    "#given a composition, find matching glycan structures in SugarBase; specific for glycan classes and taxonomy\n",
    "from glycowork.motif.tokenization import compositions_to_structures\n",
    "print(compositions_to_structures([{'Hex':3, 'HexNAc':4}], glycan_class = 'N'))\n",
    "\n",
    "#or we could calculate the mass of this composition\n",
    "from glycowork.motif.tokenization import composition_to_mass\n",
    "print(\"\\nMass of the composition Hex3HexNAc4\")\n",
    "print(composition_to_mass({'Hex':3, 'HexNAc':4}))\n",
    "print(composition_to_mass(\"H3N4\"))\n",
    "print(composition_to_mass(\"Hex3HexNAc4\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
