<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>motif – glycowork</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-000f0825b9c55ed21d14970d810c14a9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="motif – glycowork">
<meta property="og:description" content="Package for processing and analyzing glycans">
<meta property="og:image" content="https://BojarLab.github.io/glycowork/03_motif_files/figure-html/cell-3-output-1.png">
<meta property="og:site_name" content="glycowork">
<meta property="og:image:height" content="510">
<meta property="og:image:width" content="700">
<meta name="twitter:title" content="motif – glycowork">
<meta name="twitter:description" content="Package for processing and analyzing glycans">
<meta name="twitter:image" content="https://BojarLab.github.io/glycowork/03_motif_files/figure-html/cell-3-output-1.png">
<meta name="twitter:image-height" content="510">
<meta name="twitter:image-width" content="700">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">glycowork</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./motif.html">motif</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">glycowork</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./glycan_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">glycan_data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ml</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motif.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">motif</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./network.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">network</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Examples</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#draw" id="toc-draw" class="nav-link active" data-scroll-target="#draw">draw</a>
  <ul class="collapse">
  <li><a href="#glycodraw" id="toc-glycodraw" class="nav-link" data-scroll-target="#glycodraw">GlycoDraw</a></li>
  <li><a href="#annotate_figure" id="toc-annotate_figure" class="nav-link" data-scroll-target="#annotate_figure">annotate_figure</a></li>
  <li><a href="#plot_glycans_excel" id="toc-plot_glycans_excel" class="nav-link" data-scroll-target="#plot_glycans_excel">plot_glycans_excel</a></li>
  </ul></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">analysis</a>
  <ul class="collapse">
  <li><a href="#get_pvals_motifs" id="toc-get_pvals_motifs" class="nav-link" data-scroll-target="#get_pvals_motifs">get_pvals_motifs</a></li>
  <li><a href="#get_representative_substructures" id="toc-get_representative_substructures" class="nav-link" data-scroll-target="#get_representative_substructures">get_representative_substructures</a></li>
  <li><a href="#get_heatmap" id="toc-get_heatmap" class="nav-link" data-scroll-target="#get_heatmap">get_heatmap</a></li>
  <li><a href="#plot_embeddings" id="toc-plot_embeddings" class="nav-link" data-scroll-target="#plot_embeddings">plot_embeddings</a></li>
  <li><a href="#characterize_monosaccharide" id="toc-characterize_monosaccharide" class="nav-link" data-scroll-target="#characterize_monosaccharide">characterize_monosaccharide</a></li>
  <li><a href="#get_differential_expression" id="toc-get_differential_expression" class="nav-link" data-scroll-target="#get_differential_expression">get_differential_expression</a></li>
  <li><a href="#get_volcano" id="toc-get_volcano" class="nav-link" data-scroll-target="#get_volcano">get_volcano</a></li>
  <li><a href="#get_coverage" id="toc-get_coverage" class="nav-link" data-scroll-target="#get_coverage">get_coverage</a></li>
  <li><a href="#get_pca" id="toc-get_pca" class="nav-link" data-scroll-target="#get_pca">get_pca</a></li>
  <li><a href="#get_pval_distribution" id="toc-get_pval_distribution" class="nav-link" data-scroll-target="#get_pval_distribution">get_pval_distribution</a></li>
  <li><a href="#get_ma" id="toc-get_ma" class="nav-link" data-scroll-target="#get_ma">get_ma</a></li>
  <li><a href="#get_glycanova" id="toc-get_glycanova" class="nav-link" data-scroll-target="#get_glycanova">get_glycanova</a></li>
  <li><a href="#get_meta_analysis" id="toc-get_meta_analysis" class="nav-link" data-scroll-target="#get_meta_analysis">get_meta_analysis</a></li>
  <li><a href="#get_time_series" id="toc-get_time_series" class="nav-link" data-scroll-target="#get_time_series">get_time_series</a></li>
  <li><a href="#get_jtk" id="toc-get_jtk" class="nav-link" data-scroll-target="#get_jtk">get_jtk</a></li>
  <li><a href="#get_biodiversity" id="toc-get_biodiversity" class="nav-link" data-scroll-target="#get_biodiversity">get_biodiversity</a></li>
  <li><a href="#get_sparcc" id="toc-get_sparcc" class="nav-link" data-scroll-target="#get_sparcc">get_SparCC</a></li>
  <li><a href="#get_roc" id="toc-get_roc" class="nav-link" data-scroll-target="#get_roc">get_roc</a></li>
  <li><a href="#get_lectin_array" id="toc-get_lectin_array" class="nav-link" data-scroll-target="#get_lectin_array">get_lectin_array</a></li>
  <li><a href="#get_glycoshift_per_site" id="toc-get_glycoshift_per_site" class="nav-link" data-scroll-target="#get_glycoshift_per_site">get_glycoshift_per_site</a></li>
  </ul></li>
  <li><a href="#annotate" id="toc-annotate" class="nav-link" data-scroll-target="#annotate">annotate</a>
  <ul class="collapse">
  <li><a href="#annotate_glycan" id="toc-annotate_glycan" class="nav-link" data-scroll-target="#annotate_glycan">annotate_glycan</a></li>
  <li><a href="#annotate_dataset" id="toc-annotate_dataset" class="nav-link" data-scroll-target="#annotate_dataset">annotate_dataset</a></li>
  <li><a href="#quantify_motifs" id="toc-quantify_motifs" class="nav-link" data-scroll-target="#quantify_motifs">quantify_motifs</a></li>
  <li><a href="#get_k_saccharides" id="toc-get_k_saccharides" class="nav-link" data-scroll-target="#get_k_saccharides">get_k_saccharides</a></li>
  <li><a href="#get_terminal_structures" id="toc-get_terminal_structures" class="nav-link" data-scroll-target="#get_terminal_structures">get_terminal_structures</a></li>
  <li><a href="#get_molecular_properties" id="toc-get_molecular_properties" class="nav-link" data-scroll-target="#get_molecular_properties">get_molecular_properties</a></li>
  <li><a href="#get_glycan_similarity" id="toc-get_glycan_similarity" class="nav-link" data-scroll-target="#get_glycan_similarity">get_glycan_similarity</a></li>
  </ul></li>
  <li><a href="#graph" id="toc-graph" class="nav-link" data-scroll-target="#graph">graph</a>
  <ul class="collapse">
  <li><a href="#glycan_to_nxgraph" id="toc-glycan_to_nxgraph" class="nav-link" data-scroll-target="#glycan_to_nxgraph">glycan_to_nxGraph</a></li>
  <li><a href="#graph_to_string" id="toc-graph_to_string" class="nav-link" data-scroll-target="#graph_to_string">graph_to_string</a></li>
  <li><a href="#compare_glycans" id="toc-compare_glycans" class="nav-link" data-scroll-target="#compare_glycans">compare_glycans</a></li>
  <li><a href="#subgraph_isomorphism" id="toc-subgraph_isomorphism" class="nav-link" data-scroll-target="#subgraph_isomorphism">subgraph_isomorphism</a></li>
  <li><a href="#generate_graph_features" id="toc-generate_graph_features" class="nav-link" data-scroll-target="#generate_graph_features">generate_graph_features</a></li>
  <li><a href="#largest_subgraph" id="toc-largest_subgraph" class="nav-link" data-scroll-target="#largest_subgraph">largest_subgraph</a></li>
  <li><a href="#ensure_graph" id="toc-ensure_graph" class="nav-link" data-scroll-target="#ensure_graph">ensure_graph</a></li>
  <li><a href="#get_possible_topologies" id="toc-get_possible_topologies" class="nav-link" data-scroll-target="#get_possible_topologies">get_possible_topologies</a></li>
  <li><a href="#possible_topology_check" id="toc-possible_topology_check" class="nav-link" data-scroll-target="#possible_topology_check">possible_topology_check</a></li>
  <li><a href="#deduplicate_glycans" id="toc-deduplicate_glycans" class="nav-link" data-scroll-target="#deduplicate_glycans">deduplicate_glycans</a></li>
  </ul></li>
  <li><a href="#processing" id="toc-processing" class="nav-link" data-scroll-target="#processing">processing</a>
  <ul class="collapse">
  <li><a href="#min_process_glycans" id="toc-min_process_glycans" class="nav-link" data-scroll-target="#min_process_glycans">min_process_glycans</a></li>
  <li><a href="#get_lib" id="toc-get_lib" class="nav-link" data-scroll-target="#get_lib">get_lib</a></li>
  <li><a href="#expand_lib" id="toc-expand_lib" class="nav-link" data-scroll-target="#expand_lib">expand_lib</a></li>
  <li><a href="#presence_to_matrix" id="toc-presence_to_matrix" class="nav-link" data-scroll-target="#presence_to_matrix">presence_to_matrix</a></li>
  <li><a href="#enforce_class" id="toc-enforce_class" class="nav-link" data-scroll-target="#enforce_class">enforce_class</a></li>
  <li><a href="#iupac_to_smiles" id="toc-iupac_to_smiles" class="nav-link" data-scroll-target="#iupac_to_smiles">IUPAC_to_SMILES</a></li>
  <li><a href="#canonicalize_composition" id="toc-canonicalize_composition" class="nav-link" data-scroll-target="#canonicalize_composition">canonicalize_composition</a></li>
  <li><a href="#canonicalize_iupac" id="toc-canonicalize_iupac" class="nav-link" data-scroll-target="#canonicalize_iupac">canonicalize_iupac</a></li>
  <li><a href="#get_possible_linkages" id="toc-get_possible_linkages" class="nav-link" data-scroll-target="#get_possible_linkages">get_possible_linkages</a></li>
  <li><a href="#get_possible_monosaccharides" id="toc-get_possible_monosaccharides" class="nav-link" data-scroll-target="#get_possible_monosaccharides">get_possible_monosaccharides</a></li>
  <li><a href="#equal_repeats" id="toc-equal_repeats" class="nav-link" data-scroll-target="#equal_repeats">equal_repeats</a></li>
  <li><a href="#get_class" id="toc-get_class" class="nav-link" data-scroll-target="#get_class">get_class</a></li>
  </ul></li>
  <li><a href="#query" id="toc-query" class="nav-link" data-scroll-target="#query">query</a>
  <ul class="collapse">
  <li><a href="#get_insight" id="toc-get_insight" class="nav-link" data-scroll-target="#get_insight">get_insight</a></li>
  <li><a href="#glytoucan_to_glycan" id="toc-glytoucan_to_glycan" class="nav-link" data-scroll-target="#glytoucan_to_glycan">glytoucan_to_glycan</a></li>
  </ul></li>
  <li><a href="#regex" id="toc-regex" class="nav-link" data-scroll-target="#regex">regex</a>
  <ul class="collapse">
  <li><a href="#get_match" id="toc-get_match" class="nav-link" data-scroll-target="#get_match">get_match</a></li>
  <li><a href="#get_match_batch" id="toc-get_match_batch" class="nav-link" data-scroll-target="#get_match_batch">get_match_batch</a></li>
  <li><a href="#motif_to_regex" id="toc-motif_to_regex" class="nav-link" data-scroll-target="#motif_to_regex">motif_to_regex</a></li>
  </ul></li>
  <li><a href="#tokenization" id="toc-tokenization" class="nav-link" data-scroll-target="#tokenization">tokenization</a>
  <ul class="collapse">
  <li><a href="#string_to_labels" id="toc-string_to_labels" class="nav-link" data-scroll-target="#string_to_labels">string_to_labels</a></li>
  <li><a href="#pad_sequence" id="toc-pad_sequence" class="nav-link" data-scroll-target="#pad_sequence">pad_sequence</a></li>
  <li><a href="#stemify_glycan" id="toc-stemify_glycan" class="nav-link" data-scroll-target="#stemify_glycan">stemify_glycan</a></li>
  <li><a href="#stemify_dataset" id="toc-stemify_dataset" class="nav-link" data-scroll-target="#stemify_dataset">stemify_dataset</a></li>
  <li><a href="#mask_rare_glycoletters" id="toc-mask_rare_glycoletters" class="nav-link" data-scroll-target="#mask_rare_glycoletters">mask_rare_glycoletters</a></li>
  <li><a href="#mz_to_composition" id="toc-mz_to_composition" class="nav-link" data-scroll-target="#mz_to_composition">mz_to_composition</a></li>
  <li><a href="#match_composition_relaxed" id="toc-match_composition_relaxed" class="nav-link" data-scroll-target="#match_composition_relaxed">match_composition_relaxed</a></li>
  <li><a href="#condense_composition_matching" id="toc-condense_composition_matching" class="nav-link" data-scroll-target="#condense_composition_matching">condense_composition_matching</a></li>
  <li><a href="#mz_to_structures" id="toc-mz_to_structures" class="nav-link" data-scroll-target="#mz_to_structures">mz_to_structures</a></li>
  <li><a href="#compositions_to_structures" id="toc-compositions_to_structures" class="nav-link" data-scroll-target="#compositions_to_structures">compositions_to_structures</a></li>
  <li><a href="#structure_to_basic" id="toc-structure_to_basic" class="nav-link" data-scroll-target="#structure_to_basic">structure_to_basic</a></li>
  <li><a href="#glycan_to_composition" id="toc-glycan_to_composition" class="nav-link" data-scroll-target="#glycan_to_composition">glycan_to_composition</a></li>
  <li><a href="#glycan_to_mass" id="toc-glycan_to_mass" class="nav-link" data-scroll-target="#glycan_to_mass">glycan_to_mass</a></li>
  <li><a href="#composition_to_mass" id="toc-composition_to_mass" class="nav-link" data-scroll-target="#composition_to_mass">composition_to_mass</a></li>
  <li><a href="#calculate_adduct_mass" id="toc-calculate_adduct_mass" class="nav-link" data-scroll-target="#calculate_adduct_mass">calculate_adduct_mass</a></li>
  <li><a href="#get_unique_topologies" id="toc-get_unique_topologies" class="nav-link" data-scroll-target="#get_unique_topologies">get_unique_topologies</a></li>
  <li><a href="#get_random_glycan" id="toc-get_random_glycan" class="nav-link" data-scroll-target="#get_random_glycan">get_random_glycan</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BojarLab/glycowork/tree/master/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">motif</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p><code>motif</code> contains many functions to process glycans in various ways and use this processing to analyze glycans via curated motifs, graph features, and sequence features. It contains the following modules:</p>
<ul>
<li><code>draw</code> contains the GlycoDraw function to draw glycans in SNFG style</li>
<li><code>analysis</code> contains functions for downstream analyses of important glycan motifs etc.</li>
<li><code>annotate</code> contains functions to extract curated motifs, graph features, and sequence features from glycan sequences</li>
<li><code>graph</code> is used to convert glycan sequences to graphs and contains helper functions to search for motifs / check whether two sequences describe the same sequence, etc.</li>
<li><code>processing</code> contains functions to process IUPAC-condensed glycan sequences, as well as conversion functions to convert other nomenclatures into IUPAC-condensed.</li>
<li><code>regex</code> contains functionality for performing powerful regular expression-like searches on glycans; <code>get_match</code> is the user-facing function.</li>
<li><code>query</code> is used to interact with the databases contained in glycowork, delivering insights for sequences of interest</li>
<li><code>tokenization</code> has helper functions to map <em>m/z</em>–&gt;composition, composition–&gt;structure, structure–&gt;motif, and more</li>
</ul>
<section id="draw" class="level2">
<h2 class="anchored" data-anchor-id="draw">draw</h2>
<blockquote class="blockquote">
<p>drawing glycans in SNFG style</p>
</blockquote>
<hr>
<section id="glycodraw" class="level3">
<h3 class="anchored" data-anchor-id="glycodraw">GlycoDraw</h3>
<blockquote class="blockquote">
<pre><code> GlycoDraw (glycan:str, vertical:bool=False, compact:bool=False,
            show_linkage:bool=True, dim:float=50,
            highlight_motif:str|None=None, highlight_termini_list:list=[],
            highlight_linkages:list[int]|None=None,
            reverse_highlight:bool=False, repeat:bool|int|str|None=None,
            repeat_range:list[int]|None=None, draw_method:str|None=None,
            filepath:str|pathlib.Path|None=None, suppress:bool=False,
            per_residue:list=[], pdb_file:str|pathlib.Path|None=None,
            alt_text:str|None=None, libr:dict|None=None,
            reducing_end_label:str|None=None, restrict_vocab:bool=False)</code></pre>
</blockquote>
<p><em>Renders glycan structure using SNFG symbols or chemical structure representation</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>IUPAC-condensed glycan sequence</td>
</tr>
<tr class="even">
<td>vertical</td>
<td>bool</td>
<td>False</td>
<td>Draw vertically</td>
</tr>
<tr class="odd">
<td>compact</td>
<td>bool</td>
<td>False</td>
<td>Use compact style</td>
</tr>
<tr class="even">
<td>show_linkage</td>
<td>bool</td>
<td>True</td>
<td>Show linkage labels</td>
</tr>
<tr class="odd">
<td>dim</td>
<td>float</td>
<td>50</td>
<td>Base dimension for scaling</td>
</tr>
<tr class="even">
<td>highlight_motif</td>
<td>str | None</td>
<td>None</td>
<td>Motif to highlight</td>
</tr>
<tr class="odd">
<td>highlight_termini_list</td>
<td>list</td>
<td>[]</td>
<td>Terminal positions (from ‘terminal’, ‘internal’, and ‘flexible’)</td>
</tr>
<tr class="even">
<td>highlight_linkages</td>
<td>list[int] | None</td>
<td>None</td>
<td>Which linkages to highlight in a different color; indices, starting from 0, in glycan</td>
</tr>
<tr class="odd">
<td>reverse_highlight</td>
<td>bool</td>
<td>False</td>
<td>Whether to highlight everything EXCEPT highlight_motif</td>
</tr>
<tr class="even">
<td>repeat</td>
<td>bool | int | str | None</td>
<td>None</td>
<td>Repeat unit specification (True: n units, int: # of units, str: range of units)</td>
</tr>
<tr class="odd">
<td>repeat_range</td>
<td>list[int] | None</td>
<td>None</td>
<td>Repeat unit range</td>
</tr>
<tr class="even">
<td>draw_method</td>
<td>str | None</td>
<td>None</td>
<td>Drawing method: None, ‘chem2d’, ‘chem3d’</td>
</tr>
<tr class="odd">
<td>filepath</td>
<td>str | pathlib.Path | None</td>
<td>None</td>
<td>Output file path</td>
</tr>
<tr class="even">
<td>suppress</td>
<td>bool</td>
<td>False</td>
<td>Suppress display</td>
</tr>
<tr class="odd">
<td>per_residue</td>
<td>list</td>
<td>[]</td>
<td>Per-residue intensity values (order should be the same as the monosaccharides in glycan string)</td>
</tr>
<tr class="even">
<td>pdb_file</td>
<td>str | pathlib.Path | None</td>
<td>None</td>
<td>only used when draw_method=‘chem3d’; already existing glycan structure</td>
</tr>
<tr class="odd">
<td>alt_text</td>
<td>str | None</td>
<td>None</td>
<td>Custom ALT text for accessibility</td>
</tr>
<tr class="even">
<td>libr</td>
<td>dict | None</td>
<td>None</td>
<td>Can be modified for drawing too exotic monosaccharides</td>
</tr>
<tr class="odd">
<td>reducing_end_label</td>
<td>str | None</td>
<td>None</td>
<td>Label to be drawn connected to the reducing end</td>
</tr>
<tr class="even">
<td>restrict_vocab</td>
<td>bool</td>
<td>False</td>
<td>Whether only tokens present in libr can be drawn</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>Any</strong></td>
<td></td>
<td><strong>Drawing object</strong></td>
</tr>
</tbody>
</table>
<div id="intensive-receiver" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>GlycoDraw(<span class="st">"Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>         highlight_motif <span class="op">=</span> <span class="st">"GlcNAc(b1-?)Man"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-3-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="annotate_figure" class="level3">
<h3 class="anchored" data-anchor-id="annotate_figure">annotate_figure</h3>
<blockquote class="blockquote">
<pre><code> annotate_figure (svg_input:str, scale_range:tuple[int,int]=(25, 80),
                  compact:bool=False, glycan_size:str='medium',
                  filepath:str|pathlib.Path='',
                  scale_by_DE_res:pandas.core.frame.DataFrame|None=None,
                  x_thresh:float=1, y_thresh:float=0.05,
                  x_metric:str='Log2FC')</code></pre>
</blockquote>
<p><em>Replaces text labels with glycan drawings in SVG figure</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>svg_input</td>
<td>str</td>
<td></td>
<td>Input SVG file path</td>
</tr>
<tr class="even">
<td>scale_range</td>
<td>tuple</td>
<td>(25, 80)</td>
<td>Min/max glycan dimensions</td>
</tr>
<tr class="odd">
<td>compact</td>
<td>bool</td>
<td>False</td>
<td>Use compact style</td>
</tr>
<tr class="even">
<td>glycan_size</td>
<td>str</td>
<td>medium</td>
<td>Glycan size preset (‘small’, ‘medium’, ‘large’)</td>
</tr>
<tr class="odd">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Output file path</td>
</tr>
<tr class="even">
<td>scale_by_DE_res</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Differential expression results (motif_analysis.get_differential_expression)</td>
</tr>
<tr class="odd">
<td>x_thresh</td>
<td>float</td>
<td>1</td>
<td>X metric threshold</td>
</tr>
<tr class="even">
<td>y_thresh</td>
<td>float</td>
<td>0.05</td>
<td>P-value threshold</td>
</tr>
<tr class="odd">
<td>x_metric</td>
<td>str</td>
<td>Log2FC</td>
<td>X axis metric (‘Log2FC’, ‘Effect size’)</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str | None</strong></td>
<td></td>
<td><strong>Modified SVG code</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="plot_glycans_excel" class="level3">
<h3 class="anchored" data-anchor-id="plot_glycans_excel">plot_glycans_excel</h3>
<blockquote class="blockquote">
<pre><code> plot_glycans_excel (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                     folder_filepath:str|pathlib.Path,
                     glycan_col_num:int=0, scaling_factor:float=0.2,
                     compact:bool=False)</code></pre>
</blockquote>
<p><em>Creates Excel file with SNFG glycan images in a new column</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame or filepath with glycans</td>
</tr>
<tr class="even">
<td>folder_filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Output folder path</td>
</tr>
<tr class="odd">
<td>glycan_col_num</td>
<td>int</td>
<td>0</td>
<td>Glycan column index</td>
</tr>
<tr class="even">
<td>scaling_factor</td>
<td>float</td>
<td>0.2</td>
<td>Image scaling</td>
</tr>
<tr class="odd">
<td>compact</td>
<td>bool</td>
<td>False</td>
<td>Use compact style</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="analysis" class="level2">
<h2 class="anchored" data-anchor-id="analysis">analysis</h2>
<blockquote class="blockquote">
<p>downstream analyses of important glycan motifs</p>
</blockquote>
<hr>
<section id="get_pvals_motifs" class="level3">
<h3 class="anchored" data-anchor-id="get_pvals_motifs">get_pvals_motifs</h3>
<blockquote class="blockquote">
<pre><code> get_pvals_motifs (df:pandas.core.frame.DataFrame|str,
                   glycan_col_name:str='glycan',
                   label_col_name:str='target', zscores:bool=True,
                   thresh:float=1.645, sorting:bool=True,
                   feature_set:list[str]=['exhaustive'],
                   multiple_samples:bool=False,
                   motifs:pandas.core.frame.DataFrame|None=None,
                   custom_motifs:list[str]=[])</code></pre>
</blockquote>
<p><em>Identifies significantly enriched glycan motifs using Welch’s t-test with FDR correction and Cohen’s d effect size calculation, comparing samples above/below threshold</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str</td>
<td></td>
<td>Input dataframe or filepath (.csv/.xlsx)</td>
</tr>
<tr class="even">
<td>glycan_col_name</td>
<td>str</td>
<td>glycan</td>
<td>Column name for glycan sequences</td>
</tr>
<tr class="odd">
<td>label_col_name</td>
<td>str</td>
<td>target</td>
<td>Column name for labels</td>
</tr>
<tr class="even">
<td>zscores</td>
<td>bool</td>
<td>True</td>
<td>Whether data are z-scores</td>
</tr>
<tr class="odd">
<td>thresh</td>
<td>float</td>
<td>1.645</td>
<td>Threshold to separate positive/negative</td>
</tr>
<tr class="even">
<td>sorting</td>
<td>bool</td>
<td>True</td>
<td>Sort p-value dataframe</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘exhaustive’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="even">
<td>multiple_samples</td>
<td>bool</td>
<td>False</td>
<td>Multiple samples with glycan columns</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Modified motif_list</td>
</tr>
<tr class="even">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with p-values, FDR-corrected p-values, and Cohen’s d effect sizes for glycan motifs</strong></td>
</tr>
</tbody>
</table>
<div id="close-director" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>glycans <span class="op">=</span> [<span class="st">'Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc'</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">'GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcOPN(b1-6)GlcOPN'</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>          <span class="st">'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Glc(b1-3)Glc(b1-3)Glc'</span>]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>label <span class="op">=</span> [<span class="fl">3.234</span>, <span class="fl">2.423</span>, <span class="fl">0.733</span>, <span class="fl">3.102</span>, <span class="fl">0.108</span>]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>test_df <span class="op">=</span> pd.DataFrame({<span class="st">'glycan'</span>:glycans, <span class="st">'binding'</span>:label})</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Glyco-Motif enrichment p-value test"</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> get_pvals_motifs(test_df, <span class="st">'glycan'</span>, <span class="st">'binding'</span>).iloc[:<span class="dv">10</span>,:]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Glyco-Motif enrichment p-value test</code></pre>
</div>
</div>
<div id="reported-traveler" class="cell">
<div class="cell-output cell-output-display">
<style type="text/css">
#T_3d854_row0_col0, #T_3d854_row0_col1, #T_3d854_row0_col2, #T_3d854_row0_col3, #T_3d854_row1_col0, #T_3d854_row1_col1, #T_3d854_row1_col2, #T_3d854_row1_col3, #T_3d854_row2_col0, #T_3d854_row2_col1, #T_3d854_row2_col2, #T_3d854_row2_col3, #T_3d854_row3_col0, #T_3d854_row3_col1, #T_3d854_row3_col2, #T_3d854_row3_col3, #T_3d854_row4_col0, #T_3d854_row4_col1, #T_3d854_row4_col2, #T_3d854_row4_col3, #T_3d854_row5_col0, #T_3d854_row5_col1, #T_3d854_row5_col2, #T_3d854_row5_col3, #T_3d854_row6_col0, #T_3d854_row6_col1, #T_3d854_row6_col2, #T_3d854_row6_col3, #T_3d854_row7_col0, #T_3d854_row7_col1, #T_3d854_row7_col2, #T_3d854_row7_col3, #T_3d854_row8_col0, #T_3d854_row8_col1, #T_3d854_row8_col2, #T_3d854_row8_col3, #T_3d854_row9_col0, #T_3d854_row9_col1, #T_3d854_row9_col2, #T_3d854_row9_col3 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>

<table id="T_3d854" class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_3d854_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">motif</th>
<th id="T_3d854_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">pval</th>
<th id="T_3d854_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">corr_pval</th>
<th id="T_3d854_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">effect_size</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th id="T_3d854_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">4</th>
<td id="T_3d854_row0_col0" class="data row0 col0">GlcNAc</td>
<td id="T_3d854_row0_col1" class="data row0 col1">0.038120</td>
<td id="T_3d854_row0_col2" class="data row0 col2">0.205849</td>
<td id="T_3d854_row0_col3" class="data row0 col3">1.530905</td>
</tr>
<tr class="even">
<th id="T_3d854_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">8</th>
<td id="T_3d854_row1_col0" class="data row1 col0">Man</td>
<td id="T_3d854_row1_col1" class="data row1 col1">0.054356</td>
<td id="T_3d854_row1_col2" class="data row1 col2">0.234990</td>
<td id="T_3d854_row1_col3" class="data row1 col3">1.390253</td>
</tr>
<tr class="odd">
<th id="T_3d854_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">24</th>
<td id="T_3d854_row2_col0" class="data row2 col0">Man(a1-?)Man</td>
<td id="T_3d854_row2_col1" class="data row2 col1">0.060923</td>
<td id="T_3d854_row2_col2" class="data row2 col2">0.234990</td>
<td id="T_3d854_row2_col3" class="data row2 col3">1.308333</td>
</tr>
<tr class="even">
<th id="T_3d854_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">22</th>
<td id="T_3d854_row3_col0" class="data row3 col0">Man(a1-3)Man</td>
<td id="T_3d854_row3_col1" class="data row3 col1">0.034212</td>
<td id="T_3d854_row3_col2" class="data row3 col2">0.205849</td>
<td id="T_3d854_row3_col3" class="data row3 col3">1.196586</td>
</tr>
<tr class="odd">
<th id="T_3d854_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">14</th>
<td id="T_3d854_row4_col0" class="data row4 col0">GlcNAc(b1-4)GlcNAc</td>
<td id="T_3d854_row4_col1" class="data row4 col1">0.019543</td>
<td id="T_3d854_row4_col2" class="data row4 col2">0.175885</td>
<td id="T_3d854_row4_col3" class="data row4 col3">1.168815</td>
</tr>
<tr class="even">
<th id="T_3d854_level0_row5" class="row_heading level0 row5" data-quarto-table-cell-role="th">23</th>
<td id="T_3d854_row5_col0" class="data row5 col0">Man(a1-6)Man</td>
<td id="T_3d854_row5_col1" class="data row5 col1">0.019543</td>
<td id="T_3d854_row5_col2" class="data row5 col2">0.175885</td>
<td id="T_3d854_row5_col3" class="data row5 col3">1.168815</td>
</tr>
<tr class="odd">
<th id="T_3d854_level0_row6" class="row_heading level0 row6" data-quarto-table-cell-role="th">25</th>
<td id="T_3d854_row6_col0" class="data row6 col0">Man(b1-4)GlcNAc</td>
<td id="T_3d854_row6_col1" class="data row6 col1">0.019543</td>
<td id="T_3d854_row6_col2" class="data row6 col2">0.175885</td>
<td id="T_3d854_row6_col3" class="data row6 col3">1.168815</td>
</tr>
<tr class="even">
<th id="T_3d854_level0_row7" class="row_heading level0 row7" data-quarto-table-cell-role="th">7</th>
<td id="T_3d854_row7_col0" class="data row7 col0">Kdo</td>
<td id="T_3d854_row7_col1" class="data row7 col1">0.328790</td>
<td id="T_3d854_row7_col2" class="data row7 col2">0.479672</td>
<td id="T_3d854_row7_col3" class="data row7 col3">-0.811679</td>
</tr>
<tr class="odd">
<th id="T_3d854_level0_row8" class="row_heading level0 row8" data-quarto-table-cell-role="th">2</th>
<td id="T_3d854_row8_col0" class="data row8 col0">Glc</td>
<td id="T_3d854_row8_col1" class="data row8 col1">0.644180</td>
<td id="T_3d854_row8_col2" class="data row8 col2">0.668956</td>
<td id="T_3d854_row8_col3" class="data row8 col3">-0.811679</td>
</tr>
<tr class="even">
<th id="T_3d854_level0_row9" class="row_heading level0 row9" data-quarto-table-cell-role="th">21</th>
<td id="T_3d854_row9_col0" class="data row9 col0">Man(a1-2)Man</td>
<td id="T_3d854_row9_col1" class="data row9 col1">0.177461</td>
<td id="T_3d854_row9_col2" class="data row9 col2">0.479672</td>
<td id="T_3d854_row9_col3" class="data row9 col3">0.772320</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
<section id="get_representative_substructures" class="level3">
<h3 class="anchored" data-anchor-id="get_representative_substructures">get_representative_substructures</h3>
<blockquote class="blockquote">
<pre><code> get_representative_substructures
                                   (enrichment_df:pandas.core.frame.DataFr
                                   ame)</code></pre>
</blockquote>
<p><em>Constructs minimal glycan structures that represent significantly enriched motifs by optimizing for motif content while minimizing structure size using subgraph isomorphism</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>enrichment_df</td>
<td>DataFrame</td>
<td>Output from get_pvals_motifs</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td><strong>Up to 10 minimal glycans containing enriched motifs</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="get_heatmap" class="level3">
<h3 class="anchored" data-anchor-id="get_heatmap">get_heatmap</h3>
<blockquote class="blockquote">
<pre><code> get_heatmap (df:pandas.core.frame.DataFrame|str|pathlib.Path,
              motifs:bool=False, feature_set:list[str]=['known'],
              transform:str='', datatype:str='response',
              rarity_filter:float=0.05, filepath:str|pathlib.Path='',
              index_col:str='glycan', custom_motifs:list[str]=[],
              return_plot:bool=False, show_all:bool=False, **kwargs:Any)</code></pre>
</blockquote>
<p><em>Creates hierarchically clustered heatmap visualization of glycan/motif abundances</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>Input dataframe or filepath (.csv/.xlsx)</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘known’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str</td>
<td></td>
<td>Transform data before plotting</td>
</tr>
<tr class="odd">
<td>datatype</td>
<td>str</td>
<td>response</td>
<td>Data type: ‘response’ for quantitative values or ‘presence’ for presence/absence</td>
</tr>
<tr class="even">
<td>rarity_filter</td>
<td>float</td>
<td>0.05</td>
<td>Min proportion for non-zero values</td>
</tr>
<tr class="odd">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="even">
<td>index_col</td>
<td>str</td>
<td>glycan</td>
<td>Column to use as index</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>return_plot</td>
<td>bool</td>
<td>False</td>
<td>Return plot object</td>
</tr>
<tr class="odd">
<td>show_all</td>
<td>bool</td>
<td>False</td>
<td>Show all tick labels</td>
</tr>
<tr class="even">
<td>kwargs</td>
<td>Any</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>Optional</strong></td>
<td></td>
<td><strong>None or plot object if return_plot=True</strong></td>
</tr>
</tbody>
</table>
<div id="seasonal-termination" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>glycans <span class="op">=</span> [<span class="st">'Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc'</span>,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">'GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P'</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Glc(b1-3)Glc(b1-3)Glc'</span>]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>label <span class="op">=</span> [<span class="fl">3.234</span>, <span class="fl">2.423</span>, <span class="fl">0.733</span>, <span class="fl">3.102</span>, <span class="fl">0.108</span>]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>label2 <span class="op">=</span> [<span class="fl">0.134</span>, <span class="fl">0.345</span>, <span class="fl">1.15</span>, <span class="fl">0.233</span>, <span class="fl">2.981</span>]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>label3 <span class="op">=</span> [<span class="fl">0.334</span>, <span class="fl">0.245</span>, <span class="fl">1.55</span>, <span class="fl">0.133</span>, <span class="fl">2.581</span>]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>test_df <span class="op">=</span> pd.DataFrame([label, label2, label3], columns <span class="op">=</span> glycans)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>get_heatmap(test_df, motifs <span class="op">=</span> <span class="va">True</span>, feature_set <span class="op">=</span> [<span class="st">'known'</span>, <span class="st">'exhaustive'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-11-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="plot_embeddings" class="level3">
<h3 class="anchored" data-anchor-id="plot_embeddings">plot_embeddings</h3>
<blockquote class="blockquote">
<pre><code> plot_embeddings (glycans:list[str],
                  emb:dict[str,numpy.ndarray]|pandas.core.frame.DataFrame|
                  None=None, label_list:list[typing.Any]|None=None,
                  shape_feature:str|None=None,
                  filepath:str|pathlib.Path='', alpha:float=0.8,
                  palette:str='colorblind', **kwargs:Any)</code></pre>
</blockquote>
<p><em>Visualizes learned glycan embeddings using t-SNE dimensionality reduction with optional group coloring</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycans</td>
<td>list</td>
<td></td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="even">
<td>emb</td>
<td>dict[str, numpy.ndarray] | pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Glycan embeddings dict/DataFrame; defaults to SweetNet embeddings</td>
</tr>
<tr class="odd">
<td>label_list</td>
<td>list[typing.Any] | None</td>
<td>None</td>
<td>Labels for coloring points</td>
</tr>
<tr class="even">
<td>shape_feature</td>
<td>str | None</td>
<td>None</td>
<td>Monosaccharide/bond for point shapes</td>
</tr>
<tr class="odd">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="even">
<td>alpha</td>
<td>float</td>
<td>0.8</td>
<td>Point transparency</td>
</tr>
<tr class="odd">
<td>palette</td>
<td>str</td>
<td>colorblind</td>
<td>Color palette for groups</td>
</tr>
<tr class="even">
<td>kwargs</td>
<td>Any</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td><strong>Keyword args passed to seaborn scatterplot</strong></td>
</tr>
</tbody>
</table>
<div id="australian-venue" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>df_fabales <span class="op">=</span> df_species[df_species.Order <span class="op">==</span> <span class="st">'Fabales'</span>].reset_index(drop <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>plot_embeddings(df_fabales.glycan.values.tolist(), label_list <span class="op">=</span> df_fabales.Family.values.tolist())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Download completed.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-13-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="characterize_monosaccharide" class="level3">
<h3 class="anchored" data-anchor-id="characterize_monosaccharide">characterize_monosaccharide</h3>
<blockquote class="blockquote">
<pre><code> characterize_monosaccharide (sugar:str,
                              df:pandas.core.frame.DataFrame|None=None,
                              mode:str='sugar',
                              glycan_col_name:str='glycan',
                              rank:str|None=None, focus:str|None=None,
                              modifications:bool=False,
                              filepath:str|pathlib.Path='', thresh:int=10)</code></pre>
</blockquote>
<p><em>Analyzes connectivity and modification patterns of specified monosaccharides/linkages in glycan sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sugar</td>
<td>str</td>
<td></td>
<td>Monosaccharide or linkage to analyze</td>
</tr>
<tr class="even">
<td>df</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>DataFrame with glycan column ‘glycan’; defaults to df_species</td>
</tr>
<tr class="odd">
<td>mode</td>
<td>str</td>
<td>sugar</td>
<td>Analysis mode: ‘sugar’, ‘bond’, ‘sugarbond’</td>
</tr>
<tr class="even">
<td>glycan_col_name</td>
<td>str</td>
<td>glycan</td>
<td>Column name for glycan sequences</td>
</tr>
<tr class="odd">
<td>rank</td>
<td>str | None</td>
<td>None</td>
<td>Column name for group filtering</td>
</tr>
<tr class="even">
<td>focus</td>
<td>str | None</td>
<td>None</td>
<td>Row value for group filtering</td>
</tr>
<tr class="odd">
<td>modifications</td>
<td>bool</td>
<td>False</td>
<td>Consider modified monosaccharides</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td>thresh</td>
<td>int</td>
<td>10</td>
<td>Minimum count threshold for inclusion</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="mature-civilian" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>characterize_monosaccharide(<span class="st">'Rha'</span>, rank <span class="op">=</span> <span class="st">'Kingdom'</span>, focus <span class="op">=</span> <span class="st">'Fungi'</span>, modifications <span class="op">=</span> <span class="va">True</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-15-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_differential_expression" class="level3">
<h3 class="anchored" data-anchor-id="get_differential_expression">get_differential_expression</h3>
<blockquote class="blockquote">
<pre><code> get_differential_expression
                              (df:pandas.core.frame.DataFrame|str|pathlib.
                              Path, group1:list[str|int],
                              group2:list[str|int], motifs:bool=False,
                              feature_set:list[str]=['exhaustive',
                              'known'], paired:bool=False,
                              impute:bool=True, sets:bool=False,
                              set_thresh:float=0.9,
                              effect_size_variance:bool=False,
                              min_samples:float=0.1,
                              grouped_BH:bool=False,
                              custom_motifs:list[str]=[],
                              transform:str|None=None, gamma:float=0.1,
                              custom_scale:float|dict=0,
                              glycoproteomics:bool=False,
                              level:str='peptide', monte_carlo:bool=False,
                              random_state:int|numpy.random._generator.Gen
                              erator|None=None)</code></pre>
</blockquote>
<p><em>Performs differential expression analysis using Welch’s t-test (or Hotelling’s T2 for sets) with multiple testing correction on glycomics abundance data</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1) and abundance values in subsequent columns</td>
</tr>
<tr class="even">
<td>group1</td>
<td>list</td>
<td></td>
<td>Column indices/names for first group</td>
</tr>
<tr class="odd">
<td>group2</td>
<td>list</td>
<td></td>
<td>Column indices/names for second group</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘exhaustive’, ‘known’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="even">
<td>paired</td>
<td>bool</td>
<td>False</td>
<td>Whether samples are paired</td>
</tr>
<tr class="odd">
<td>impute</td>
<td>bool</td>
<td>True</td>
<td>Replace zeros with Random Forest model</td>
</tr>
<tr class="even">
<td>sets</td>
<td>bool</td>
<td>False</td>
<td>Identify clusters of correlated glycans</td>
</tr>
<tr class="odd">
<td>set_thresh</td>
<td>float</td>
<td>0.9</td>
<td>Correlation threshold for clusters</td>
</tr>
<tr class="even">
<td>effect_size_variance</td>
<td>bool</td>
<td>False</td>
<td>Calculate effect size variance</td>
</tr>
<tr class="odd">
<td>min_samples</td>
<td>float</td>
<td>0.1</td>
<td>Min percent of non-zero samples required</td>
</tr>
<tr class="even">
<td>grouped_BH</td>
<td>bool</td>
<td>False</td>
<td>Use two-stage adaptive Benjamini-Hochberg</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>custom_scale</td>
<td>float | dict</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="odd">
<td>glycoproteomics</td>
<td>bool</td>
<td>False</td>
<td>Whether data is from glycoproteomics</td>
</tr>
<tr class="even">
<td>level</td>
<td>str</td>
<td>peptide</td>
<td>Analysis level for glycoproteomics</td>
</tr>
<tr class="odd">
<td>monte_carlo</td>
<td>bool</td>
<td>False</td>
<td>Use Monte Carlo for technical variation</td>
</tr>
<tr class="even">
<td>random_state</td>
<td>int | numpy.random._generator.Generator | None</td>
<td>None</td>
<td>optional random state for reproducibility</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with log2FC, p-values, FDR-corrected p-values, and Cohen’s d/Mahalanobis distance effect sizes</strong></td>
</tr>
</tbody>
</table>
<div id="difficult-index" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>test_df <span class="op">=</span> glycomics_data_loader.human_skin_O_PMC5871710_BCC</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> get_differential_expression(test_df, group1 <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>,<span class="dv">15</span>,<span class="dv">17</span>,<span class="dv">19</span>,<span class="dv">21</span>,<span class="dv">23</span>,<span class="dv">25</span>,<span class="dv">27</span>,<span class="dv">29</span>,<span class="dv">31</span>,<span class="dv">33</span>,<span class="dv">35</span>,<span class="dv">37</span>,<span class="dv">39</span>],</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                                  group2 <span class="op">=</span> [<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>,<span class="dv">16</span>,<span class="dv">18</span>,<span class="dv">20</span>,<span class="dv">22</span>,<span class="dv">24</span>,<span class="dv">26</span>,<span class="dv">28</span>,<span class="dv">30</span>,<span class="dv">32</span>,<span class="dv">34</span>,<span class="dv">36</span>,<span class="dv">38</span>,<span class="dv">40</span>], motifs <span class="op">=</span> <span class="va">True</span>, paired <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>res</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.044390023979542614 that has been adjusted for your sample size of 40.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Glycan</th>
<th data-quarto-table-cell-role="th">Mean abundance</th>
<th data-quarto-table-cell-role="th">Log2FC</th>
<th data-quarto-table-cell-role="th">p-val</th>
<th data-quarto-table-cell-role="th">corr p-val</th>
<th data-quarto-table-cell-role="th">significant</th>
<th data-quarto-table-cell-role="th">corr Levene p-val</th>
<th data-quarto-table-cell-role="th">Effect size</th>
<th data-quarto-table-cell-role="th">Equivalence p-val</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">8</th>
<td>GalOS(b1-3)GalNAc</td>
<td>0.159900</td>
<td>-0.930987</td>
<td>0.000848</td>
<td>0.005641</td>
<td>True</td>
<td>0.940575</td>
<td>-0.884455</td>
<td>1.000000</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">0</th>
<td>H_antigen_type2</td>
<td>0.247156</td>
<td>-0.689804</td>
<td>0.002058</td>
<td>0.005641</td>
<td>True</td>
<td>0.940575</td>
<td>-0.797572</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">1</th>
<td>Internal_LacNAc_type2</td>
<td>2.328696</td>
<td>0.467133</td>
<td>0.002552</td>
<td>0.005641</td>
<td>True</td>
<td>0.940575</td>
<td>0.776382</td>
<td>1.000000</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">9</th>
<td>GlcNAc6S(b1-6)GalNAc</td>
<td>1.046247</td>
<td>0.894961</td>
<td>0.002820</td>
<td>0.005641</td>
<td>True</td>
<td>0.940575</td>
<td>0.766511</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">12</th>
<td>Neu5Ac(a2-3)Gal</td>
<td>12.345115</td>
<td>0.250839</td>
<td>0.004655</td>
<td>0.007449</td>
<td>True</td>
<td>0.940575</td>
<td>0.716794</td>
<td>1.000000</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">2</th>
<td>Terminal_LacNAc_type2</td>
<td>2.440640</td>
<td>-0.475258</td>
<td>0.007864</td>
<td>0.010485</td>
<td>True</td>
<td>0.940575</td>
<td>-0.664152</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">14</th>
<td>Neu5Ac(a2-8)Neu5Ac</td>
<td>0.038663</td>
<td>-0.635598</td>
<td>0.018773</td>
<td>0.021455</td>
<td>True</td>
<td>0.940575</td>
<td>-0.574518</td>
<td>1.000000</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">10</th>
<td>Neu5Ac</td>
<td>16.553504</td>
<td>0.165081</td>
<td>0.048618</td>
<td>0.048618</td>
<td>False</td>
<td>0.940575</td>
<td>0.471179</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">6</th>
<td>Gal</td>
<td>18.541780</td>
<td>0.111858</td>
<td>0.073880</td>
<td>0.073880</td>
<td>False</td>
<td>0.940575</td>
<td>0.422987</td>
<td>0.739562</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">4</th>
<td>Oglycan_core1</td>
<td>7.956860</td>
<td>0.162683</td>
<td>0.084723</td>
<td>0.084723</td>
<td>False</td>
<td>0.940575</td>
<td>0.406721</td>
<td>0.739562</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">11</th>
<td>Gal(b1-3)GalNAc</td>
<td>12.726196</td>
<td>0.095097</td>
<td>0.133433</td>
<td>0.133433</td>
<td>False</td>
<td>0.940575</td>
<td>0.350569</td>
<td>0.739562</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">7</th>
<td>GalNAc</td>
<td>12.886096</td>
<td>0.085435</td>
<td>0.168975</td>
<td>0.168975</td>
<td>False</td>
<td>0.940575</td>
<td>0.319743</td>
<td>0.739562</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">13</th>
<td>Neu5Ac(a2-6)GalNAc</td>
<td>4.169726</td>
<td>-0.118513</td>
<td>0.312657</td>
<td>0.312657</td>
<td>False</td>
<td>0.940575</td>
<td>-0.231928</td>
<td>0.739562</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Disialyl_T_antigen</td>
<td>3.790085</td>
<td>-0.073129</td>
<td>0.543318</td>
<td>0.543318</td>
<td>False</td>
<td>0.940575</td>
<td>-0.138395</td>
<td>0.739562</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">5</th>
<td>Mucin_elongated_core2</td>
<td>4.769337</td>
<td>-0.012072</td>
<td>0.918651</td>
<td>0.918651</td>
<td>False</td>
<td>0.940575</td>
<td>-0.023143</td>
<td>0.739562</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_volcano" class="level3">
<h3 class="anchored" data-anchor-id="get_volcano">get_volcano</h3>
<blockquote class="blockquote">
<pre><code> get_volcano (df_res:pandas.core.frame.DataFrame|str|pathlib.Path,
              y_thresh:float=0.05, x_thresh:float=0, n:int|None=None,
              label_changed:bool=True, x_metric:str='Log2FC',
              annotate_volcano:bool=False, filepath:str='', **kwargs:Any)</code></pre>
</blockquote>
<p><em>Creates volcano plot showing -log10(FDR-corrected p-values) vs Log2FC or effect size</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df_res</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame from get_differential_expression with columns [Glycan, Log2FC, p-val, corr p-val]</td>
</tr>
<tr class="even">
<td>y_thresh</td>
<td>float</td>
<td>0.05</td>
<td>Corrected p threshold for labeling</td>
</tr>
<tr class="odd">
<td>x_thresh</td>
<td>float</td>
<td>0</td>
<td>Absolute x metric threshold for labeling</td>
</tr>
<tr class="even">
<td>n</td>
<td>int | None</td>
<td>None</td>
<td>Sample size for Bayesian-Adaptive Alpha</td>
</tr>
<tr class="odd">
<td>label_changed</td>
<td>bool</td>
<td>True</td>
<td>Add text labels to significant points</td>
</tr>
<tr class="even">
<td>x_metric</td>
<td>str</td>
<td>Log2FC</td>
<td>x-axis metric: ‘Log2FC’ or ‘Effect size’</td>
</tr>
<tr class="odd">
<td>annotate_volcano</td>
<td>bool</td>
<td>False</td>
<td>Annotate dots with SNFG images</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td>kwargs</td>
<td>Any</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td><strong>Displays volcano plot</strong></td>
</tr>
</tbody>
</table>
<div id="electronic-iceland" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>get_volcano(res)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with a default alpha of 0.05. Set sample size (n = ...) for Bayesian-Adaptive Alpha Adjustment</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-19-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_coverage" class="level3">
<h3 class="anchored" data-anchor-id="get_coverage">get_coverage</h3>
<blockquote class="blockquote">
<pre><code> get_coverage (df:pandas.core.frame.DataFrame|str|pathlib.Path,
               filepath:str='')</code></pre>
</blockquote>
<p><em>Visualizes glycan detection frequency across samples with intensity-based ordering</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1), abundances in columns</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="20b362fc" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>test_df <span class="op">=</span> pd.concat([test_df.iloc[:, <span class="dv">0</span>], test_df[test_df.columns[<span class="dv">1</span>:]].astype(<span class="bu">float</span>)], axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>get_coverage(test_df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-21-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_pca" class="level3">
<h3 class="anchored" data-anchor-id="get_pca">get_pca</h3>
<blockquote class="blockquote">
<pre><code> get_pca (df:pandas.core.frame.DataFrame|str|pathlib.Path,
          groups:list[int]|pandas.core.frame.DataFrame|None=None,
          motifs:bool=False, feature_set:list[str]=['known',
          'exhaustive'], pc_x:int=1, pc_y:int=2, color:str|None=None,
          shape:str|None=None, size:str|None=None,
          filepath:str|pathlib.Path='', custom_motifs:list[str]=[],
          transform:str|None=None, rarity_filter:float=0.05)</code></pre>
</blockquote>
<p><em>Performs PCA on glycan/motif abundance data with group-based visualization</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1), abundances in columns</td>
</tr>
<tr class="even">
<td>groups</td>
<td>list[int] | pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Group labels (e.g., [1,1,1,2,2,2,3,3,3]) or metadata DataFrame with ‘id’ column</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="odd">
<td>pc_x</td>
<td>int</td>
<td>1</td>
<td>Principal component for x-axis</td>
</tr>
<tr class="even">
<td>pc_y</td>
<td>int</td>
<td>2</td>
<td>Principal component for y-axis</td>
</tr>
<tr class="odd">
<td>color</td>
<td>str | None</td>
<td>None</td>
<td>Column in metadata for color grouping; recommended to be categorical</td>
</tr>
<tr class="even">
<td>shape</td>
<td>str | None</td>
<td>None</td>
<td>Column in metadata for shape grouping; recommended to be categorical</td>
</tr>
<tr class="odd">
<td>size</td>
<td>str | None</td>
<td>None</td>
<td>Column in metadata for point size control; recommended to be scalar</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>rarity_filter</td>
<td>float</td>
<td>0.05</td>
<td>Min proportion for non-zero values</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="7430cc2c" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>get_pca(test_df, motifs <span class="op">=</span> <span class="va">True</span>, groups <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-23-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_pval_distribution" class="level3">
<h3 class="anchored" data-anchor-id="get_pval_distribution">get_pval_distribution</h3>
<blockquote class="blockquote">
<pre><code> get_pval_distribution
                        (df_res:pandas.core.frame.DataFrame|str|pathlib.Pa
                        th, filepath:str|pathlib.Path='')</code></pre>
</blockquote>
<p><em>Creates histogram of p-values from differential expression analysis</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df_res</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>Output DataFrame from get_differential_expression</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="752a3f7a" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>get_pval_distribution(res)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-25-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_ma" class="level3">
<h3 class="anchored" data-anchor-id="get_ma">get_ma</h3>
<blockquote class="blockquote">
<pre><code> get_ma (df_res:pandas.core.frame.DataFrame|str|pathlib.Path,
         log2fc_thresh:int=1, sig_thresh:float=0.05,
         filepath:str|pathlib.Path='')</code></pre>
</blockquote>
<p><em>Generates MA plot (mean abundance vs log2 fold change) from differential expression results</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df_res</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>Output DataFrame from get_differential_expression</td>
</tr>
<tr class="even">
<td>log2fc_thresh</td>
<td>int</td>
<td>1</td>
<td>Log2FC threshold for highlighting</td>
</tr>
<tr class="odd">
<td>sig_thresh</td>
<td>float</td>
<td>0.05</td>
<td>Significance threshold for highlighting</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save plot</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="b9ea35c3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>get_ma(res)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-27-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_glycanova" class="level3">
<h3 class="anchored" data-anchor-id="get_glycanova">get_glycanova</h3>
<blockquote class="blockquote">
<pre><code> get_glycanova (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                groups:list[typing.Any], impute:bool=True,
                motifs:bool=False, feature_set:list[str]=['exhaustive',
                'known'], min_samples:float=0.1, posthoc:bool=True,
                custom_motifs:list[str]=[], transform:str|None=None,
                gamma:float=0.1, custom_scale:float=0, random_state:int|nu
                mpy.random._generator.Generator|None=None)</code></pre>
</blockquote>
<p><em>Performs one-way ANOVA with omega-squared effect size calculation and optional Tukey’s HSD post-hoc testing on glycomics data across multiple groups</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1) and abundance values in columns</td>
</tr>
<tr class="even">
<td>groups</td>
<td>list</td>
<td></td>
<td>Group labels for samples (e.g., [1,1,1,2,2,2,3,3,3])</td>
</tr>
<tr class="odd">
<td>impute</td>
<td>bool</td>
<td>True</td>
<td>Replace zeros with Random Forest model</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘exhaustive’, ‘known’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="even">
<td>min_samples</td>
<td>float</td>
<td>0.1</td>
<td>Min percent of non-zero samples required</td>
</tr>
<tr class="odd">
<td>posthoc</td>
<td>bool</td>
<td>True</td>
<td>Perform Tukey’s HSD test post-hoc</td>
</tr>
<tr class="even">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="odd">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="even">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="odd">
<td>custom_scale</td>
<td>float</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="even">
<td>random_state</td>
<td>int | numpy.random._generator.Generator | None</td>
<td>None</td>
<td>optional random state for reproducibility</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>tuple</strong></td>
<td></td>
<td><strong>(ANOVA results with F-stats and omega-squared effect sizes, post-hoc results)</strong></td>
</tr>
</tbody>
</table>
<div id="capable-rachel" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>test_df2 <span class="op">=</span> glycomics_data_loader.HIV_gagtransfection_O_PMID35112714</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>anv, ph <span class="op">=</span> get_glycanova(test_df2, [<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">3</span>], motifs <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>anv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.06364810000741428 that has been adjusted for your sample size of 12.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Glycan</th>
<th data-quarto-table-cell-role="th">F statistic</th>
<th data-quarto-table-cell-role="th">p-val</th>
<th data-quarto-table-cell-role="th">corr p-val</th>
<th data-quarto-table-cell-role="th">significant</th>
<th data-quarto-table-cell-role="th">Effect size</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>3.337354</td>
<td>0.082356</td>
<td>0.255515</td>
<td>False</td>
<td>0.182074</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc</td>
<td>2.264846</td>
<td>0.159697</td>
<td>0.255515</td>
<td>False</td>
<td>0.107511</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Neu5Ac(a2-3)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Ga...</td>
<td>2.987894</td>
<td>0.101120</td>
<td>0.255515</td>
<td>False</td>
<td>-0.097792</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>4.902553</td>
<td>0.036295</td>
<td>0.255515</td>
<td>False</td>
<td>0.159186</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">7</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc6S(b1-6)[Neu5Ac(a2-...</td>
<td>2.442388</td>
<td>0.142124</td>
<td>0.255515</td>
<td>False</td>
<td>-0.072202</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">4</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3/6)[GlcNAc(b1-...</td>
<td>1.346603</td>
<td>0.307886</td>
<td>0.368114</td>
<td>False</td>
<td>0.270963</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">5</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]Ga...</td>
<td>1.288259</td>
<td>0.322100</td>
<td>0.368114</td>
<td>False</td>
<td>0.031955</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">6</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Neu5Ac(a2-3)...</td>
<td>0.292927</td>
<td>0.752927</td>
<td>0.752927</td>
<td>False</td>
<td>0.026720</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">8</th>
<td>Neu5Ac(a2-3)Gal(b1-3)GalNAc</td>
<td>0.000000</td>
<td>1.000000</td>
<td>1.000000</td>
<td>False</td>
<td>0.120779</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_meta_analysis" class="level3">
<h3 class="anchored" data-anchor-id="get_meta_analysis">get_meta_analysis</h3>
<blockquote class="blockquote">
<pre><code> get_meta_analysis (effect_sizes:numpy.ndarray|list[float],
                    variances:numpy.ndarray|list[float],
                    model:str='fixed', filepath:str='',
                    study_names:list[str]=[])</code></pre>
</blockquote>
<p><em>Performs fixed/random effects meta-analysis using DerSimonian-Laird method for between-study variance estimation, with optional Forest plot visualization</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>effect_sizes</td>
<td>numpy.ndarray | list[float]</td>
<td></td>
<td>List of Cohen’s d/other effect sizes</td>
</tr>
<tr class="even">
<td>variances</td>
<td>numpy.ndarray | list[float]</td>
<td></td>
<td>Associated variance estimates</td>
</tr>
<tr class="odd">
<td>model</td>
<td>str</td>
<td>fixed</td>
<td>‘fixed’ or ‘random’ effects model</td>
</tr>
<tr class="even">
<td>filepath</td>
<td>str</td>
<td></td>
<td>Path to save Forest plot</td>
</tr>
<tr class="odd">
<td>study_names</td>
<td>list</td>
<td>[]</td>
<td>Names corresponding to each effect size</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>tuple</strong></td>
<td></td>
<td><strong>(combined effect size, two-tailed p-value)</strong></td>
</tr>
</tbody>
</table>
<div id="extreme-logan" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>get_meta_analysis([<span class="op">-</span><span class="fl">8.759</span>, <span class="op">-</span><span class="fl">6.363</span>, <span class="op">-</span><span class="fl">5.199</span>, <span class="op">-</span><span class="fl">3.952</span>],</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                 [<span class="fl">7.061</span>, <span class="fl">4.041</span>, <span class="fl">2.919</span>, <span class="fl">1.968</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>(-5.326913553837341, 3.005077298112724e-09)</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_time_series" class="level3">
<h3 class="anchored" data-anchor-id="get_time_series">get_time_series</h3>
<blockquote class="blockquote">
<pre><code> get_time_series (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                  impute:bool=True, motifs:bool=False,
                  feature_set:list[str]=['known', 'exhaustive'],
                  degree:int=1, min_samples:float=0.1,
                  custom_motifs:list[str]=[], transform:str|None=None,
                  gamma:float=0.1, custom_scale:float|dict=0)</code></pre>
</blockquote>
<p><em>Analyzes time series glycomics data using polynomial regression</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with sample IDs as ‘sampleID_timepoint_replicate’ in col 1 (e.g., T1_h5_r1)</td>
</tr>
<tr class="even">
<td>impute</td>
<td>bool</td>
<td>True</td>
<td>Replace zeros with Random Forest model</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="odd">
<td>degree</td>
<td>int</td>
<td>1</td>
<td>Polynomial degree for regression</td>
</tr>
<tr class="even">
<td>min_samples</td>
<td>float</td>
<td>0.1</td>
<td>Min percent of non-zero samples required</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>custom_scale</td>
<td>float | dict</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with regression coefficients and FDR-corrected p-values</strong></td>
</tr>
</tbody>
</table>
<div id="3878cbcf" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>t_dic <span class="op">=</span> {}</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"ID"</span>] <span class="op">=</span> [<span class="st">"D1_h5_r1"</span>, <span class="st">"D1_h5_r2"</span>, <span class="st">"D1_h5_r3"</span>, <span class="st">"D1_h10_r1"</span>, <span class="st">"D1_h10_r2"</span>, <span class="st">"D1_h10_r3"</span>, <span class="st">"D1_h15_r1"</span>, <span class="st">"D1_h15_r2"</span>, <span class="st">"D1_h15_r3"</span>]</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.33</span>, <span class="fl">0.31</span>, <span class="fl">0.35</span>, <span class="fl">1.51</span>, <span class="fl">1.57</span>, <span class="fl">1.66</span>, <span class="fl">2.11</span>, <span class="fl">2.04</span>, <span class="fl">2.09</span>]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.78</span>, <span class="fl">1.01</span>, <span class="fl">0.98</span>, <span class="fl">0.88</span>, <span class="fl">1.11</span>, <span class="fl">0.72</span>, <span class="fl">1.22</span>, <span class="fl">1.00</span>, <span class="fl">0.54</span>]</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Neu5Ac(a2-6)GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.11</span>, <span class="fl">0.09</span>, <span class="fl">0.14</span>, <span class="fl">0.02</span>, <span class="fl">0.07</span>, <span class="fl">0.10</span>, <span class="fl">0.11</span>, <span class="fl">0.09</span>, <span class="fl">0.08</span>]</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>get_time_series(pd.DataFrame(t_dic).set_index(<span class="st">"ID"</span>).T)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.0694557066556809 that has been adjusted for your sample size of 9.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Glycan</th>
<th data-quarto-table-cell-role="th">Change</th>
<th data-quarto-table-cell-role="th">p-val</th>
<th data-quarto-table-cell-role="th">corr p-val</th>
<th data-quarto-table-cell-role="th">significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Fuc(a1-2)Gal(b1-3)GalNAc</td>
<td>-0.009300</td>
<td>0.415220</td>
<td>0.633796</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]Ga...</td>
<td>0.005395</td>
<td>0.422530</td>
<td>0.633796</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Neu5Ac(a2-6)GalNAc</td>
<td>-0.001835</td>
<td>0.843457</td>
<td>0.843457</td>
<td>False</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_jtk" class="level3">
<h3 class="anchored" data-anchor-id="get_jtk">get_jtk</h3>
<blockquote class="blockquote">
<pre><code> get_jtk (df_in:pandas.core.frame.DataFrame|str|pathlib.Path,
          timepoints:int, interval:int, periods:list[int]=[12, 24],
          motifs:bool=False, feature_set:list[str]=['known', 'exhaustive',
          'terminal'], custom_motifs:list[str]=[],
          transform:str|None=None, gamma:float=0.1,
          correction_method:str='two-stage')</code></pre>
</blockquote>
<p><em>Identifies rhythmically expressed glycans using Jonckheere-Terpstra-Kendall algorithm for time series analysis</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df_in</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (first column), then groups arranged by ascending timepoints</td>
</tr>
<tr class="even">
<td>timepoints</td>
<td>int</td>
<td></td>
<td>Number of timepoints (each must have same number of replicates)</td>
</tr>
<tr class="odd">
<td>interval</td>
<td>int</td>
<td></td>
<td>Time units between experimental timepoints</td>
</tr>
<tr class="even">
<td>periods</td>
<td>list</td>
<td>[12, 24]</td>
<td>Timepoints per cycle to test</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’, ‘terminal’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>correction_method</td>
<td>str</td>
<td>two-stage</td>
<td>Multiple testing correction method</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with JTK results: adjusted p-values, period length, lag phase, amplitude</strong></td>
</tr>
</tbody>
</table>
<div id="fad4928d-6da4-47ef-b0f7-1d92ade8b236" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>t_dic <span class="op">=</span> {}</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Neu5Ac(a2-3)Gal(b1-3)GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.433138901</span>, <span class="fl">0.149729209</span>, <span class="fl">0.358018822</span>, <span class="fl">0.537641256</span>, <span class="fl">1.526963756</span>, <span class="fl">1.349986672</span>, <span class="fl">0.75156406</span>, <span class="fl">0.736710183</span>]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Gal(b1-3)GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.919762334</span>, <span class="fl">0.760237184</span>, <span class="fl">0.725566662</span>, <span class="fl">0.459945797</span>, <span class="fl">0.523801515</span>, <span class="fl">0.695106926</span>, <span class="fl">0.627632047</span>, <span class="fl">1.183511209</span>]</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Gal(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>] <span class="op">=</span> [<span class="fl">0.533138901</span>, <span class="fl">0.119729209</span>, <span class="fl">0.458018822</span>, <span class="fl">0.637641256</span>, <span class="fl">1.726963756</span>, <span class="fl">1.249986672</span>, <span class="fl">0.55156406</span>, <span class="fl">0.436710183</span>]</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>t_dic[<span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>] <span class="op">=</span> [<span class="fl">3.862169504</span>, <span class="fl">5.455032837</span>, <span class="fl">3.858163289</span>, <span class="fl">5.614650335</span>, <span class="fl">3.124254095</span>, <span class="fl">4.189550337</span>, <span class="fl">4.641831312</span>, <span class="fl">4.19538484</span>]</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>tps <span class="op">=</span> <span class="dv">8</span>  <span class="co"># number of timepoints in experiment</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>periods <span class="op">=</span> [<span class="dv">8</span>]  <span class="co"># potential cycles to test</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>interval <span class="op">=</span> <span class="dv">3</span>  <span class="co"># units of time between experimental timepoints</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>t_df <span class="op">=</span> pd.DataFrame(t_dic).T</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>t_df.columns <span class="op">=</span> [<span class="st">"T3"</span>, <span class="st">"T6"</span>, <span class="st">"T9"</span>, <span class="st">"T12"</span>, <span class="st">"T15"</span>, <span class="st">"T18"</span>, <span class="st">"T21"</span>, <span class="st">"T24"</span>]</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>get_jtk(t_df.reset_index(), tps, interval, periods <span class="op">=</span> periods)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.22004505213567527 that has been adjusted for your sample size of 1.
Significance inflation detected. The CLR/ALR transformation possibly cannot handle this dataset. Consider running again with a higher gamma value.             Proceed with caution; for now switching to Bonferroni correction to be conservative about this.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Molecule_Name</th>
<th data-quarto-table-cell-role="th">Adjusted_P_value</th>
<th data-quarto-table-cell-role="th">Period_Length</th>
<th data-quarto-table-cell-role="th">Lag_Phase</th>
<th data-quarto-table-cell-role="th">Amplitude</th>
<th data-quarto-table-cell-role="th">significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Gal(b1-3)GalNAc</td>
<td>0.037499</td>
<td>8</td>
<td>12</td>
<td>0.785714</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>0.037499</td>
<td>8</td>
<td>12</td>
<td>0.785714</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Neu5Ac(a2-3)Gal(b1-3)GalNAc</td>
<td>0.431049</td>
<td>8</td>
<td>9</td>
<td>0.500000</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Fuc(a1-2)Gal(b1-3)GalNAc</td>
<td>0.694185</td>
<td>8</td>
<td>9</td>
<td>0.428571</td>
<td>False</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="331c74a4-6aee-4915-b427-6c88c00648a3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>get_jtk(t_df.reset_index(), tps, interval, periods <span class="op">=</span> periods, motifs <span class="op">=</span> <span class="va">True</span>, feature_set <span class="op">=</span> [<span class="st">'terminal'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.22004505213567527 that has been adjusted for your sample size of 1.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Molecule_Name</th>
<th data-quarto-table-cell-role="th">Adjusted_P_value</th>
<th data-quarto-table-cell-role="th">Period_Length</th>
<th data-quarto-table-cell-role="th">Lag_Phase</th>
<th data-quarto-table-cell-role="th">Amplitude</th>
<th data-quarto-table-cell-role="th">significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Terminal_Gal(b1-3)</td>
<td>0.059080</td>
<td>8</td>
<td>3</td>
<td>0.642857</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Terminal_Neu5Ac(a2-6)</td>
<td>0.059080</td>
<td>8</td>
<td>12</td>
<td>0.642857</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Terminal_Neu5Ac(a2-?)</td>
<td>0.059080</td>
<td>8</td>
<td>12</td>
<td>0.714286</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Terminal_Neu5Ac(a2-3)</td>
<td>0.216933</td>
<td>8</td>
<td>9</td>
<td>0.428571</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Terminal_Fuc(a1-2)</td>
<td>0.386476</td>
<td>8</td>
<td>3</td>
<td>0.285714</td>
<td>False</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_biodiversity" class="level3">
<h3 class="anchored" data-anchor-id="get_biodiversity">get_biodiversity</h3>
<blockquote class="blockquote">
<pre><code> get_biodiversity (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                   group1:list[str|int], group2:list[str|int],
                   metrics:list[str]=['alpha', 'beta'], motifs:bool=False,
                   feature_set:list[str]=['exhaustive', 'known'],
                   custom_motifs:list[str]=[], paired:bool=False,
                   permutations:int=999, transform:str|None=None,
                   gamma:float=0.1, custom_scale:float|dict=0, random_stat
                   e:int|numpy.random._generator.Generator|None=None)</code></pre>
</blockquote>
<p><em>Calculates alpha (Shannon/Simpson) and beta (ANOSIM/PERMANOVA) diversity measures from glycomics data</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1), abundances in columns</td>
</tr>
<tr class="even">
<td>group1</td>
<td>list</td>
<td></td>
<td>First group column indices or group labels</td>
</tr>
<tr class="odd">
<td>group2</td>
<td>list</td>
<td></td>
<td>Second group indices or additional group labels</td>
</tr>
<tr class="even">
<td>metrics</td>
<td>list</td>
<td>[‘alpha’, ‘beta’]</td>
<td>Diversity metrics to calculate</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘exhaustive’, ‘known’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>paired</td>
<td>bool</td>
<td>False</td>
<td>Whether samples are paired</td>
</tr>
<tr class="odd">
<td>permutations</td>
<td>int</td>
<td>999</td>
<td>Number of permutations for ANOSIM/PERMANOVA</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>custom_scale</td>
<td>float | dict</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="odd">
<td>random_state</td>
<td>int | numpy.random._generator.Generator | None</td>
<td>None</td>
<td>optional random state for reproducibility</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with diversity indices and test statistics</strong></td>
</tr>
</tbody>
</table>
<div id="d2725d8a-5c15-458e-9e40-c3357f470364" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> get_biodiversity(test_df, group1 <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>,<span class="dv">15</span>,<span class="dv">17</span>,<span class="dv">19</span>,<span class="dv">21</span>,<span class="dv">23</span>,<span class="dv">25</span>,<span class="dv">27</span>,<span class="dv">29</span>,<span class="dv">31</span>,<span class="dv">33</span>,<span class="dv">35</span>,<span class="dv">37</span>,<span class="dv">39</span>],</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>                                  group2 <span class="op">=</span> [<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>,<span class="dv">16</span>,<span class="dv">18</span>,<span class="dv">20</span>,<span class="dv">22</span>,<span class="dv">24</span>,<span class="dv">26</span>,<span class="dv">28</span>,<span class="dv">30</span>,<span class="dv">32</span>,<span class="dv">34</span>,<span class="dv">36</span>,<span class="dv">38</span>,<span class="dv">40</span>], motifs <span class="op">=</span> <span class="va">True</span>, paired <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>res</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.044390023979542614 that has been adjusted for your sample size of 40.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Metric</th>
<th data-quarto-table-cell-role="th">Group1 mean</th>
<th data-quarto-table-cell-role="th">Group2 mean</th>
<th data-quarto-table-cell-role="th">p-val</th>
<th data-quarto-table-cell-role="th">Effect size</th>
<th data-quarto-table-cell-role="th">corr p-val</th>
<th data-quarto-table-cell-role="th">significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>simpson_diversity</td>
<td>0.876756</td>
<td>0.874348</td>
<td>0.000443</td>
<td>-0.948203</td>
<td>0.000443</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>shannon_diversity</td>
<td>2.244523</td>
<td>2.225758</td>
<td>0.001255</td>
<td>-0.846077</td>
<td>0.001255</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Beta diversity (PERMANOVA)</td>
<td>NaN</td>
<td>NaN</td>
<td>0.003003</td>
<td>43.641775</td>
<td>0.003003</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Beta diversity (ANOSIM)</td>
<td>NaN</td>
<td>NaN</td>
<td>0.004004</td>
<td>0.146395</td>
<td>0.004004</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>species_richness</td>
<td>15.000000</td>
<td>15.000000</td>
<td>1.000000</td>
<td>0.000000</td>
<td>1.000000</td>
<td>False</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_sparcc" class="level3">
<h3 class="anchored" data-anchor-id="get_sparcc">get_SparCC</h3>
<blockquote class="blockquote">
<pre><code> get_SparCC (df1:pandas.core.frame.DataFrame|str|pathlib.Path,
             df2:pandas.core.frame.DataFrame|str|pathlib.Path,
             motifs:bool=False, feature_set:list[str]=['known',
             'exhaustive'], custom_motifs:list[str]=[],
             transform:str|None=None, gamma:float=0.1,
             partial_correlations:bool=False)</code></pre>
</blockquote>
<p><em>Calculates SparCC (Sparse Correlations for Compositional Data) between two matching datasets (e.g., glycomics)</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df1</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>First DataFrame with glycans in rows (col 1) and abundances in columns</td>
</tr>
<tr class="even">
<td>df2</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>Second DataFrame with same format as df1</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>partial_correlations</td>
<td>bool</td>
<td>False</td>
<td>Use regularized partial correlations</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>tuple</strong></td>
<td></td>
<td><strong>(Spearman correlation matrix, FDR-corrected p-value matrix)</strong></td>
</tr>
</tbody>
</table>
<div id="01059f59-1e20-4551-a38b-29504e57b8cd" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>df1 <span class="op">=</span> glycomics_data_loader.time_series_N_PMID32149347</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>df2 <span class="op">=</span> glycomics_data_loader.time_series_O_PMID32149347</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>df1 <span class="op">=</span> pd.merge(df1, df2[[<span class="st">'ID'</span>]], on <span class="op">=</span> <span class="st">'ID'</span>, how <span class="op">=</span> <span class="st">'inner'</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>df2 <span class="op">=</span> pd.merge(df2, df1[[<span class="st">'ID'</span>]], on <span class="op">=</span> <span class="st">'ID'</span>, how <span class="op">=</span> <span class="st">'inner'</span>)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>df1 <span class="op">=</span> df1.set_index(df1.columns.tolist()[<span class="dv">0</span>]).T.reset_index()</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>df2 <span class="op">=</span> df2.set_index(df2.columns.tolist()[<span class="dv">0</span>]).T.reset_index()</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>corr, pval <span class="op">=</span> get_SparCC(df1, df2, motifs <span class="op">=</span> <span class="va">True</span>, transform <span class="op">=</span> <span class="st">"CLR"</span>)</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>sns.clustermap(corr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.04787928055709467 that has been adjusted for your sample size of 31.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-40-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_roc" class="level3">
<h3 class="anchored" data-anchor-id="get_roc">get_roc</h3>
<blockquote class="blockquote">
<pre><code> get_roc (df:pandas.core.frame.DataFrame|str|pathlib.Path,
          group1:list[str|int], group2:list[str|int], motifs:bool=False,
          feature_set:list[str]=['known', 'exhaustive'],
          paired:bool=False, impute:bool=True, min_samples:float=0.1,
          custom_motifs:list[str]=[], transform:str|None=None,
          gamma:float=0.1, custom_scale:float|dict=0,
          filepath:str|pathlib.Path='', multi_score:bool=False,
          random_state:int|numpy.random._generator.Generator|None=None)</code></pre>
</blockquote>
<p><em>Calculates ROC curves and AUC scores for glycans/motifs or multi-glycan classifiers</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with glycans in rows (col 1), abundances in columns</td>
</tr>
<tr class="even">
<td>group1</td>
<td>list</td>
<td></td>
<td>First group indices/names</td>
</tr>
<tr class="odd">
<td>group2</td>
<td>list</td>
<td></td>
<td>Second group indices/names</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>bool</td>
<td>False</td>
<td>Analyze motifs instead of sequences</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’]</td>
<td>Feature sets to use; exhaustive, known, terminal1, terminal2, terminal3, chemical, graph, custom, size_branch</td>
</tr>
<tr class="even">
<td>paired</td>
<td>bool</td>
<td>False</td>
<td>Whether samples are paired</td>
</tr>
<tr class="odd">
<td>impute</td>
<td>bool</td>
<td>True</td>
<td>Replace zeros with Random Forest model</td>
</tr>
<tr class="even">
<td>min_samples</td>
<td>float</td>
<td>0.1</td>
<td>Min percent of non-zero samples required</td>
</tr>
<tr class="odd">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs if using ‘custom’ feature set</td>
</tr>
<tr class="even">
<td>transform</td>
<td>str | None</td>
<td>None</td>
<td>Transformation type: “CLR” or “ALR”</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>custom_scale</td>
<td>float | dict</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="odd">
<td>filepath</td>
<td>str | pathlib.Path</td>
<td></td>
<td>Path to save ROC plot</td>
</tr>
<tr class="even">
<td>multi_score</td>
<td>bool</td>
<td>False</td>
<td>Find best multi-glycan score</td>
</tr>
<tr class="odd">
<td>random_state</td>
<td>int | numpy.random._generator.Generator | None</td>
<td>None</td>
<td>optional random state for reproducibility</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list[tuple[str, float]] | dict[typing.Any, tuple[str, float]] | tuple[sklearn.linear_model._logistic.LogisticRegression, float]</strong></td>
<td></td>
<td><strong>(Feature scores with ROC AUC values)</strong></td>
</tr>
</tbody>
</table>
<div id="5967bfd1-5fac-4817-96a2-c6e4aeac81d2" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>get_roc(test_df, group1 <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>,<span class="dv">15</span>,<span class="dv">17</span>,<span class="dv">19</span>,<span class="dv">21</span>,<span class="dv">23</span>,<span class="dv">25</span>,<span class="dv">27</span>,<span class="dv">29</span>,<span class="dv">31</span>,<span class="dv">33</span>,<span class="dv">35</span>,<span class="dv">37</span>,<span class="dv">39</span>],</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                                  group2 <span class="op">=</span> [<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>,<span class="dv">16</span>,<span class="dv">18</span>,<span class="dv">20</span>,<span class="dv">22</span>,<span class="dv">24</span>,<span class="dv">26</span>,<span class="dv">28</span>,<span class="dv">30</span>,<span class="dv">32</span>,<span class="dv">34</span>,<span class="dv">36</span>,<span class="dv">38</span>,<span class="dv">40</span>], motifs <span class="op">=</span> <span class="va">True</span>, paired <span class="op">=</span> <span class="va">True</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>[('GlcNAc6S(b1-6)GalNAc', 0.7599999999999999),
 ('Internal_LacNAc_type2', 0.715),
 ('Neu5Ac(a2-3)Gal', 0.6699999999999999),
 ('Neu5Ac', 0.65),
 ('Oglycan_core1', 0.6224999999999999),
 ('Gal', 0.6),
 ('Gal(b1-3)GalNAc', 0.5874999999999999),
 ('GalNAc', 0.585),
 ('Mucin_elongated_core2', 0.4975),
 ('Disialyl_T_antigen', 0.465),
 ('Neu5Ac(a2-6)GalNAc', 0.45),
 ('Neu5Ac(a2-8)Neu5Ac', 0.37000000000000005),
 ('H_antigen_type2', 0.2625),
 ('Terminal_LacNAc_type2', 0.2625),
 ('GalOS(b1-3)GalNAc', 0.2475)]</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="03_motif_files/figure-html/cell-42-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_lectin_array" class="level3">
<h3 class="anchored" data-anchor-id="get_lectin_array">get_lectin_array</h3>
<blockquote class="blockquote">
<pre><code> get_lectin_array (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                   group1:list[str|int], group2:list[str|int],
                   paired:bool=False, transform:str='')</code></pre>
</blockquote>
<p><em>Analyzes lectin microarray data by mapping lectin binding patterns to glycan motifs, calculating Cohen’s d effect sizes between groups and clustering results by significance</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with samples as rows and lectins as columns, first column containing sample IDs</td>
</tr>
<tr class="even">
<td>group1</td>
<td>list</td>
<td></td>
<td>First group indices/names</td>
</tr>
<tr class="odd">
<td>group2</td>
<td>list</td>
<td></td>
<td>Second group indices/names</td>
</tr>
<tr class="even">
<td>paired</td>
<td>bool</td>
<td>False</td>
<td>Whether samples are paired</td>
</tr>
<tr class="odd">
<td>transform</td>
<td>str</td>
<td></td>
<td>Optional log2 transformation</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with altered glycan motifs, supporting lectins, and effect sizes</strong></td>
</tr>
</tbody>
</table>
<div id="ffb3a8cc-1b17-4a5c-9f37-5bb52f66eeba" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>lectin_df <span class="op">=</span> lectin_array_data_loader.A549_influenza_PMID33046650</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>get_lectin_array(lectin_df, [<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>], [<span class="dv">8</span>,<span class="dv">9</span>,<span class="dv">10</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Lectin "Ab-LeB-1" is not found in our annotated lectin library and is excluded from analysis.
Lectin "APA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "APP" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Blood Group B [CLCP-19B]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Blood Group H2" is not found in our annotated lectin library and is excluded from analysis.
Lectin "CA19-9 [121SLE]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "CCA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "CD15 [ICRF29-2]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "CD15 [MY-1]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "CD15 [SP-159]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Forssman" is not found in our annotated lectin library and is excluded from analysis.
Lectin "IAA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "IRA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Le X [P12]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Lewis A [7LE]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Lewis B [218]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "Lewis Y [F3]" is not found in our annotated lectin library and is excluded from analysis.
Lectin "LFA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "LPA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "MNA-M " is not found in our annotated lectin library and is excluded from analysis.
Lectin "MUC5Ac Ab" is not found in our annotated lectin library and is excluded from analysis.
Lectin "PMA" is not found in our annotated lectin library and is excluded from analysis.
Lectin "PTA_1" is not found in our annotated lectin library and is excluded from analysis.
Lectin "PTA_2" is not found in our annotated lectin library and is excluded from analysis.
Lectin "SNA-S" is not found in our annotated lectin library and is excluded from analysis.
Lectin "SNA-V" is not found in our annotated lectin library and is excluded from analysis.
Lectin "VFA" is not found in our annotated lectin library and is excluded from analysis.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif</th>
<th data-quarto-table-cell-role="th">named_motifs</th>
<th data-quarto-table-cell-role="th">lectin(s)</th>
<th data-quarto-table-cell-role="th">change</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">39</th>
<td>Neu5Ac(a2-6)Gal(b1-3)GlcNAc</td>
<td>[Internal_LacNAc_type1]</td>
<td>PSL, SNA, TJA-I, BDA, BPA, WGA_1, WGA_2</td>
<td>down</td>
<td>11.32</td>
<td>highly significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">38</th>
<td>Neu5Ac(a2-6)Gal(b1-4)GlcNAc</td>
<td>[Internal_LacNAc_type2]</td>
<td>PSL, SNA, TJA-I, BDA, BPA, ECA, RCA120, Ricin ...</td>
<td>down</td>
<td>10.81</td>
<td>highly significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">7</th>
<td>Man(a1-2)</td>
<td>[]</td>
<td>ASA, Con A, CVN, HHL, SVN_1, GRFT, SVN_2, SNA-...</td>
<td>up</td>
<td>4.83</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">14</th>
<td>Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc...</td>
<td>[Chitobiose, Trimannosylcore, Terminal_LacNAc_...</td>
<td>CA, CAA, DSA_1, DSA_2, DSA_3, AMA, BDA, BPA, C...</td>
<td>up</td>
<td>3.51</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Gal(b1-3)GalNAc</td>
<td>[Oglycan_core1]</td>
<td>ACA, AIA, MPA, PNA_1, PNA_2, BDA, BPA</td>
<td>up</td>
<td>3.48</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">43</th>
<td>Neu5Ac(a2-6)GalNAc(b1-4)GlcNAc</td>
<td>[Internal_LacdiNAc_type2]</td>
<td>SNA, CSA, SBA, VVA_1, VVA_2, WFA, BPA, ECA, ST...</td>
<td>down</td>
<td>2.86</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">10</th>
<td>Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][G...</td>
<td>[Chitobiose, Trimannosylcore, Terminal_LacNAc_...</td>
<td>Blackbean, Calsepa, PHA-E_1, PHA-E_2, AMA, BDA...</td>
<td>up</td>
<td>2.70</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">16</th>
<td>Fuc(a1-2)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Ga...</td>
<td>[Internal_LacNAc_type2, H_type3]</td>
<td>Cholera Toxin, AAA, AAL, ACA, AIA, AOL, BDA, B...</td>
<td>up</td>
<td>2.51</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">15</th>
<td>Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Gl...</td>
<td>[Internal_LacNAc_type2]</td>
<td>Cholera Toxin, ACA, AIA, BDA, BPA, CSA, ECA, L...</td>
<td>up</td>
<td>2.46</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">47</th>
<td>GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Ma...</td>
<td>[Chitobiose, Trimannosylcore, core_fucose]</td>
<td>TL, AAL, AMA, AOL, Con A, GNA, GNL, HHL, LcH, ...</td>
<td>up</td>
<td>2.36</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">18</th>
<td>Man(a1-6)</td>
<td>[]</td>
<td>Con A, GNA, GNL, HHL, NPA, SNA-II, UDA</td>
<td>up</td>
<td>2.30</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">17</th>
<td>Man(a1-3)</td>
<td>[]</td>
<td>Con A, GNA, GNL, HHL, NPA, SNA-II, UDA</td>
<td>up</td>
<td>2.30</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">22</th>
<td>Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Ma...</td>
<td>[Chitobiose, Trimannosylcore, Terminal_LacNAc_...</td>
<td>DSA_1, DSA_2, DSA_3, AMA, BDA, Blackbean, BPA,...</td>
<td>up</td>
<td>2.05</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">46</th>
<td>Fuc(a1-2)Gal(b1-3)GalNAc</td>
<td>[H_type3, Oglycan_core1]</td>
<td>TJA-II, AAA, AAL, ACA, AIA, AOL, BDA, BPA, MPA...</td>
<td>up</td>
<td>1.96</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">3</th>
<td>Fuc(a1-6)</td>
<td>[]</td>
<td>AAL, AOL, LcH, PSA</td>
<td>up</td>
<td>1.70</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">34</th>
<td>Neu5Ac(a2-3)Gal(b1-3)GalNAc</td>
<td>[Oglycan_core1]</td>
<td>MAL-II, ACA, AIA, BDA, BPA, MPA, PNA_1, PNA_2,...</td>
<td>up</td>
<td>1.59</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">6</th>
<td>Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</td>
<td>[Chitobiose, Trimannosylcore]</td>
<td>AMA, Con A, GNA, GNL, HHL, NPA, SNA-II, UDA, W...</td>
<td>up</td>
<td>1.58</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">11</th>
<td>GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)[GlcNAc(b1-6...</td>
<td>[Chitobiose, Trimannosylcore]</td>
<td>Blackbean, PHA-L, AMA, Con A, GNA, GNL, HHL, N...</td>
<td>up</td>
<td>1.44</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">42</th>
<td>GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-...</td>
<td>[Chitobiose, Trimannosylcore, bisectingGlcNAc]</td>
<td>RPA, AMA, Blackbean, Con A, GNA, GNL, HHL, NPA...</td>
<td>up</td>
<td>1.40</td>
<td>moderately significant</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">41</th>
<td>GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-...</td>
<td>[Chitobiose, Trimannosylcore, bisectingGlcNAc]</td>
<td>RPA, AMA, Con A, GNA, GNL, HHL, NPA, SNA-II, U...</td>
<td>up</td>
<td>1.36</td>
<td>moderately significant</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">23</th>
<td>Gal(b1-4)GlcNAc</td>
<td>[Terminal_LacNAc_type2]</td>
<td>ECA, RCA120, Ricin B Chain, SJA, BDA, BPA</td>
<td>up</td>
<td>1.05</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">5</th>
<td>GlcNAc(b1-3)GalNAc</td>
<td>[Oglycan_core3]</td>
<td>AIA, UEA-II, WGA_1, WGA_2</td>
<td>up</td>
<td>0.86</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">26</th>
<td>Gal(a1-3)</td>
<td>[]</td>
<td>GS-I_1, GS-I_2, GS-I_3, GS-I_4, MNA-G, PA-IL</td>
<td>up</td>
<td>0.83</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">27</th>
<td>Gal(a1-4)</td>
<td>[]</td>
<td>GS-I_1, GS-I_2, GS-I_3, GS-I_4, MNA-G, PA-IL</td>
<td>up</td>
<td>0.83</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">30</th>
<td>Gal(b1-4)GlcNAc(b1-3)</td>
<td>[Terminal_LacNAc_type2]</td>
<td>LEA_1, LEA_2, STA, BDA, BPA, ECA, RCA120, Rici...</td>
<td>up</td>
<td>0.54</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">25</th>
<td>Gal(a1-3)Gal</td>
<td>[]</td>
<td>EEA, EEL, MOA, GS-I_1, GS-I_2, GS-I_3, GS-I_4,...</td>
<td>up</td>
<td>0.51</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">33</th>
<td>Neu5Ac(a2-3)Gal(b1-4)GlcNAc</td>
<td>[Internal_LacNAc_type2]</td>
<td>MAA_1, MAA_2, MAL-I, BDA, BPA, ECA, RCA120, Ri...</td>
<td>up</td>
<td>0.49</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">37</th>
<td>Gal(a1-3)GalNAc</td>
<td>[]</td>
<td>MOA, EEA, EEL, GS-I_1, GS-I_2, GS-I_3, GS-I_4,...</td>
<td>up</td>
<td>0.46</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">20</th>
<td>GalNAc(a1-4)</td>
<td>[]</td>
<td>GHA, HAA, HPA, CSA, GS-I_1, GS-I_2, GS-I_3, GS...</td>
<td>up</td>
<td>0.39</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">19</th>
<td>GalNAc(a1-3)</td>
<td>[]</td>
<td>GHA, HAA, HPA, CSA, GS-I_1, GS-I_2, GS-I_3, GS...</td>
<td>up</td>
<td>0.39</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">21</th>
<td>GalNAc(a1-3)GalNAc(b1-3)</td>
<td>[]</td>
<td>DBA, SBA, CSA, GHA, HAA, HPA, VVA_1, VVA_2, WF...</td>
<td>up</td>
<td>0.25</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">24</th>
<td>GalNAc(b1-4)GlcNAc</td>
<td>[Terminal_LacdiNAc_type2]</td>
<td>ECA, STA, CSA, SBA, VVA_1, VVA_2, WFA, BPA, WG...</td>
<td>up</td>
<td>0.20</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">44</th>
<td>Fuc(a1-2)Gal(b1-4)GalNAc(b1-3)</td>
<td>[]</td>
<td>SNA-II, AAA, AAL, AOL, BDA, BPA, CSA, SBA, VVA...</td>
<td>up</td>
<td>0.16</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">13</th>
<td>GalNAc(b1-4)</td>
<td>[]</td>
<td>CSA, SBA, VVA_1, VVA_2, WFA, BPA, WGA_1, WGA_2</td>
<td>up</td>
<td>0.13</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">12</th>
<td>GalNAc(b1-3)</td>
<td>[]</td>
<td>CSA, SBA, VVA_1, VVA_2, WFA, BPA, WGA_1, WGA_2</td>
<td>up</td>
<td>0.13</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">40</th>
<td>Fuc(a1-2)Gal(b1-4)GlcNAc</td>
<td>[H_antigen_type2, Internal_LacNAc_type2]</td>
<td>PTL-II, TJA-II, UEA-I, UEA-II, AAA, AAL, AOL, ...</td>
<td>up</td>
<td>0.13</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">28</th>
<td>GlcNAc(a1-3)</td>
<td>[]</td>
<td>HAA, HPA, WGA_1, WGA_2</td>
<td>up</td>
<td>0.12</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">29</th>
<td>GlcNAc(a1-4)</td>
<td>[]</td>
<td>HAA, HPA, WGA_1, WGA_2</td>
<td>up</td>
<td>0.12</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">32</th>
<td>Gal3S(b1-4)GlcNAc</td>
<td>[]</td>
<td>MAA_1, MAA_2, MAL-I, MAL-II</td>
<td>down</td>
<td>0.12</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">0</th>
<td>Fuc(a1-2)</td>
<td>[]</td>
<td>AAA, AAL, AOL</td>
<td>up</td>
<td>0.09</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">36</th>
<td>Gal3S(b1-4)</td>
<td>[]</td>
<td>MAL-II</td>
<td>down</td>
<td>0.08</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">35</th>
<td>Gal3S(b1-3)</td>
<td>[]</td>
<td>MAL-II</td>
<td>down</td>
<td>0.08</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">49</th>
<td>Fuc(a1-2)Gal(b1-4)GalNAc</td>
<td>[]</td>
<td>UEA-II, AAA, AAL, AOL, BDA, BPA</td>
<td>up</td>
<td>0.07</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">9</th>
<td>Gal(b1-4)</td>
<td>[]</td>
<td>BDA, BPA</td>
<td>up</td>
<td>0.05</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">8</th>
<td>Gal(b1-3)</td>
<td>[]</td>
<td>BDA, BPA</td>
<td>up</td>
<td>0.05</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Fuc(a1-3)</td>
<td>[]</td>
<td>AAL, AOL, Lotus</td>
<td>down</td>
<td>0.03</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Fuc(a1-4)</td>
<td>[]</td>
<td>AAL, AOL</td>
<td>down</td>
<td>0.03</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">31</th>
<td>GlcNAc(b1-4)GlcNAc(b1-4)</td>
<td>[Chitobiose]</td>
<td>LEA_1, LEA_2, WGA_1, WGA_2</td>
<td>down</td>
<td>0.01</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">50</th>
<td>GlcNAc(b1-3)</td>
<td>[]</td>
<td>WGA_1, WGA_2</td>
<td>down</td>
<td>0.01</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">51</th>
<td>GlcNAc(b1-4)</td>
<td>[]</td>
<td>WGA_1, WGA_2</td>
<td>down</td>
<td>0.01</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">45</th>
<td>GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)</td>
<td>[Chitobiose]</td>
<td>STA, LEA_1, LEA_2, WGA_1, WGA_2</td>
<td>down</td>
<td>0.00</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">48</th>
<td>GlcNAc(b1-3)Gal</td>
<td>[]</td>
<td>UEA-II, WGA_1, WGA_2</td>
<td>up</td>
<td>0.00</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">52</th>
<td>Neu5Ac(a2-3)</td>
<td>[]</td>
<td>WGA_1, WGA_2</td>
<td>down</td>
<td>0.00</td>
<td>low significance</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">53</th>
<td>Neu5Ac(a2-6)</td>
<td>[]</td>
<td>WGA_1, WGA_2</td>
<td>down</td>
<td>0.00</td>
<td>low significance</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">54</th>
<td>Neu5Ac(a2-8)</td>
<td>[]</td>
<td>WGA_1, WGA_2</td>
<td>down</td>
<td>0.00</td>
<td>low significance</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_glycoshift_per_site" class="level3">
<h3 class="anchored" data-anchor-id="get_glycoshift_per_site">get_glycoshift_per_site</h3>
<blockquote class="blockquote">
<pre><code> get_glycoshift_per_site (df:pandas.core.frame.DataFrame|str|pathlib.Path,
                          group1:list[str|int], group2:list[str|int],
                          paired:bool=False, impute:bool=True,
                          min_samples:float=0.2, gamma:float=0.1,
                          custom_scale:float|dict=0, random_state:int|nump
                          y.random._generator.Generator|None=None)</code></pre>
</blockquote>
<p><em>Analyzes site-specific glycosylation changes in glycoproteomics data using generalized linear models (GLM) with compositional data normalization</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>pandas.core.frame.DataFrame | str | pathlib.Path</td>
<td></td>
<td>DataFrame with rows formatted as ‘protein_site_composition’ in col 1, abundances in remaining cols</td>
</tr>
<tr class="even">
<td>group1</td>
<td>list</td>
<td></td>
<td>First group indices/names or group labels for multi-group</td>
</tr>
<tr class="odd">
<td>group2</td>
<td>list</td>
<td></td>
<td>Second group indices/names</td>
</tr>
<tr class="even">
<td>paired</td>
<td>bool</td>
<td>False</td>
<td>Whether samples are paired</td>
</tr>
<tr class="odd">
<td>impute</td>
<td>bool</td>
<td>True</td>
<td>Replace zeros with Random Forest model</td>
</tr>
<tr class="even">
<td>min_samples</td>
<td>float</td>
<td>0.2</td>
<td>Min percent of non-zero samples required</td>
</tr>
<tr class="odd">
<td>gamma</td>
<td>float</td>
<td>0.1</td>
<td>Uncertainty parameter for CLR transform</td>
</tr>
<tr class="even">
<td>custom_scale</td>
<td>float | dict</td>
<td>0</td>
<td>Ratio of total signal in group2/group1 for an informed scale model (or group_idx: mean(group)/min(mean(groups)) signal dict for multivariate)</td>
</tr>
<tr class="odd">
<td>random_state</td>
<td>int | numpy.random._generator.Generator | None</td>
<td>None</td>
<td>optional random state for reproducibility</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with GLM coefficients and FDR-corrected p-values</strong></td>
</tr>
</tbody>
</table>
<div id="155f86f0-ea8d-4a3d-a773-012ea10c9e84" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>df_milk <span class="op">=</span> glycoproteomics_data_loader.human_milk_N_PMID34087070</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>get_glycoshift_per_site(df_milk, [<span class="st">'Colostrum1'</span>, <span class="st">'Colostrum2'</span>, <span class="st">'Colostrum3'</span>], [<span class="st">'Mature1'</span>, <span class="st">'Mature2'</span>, <span class="st">'Mature3'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>You're working with an alpha of 0.07862467893233027 that has been adjusted for your sample size of 6.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Condition_coefficient</th>
<th data-quarto-table-cell-role="th">Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">Condition_significant</th>
<th data-quarto-table-cell-role="th">antennary_Fuc_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">antennary_Fuc_Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">antennary_Fuc_Condition_significant</th>
<th data-quarto-table-cell-role="th">dHex_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">dHex_Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">dHex_Condition_significant</th>
<th data-quarto-table-cell-role="th">complex_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">high_Man_Condition_significant</th>
<th data-quarto-table-cell-role="th">hybrid_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">hybrid_Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">hybrid_Condition_significant</th>
<th data-quarto-table-cell-role="th">Neu5Ac_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">Neu5Ac_Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">Neu5Ac_Condition_significant</th>
<th data-quarto-table-cell-role="th">Hex_Condition_coefficient</th>
<th data-quarto-table-cell-role="th">Hex_Condition_corr_pval</th>
<th data-quarto-table-cell-role="th">Hex_Condition_significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P47710|CASA1_69</th>
<td>0.351306</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>3.651421</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.351306</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.351306</td>
<td>0.000000e+00</td>
<td>True</td>
<td>-1.543583</td>
<td>0.000000e+00</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P01024|CO3_85</th>
<td>-13.741464</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>True</td>
<td>-13.741464</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>12.821387</td>
<td>0.000000e+00</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P10909|CLUS_103</th>
<td>-0.148812</td>
<td>0.000000e+00</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.148812</td>
<td>0.000000e+00</td>
<td>True</td>
<td>4.461066</td>
<td>...</td>
<td>False</td>
<td>-4.609878</td>
<td>0.000000e+00</td>
<td>True</td>
<td>4.312254</td>
<td>0.000000e+00</td>
<td>True</td>
<td>-0.744062</td>
<td>0.000000e+00</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|Q13410|BT1A1_55</th>
<td>-13.032160</td>
<td>1.632815e-86</td>
<td>True</td>
<td>-13.621108</td>
<td>2.312512e-126</td>
<td>True</td>
<td>0.530115</td>
<td>4.007150e-01</td>
<td>False</td>
<td>-4.158213</td>
<td>...</td>
<td>False</td>
<td>-8.873947</td>
<td>1.136641e-42</td>
<td>True</td>
<td>-17.190373</td>
<td>1.192223e-132</td>
<td>True</td>
<td>-0.631943</td>
<td>1.290904e-18</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P01011|AACT_106</th>
<td>-0.027180</td>
<td>8.973881e-15</td>
<td>True</td>
<td>2.529792</td>
<td>0.000000e+00</td>
<td>True</td>
<td>2.502612</td>
<td>0.000000e+00</td>
<td>True</td>
<td>-2.556972</td>
<td>...</td>
<td>False</td>
<td>2.529792</td>
<td>0.000000e+00</td>
<td>True</td>
<td>-2.584152</td>
<td>0.000000e+00</td>
<td>True</td>
<td>-0.135901</td>
<td>8.973881e-15</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P00709|LALBA_90</th>
<td>-1.256621</td>
<td>1.737034e-08</td>
<td>True</td>
<td>-1.295395</td>
<td>5.342502e-01</td>
<td>False</td>
<td>7.335961</td>
<td>2.957124e-28</td>
<td>True</td>
<td>-0.554787</td>
<td>...</td>
<td>False</td>
<td>-0.701834</td>
<td>4.151344e-01</td>
<td>False</td>
<td>-1.811408</td>
<td>4.231494e-06</td>
<td>True</td>
<td>3.643645</td>
<td>1.656325e-06</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P02749|APOH_253</th>
<td>-0.002492</td>
<td>2.356643e-06</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.002492</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.004984</td>
<td>3.167328e-06</td>
<td>True</td>
<td>-0.012460</td>
<td>2.356643e-06</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P00738|HPT_241</th>
<td>0.001144</td>
<td>1.133736e-04</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.001144</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.002289</td>
<td>1.243870e-04</td>
<td>True</td>
<td>0.005722</td>
<td>1.133736e-04</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P02765|FETUA_156</th>
<td>-0.002672</td>
<td>3.394518e-03</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.002672</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.005343</td>
<td>3.666080e-03</td>
<td>True</td>
<td>-0.013358</td>
<td>3.394518e-03</td>
<td>True</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P01871|IGHM_46</th>
<td>-0.000319</td>
<td>7.588850e-02</td>
<td>True</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000319</td>
<td>1.214216e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.000319</td>
<td>9.443902e-02</td>
<td>False</td>
<td>-0.000319</td>
<td>5.604074e-02</td>
<td>True</td>
<td>-0.001594</td>
<td>7.588850e-02</td>
<td>True</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|Q08431|MFGM_238</th>
<td>0.153221</td>
<td>1.005956e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.153221</td>
<td>1.239337e-01</td>
<td>False</td>
<td>-0.315619</td>
<td>6.920843e-02</td>
<td>True</td>
<td>-0.010198</td>
<td>7.085456e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P25311|ZA2G_109</th>
<td>0.009796</td>
<td>1.359146e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.009796</td>
<td>2.283365e-01</td>
<td>False</td>
<td>-0.263924</td>
<td>...</td>
<td>False</td>
<td>0.273720</td>
<td>7.532107e-02</td>
<td>True</td>
<td>-0.254128</td>
<td>4.310116e-02</td>
<td>True</td>
<td>0.048980</td>
<td>1.482705e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P08571|CD14_151</th>
<td>0.001329</td>
<td>3.366071e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.001329</td>
<td>4.084166e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.007973</td>
<td>2.734933e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P10909|CLUS_291</th>
<td>-0.000684</td>
<td>4.277733e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.000684</td>
<td>4.471657e-01</td>
<td>False</td>
<td>-0.000684</td>
<td>3.381925e-01</td>
<td>False</td>
<td>-0.003422</td>
<td>3.522839e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P06858|LIPL_70</th>
<td>-0.000554</td>
<td>5.525567e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000554</td>
<td>6.630680e-01</td>
<td>False</td>
<td>-0.000554</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.001107</td>
<td>4.420453e-01</td>
<td>False</td>
<td>-0.002768</td>
<td>4.362290e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P07602|SAP_101</th>
<td>-0.001426</td>
<td>5.635031e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.001426</td>
<td>6.731984e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.001426</td>
<td>6.311235e-01</td>
<td>False</td>
<td>-0.001426</td>
<td>4.555478e-01</td>
<td>False</td>
<td>-0.007129</td>
<td>4.508025e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P02788|TRFL_156</th>
<td>-3.509734</td>
<td>6.172068e-01</td>
<td>False</td>
<td>0.781505</td>
<td>9.117647e-01</td>
<td>False</td>
<td>4.237603</td>
<td>3.139568e-04</td>
<td>True</td>
<td>-0.558968</td>
<td>...</td>
<td>False</td>
<td>-2.950766</td>
<td>4.471657e-01</td>
<td>False</td>
<td>0.726250</td>
<td>6.386663e-01</td>
<td>False</td>
<td>2.716540</td>
<td>2.734301e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P02788|TRFL_497</th>
<td>0.282217</td>
<td>6.343135e-01</td>
<td>False</td>
<td>1.366091</td>
<td>9.117647e-01</td>
<td>False</td>
<td>-1.143372</td>
<td>6.991406e-01</td>
<td>False</td>
<td>-11.362849</td>
<td>...</td>
<td>False</td>
<td>-13.834092</td>
<td>5.356118e-07</td>
<td>True</td>
<td>-2.046556</td>
<td>2.146705e-01</td>
<td>False</td>
<td>1.888504</td>
<td>2.734301e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P07602|SAP_215</th>
<td>-0.001385</td>
<td>6.588839e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.001385</td>
<td>8.235294e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.002770</td>
<td>6.108151e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P10909|CLUS_374</th>
<td>0.000707</td>
<td>6.588839e-01</td>
<td>False</td>
<td>0.000707</td>
<td>9.117647e-01</td>
<td>False</td>
<td>0.001414</td>
<td>8.235294e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.000707</td>
<td>8.058303e-01</td>
<td>False</td>
<td>0.000707</td>
<td>6.386663e-01</td>
<td>False</td>
<td>0.003535</td>
<td>6.120552e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P0C0L5|CO4B_HUMAN/sp|P0C0L4|CO4A</th>
<td>0.000433</td>
<td>6.588839e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.000433</td>
<td>7.456105e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.003901</td>
<td>6.108151e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P02790|HEMO_453</th>
<td>0.000962</td>
<td>6.588839e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000962</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.001924</td>
<td>6.331523e-01</td>
<td>False</td>
<td>0.004810</td>
<td>6.108151e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P01876|IGHA1_340</th>
<td>4.689804</td>
<td>6.588839e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-2.429746</td>
<td>3.616176e-01</td>
<td>False</td>
<td>-4.685317</td>
<td>...</td>
<td>False</td>
<td>-2.193530</td>
<td>6.529675e-01</td>
<td>False</td>
<td>4.191721</td>
<td>3.152153e-02</td>
<td>True</td>
<td>-0.676016</td>
<td>6.927812e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P0DOX2|IGA2_HUMAN/sp|P01877|IGHA2</th>
<td>-1.970855</td>
<td>6.754175e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-1.350375</td>
<td>3.616176e-01</td>
<td>False</td>
<td>-5.021428</td>
<td>...</td>
<td>False</td>
<td>-3.149798</td>
<td>1.501009e-01</td>
<td>False</td>
<td>-2.029984</td>
<td>5.604074e-02</td>
<td>True</td>
<td>1.473677</td>
<td>2.656424e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P07602|SAP_426</th>
<td>0.001044</td>
<td>6.862599e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.001044</td>
<td>8.235294e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.005222</td>
<td>6.862599e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P01877|IGHA2_327</th>
<td>-0.727920</td>
<td>7.386459e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.727920</td>
<td>8.235294e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.217980</td>
<td>9.310474e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P19652|A1AG2_HUMAN/sp|P02763|A1AG1</th>
<td>-0.000394</td>
<td>7.493023e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000394</td>
<td>...</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000788</td>
<td>7.493023e-01</td>
<td>False</td>
<td>-0.001971</td>
<td>7.493023e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P01833|PIGR_186</th>
<td>-0.001645</td>
<td>7.727110e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>0.025357</td>
<td>8.235294e-01</td>
<td>False</td>
<td>-0.027002</td>
<td>...</td>
<td>False</td>
<td>0.025357</td>
<td>8.058303e-01</td>
<td>False</td>
<td>-0.028648</td>
<td>6.514894e-01</td>
<td>False</td>
<td>0.017131</td>
<td>6.108151e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P01833|PIGR_499</th>
<td>-2.947551</td>
<td>7.916424e-01</td>
<td>False</td>
<td>1.517432</td>
<td>9.117647e-01</td>
<td>False</td>
<td>2.930012</td>
<td>3.317594e-01</td>
<td>False</td>
<td>-1.813514</td>
<td>...</td>
<td>False</td>
<td>-2.813763</td>
<td>7.196543e-01</td>
<td>False</td>
<td>-3.178778</td>
<td>1.651305e-01</td>
<td>False</td>
<td>-0.871071</td>
<td>6.542124e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|Q08380|LG3BP_125</th>
<td>-0.000227</td>
<td>8.418318e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000227</td>
<td>8.418318e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.000227</td>
<td>8.418318e-01</td>
<td>False</td>
<td>-0.000227</td>
<td>8.418318e-01</td>
<td>False</td>
<td>-0.001135</td>
<td>8.418318e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P01591|IGJ_71</th>
<td>0.986725</td>
<td>8.438826e-01</td>
<td>False</td>
<td>-1.068357</td>
<td>9.117647e-01</td>
<td>False</td>
<td>1.155134</td>
<td>3.037004e-01</td>
<td>False</td>
<td>-0.341331</td>
<td>...</td>
<td>False</td>
<td>1.031008</td>
<td>6.529675e-01</td>
<td>False</td>
<td>0.292119</td>
<td>7.150544e-01</td>
<td>False</td>
<td>-0.877081</td>
<td>4.362290e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P01833|PIGR_421</th>
<td>0.364442</td>
<td>8.577440e-01</td>
<td>False</td>
<td>1.445412</td>
<td>9.117647e-01</td>
<td>False</td>
<td>-0.486043</td>
<td>8.235294e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>0.364442</td>
<td>8.577440e-01</td>
<td>False</td>
<td>0.612043</td>
<td>4.889933e-01</td>
<td>False</td>
<td>0.129271</td>
<td>8.937029e-01</td>
<td>False</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sp|P01833|PIGR_469</th>
<td>0.917714</td>
<td>8.697268e-01</td>
<td>False</td>
<td>8.789078</td>
<td>9.836502e-04</td>
<td>True</td>
<td>-3.064088</td>
<td>1.096962e-02</td>
<td>True</td>
<td>14.562257</td>
<td>...</td>
<td>False</td>
<td>6.868686</td>
<td>7.890305e-02</td>
<td>False</td>
<td>-4.112379</td>
<td>2.455828e-02</td>
<td>True</td>
<td>-2.408593</td>
<td>2.656424e-01</td>
<td>False</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">sp|P10909|CLUS_86</th>
<td>-0.000104</td>
<td>9.648675e-01</td>
<td>False</td>
<td>0.000000</td>
<td>1.000000e+00</td>
<td>False</td>
<td>-0.000104</td>
<td>9.648675e-01</td>
<td>False</td>
<td>0.000000</td>
<td>...</td>
<td>False</td>
<td>-0.000104</td>
<td>9.648675e-01</td>
<td>False</td>
<td>-0.000104</td>
<td>9.648675e-01</td>
<td>False</td>
<td>-0.000518</td>
<td>9.648675e-01</td>
<td>False</td>
</tr>
</tbody>
</table>

<p>34 rows × 27 columns</p>
</div>
</div>
</div>
</section>
</section>
<section id="annotate" class="level2">
<h2 class="anchored" data-anchor-id="annotate">annotate</h2>
<blockquote class="blockquote">
<p>extract curated motifs, graph features, and sequence features from glycan sequences</p>
</blockquote>
<hr>
<section id="annotate_glycan" class="level3">
<h3 class="anchored" data-anchor-id="annotate_glycan">annotate_glycan</h3>
<blockquote class="blockquote">
<pre><code> annotate_glycan (glycan:str|networkx.classes.digraph.DiGraph,
                  motifs:pandas.core.frame.DataFrame|None=None,
                  termini_list:list=[], gmotifs:list[networkx.classes.digr
                  aph.DiGraph]|None=None)</code></pre>
</blockquote>
<p><em>Counts occurrences of known motifs in a glycan structure using subgraph isomorphism</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>IUPAC-condensed glycan sequence or NetworkX graph</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Motif dataframe (name + sequence); defaults to motif_list</td>
</tr>
<tr class="odd">
<td>termini_list</td>
<td>list</td>
<td>[]</td>
<td>Monosaccharide positions: ‘terminal’, ‘internal’, or ‘flexible’</td>
</tr>
<tr class="even">
<td>gmotifs</td>
<td>list[networkx.classes.digraph.DiGraph] | None</td>
<td>None</td>
<td>Precalculated motif graphs for speed</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with motif counts for the glycan</strong></td>
</tr>
</tbody>
</table>
<div id="champion-booth" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>annotate_glycan(<span class="st">"Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">motif_name</th>
<th data-quarto-table-cell-role="th">Terminal_LewisX</th>
<th data-quarto-table-cell-role="th">Internal_LewisX</th>
<th data-quarto-table-cell-role="th">LewisY</th>
<th data-quarto-table-cell-role="th">SialylLewisX</th>
<th data-quarto-table-cell-role="th">SulfoSialylLewisX</th>
<th data-quarto-table-cell-role="th">Terminal_LewisA</th>
<th data-quarto-table-cell-role="th">Internal_LewisA</th>
<th data-quarto-table-cell-role="th">LewisB</th>
<th data-quarto-table-cell-role="th">SialylLewisA</th>
<th data-quarto-table-cell-role="th">SulfoLewisA</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Mucin_elongated_core2</th>
<th data-quarto-table-cell-role="th">Fucoidan</th>
<th data-quarto-table-cell-role="th">Alginate</th>
<th data-quarto-table-cell-role="th">FG</th>
<th data-quarto-table-cell-role="th">XX</th>
<th data-quarto-table-cell-role="th">Difucosylated_core</th>
<th data-quarto-table-cell-role="th">GalFuc_core</th>
<th data-quarto-table-cell-role="th">DisialylLewisC</th>
<th data-quarto-table-cell-role="th">RM2</th>
<th data-quarto-table-cell-role="th">DisialylLewisA</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>

<p>1 rows × 165 columns</p>
</div>
</div>
</div>
<hr>
</section>
<section id="annotate_dataset" class="level3">
<h3 class="anchored" data-anchor-id="annotate_dataset">annotate_dataset</h3>
<blockquote class="blockquote">
<pre><code> annotate_dataset (glycans:list[str],
                   motifs:pandas.core.frame.DataFrame|None=None,
                   feature_set:list[str]=['known'], termini_list:list=[],
                   condense:bool=False, custom_motifs:list=[])</code></pre>
</blockquote>
<p><em>Comprehensive glycan annotation combining multiple feature types: structural motifs, graph properties, terminal sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycans</td>
<td>list</td>
<td></td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Motif dataframe (name + sequence); defaults to motif_list</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td>[‘known’]</td>
<td>Feature types to analyze: known, graph, exhaustive, terminal(1-3), custom, chemical, size_branch</td>
</tr>
<tr class="even">
<td>termini_list</td>
<td>list</td>
<td>[]</td>
<td>Monosaccharide positions: ‘terminal’, ‘internal’, or ‘flexible’</td>
</tr>
<tr class="odd">
<td>condense</td>
<td>bool</td>
<td>False</td>
<td>Remove columns with only zeros</td>
</tr>
<tr class="even">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs when using ‘custom’ feature set</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame mapping glycans to presence/absence of motifs</strong></td>
</tr>
</tbody>
</table>
<div id="worthy-holly" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>glycans <span class="op">=</span> [<span class="st">'Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc'</span>,</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">'GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P'</span>]</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Annotate Test"</span>)</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> annotate_dataset(glycans)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Annotate Test</code></pre>
</div>
</div>
<div id="numerous-consortium" class="cell">
<div class="cell-output cell-output-display">
<style type="text/css">
#T_e2eab_row0_col0, #T_e2eab_row0_col1, #T_e2eab_row0_col2, #T_e2eab_row0_col3, #T_e2eab_row0_col4, #T_e2eab_row0_col5, #T_e2eab_row0_col6, #T_e2eab_row0_col7, #T_e2eab_row0_col8, #T_e2eab_row0_col9, #T_e2eab_row0_col10, #T_e2eab_row0_col11, #T_e2eab_row0_col12, #T_e2eab_row0_col13, #T_e2eab_row0_col14, #T_e2eab_row0_col15, #T_e2eab_row0_col16, #T_e2eab_row0_col17, #T_e2eab_row0_col18, #T_e2eab_row0_col19, #T_e2eab_row0_col20, #T_e2eab_row0_col21, #T_e2eab_row0_col22, #T_e2eab_row0_col23, #T_e2eab_row0_col24, #T_e2eab_row0_col25, #T_e2eab_row0_col26, #T_e2eab_row0_col27, #T_e2eab_row0_col28, #T_e2eab_row0_col29, #T_e2eab_row0_col30, #T_e2eab_row0_col31, #T_e2eab_row0_col32, #T_e2eab_row0_col33, #T_e2eab_row0_col34, #T_e2eab_row0_col35, #T_e2eab_row0_col36, #T_e2eab_row0_col37, #T_e2eab_row0_col38, #T_e2eab_row0_col39, #T_e2eab_row0_col40, #T_e2eab_row0_col41, #T_e2eab_row0_col42, #T_e2eab_row0_col43, #T_e2eab_row0_col44, #T_e2eab_row0_col45, #T_e2eab_row0_col46, #T_e2eab_row0_col47, #T_e2eab_row0_col48, #T_e2eab_row0_col49, #T_e2eab_row0_col50, #T_e2eab_row0_col51, #T_e2eab_row0_col52, #T_e2eab_row0_col53, #T_e2eab_row0_col54, #T_e2eab_row0_col55, #T_e2eab_row0_col56, #T_e2eab_row0_col57, #T_e2eab_row0_col58, #T_e2eab_row0_col59, #T_e2eab_row0_col60, #T_e2eab_row0_col61, #T_e2eab_row0_col62, #T_e2eab_row0_col63, #T_e2eab_row0_col64, #T_e2eab_row0_col65, #T_e2eab_row0_col66, #T_e2eab_row0_col67, #T_e2eab_row0_col68, #T_e2eab_row0_col69, #T_e2eab_row0_col70, #T_e2eab_row0_col71, #T_e2eab_row0_col72, #T_e2eab_row0_col73, #T_e2eab_row0_col74, #T_e2eab_row0_col75, #T_e2eab_row0_col76, #T_e2eab_row0_col77, #T_e2eab_row0_col78, #T_e2eab_row0_col79, #T_e2eab_row0_col80, #T_e2eab_row0_col81, #T_e2eab_row0_col82, #T_e2eab_row0_col83, #T_e2eab_row0_col84, #T_e2eab_row0_col85, #T_e2eab_row0_col86, #T_e2eab_row0_col87, #T_e2eab_row0_col88, #T_e2eab_row0_col89, #T_e2eab_row0_col90, #T_e2eab_row0_col91, #T_e2eab_row0_col92, #T_e2eab_row0_col93, #T_e2eab_row0_col94, #T_e2eab_row0_col95, #T_e2eab_row0_col96, #T_e2eab_row0_col97, #T_e2eab_row0_col98, #T_e2eab_row0_col99, #T_e2eab_row0_col100, #T_e2eab_row0_col101, #T_e2eab_row0_col102, #T_e2eab_row0_col103, #T_e2eab_row0_col104, #T_e2eab_row0_col105, #T_e2eab_row0_col106, #T_e2eab_row0_col107, #T_e2eab_row0_col108, #T_e2eab_row0_col109, #T_e2eab_row0_col110, #T_e2eab_row0_col111, #T_e2eab_row0_col112, #T_e2eab_row0_col113, #T_e2eab_row0_col114, #T_e2eab_row0_col115, #T_e2eab_row0_col116, #T_e2eab_row0_col117, #T_e2eab_row0_col118, #T_e2eab_row0_col119, #T_e2eab_row0_col120, #T_e2eab_row0_col121, #T_e2eab_row0_col122, #T_e2eab_row0_col123, #T_e2eab_row0_col124, #T_e2eab_row0_col125, #T_e2eab_row0_col126, #T_e2eab_row0_col127, #T_e2eab_row0_col128, #T_e2eab_row0_col129, #T_e2eab_row0_col130, #T_e2eab_row0_col131, #T_e2eab_row0_col132, #T_e2eab_row0_col133, #T_e2eab_row0_col134, #T_e2eab_row0_col135, #T_e2eab_row0_col136, #T_e2eab_row0_col137, #T_e2eab_row0_col138, #T_e2eab_row0_col139, #T_e2eab_row0_col140, #T_e2eab_row0_col141, #T_e2eab_row0_col142, #T_e2eab_row0_col143, #T_e2eab_row0_col144, #T_e2eab_row0_col145, #T_e2eab_row0_col146, #T_e2eab_row0_col147, #T_e2eab_row0_col148, #T_e2eab_row0_col149, #T_e2eab_row0_col150, #T_e2eab_row0_col151, #T_e2eab_row0_col152, #T_e2eab_row0_col153, #T_e2eab_row0_col154, #T_e2eab_row0_col155, #T_e2eab_row0_col156, #T_e2eab_row0_col157, #T_e2eab_row0_col158, #T_e2eab_row0_col159, #T_e2eab_row0_col160, #T_e2eab_row0_col161, #T_e2eab_row0_col162, #T_e2eab_row0_col163, #T_e2eab_row0_col164, #T_e2eab_row1_col0, #T_e2eab_row1_col1, #T_e2eab_row1_col2, #T_e2eab_row1_col3, #T_e2eab_row1_col4, #T_e2eab_row1_col5, #T_e2eab_row1_col6, #T_e2eab_row1_col7, #T_e2eab_row1_col8, #T_e2eab_row1_col9, #T_e2eab_row1_col10, #T_e2eab_row1_col11, #T_e2eab_row1_col12, #T_e2eab_row1_col13, #T_e2eab_row1_col14, #T_e2eab_row1_col15, #T_e2eab_row1_col16, #T_e2eab_row1_col17, #T_e2eab_row1_col18, #T_e2eab_row1_col19, #T_e2eab_row1_col20, #T_e2eab_row1_col21, #T_e2eab_row1_col22, #T_e2eab_row1_col23, #T_e2eab_row1_col24, #T_e2eab_row1_col25, #T_e2eab_row1_col26, #T_e2eab_row1_col27, #T_e2eab_row1_col28, #T_e2eab_row1_col29, #T_e2eab_row1_col30, #T_e2eab_row1_col31, #T_e2eab_row1_col32, #T_e2eab_row1_col33, #T_e2eab_row1_col34, #T_e2eab_row1_col35, #T_e2eab_row1_col36, #T_e2eab_row1_col37, #T_e2eab_row1_col38, #T_e2eab_row1_col39, #T_e2eab_row1_col40, #T_e2eab_row1_col41, #T_e2eab_row1_col42, #T_e2eab_row1_col43, #T_e2eab_row1_col44, #T_e2eab_row1_col45, #T_e2eab_row1_col46, #T_e2eab_row1_col47, #T_e2eab_row1_col48, #T_e2eab_row1_col49, #T_e2eab_row1_col50, #T_e2eab_row1_col51, #T_e2eab_row1_col52, #T_e2eab_row1_col53, #T_e2eab_row1_col54, #T_e2eab_row1_col55, #T_e2eab_row1_col56, #T_e2eab_row1_col57, #T_e2eab_row1_col58, #T_e2eab_row1_col59, #T_e2eab_row1_col60, #T_e2eab_row1_col61, #T_e2eab_row1_col62, #T_e2eab_row1_col63, #T_e2eab_row1_col64, #T_e2eab_row1_col65, #T_e2eab_row1_col66, #T_e2eab_row1_col67, #T_e2eab_row1_col68, #T_e2eab_row1_col69, #T_e2eab_row1_col70, #T_e2eab_row1_col71, #T_e2eab_row1_col72, #T_e2eab_row1_col73, #T_e2eab_row1_col74, #T_e2eab_row1_col75, #T_e2eab_row1_col76, #T_e2eab_row1_col77, #T_e2eab_row1_col78, #T_e2eab_row1_col79, #T_e2eab_row1_col80, #T_e2eab_row1_col81, #T_e2eab_row1_col82, #T_e2eab_row1_col83, #T_e2eab_row1_col84, #T_e2eab_row1_col85, #T_e2eab_row1_col86, #T_e2eab_row1_col87, #T_e2eab_row1_col88, #T_e2eab_row1_col89, #T_e2eab_row1_col90, #T_e2eab_row1_col91, #T_e2eab_row1_col92, #T_e2eab_row1_col93, #T_e2eab_row1_col94, #T_e2eab_row1_col95, #T_e2eab_row1_col96, #T_e2eab_row1_col97, #T_e2eab_row1_col98, #T_e2eab_row1_col99, #T_e2eab_row1_col100, #T_e2eab_row1_col101, #T_e2eab_row1_col102, #T_e2eab_row1_col103, #T_e2eab_row1_col104, #T_e2eab_row1_col105, #T_e2eab_row1_col106, #T_e2eab_row1_col107, #T_e2eab_row1_col108, #T_e2eab_row1_col109, #T_e2eab_row1_col110, #T_e2eab_row1_col111, #T_e2eab_row1_col112, #T_e2eab_row1_col113, #T_e2eab_row1_col114, #T_e2eab_row1_col115, #T_e2eab_row1_col116, #T_e2eab_row1_col117, #T_e2eab_row1_col118, #T_e2eab_row1_col119, #T_e2eab_row1_col120, #T_e2eab_row1_col121, #T_e2eab_row1_col122, #T_e2eab_row1_col123, #T_e2eab_row1_col124, #T_e2eab_row1_col125, #T_e2eab_row1_col126, #T_e2eab_row1_col127, #T_e2eab_row1_col128, #T_e2eab_row1_col129, #T_e2eab_row1_col130, #T_e2eab_row1_col131, #T_e2eab_row1_col132, #T_e2eab_row1_col133, #T_e2eab_row1_col134, #T_e2eab_row1_col135, #T_e2eab_row1_col136, #T_e2eab_row1_col137, #T_e2eab_row1_col138, #T_e2eab_row1_col139, #T_e2eab_row1_col140, #T_e2eab_row1_col141, #T_e2eab_row1_col142, #T_e2eab_row1_col143, #T_e2eab_row1_col144, #T_e2eab_row1_col145, #T_e2eab_row1_col146, #T_e2eab_row1_col147, #T_e2eab_row1_col148, #T_e2eab_row1_col149, #T_e2eab_row1_col150, #T_e2eab_row1_col151, #T_e2eab_row1_col152, #T_e2eab_row1_col153, #T_e2eab_row1_col154, #T_e2eab_row1_col155, #T_e2eab_row1_col156, #T_e2eab_row1_col157, #T_e2eab_row1_col158, #T_e2eab_row1_col159, #T_e2eab_row1_col160, #T_e2eab_row1_col161, #T_e2eab_row1_col162, #T_e2eab_row1_col163, #T_e2eab_row1_col164, #T_e2eab_row2_col0, #T_e2eab_row2_col1, #T_e2eab_row2_col2, #T_e2eab_row2_col3, #T_e2eab_row2_col4, #T_e2eab_row2_col5, #T_e2eab_row2_col6, #T_e2eab_row2_col7, #T_e2eab_row2_col8, #T_e2eab_row2_col9, #T_e2eab_row2_col10, #T_e2eab_row2_col11, #T_e2eab_row2_col12, #T_e2eab_row2_col13, #T_e2eab_row2_col14, #T_e2eab_row2_col15, #T_e2eab_row2_col16, #T_e2eab_row2_col17, #T_e2eab_row2_col18, #T_e2eab_row2_col19, #T_e2eab_row2_col20, #T_e2eab_row2_col21, #T_e2eab_row2_col22, #T_e2eab_row2_col23, #T_e2eab_row2_col24, #T_e2eab_row2_col25, #T_e2eab_row2_col26, #T_e2eab_row2_col27, #T_e2eab_row2_col28, #T_e2eab_row2_col29, #T_e2eab_row2_col30, #T_e2eab_row2_col31, #T_e2eab_row2_col32, #T_e2eab_row2_col33, #T_e2eab_row2_col34, #T_e2eab_row2_col35, #T_e2eab_row2_col36, #T_e2eab_row2_col37, #T_e2eab_row2_col38, #T_e2eab_row2_col39, #T_e2eab_row2_col40, #T_e2eab_row2_col41, #T_e2eab_row2_col42, #T_e2eab_row2_col43, #T_e2eab_row2_col44, #T_e2eab_row2_col45, #T_e2eab_row2_col46, #T_e2eab_row2_col47, #T_e2eab_row2_col48, #T_e2eab_row2_col49, #T_e2eab_row2_col50, #T_e2eab_row2_col51, #T_e2eab_row2_col52, #T_e2eab_row2_col53, #T_e2eab_row2_col54, #T_e2eab_row2_col55, #T_e2eab_row2_col56, #T_e2eab_row2_col57, #T_e2eab_row2_col58, #T_e2eab_row2_col59, #T_e2eab_row2_col60, #T_e2eab_row2_col61, #T_e2eab_row2_col62, #T_e2eab_row2_col63, #T_e2eab_row2_col64, #T_e2eab_row2_col65, #T_e2eab_row2_col66, #T_e2eab_row2_col67, #T_e2eab_row2_col68, #T_e2eab_row2_col69, #T_e2eab_row2_col70, #T_e2eab_row2_col71, #T_e2eab_row2_col72, #T_e2eab_row2_col73, #T_e2eab_row2_col74, #T_e2eab_row2_col75, #T_e2eab_row2_col76, #T_e2eab_row2_col77, #T_e2eab_row2_col78, #T_e2eab_row2_col79, #T_e2eab_row2_col80, #T_e2eab_row2_col81, #T_e2eab_row2_col82, #T_e2eab_row2_col83, #T_e2eab_row2_col84, #T_e2eab_row2_col85, #T_e2eab_row2_col86, #T_e2eab_row2_col87, #T_e2eab_row2_col88, #T_e2eab_row2_col89, #T_e2eab_row2_col90, #T_e2eab_row2_col91, #T_e2eab_row2_col92, #T_e2eab_row2_col93, #T_e2eab_row2_col94, #T_e2eab_row2_col95, #T_e2eab_row2_col96, #T_e2eab_row2_col97, #T_e2eab_row2_col98, #T_e2eab_row2_col99, #T_e2eab_row2_col100, #T_e2eab_row2_col101, #T_e2eab_row2_col102, #T_e2eab_row2_col103, #T_e2eab_row2_col104, #T_e2eab_row2_col105, #T_e2eab_row2_col106, #T_e2eab_row2_col107, #T_e2eab_row2_col108, #T_e2eab_row2_col109, #T_e2eab_row2_col110, #T_e2eab_row2_col111, #T_e2eab_row2_col112, #T_e2eab_row2_col113, #T_e2eab_row2_col114, #T_e2eab_row2_col115, #T_e2eab_row2_col116, #T_e2eab_row2_col117, #T_e2eab_row2_col118, #T_e2eab_row2_col119, #T_e2eab_row2_col120, #T_e2eab_row2_col121, #T_e2eab_row2_col122, #T_e2eab_row2_col123, #T_e2eab_row2_col124, #T_e2eab_row2_col125, #T_e2eab_row2_col126, #T_e2eab_row2_col127, #T_e2eab_row2_col128, #T_e2eab_row2_col129, #T_e2eab_row2_col130, #T_e2eab_row2_col131, #T_e2eab_row2_col132, #T_e2eab_row2_col133, #T_e2eab_row2_col134, #T_e2eab_row2_col135, #T_e2eab_row2_col136, #T_e2eab_row2_col137, #T_e2eab_row2_col138, #T_e2eab_row2_col139, #T_e2eab_row2_col140, #T_e2eab_row2_col141, #T_e2eab_row2_col142, #T_e2eab_row2_col143, #T_e2eab_row2_col144, #T_e2eab_row2_col145, #T_e2eab_row2_col146, #T_e2eab_row2_col147, #T_e2eab_row2_col148, #T_e2eab_row2_col149, #T_e2eab_row2_col150, #T_e2eab_row2_col151, #T_e2eab_row2_col152, #T_e2eab_row2_col153, #T_e2eab_row2_col154, #T_e2eab_row2_col155, #T_e2eab_row2_col156, #T_e2eab_row2_col157, #T_e2eab_row2_col158, #T_e2eab_row2_col159, #T_e2eab_row2_col160, #T_e2eab_row2_col161, #T_e2eab_row2_col162, #T_e2eab_row2_col163, #T_e2eab_row2_col164 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>

<table id="T_e2eab" class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th class="index_name level0" data-quarto-table-cell-role="th">motif_name</th>
<th id="T_e2eab_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">Terminal_LewisX</th>
<th id="T_e2eab_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">Internal_LewisX</th>
<th id="T_e2eab_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">LewisY</th>
<th id="T_e2eab_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">SialylLewisX</th>
<th id="T_e2eab_level0_col4" class="col_heading level0 col4" data-quarto-table-cell-role="th">SulfoSialylLewisX</th>
<th id="T_e2eab_level0_col5" class="col_heading level0 col5" data-quarto-table-cell-role="th">Terminal_LewisA</th>
<th id="T_e2eab_level0_col6" class="col_heading level0 col6" data-quarto-table-cell-role="th">Internal_LewisA</th>
<th id="T_e2eab_level0_col7" class="col_heading level0 col7" data-quarto-table-cell-role="th">LewisB</th>
<th id="T_e2eab_level0_col8" class="col_heading level0 col8" data-quarto-table-cell-role="th">SialylLewisA</th>
<th id="T_e2eab_level0_col9" class="col_heading level0 col9" data-quarto-table-cell-role="th">SulfoLewisA</th>
<th id="T_e2eab_level0_col10" class="col_heading level0 col10" data-quarto-table-cell-role="th">H_antigen_type2</th>
<th id="T_e2eab_level0_col11" class="col_heading level0 col11" data-quarto-table-cell-role="th">H_antigen_type1</th>
<th id="T_e2eab_level0_col12" class="col_heading level0 col12" data-quarto-table-cell-role="th">H_antigen_type3</th>
<th id="T_e2eab_level0_col13" class="col_heading level0 col13" data-quarto-table-cell-role="th">A_antigen</th>
<th id="T_e2eab_level0_col14" class="col_heading level0 col14" data-quarto-table-cell-role="th">A_antigen_type1</th>
<th id="T_e2eab_level0_col15" class="col_heading level0 col15" data-quarto-table-cell-role="th">A_antigen_type2</th>
<th id="T_e2eab_level0_col16" class="col_heading level0 col16" data-quarto-table-cell-role="th">A_antigen_type3</th>
<th id="T_e2eab_level0_col17" class="col_heading level0 col17" data-quarto-table-cell-role="th">B_antigen</th>
<th id="T_e2eab_level0_col18" class="col_heading level0 col18" data-quarto-table-cell-role="th">B_antigen_type1</th>
<th id="T_e2eab_level0_col19" class="col_heading level0 col19" data-quarto-table-cell-role="th">B_antigen_type2</th>
<th id="T_e2eab_level0_col20" class="col_heading level0 col20" data-quarto-table-cell-role="th">ExtB</th>
<th id="T_e2eab_level0_col21" class="col_heading level0 col21" data-quarto-table-cell-role="th">Galili_antigen</th>
<th id="T_e2eab_level0_col22" class="col_heading level0 col22" data-quarto-table-cell-role="th">GloboH</th>
<th id="T_e2eab_level0_col23" class="col_heading level0 col23" data-quarto-table-cell-role="th">Gb5</th>
<th id="T_e2eab_level0_col24" class="col_heading level0 col24" data-quarto-table-cell-role="th">Gb4</th>
<th id="T_e2eab_level0_col25" class="col_heading level0 col25" data-quarto-table-cell-role="th">Gb3</th>
<th id="T_e2eab_level0_col26" class="col_heading level0 col26" data-quarto-table-cell-role="th">3SGb3</th>
<th id="T_e2eab_level0_col27" class="col_heading level0 col27" data-quarto-table-cell-role="th">8DSGb3</th>
<th id="T_e2eab_level0_col28" class="col_heading level0 col28" data-quarto-table-cell-role="th">3SGb4</th>
<th id="T_e2eab_level0_col29" class="col_heading level0 col29" data-quarto-table-cell-role="th">8DSGb4</th>
<th id="T_e2eab_level0_col30" class="col_heading level0 col30" data-quarto-table-cell-role="th">6DSGb4</th>
<th id="T_e2eab_level0_col31" class="col_heading level0 col31" data-quarto-table-cell-role="th">3SGb5</th>
<th id="T_e2eab_level0_col32" class="col_heading level0 col32" data-quarto-table-cell-role="th">8DSGb5</th>
<th id="T_e2eab_level0_col33" class="col_heading level0 col33" data-quarto-table-cell-role="th">6DSGb5</th>
<th id="T_e2eab_level0_col34" class="col_heading level0 col34" data-quarto-table-cell-role="th">6DSGb5_2</th>
<th id="T_e2eab_level0_col35" class="col_heading level0 col35" data-quarto-table-cell-role="th">6SGb3</th>
<th id="T_e2eab_level0_col36" class="col_heading level0 col36" data-quarto-table-cell-role="th">8DSGb3_2</th>
<th id="T_e2eab_level0_col37" class="col_heading level0 col37" data-quarto-table-cell-role="th">6SGb4</th>
<th id="T_e2eab_level0_col38" class="col_heading level0 col38" data-quarto-table-cell-role="th">8DSGb4_2</th>
<th id="T_e2eab_level0_col39" class="col_heading level0 col39" data-quarto-table-cell-role="th">6SGb5</th>
<th id="T_e2eab_level0_col40" class="col_heading level0 col40" data-quarto-table-cell-role="th">8DSGb5_2</th>
<th id="T_e2eab_level0_col41" class="col_heading level0 col41" data-quarto-table-cell-role="th">66DSGb5</th>
<th id="T_e2eab_level0_col42" class="col_heading level0 col42" data-quarto-table-cell-role="th">Forssman_antigen</th>
<th id="T_e2eab_level0_col43" class="col_heading level0 col43" data-quarto-table-cell-role="th">iGb3</th>
<th id="T_e2eab_level0_col44" class="col_heading level0 col44" data-quarto-table-cell-role="th">I_antigen</th>
<th id="T_e2eab_level0_col45" class="col_heading level0 col45" data-quarto-table-cell-role="th">i_antigen</th>
<th id="T_e2eab_level0_col46" class="col_heading level0 col46" data-quarto-table-cell-role="th">PI_antigen</th>
<th id="T_e2eab_level0_col47" class="col_heading level0 col47" data-quarto-table-cell-role="th">Chitobiose</th>
<th id="T_e2eab_level0_col48" class="col_heading level0 col48" data-quarto-table-cell-role="th">Trimannosylcore</th>
<th id="T_e2eab_level0_col49" class="col_heading level0 col49" data-quarto-table-cell-role="th">Internal_LacNAc_type1</th>
<th id="T_e2eab_level0_col50" class="col_heading level0 col50" data-quarto-table-cell-role="th">Terminal_LacNAc_type1</th>
<th id="T_e2eab_level0_col51" class="col_heading level0 col51" data-quarto-table-cell-role="th">Internal_LacNAc_type2</th>
<th id="T_e2eab_level0_col52" class="col_heading level0 col52" data-quarto-table-cell-role="th">Terminal_LacNAc_type2</th>
<th id="T_e2eab_level0_col53" class="col_heading level0 col53" data-quarto-table-cell-role="th">Internal_LacdiNAc_type1</th>
<th id="T_e2eab_level0_col54" class="col_heading level0 col54" data-quarto-table-cell-role="th">Terminal_LacdiNAc_type1</th>
<th id="T_e2eab_level0_col55" class="col_heading level0 col55" data-quarto-table-cell-role="th">Internal_LacdiNAc_type2</th>
<th id="T_e2eab_level0_col56" class="col_heading level0 col56" data-quarto-table-cell-role="th">Terminal_LacdiNAc_type2</th>
<th id="T_e2eab_level0_col57" class="col_heading level0 col57" data-quarto-table-cell-role="th">bisectingGlcNAc</th>
<th id="T_e2eab_level0_col58" class="col_heading level0 col58" data-quarto-table-cell-role="th">VIM</th>
<th id="T_e2eab_level0_col59" class="col_heading level0 col59" data-quarto-table-cell-role="th">PolyLacNAc</th>
<th id="T_e2eab_level0_col60" class="col_heading level0 col60" data-quarto-table-cell-role="th">Ganglio_Series</th>
<th id="T_e2eab_level0_col61" class="col_heading level0 col61" data-quarto-table-cell-role="th">Lacto_Series(LewisC)</th>
<th id="T_e2eab_level0_col62" class="col_heading level0 col62" data-quarto-table-cell-role="th">NeoLacto_Series</th>
<th id="T_e2eab_level0_col63" class="col_heading level0 col63" data-quarto-table-cell-role="th">betaGlucan</th>
<th id="T_e2eab_level0_col64" class="col_heading level0 col64" data-quarto-table-cell-role="th">KeratanSulfate</th>
<th id="T_e2eab_level0_col65" class="col_heading level0 col65" data-quarto-table-cell-role="th">Hyaluronan</th>
<th id="T_e2eab_level0_col66" class="col_heading level0 col66" data-quarto-table-cell-role="th">Mollu_series</th>
<th id="T_e2eab_level0_col67" class="col_heading level0 col67" data-quarto-table-cell-role="th">Arthro_series</th>
<th id="T_e2eab_level0_col68" class="col_heading level0 col68" data-quarto-table-cell-role="th">Cellulose_like</th>
<th id="T_e2eab_level0_col69" class="col_heading level0 col69" data-quarto-table-cell-role="th">Chondroitin_4S</th>
<th id="T_e2eab_level0_col70" class="col_heading level0 col70" data-quarto-table-cell-role="th">GPI_anchor</th>
<th id="T_e2eab_level0_col71" class="col_heading level0 col71" data-quarto-table-cell-role="th">Isoglobo_series</th>
<th id="T_e2eab_level0_col72" class="col_heading level0 col72" data-quarto-table-cell-role="th">LewisD</th>
<th id="T_e2eab_level0_col73" class="col_heading level0 col73" data-quarto-table-cell-role="th">Globo_series</th>
<th id="T_e2eab_level0_col74" class="col_heading level0 col74" data-quarto-table-cell-role="th">Sda</th>
<th id="T_e2eab_level0_col75" class="col_heading level0 col75" data-quarto-table-cell-role="th">SDA</th>
<th id="T_e2eab_level0_col76" class="col_heading level0 col76" data-quarto-table-cell-role="th">Muco_series</th>
<th id="T_e2eab_level0_col77" class="col_heading level0 col77" data-quarto-table-cell-role="th">Heparin</th>
<th id="T_e2eab_level0_col78" class="col_heading level0 col78" data-quarto-table-cell-role="th">Peptidoglycan</th>
<th id="T_e2eab_level0_col79" class="col_heading level0 col79" data-quarto-table-cell-role="th">Dermatansulfate</th>
<th id="T_e2eab_level0_col80" class="col_heading level0 col80" data-quarto-table-cell-role="th">CAD</th>
<th id="T_e2eab_level0_col81" class="col_heading level0 col81" data-quarto-table-cell-role="th">Lactosylceramide</th>
<th id="T_e2eab_level0_col82" class="col_heading level0 col82" data-quarto-table-cell-role="th">Lactotriaosylceramide</th>
<th id="T_e2eab_level0_col83" class="col_heading level0 col83" data-quarto-table-cell-role="th">LexLex</th>
<th id="T_e2eab_level0_col84" class="col_heading level0 col84" data-quarto-table-cell-role="th">GM3</th>
<th id="T_e2eab_level0_col85" class="col_heading level0 col85" data-quarto-table-cell-role="th">H_type3</th>
<th id="T_e2eab_level0_col86" class="col_heading level0 col86" data-quarto-table-cell-role="th">GM2</th>
<th id="T_e2eab_level0_col87" class="col_heading level0 col87" data-quarto-table-cell-role="th">GM1</th>
<th id="T_e2eab_level0_col88" class="col_heading level0 col88" data-quarto-table-cell-role="th">cisGM1</th>
<th id="T_e2eab_level0_col89" class="col_heading level0 col89" data-quarto-table-cell-role="th">VIM2</th>
<th id="T_e2eab_level0_col90" class="col_heading level0 col90" data-quarto-table-cell-role="th">GD3</th>
<th id="T_e2eab_level0_col91" class="col_heading level0 col91" data-quarto-table-cell-role="th">GD1a</th>
<th id="T_e2eab_level0_col92" class="col_heading level0 col92" data-quarto-table-cell-role="th">GD2</th>
<th id="T_e2eab_level0_col93" class="col_heading level0 col93" data-quarto-table-cell-role="th">GD1b</th>
<th id="T_e2eab_level0_col94" class="col_heading level0 col94" data-quarto-table-cell-role="th">SDLex</th>
<th id="T_e2eab_level0_col95" class="col_heading level0 col95" data-quarto-table-cell-role="th">Fuc_LN3</th>
<th id="T_e2eab_level0_col96" class="col_heading level0 col96" data-quarto-table-cell-role="th">GT1b</th>
<th id="T_e2eab_level0_col97" class="col_heading level0 col97" data-quarto-table-cell-role="th">GD1</th>
<th id="T_e2eab_level0_col98" class="col_heading level0 col98" data-quarto-table-cell-role="th">GD1a_2</th>
<th id="T_e2eab_level0_col99" class="col_heading level0 col99" data-quarto-table-cell-role="th">LcGg4</th>
<th id="T_e2eab_level0_col100" class="col_heading level0 col100" data-quarto-table-cell-role="th">GT3</th>
<th id="T_e2eab_level0_col101" class="col_heading level0 col101" data-quarto-table-cell-role="th">Disialyl_T_antigen</th>
<th id="T_e2eab_level0_col102" class="col_heading level0 col102" data-quarto-table-cell-role="th">GT1a</th>
<th id="T_e2eab_level0_col103" class="col_heading level0 col103" data-quarto-table-cell-role="th">GT2</th>
<th id="T_e2eab_level0_col104" class="col_heading level0 col104" data-quarto-table-cell-role="th">GT1c</th>
<th id="T_e2eab_level0_col105" class="col_heading level0 col105" data-quarto-table-cell-role="th">2Fuc_GM1</th>
<th id="T_e2eab_level0_col106" class="col_heading level0 col106" data-quarto-table-cell-role="th">GQ1c</th>
<th id="T_e2eab_level0_col107" class="col_heading level0 col107" data-quarto-table-cell-role="th">O_linked_mannose</th>
<th id="T_e2eab_level0_col108" class="col_heading level0 col108" data-quarto-table-cell-role="th">GT1aa</th>
<th id="T_e2eab_level0_col109" class="col_heading level0 col109" data-quarto-table-cell-role="th">GQ1b</th>
<th id="T_e2eab_level0_col110" class="col_heading level0 col110" data-quarto-table-cell-role="th">HNK1</th>
<th id="T_e2eab_level0_col111" class="col_heading level0 col111" data-quarto-table-cell-role="th">GQ1ba</th>
<th id="T_e2eab_level0_col112" class="col_heading level0 col112" data-quarto-table-cell-role="th">O_mannose_Lex</th>
<th id="T_e2eab_level0_col113" class="col_heading level0 col113" data-quarto-table-cell-role="th">2Fuc_GD1b</th>
<th id="T_e2eab_level0_col114" class="col_heading level0 col114" data-quarto-table-cell-role="th">Sialopentaosylceramide</th>
<th id="T_e2eab_level0_col115" class="col_heading level0 col115" data-quarto-table-cell-role="th">Sulfogangliotetraosylceramide</th>
<th id="T_e2eab_level0_col116" class="col_heading level0 col116" data-quarto-table-cell-role="th">B-GM1</th>
<th id="T_e2eab_level0_col117" class="col_heading level0 col117" data-quarto-table-cell-role="th">GQ1aa</th>
<th id="T_e2eab_level0_col118" class="col_heading level0 col118" data-quarto-table-cell-role="th">bisSulfo-Lewis x</th>
<th id="T_e2eab_level0_col119" class="col_heading level0 col119" data-quarto-table-cell-role="th">para-Forssman</th>
<th id="T_e2eab_level0_col120" class="col_heading level0 col120" data-quarto-table-cell-role="th">core_fucose</th>
<th id="T_e2eab_level0_col121" class="col_heading level0 col121" data-quarto-table-cell-role="th">core_fucose(a1-3)</th>
<th id="T_e2eab_level0_col122" class="col_heading level0 col122" data-quarto-table-cell-role="th">GP1c</th>
<th id="T_e2eab_level0_col123" class="col_heading level0 col123" data-quarto-table-cell-role="th">B-GD1b</th>
<th id="T_e2eab_level0_col124" class="col_heading level0 col124" data-quarto-table-cell-role="th">GP1ca</th>
<th id="T_e2eab_level0_col125" class="col_heading level0 col125" data-quarto-table-cell-role="th">Isoglobotetraosylceramide</th>
<th id="T_e2eab_level0_col126" class="col_heading level0 col126" data-quarto-table-cell-role="th">polySia</th>
<th id="T_e2eab_level0_col127" class="col_heading level0 col127" data-quarto-table-cell-role="th">high_mannose</th>
<th id="T_e2eab_level0_col128" class="col_heading level0 col128" data-quarto-table-cell-role="th">Gala_series</th>
<th id="T_e2eab_level0_col129" class="col_heading level0 col129" data-quarto-table-cell-role="th">LPS_core</th>
<th id="T_e2eab_level0_col130" class="col_heading level0 col130" data-quarto-table-cell-role="th">Nglycan_complex</th>
<th id="T_e2eab_level0_col131" class="col_heading level0 col131" data-quarto-table-cell-role="th">Nglycan_complex2</th>
<th id="T_e2eab_level0_col132" class="col_heading level0 col132" data-quarto-table-cell-role="th">Oglycan_core1</th>
<th id="T_e2eab_level0_col133" class="col_heading level0 col133" data-quarto-table-cell-role="th">Oglycan_core2</th>
<th id="T_e2eab_level0_col134" class="col_heading level0 col134" data-quarto-table-cell-role="th">Oglycan_core3</th>
<th id="T_e2eab_level0_col135" class="col_heading level0 col135" data-quarto-table-cell-role="th">Oglycan_core4</th>
<th id="T_e2eab_level0_col136" class="col_heading level0 col136" data-quarto-table-cell-role="th">Oglycan_core5</th>
<th id="T_e2eab_level0_col137" class="col_heading level0 col137" data-quarto-table-cell-role="th">Oglycan_core6</th>
<th id="T_e2eab_level0_col138" class="col_heading level0 col138" data-quarto-table-cell-role="th">Oglycan_core7</th>
<th id="T_e2eab_level0_col139" class="col_heading level0 col139" data-quarto-table-cell-role="th">Xylogalacturonan</th>
<th id="T_e2eab_level0_col140" class="col_heading level0 col140" data-quarto-table-cell-role="th">Sialosylparagloboside</th>
<th id="T_e2eab_level0_col141" class="col_heading level0 col141" data-quarto-table-cell-role="th">LDNF</th>
<th id="T_e2eab_level0_col142" class="col_heading level0 col142" data-quarto-table-cell-role="th">OFuc</th>
<th id="T_e2eab_level0_col143" class="col_heading level0 col143" data-quarto-table-cell-role="th">Arabinogalactan_type2</th>
<th id="T_e2eab_level0_col144" class="col_heading level0 col144" data-quarto-table-cell-role="th">EGF_repeat</th>
<th id="T_e2eab_level0_col145" class="col_heading level0 col145" data-quarto-table-cell-role="th">Nglycan_hybrid</th>
<th id="T_e2eab_level0_col146" class="col_heading level0 col146" data-quarto-table-cell-role="th">Arabinan</th>
<th id="T_e2eab_level0_col147" class="col_heading level0 col147" data-quarto-table-cell-role="th">Xyloglucan</th>
<th id="T_e2eab_level0_col148" class="col_heading level0 col148" data-quarto-table-cell-role="th">Acharan_Sulfate</th>
<th id="T_e2eab_level0_col149" class="col_heading level0 col149" data-quarto-table-cell-role="th">M3FX</th>
<th id="T_e2eab_level0_col150" class="col_heading level0 col150" data-quarto-table-cell-role="th">M3X</th>
<th id="T_e2eab_level0_col151" class="col_heading level0 col151" data-quarto-table-cell-role="th">1-6betaGalactan</th>
<th id="T_e2eab_level0_col152" class="col_heading level0 col152" data-quarto-table-cell-role="th">Arabinogalactan_type1</th>
<th id="T_e2eab_level0_col153" class="col_heading level0 col153" data-quarto-table-cell-role="th">Galactomannan</th>
<th id="T_e2eab_level0_col154" class="col_heading level0 col154" data-quarto-table-cell-role="th">Tetraantennary_Nglycan</th>
<th id="T_e2eab_level0_col155" class="col_heading level0 col155" data-quarto-table-cell-role="th">Mucin_elongated_core2</th>
<th id="T_e2eab_level0_col156" class="col_heading level0 col156" data-quarto-table-cell-role="th">Fucoidan</th>
<th id="T_e2eab_level0_col157" class="col_heading level0 col157" data-quarto-table-cell-role="th">Alginate</th>
<th id="T_e2eab_level0_col158" class="col_heading level0 col158" data-quarto-table-cell-role="th">FG</th>
<th id="T_e2eab_level0_col159" class="col_heading level0 col159" data-quarto-table-cell-role="th">XX</th>
<th id="T_e2eab_level0_col160" class="col_heading level0 col160" data-quarto-table-cell-role="th">Difucosylated_core</th>
<th id="T_e2eab_level0_col161" class="col_heading level0 col161" data-quarto-table-cell-role="th">GalFuc_core</th>
<th id="T_e2eab_level0_col162" class="col_heading level0 col162" data-quarto-table-cell-role="th">DisialylLewisC</th>
<th id="T_e2eab_level0_col163" class="col_heading level0 col163" data-quarto-table-cell-role="th">RM2</th>
<th id="T_e2eab_level0_col164" class="col_heading level0 col164" data-quarto-table-cell-role="th">DisialylLewisA</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th id="T_e2eab_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<td id="T_e2eab_row0_col0" class="data row0 col0">0</td>
<td id="T_e2eab_row0_col1" class="data row0 col1">0</td>
<td id="T_e2eab_row0_col2" class="data row0 col2">0</td>
<td id="T_e2eab_row0_col3" class="data row0 col3">0</td>
<td id="T_e2eab_row0_col4" class="data row0 col4">0</td>
<td id="T_e2eab_row0_col5" class="data row0 col5">0</td>
<td id="T_e2eab_row0_col6" class="data row0 col6">0</td>
<td id="T_e2eab_row0_col7" class="data row0 col7">0</td>
<td id="T_e2eab_row0_col8" class="data row0 col8">0</td>
<td id="T_e2eab_row0_col9" class="data row0 col9">0</td>
<td id="T_e2eab_row0_col10" class="data row0 col10">0</td>
<td id="T_e2eab_row0_col11" class="data row0 col11">0</td>
<td id="T_e2eab_row0_col12" class="data row0 col12">0</td>
<td id="T_e2eab_row0_col13" class="data row0 col13">0</td>
<td id="T_e2eab_row0_col14" class="data row0 col14">0</td>
<td id="T_e2eab_row0_col15" class="data row0 col15">0</td>
<td id="T_e2eab_row0_col16" class="data row0 col16">0</td>
<td id="T_e2eab_row0_col17" class="data row0 col17">0</td>
<td id="T_e2eab_row0_col18" class="data row0 col18">0</td>
<td id="T_e2eab_row0_col19" class="data row0 col19">0</td>
<td id="T_e2eab_row0_col20" class="data row0 col20">0</td>
<td id="T_e2eab_row0_col21" class="data row0 col21">0</td>
<td id="T_e2eab_row0_col22" class="data row0 col22">0</td>
<td id="T_e2eab_row0_col23" class="data row0 col23">0</td>
<td id="T_e2eab_row0_col24" class="data row0 col24">0</td>
<td id="T_e2eab_row0_col25" class="data row0 col25">0</td>
<td id="T_e2eab_row0_col26" class="data row0 col26">0</td>
<td id="T_e2eab_row0_col27" class="data row0 col27">0</td>
<td id="T_e2eab_row0_col28" class="data row0 col28">0</td>
<td id="T_e2eab_row0_col29" class="data row0 col29">0</td>
<td id="T_e2eab_row0_col30" class="data row0 col30">0</td>
<td id="T_e2eab_row0_col31" class="data row0 col31">0</td>
<td id="T_e2eab_row0_col32" class="data row0 col32">0</td>
<td id="T_e2eab_row0_col33" class="data row0 col33">0</td>
<td id="T_e2eab_row0_col34" class="data row0 col34">0</td>
<td id="T_e2eab_row0_col35" class="data row0 col35">0</td>
<td id="T_e2eab_row0_col36" class="data row0 col36">0</td>
<td id="T_e2eab_row0_col37" class="data row0 col37">0</td>
<td id="T_e2eab_row0_col38" class="data row0 col38">0</td>
<td id="T_e2eab_row0_col39" class="data row0 col39">0</td>
<td id="T_e2eab_row0_col40" class="data row0 col40">0</td>
<td id="T_e2eab_row0_col41" class="data row0 col41">0</td>
<td id="T_e2eab_row0_col42" class="data row0 col42">0</td>
<td id="T_e2eab_row0_col43" class="data row0 col43">0</td>
<td id="T_e2eab_row0_col44" class="data row0 col44">0</td>
<td id="T_e2eab_row0_col45" class="data row0 col45">0</td>
<td id="T_e2eab_row0_col46" class="data row0 col46">0</td>
<td id="T_e2eab_row0_col47" class="data row0 col47">1</td>
<td id="T_e2eab_row0_col48" class="data row0 col48">1</td>
<td id="T_e2eab_row0_col49" class="data row0 col49">0</td>
<td id="T_e2eab_row0_col50" class="data row0 col50">0</td>
<td id="T_e2eab_row0_col51" class="data row0 col51">0</td>
<td id="T_e2eab_row0_col52" class="data row0 col52">0</td>
<td id="T_e2eab_row0_col53" class="data row0 col53">0</td>
<td id="T_e2eab_row0_col54" class="data row0 col54">0</td>
<td id="T_e2eab_row0_col55" class="data row0 col55">0</td>
<td id="T_e2eab_row0_col56" class="data row0 col56">0</td>
<td id="T_e2eab_row0_col57" class="data row0 col57">0</td>
<td id="T_e2eab_row0_col58" class="data row0 col58">0</td>
<td id="T_e2eab_row0_col59" class="data row0 col59">0</td>
<td id="T_e2eab_row0_col60" class="data row0 col60">0</td>
<td id="T_e2eab_row0_col61" class="data row0 col61">0</td>
<td id="T_e2eab_row0_col62" class="data row0 col62">0</td>
<td id="T_e2eab_row0_col63" class="data row0 col63">0</td>
<td id="T_e2eab_row0_col64" class="data row0 col64">0</td>
<td id="T_e2eab_row0_col65" class="data row0 col65">0</td>
<td id="T_e2eab_row0_col66" class="data row0 col66">0</td>
<td id="T_e2eab_row0_col67" class="data row0 col67">0</td>
<td id="T_e2eab_row0_col68" class="data row0 col68">0</td>
<td id="T_e2eab_row0_col69" class="data row0 col69">0</td>
<td id="T_e2eab_row0_col70" class="data row0 col70">0</td>
<td id="T_e2eab_row0_col71" class="data row0 col71">0</td>
<td id="T_e2eab_row0_col72" class="data row0 col72">0</td>
<td id="T_e2eab_row0_col73" class="data row0 col73">0</td>
<td id="T_e2eab_row0_col74" class="data row0 col74">0</td>
<td id="T_e2eab_row0_col75" class="data row0 col75">0</td>
<td id="T_e2eab_row0_col76" class="data row0 col76">0</td>
<td id="T_e2eab_row0_col77" class="data row0 col77">0</td>
<td id="T_e2eab_row0_col78" class="data row0 col78">0</td>
<td id="T_e2eab_row0_col79" class="data row0 col79">0</td>
<td id="T_e2eab_row0_col80" class="data row0 col80">0</td>
<td id="T_e2eab_row0_col81" class="data row0 col81">0</td>
<td id="T_e2eab_row0_col82" class="data row0 col82">0</td>
<td id="T_e2eab_row0_col83" class="data row0 col83">0</td>
<td id="T_e2eab_row0_col84" class="data row0 col84">0</td>
<td id="T_e2eab_row0_col85" class="data row0 col85">0</td>
<td id="T_e2eab_row0_col86" class="data row0 col86">0</td>
<td id="T_e2eab_row0_col87" class="data row0 col87">0</td>
<td id="T_e2eab_row0_col88" class="data row0 col88">0</td>
<td id="T_e2eab_row0_col89" class="data row0 col89">0</td>
<td id="T_e2eab_row0_col90" class="data row0 col90">0</td>
<td id="T_e2eab_row0_col91" class="data row0 col91">0</td>
<td id="T_e2eab_row0_col92" class="data row0 col92">0</td>
<td id="T_e2eab_row0_col93" class="data row0 col93">0</td>
<td id="T_e2eab_row0_col94" class="data row0 col94">0</td>
<td id="T_e2eab_row0_col95" class="data row0 col95">0</td>
<td id="T_e2eab_row0_col96" class="data row0 col96">0</td>
<td id="T_e2eab_row0_col97" class="data row0 col97">0</td>
<td id="T_e2eab_row0_col98" class="data row0 col98">0</td>
<td id="T_e2eab_row0_col99" class="data row0 col99">0</td>
<td id="T_e2eab_row0_col100" class="data row0 col100">0</td>
<td id="T_e2eab_row0_col101" class="data row0 col101">0</td>
<td id="T_e2eab_row0_col102" class="data row0 col102">0</td>
<td id="T_e2eab_row0_col103" class="data row0 col103">0</td>
<td id="T_e2eab_row0_col104" class="data row0 col104">0</td>
<td id="T_e2eab_row0_col105" class="data row0 col105">0</td>
<td id="T_e2eab_row0_col106" class="data row0 col106">0</td>
<td id="T_e2eab_row0_col107" class="data row0 col107">0</td>
<td id="T_e2eab_row0_col108" class="data row0 col108">0</td>
<td id="T_e2eab_row0_col109" class="data row0 col109">0</td>
<td id="T_e2eab_row0_col110" class="data row0 col110">0</td>
<td id="T_e2eab_row0_col111" class="data row0 col111">0</td>
<td id="T_e2eab_row0_col112" class="data row0 col112">0</td>
<td id="T_e2eab_row0_col113" class="data row0 col113">0</td>
<td id="T_e2eab_row0_col114" class="data row0 col114">0</td>
<td id="T_e2eab_row0_col115" class="data row0 col115">0</td>
<td id="T_e2eab_row0_col116" class="data row0 col116">0</td>
<td id="T_e2eab_row0_col117" class="data row0 col117">0</td>
<td id="T_e2eab_row0_col118" class="data row0 col118">0</td>
<td id="T_e2eab_row0_col119" class="data row0 col119">0</td>
<td id="T_e2eab_row0_col120" class="data row0 col120">0</td>
<td id="T_e2eab_row0_col121" class="data row0 col121">1</td>
<td id="T_e2eab_row0_col122" class="data row0 col122">0</td>
<td id="T_e2eab_row0_col123" class="data row0 col123">0</td>
<td id="T_e2eab_row0_col124" class="data row0 col124">0</td>
<td id="T_e2eab_row0_col125" class="data row0 col125">0</td>
<td id="T_e2eab_row0_col126" class="data row0 col126">0</td>
<td id="T_e2eab_row0_col127" class="data row0 col127">0</td>
<td id="T_e2eab_row0_col128" class="data row0 col128">0</td>
<td id="T_e2eab_row0_col129" class="data row0 col129">0</td>
<td id="T_e2eab_row0_col130" class="data row0 col130">0</td>
<td id="T_e2eab_row0_col131" class="data row0 col131">0</td>
<td id="T_e2eab_row0_col132" class="data row0 col132">0</td>
<td id="T_e2eab_row0_col133" class="data row0 col133">0</td>
<td id="T_e2eab_row0_col134" class="data row0 col134">0</td>
<td id="T_e2eab_row0_col135" class="data row0 col135">0</td>
<td id="T_e2eab_row0_col136" class="data row0 col136">0</td>
<td id="T_e2eab_row0_col137" class="data row0 col137">0</td>
<td id="T_e2eab_row0_col138" class="data row0 col138">0</td>
<td id="T_e2eab_row0_col139" class="data row0 col139">0</td>
<td id="T_e2eab_row0_col140" class="data row0 col140">0</td>
<td id="T_e2eab_row0_col141" class="data row0 col141">0</td>
<td id="T_e2eab_row0_col142" class="data row0 col142">0</td>
<td id="T_e2eab_row0_col143" class="data row0 col143">0</td>
<td id="T_e2eab_row0_col144" class="data row0 col144">0</td>
<td id="T_e2eab_row0_col145" class="data row0 col145">0</td>
<td id="T_e2eab_row0_col146" class="data row0 col146">0</td>
<td id="T_e2eab_row0_col147" class="data row0 col147">0</td>
<td id="T_e2eab_row0_col148" class="data row0 col148">0</td>
<td id="T_e2eab_row0_col149" class="data row0 col149">0</td>
<td id="T_e2eab_row0_col150" class="data row0 col150">0</td>
<td id="T_e2eab_row0_col151" class="data row0 col151">0</td>
<td id="T_e2eab_row0_col152" class="data row0 col152">0</td>
<td id="T_e2eab_row0_col153" class="data row0 col153">0</td>
<td id="T_e2eab_row0_col154" class="data row0 col154">0</td>
<td id="T_e2eab_row0_col155" class="data row0 col155">0</td>
<td id="T_e2eab_row0_col156" class="data row0 col156">0</td>
<td id="T_e2eab_row0_col157" class="data row0 col157">0</td>
<td id="T_e2eab_row0_col158" class="data row0 col158">0</td>
<td id="T_e2eab_row0_col159" class="data row0 col159">0</td>
<td id="T_e2eab_row0_col160" class="data row0 col160">0</td>
<td id="T_e2eab_row0_col161" class="data row0 col161">0</td>
<td id="T_e2eab_row0_col162" class="data row0 col162">0</td>
<td id="T_e2eab_row0_col163" class="data row0 col163">0</td>
<td id="T_e2eab_row0_col164" class="data row0 col164">0</td>
</tr>
<tr class="even">
<th id="T_e2eab_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<td id="T_e2eab_row1_col0" class="data row1 col0">0</td>
<td id="T_e2eab_row1_col1" class="data row1 col1">0</td>
<td id="T_e2eab_row1_col2" class="data row1 col2">0</td>
<td id="T_e2eab_row1_col3" class="data row1 col3">0</td>
<td id="T_e2eab_row1_col4" class="data row1 col4">0</td>
<td id="T_e2eab_row1_col5" class="data row1 col5">0</td>
<td id="T_e2eab_row1_col6" class="data row1 col6">0</td>
<td id="T_e2eab_row1_col7" class="data row1 col7">0</td>
<td id="T_e2eab_row1_col8" class="data row1 col8">0</td>
<td id="T_e2eab_row1_col9" class="data row1 col9">0</td>
<td id="T_e2eab_row1_col10" class="data row1 col10">0</td>
<td id="T_e2eab_row1_col11" class="data row1 col11">0</td>
<td id="T_e2eab_row1_col12" class="data row1 col12">0</td>
<td id="T_e2eab_row1_col13" class="data row1 col13">0</td>
<td id="T_e2eab_row1_col14" class="data row1 col14">0</td>
<td id="T_e2eab_row1_col15" class="data row1 col15">0</td>
<td id="T_e2eab_row1_col16" class="data row1 col16">0</td>
<td id="T_e2eab_row1_col17" class="data row1 col17">0</td>
<td id="T_e2eab_row1_col18" class="data row1 col18">0</td>
<td id="T_e2eab_row1_col19" class="data row1 col19">0</td>
<td id="T_e2eab_row1_col20" class="data row1 col20">0</td>
<td id="T_e2eab_row1_col21" class="data row1 col21">0</td>
<td id="T_e2eab_row1_col22" class="data row1 col22">0</td>
<td id="T_e2eab_row1_col23" class="data row1 col23">0</td>
<td id="T_e2eab_row1_col24" class="data row1 col24">0</td>
<td id="T_e2eab_row1_col25" class="data row1 col25">0</td>
<td id="T_e2eab_row1_col26" class="data row1 col26">0</td>
<td id="T_e2eab_row1_col27" class="data row1 col27">0</td>
<td id="T_e2eab_row1_col28" class="data row1 col28">0</td>
<td id="T_e2eab_row1_col29" class="data row1 col29">0</td>
<td id="T_e2eab_row1_col30" class="data row1 col30">0</td>
<td id="T_e2eab_row1_col31" class="data row1 col31">0</td>
<td id="T_e2eab_row1_col32" class="data row1 col32">0</td>
<td id="T_e2eab_row1_col33" class="data row1 col33">0</td>
<td id="T_e2eab_row1_col34" class="data row1 col34">0</td>
<td id="T_e2eab_row1_col35" class="data row1 col35">0</td>
<td id="T_e2eab_row1_col36" class="data row1 col36">0</td>
<td id="T_e2eab_row1_col37" class="data row1 col37">0</td>
<td id="T_e2eab_row1_col38" class="data row1 col38">0</td>
<td id="T_e2eab_row1_col39" class="data row1 col39">0</td>
<td id="T_e2eab_row1_col40" class="data row1 col40">0</td>
<td id="T_e2eab_row1_col41" class="data row1 col41">0</td>
<td id="T_e2eab_row1_col42" class="data row1 col42">0</td>
<td id="T_e2eab_row1_col43" class="data row1 col43">0</td>
<td id="T_e2eab_row1_col44" class="data row1 col44">0</td>
<td id="T_e2eab_row1_col45" class="data row1 col45">0</td>
<td id="T_e2eab_row1_col46" class="data row1 col46">0</td>
<td id="T_e2eab_row1_col47" class="data row1 col47">1</td>
<td id="T_e2eab_row1_col48" class="data row1 col48">1</td>
<td id="T_e2eab_row1_col49" class="data row1 col49">0</td>
<td id="T_e2eab_row1_col50" class="data row1 col50">0</td>
<td id="T_e2eab_row1_col51" class="data row1 col51">0</td>
<td id="T_e2eab_row1_col52" class="data row1 col52">0</td>
<td id="T_e2eab_row1_col53" class="data row1 col53">0</td>
<td id="T_e2eab_row1_col54" class="data row1 col54">0</td>
<td id="T_e2eab_row1_col55" class="data row1 col55">0</td>
<td id="T_e2eab_row1_col56" class="data row1 col56">0</td>
<td id="T_e2eab_row1_col57" class="data row1 col57">0</td>
<td id="T_e2eab_row1_col58" class="data row1 col58">0</td>
<td id="T_e2eab_row1_col59" class="data row1 col59">0</td>
<td id="T_e2eab_row1_col60" class="data row1 col60">0</td>
<td id="T_e2eab_row1_col61" class="data row1 col61">0</td>
<td id="T_e2eab_row1_col62" class="data row1 col62">0</td>
<td id="T_e2eab_row1_col63" class="data row1 col63">0</td>
<td id="T_e2eab_row1_col64" class="data row1 col64">0</td>
<td id="T_e2eab_row1_col65" class="data row1 col65">0</td>
<td id="T_e2eab_row1_col66" class="data row1 col66">0</td>
<td id="T_e2eab_row1_col67" class="data row1 col67">0</td>
<td id="T_e2eab_row1_col68" class="data row1 col68">0</td>
<td id="T_e2eab_row1_col69" class="data row1 col69">0</td>
<td id="T_e2eab_row1_col70" class="data row1 col70">0</td>
<td id="T_e2eab_row1_col71" class="data row1 col71">0</td>
<td id="T_e2eab_row1_col72" class="data row1 col72">0</td>
<td id="T_e2eab_row1_col73" class="data row1 col73">0</td>
<td id="T_e2eab_row1_col74" class="data row1 col74">0</td>
<td id="T_e2eab_row1_col75" class="data row1 col75">0</td>
<td id="T_e2eab_row1_col76" class="data row1 col76">0</td>
<td id="T_e2eab_row1_col77" class="data row1 col77">0</td>
<td id="T_e2eab_row1_col78" class="data row1 col78">0</td>
<td id="T_e2eab_row1_col79" class="data row1 col79">0</td>
<td id="T_e2eab_row1_col80" class="data row1 col80">0</td>
<td id="T_e2eab_row1_col81" class="data row1 col81">0</td>
<td id="T_e2eab_row1_col82" class="data row1 col82">0</td>
<td id="T_e2eab_row1_col83" class="data row1 col83">0</td>
<td id="T_e2eab_row1_col84" class="data row1 col84">0</td>
<td id="T_e2eab_row1_col85" class="data row1 col85">0</td>
<td id="T_e2eab_row1_col86" class="data row1 col86">0</td>
<td id="T_e2eab_row1_col87" class="data row1 col87">0</td>
<td id="T_e2eab_row1_col88" class="data row1 col88">0</td>
<td id="T_e2eab_row1_col89" class="data row1 col89">0</td>
<td id="T_e2eab_row1_col90" class="data row1 col90">0</td>
<td id="T_e2eab_row1_col91" class="data row1 col91">0</td>
<td id="T_e2eab_row1_col92" class="data row1 col92">0</td>
<td id="T_e2eab_row1_col93" class="data row1 col93">0</td>
<td id="T_e2eab_row1_col94" class="data row1 col94">0</td>
<td id="T_e2eab_row1_col95" class="data row1 col95">0</td>
<td id="T_e2eab_row1_col96" class="data row1 col96">0</td>
<td id="T_e2eab_row1_col97" class="data row1 col97">0</td>
<td id="T_e2eab_row1_col98" class="data row1 col98">0</td>
<td id="T_e2eab_row1_col99" class="data row1 col99">0</td>
<td id="T_e2eab_row1_col100" class="data row1 col100">0</td>
<td id="T_e2eab_row1_col101" class="data row1 col101">0</td>
<td id="T_e2eab_row1_col102" class="data row1 col102">0</td>
<td id="T_e2eab_row1_col103" class="data row1 col103">0</td>
<td id="T_e2eab_row1_col104" class="data row1 col104">0</td>
<td id="T_e2eab_row1_col105" class="data row1 col105">0</td>
<td id="T_e2eab_row1_col106" class="data row1 col106">0</td>
<td id="T_e2eab_row1_col107" class="data row1 col107">0</td>
<td id="T_e2eab_row1_col108" class="data row1 col108">0</td>
<td id="T_e2eab_row1_col109" class="data row1 col109">0</td>
<td id="T_e2eab_row1_col110" class="data row1 col110">0</td>
<td id="T_e2eab_row1_col111" class="data row1 col111">0</td>
<td id="T_e2eab_row1_col112" class="data row1 col112">0</td>
<td id="T_e2eab_row1_col113" class="data row1 col113">0</td>
<td id="T_e2eab_row1_col114" class="data row1 col114">0</td>
<td id="T_e2eab_row1_col115" class="data row1 col115">0</td>
<td id="T_e2eab_row1_col116" class="data row1 col116">0</td>
<td id="T_e2eab_row1_col117" class="data row1 col117">0</td>
<td id="T_e2eab_row1_col118" class="data row1 col118">0</td>
<td id="T_e2eab_row1_col119" class="data row1 col119">0</td>
<td id="T_e2eab_row1_col120" class="data row1 col120">0</td>
<td id="T_e2eab_row1_col121" class="data row1 col121">0</td>
<td id="T_e2eab_row1_col122" class="data row1 col122">0</td>
<td id="T_e2eab_row1_col123" class="data row1 col123">0</td>
<td id="T_e2eab_row1_col124" class="data row1 col124">0</td>
<td id="T_e2eab_row1_col125" class="data row1 col125">0</td>
<td id="T_e2eab_row1_col126" class="data row1 col126">0</td>
<td id="T_e2eab_row1_col127" class="data row1 col127">0</td>
<td id="T_e2eab_row1_col128" class="data row1 col128">0</td>
<td id="T_e2eab_row1_col129" class="data row1 col129">0</td>
<td id="T_e2eab_row1_col130" class="data row1 col130">0</td>
<td id="T_e2eab_row1_col131" class="data row1 col131">0</td>
<td id="T_e2eab_row1_col132" class="data row1 col132">0</td>
<td id="T_e2eab_row1_col133" class="data row1 col133">0</td>
<td id="T_e2eab_row1_col134" class="data row1 col134">0</td>
<td id="T_e2eab_row1_col135" class="data row1 col135">0</td>
<td id="T_e2eab_row1_col136" class="data row1 col136">0</td>
<td id="T_e2eab_row1_col137" class="data row1 col137">0</td>
<td id="T_e2eab_row1_col138" class="data row1 col138">0</td>
<td id="T_e2eab_row1_col139" class="data row1 col139">0</td>
<td id="T_e2eab_row1_col140" class="data row1 col140">0</td>
<td id="T_e2eab_row1_col141" class="data row1 col141">0</td>
<td id="T_e2eab_row1_col142" class="data row1 col142">0</td>
<td id="T_e2eab_row1_col143" class="data row1 col143">0</td>
<td id="T_e2eab_row1_col144" class="data row1 col144">0</td>
<td id="T_e2eab_row1_col145" class="data row1 col145">0</td>
<td id="T_e2eab_row1_col146" class="data row1 col146">0</td>
<td id="T_e2eab_row1_col147" class="data row1 col147">0</td>
<td id="T_e2eab_row1_col148" class="data row1 col148">0</td>
<td id="T_e2eab_row1_col149" class="data row1 col149">0</td>
<td id="T_e2eab_row1_col150" class="data row1 col150">0</td>
<td id="T_e2eab_row1_col151" class="data row1 col151">0</td>
<td id="T_e2eab_row1_col152" class="data row1 col152">0</td>
<td id="T_e2eab_row1_col153" class="data row1 col153">0</td>
<td id="T_e2eab_row1_col154" class="data row1 col154">0</td>
<td id="T_e2eab_row1_col155" class="data row1 col155">0</td>
<td id="T_e2eab_row1_col156" class="data row1 col156">0</td>
<td id="T_e2eab_row1_col157" class="data row1 col157">0</td>
<td id="T_e2eab_row1_col158" class="data row1 col158">0</td>
<td id="T_e2eab_row1_col159" class="data row1 col159">0</td>
<td id="T_e2eab_row1_col160" class="data row1 col160">0</td>
<td id="T_e2eab_row1_col161" class="data row1 col161">0</td>
<td id="T_e2eab_row1_col162" class="data row1 col162">0</td>
<td id="T_e2eab_row1_col163" class="data row1 col163">0</td>
<td id="T_e2eab_row1_col164" class="data row1 col164">0</td>
</tr>
<tr class="odd">
<th id="T_e2eab_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P</th>
<td id="T_e2eab_row2_col0" class="data row2 col0">0</td>
<td id="T_e2eab_row2_col1" class="data row2 col1">0</td>
<td id="T_e2eab_row2_col2" class="data row2 col2">0</td>
<td id="T_e2eab_row2_col3" class="data row2 col3">0</td>
<td id="T_e2eab_row2_col4" class="data row2 col4">0</td>
<td id="T_e2eab_row2_col5" class="data row2 col5">0</td>
<td id="T_e2eab_row2_col6" class="data row2 col6">0</td>
<td id="T_e2eab_row2_col7" class="data row2 col7">0</td>
<td id="T_e2eab_row2_col8" class="data row2 col8">0</td>
<td id="T_e2eab_row2_col9" class="data row2 col9">0</td>
<td id="T_e2eab_row2_col10" class="data row2 col10">0</td>
<td id="T_e2eab_row2_col11" class="data row2 col11">0</td>
<td id="T_e2eab_row2_col12" class="data row2 col12">0</td>
<td id="T_e2eab_row2_col13" class="data row2 col13">0</td>
<td id="T_e2eab_row2_col14" class="data row2 col14">0</td>
<td id="T_e2eab_row2_col15" class="data row2 col15">0</td>
<td id="T_e2eab_row2_col16" class="data row2 col16">0</td>
<td id="T_e2eab_row2_col17" class="data row2 col17">0</td>
<td id="T_e2eab_row2_col18" class="data row2 col18">0</td>
<td id="T_e2eab_row2_col19" class="data row2 col19">0</td>
<td id="T_e2eab_row2_col20" class="data row2 col20">0</td>
<td id="T_e2eab_row2_col21" class="data row2 col21">0</td>
<td id="T_e2eab_row2_col22" class="data row2 col22">0</td>
<td id="T_e2eab_row2_col23" class="data row2 col23">0</td>
<td id="T_e2eab_row2_col24" class="data row2 col24">0</td>
<td id="T_e2eab_row2_col25" class="data row2 col25">0</td>
<td id="T_e2eab_row2_col26" class="data row2 col26">0</td>
<td id="T_e2eab_row2_col27" class="data row2 col27">0</td>
<td id="T_e2eab_row2_col28" class="data row2 col28">0</td>
<td id="T_e2eab_row2_col29" class="data row2 col29">0</td>
<td id="T_e2eab_row2_col30" class="data row2 col30">0</td>
<td id="T_e2eab_row2_col31" class="data row2 col31">0</td>
<td id="T_e2eab_row2_col32" class="data row2 col32">0</td>
<td id="T_e2eab_row2_col33" class="data row2 col33">0</td>
<td id="T_e2eab_row2_col34" class="data row2 col34">0</td>
<td id="T_e2eab_row2_col35" class="data row2 col35">0</td>
<td id="T_e2eab_row2_col36" class="data row2 col36">0</td>
<td id="T_e2eab_row2_col37" class="data row2 col37">0</td>
<td id="T_e2eab_row2_col38" class="data row2 col38">0</td>
<td id="T_e2eab_row2_col39" class="data row2 col39">0</td>
<td id="T_e2eab_row2_col40" class="data row2 col40">0</td>
<td id="T_e2eab_row2_col41" class="data row2 col41">0</td>
<td id="T_e2eab_row2_col42" class="data row2 col42">0</td>
<td id="T_e2eab_row2_col43" class="data row2 col43">0</td>
<td id="T_e2eab_row2_col44" class="data row2 col44">0</td>
<td id="T_e2eab_row2_col45" class="data row2 col45">0</td>
<td id="T_e2eab_row2_col46" class="data row2 col46">0</td>
<td id="T_e2eab_row2_col47" class="data row2 col47">0</td>
<td id="T_e2eab_row2_col48" class="data row2 col48">0</td>
<td id="T_e2eab_row2_col49" class="data row2 col49">0</td>
<td id="T_e2eab_row2_col50" class="data row2 col50">0</td>
<td id="T_e2eab_row2_col51" class="data row2 col51">0</td>
<td id="T_e2eab_row2_col52" class="data row2 col52">0</td>
<td id="T_e2eab_row2_col53" class="data row2 col53">0</td>
<td id="T_e2eab_row2_col54" class="data row2 col54">0</td>
<td id="T_e2eab_row2_col55" class="data row2 col55">0</td>
<td id="T_e2eab_row2_col56" class="data row2 col56">0</td>
<td id="T_e2eab_row2_col57" class="data row2 col57">0</td>
<td id="T_e2eab_row2_col58" class="data row2 col58">0</td>
<td id="T_e2eab_row2_col59" class="data row2 col59">0</td>
<td id="T_e2eab_row2_col60" class="data row2 col60">0</td>
<td id="T_e2eab_row2_col61" class="data row2 col61">0</td>
<td id="T_e2eab_row2_col62" class="data row2 col62">0</td>
<td id="T_e2eab_row2_col63" class="data row2 col63">0</td>
<td id="T_e2eab_row2_col64" class="data row2 col64">0</td>
<td id="T_e2eab_row2_col65" class="data row2 col65">0</td>
<td id="T_e2eab_row2_col66" class="data row2 col66">0</td>
<td id="T_e2eab_row2_col67" class="data row2 col67">0</td>
<td id="T_e2eab_row2_col68" class="data row2 col68">0</td>
<td id="T_e2eab_row2_col69" class="data row2 col69">0</td>
<td id="T_e2eab_row2_col70" class="data row2 col70">0</td>
<td id="T_e2eab_row2_col71" class="data row2 col71">0</td>
<td id="T_e2eab_row2_col72" class="data row2 col72">0</td>
<td id="T_e2eab_row2_col73" class="data row2 col73">0</td>
<td id="T_e2eab_row2_col74" class="data row2 col74">0</td>
<td id="T_e2eab_row2_col75" class="data row2 col75">0</td>
<td id="T_e2eab_row2_col76" class="data row2 col76">0</td>
<td id="T_e2eab_row2_col77" class="data row2 col77">0</td>
<td id="T_e2eab_row2_col78" class="data row2 col78">0</td>
<td id="T_e2eab_row2_col79" class="data row2 col79">0</td>
<td id="T_e2eab_row2_col80" class="data row2 col80">0</td>
<td id="T_e2eab_row2_col81" class="data row2 col81">0</td>
<td id="T_e2eab_row2_col82" class="data row2 col82">0</td>
<td id="T_e2eab_row2_col83" class="data row2 col83">0</td>
<td id="T_e2eab_row2_col84" class="data row2 col84">0</td>
<td id="T_e2eab_row2_col85" class="data row2 col85">0</td>
<td id="T_e2eab_row2_col86" class="data row2 col86">0</td>
<td id="T_e2eab_row2_col87" class="data row2 col87">0</td>
<td id="T_e2eab_row2_col88" class="data row2 col88">0</td>
<td id="T_e2eab_row2_col89" class="data row2 col89">0</td>
<td id="T_e2eab_row2_col90" class="data row2 col90">0</td>
<td id="T_e2eab_row2_col91" class="data row2 col91">0</td>
<td id="T_e2eab_row2_col92" class="data row2 col92">0</td>
<td id="T_e2eab_row2_col93" class="data row2 col93">0</td>
<td id="T_e2eab_row2_col94" class="data row2 col94">0</td>
<td id="T_e2eab_row2_col95" class="data row2 col95">0</td>
<td id="T_e2eab_row2_col96" class="data row2 col96">0</td>
<td id="T_e2eab_row2_col97" class="data row2 col97">0</td>
<td id="T_e2eab_row2_col98" class="data row2 col98">0</td>
<td id="T_e2eab_row2_col99" class="data row2 col99">0</td>
<td id="T_e2eab_row2_col100" class="data row2 col100">0</td>
<td id="T_e2eab_row2_col101" class="data row2 col101">0</td>
<td id="T_e2eab_row2_col102" class="data row2 col102">0</td>
<td id="T_e2eab_row2_col103" class="data row2 col103">0</td>
<td id="T_e2eab_row2_col104" class="data row2 col104">0</td>
<td id="T_e2eab_row2_col105" class="data row2 col105">0</td>
<td id="T_e2eab_row2_col106" class="data row2 col106">0</td>
<td id="T_e2eab_row2_col107" class="data row2 col107">0</td>
<td id="T_e2eab_row2_col108" class="data row2 col108">0</td>
<td id="T_e2eab_row2_col109" class="data row2 col109">0</td>
<td id="T_e2eab_row2_col110" class="data row2 col110">0</td>
<td id="T_e2eab_row2_col111" class="data row2 col111">0</td>
<td id="T_e2eab_row2_col112" class="data row2 col112">0</td>
<td id="T_e2eab_row2_col113" class="data row2 col113">0</td>
<td id="T_e2eab_row2_col114" class="data row2 col114">0</td>
<td id="T_e2eab_row2_col115" class="data row2 col115">0</td>
<td id="T_e2eab_row2_col116" class="data row2 col116">0</td>
<td id="T_e2eab_row2_col117" class="data row2 col117">0</td>
<td id="T_e2eab_row2_col118" class="data row2 col118">0</td>
<td id="T_e2eab_row2_col119" class="data row2 col119">0</td>
<td id="T_e2eab_row2_col120" class="data row2 col120">0</td>
<td id="T_e2eab_row2_col121" class="data row2 col121">0</td>
<td id="T_e2eab_row2_col122" class="data row2 col122">0</td>
<td id="T_e2eab_row2_col123" class="data row2 col123">0</td>
<td id="T_e2eab_row2_col124" class="data row2 col124">0</td>
<td id="T_e2eab_row2_col125" class="data row2 col125">0</td>
<td id="T_e2eab_row2_col126" class="data row2 col126">0</td>
<td id="T_e2eab_row2_col127" class="data row2 col127">0</td>
<td id="T_e2eab_row2_col128" class="data row2 col128">0</td>
<td id="T_e2eab_row2_col129" class="data row2 col129">0</td>
<td id="T_e2eab_row2_col130" class="data row2 col130">0</td>
<td id="T_e2eab_row2_col131" class="data row2 col131">0</td>
<td id="T_e2eab_row2_col132" class="data row2 col132">0</td>
<td id="T_e2eab_row2_col133" class="data row2 col133">0</td>
<td id="T_e2eab_row2_col134" class="data row2 col134">0</td>
<td id="T_e2eab_row2_col135" class="data row2 col135">0</td>
<td id="T_e2eab_row2_col136" class="data row2 col136">0</td>
<td id="T_e2eab_row2_col137" class="data row2 col137">0</td>
<td id="T_e2eab_row2_col138" class="data row2 col138">0</td>
<td id="T_e2eab_row2_col139" class="data row2 col139">0</td>
<td id="T_e2eab_row2_col140" class="data row2 col140">0</td>
<td id="T_e2eab_row2_col141" class="data row2 col141">0</td>
<td id="T_e2eab_row2_col142" class="data row2 col142">0</td>
<td id="T_e2eab_row2_col143" class="data row2 col143">0</td>
<td id="T_e2eab_row2_col144" class="data row2 col144">0</td>
<td id="T_e2eab_row2_col145" class="data row2 col145">0</td>
<td id="T_e2eab_row2_col146" class="data row2 col146">0</td>
<td id="T_e2eab_row2_col147" class="data row2 col147">0</td>
<td id="T_e2eab_row2_col148" class="data row2 col148">0</td>
<td id="T_e2eab_row2_col149" class="data row2 col149">0</td>
<td id="T_e2eab_row2_col150" class="data row2 col150">0</td>
<td id="T_e2eab_row2_col151" class="data row2 col151">0</td>
<td id="T_e2eab_row2_col152" class="data row2 col152">0</td>
<td id="T_e2eab_row2_col153" class="data row2 col153">0</td>
<td id="T_e2eab_row2_col154" class="data row2 col154">0</td>
<td id="T_e2eab_row2_col155" class="data row2 col155">0</td>
<td id="T_e2eab_row2_col156" class="data row2 col156">0</td>
<td id="T_e2eab_row2_col157" class="data row2 col157">0</td>
<td id="T_e2eab_row2_col158" class="data row2 col158">0</td>
<td id="T_e2eab_row2_col159" class="data row2 col159">0</td>
<td id="T_e2eab_row2_col160" class="data row2 col160">0</td>
<td id="T_e2eab_row2_col161" class="data row2 col161">0</td>
<td id="T_e2eab_row2_col162" class="data row2 col162">0</td>
<td id="T_e2eab_row2_col163" class="data row2 col163">0</td>
<td id="T_e2eab_row2_col164" class="data row2 col164">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
<section id="quantify_motifs" class="level3">
<h3 class="anchored" data-anchor-id="quantify_motifs">quantify_motifs</h3>
<blockquote class="blockquote">
<pre><code> quantify_motifs (df:str|pandas.core.frame.DataFrame, glycans:list[str],
                  feature_set:list[str], custom_motifs:list=[],
                  remove_redundant:bool=True)</code></pre>
</blockquote>
<p><em>Extracts and quantifies motif abundances from glycan abundance data by weighting motif occurrences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>str | pandas.core.frame.DataFrame</td>
<td></td>
<td>DataFrame or filepath with samples as columns, abundances as values</td>
</tr>
<tr class="even">
<td>glycans</td>
<td>list</td>
<td></td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="odd">
<td>feature_set</td>
<td>list</td>
<td></td>
<td>Feature types to analyze: known, graph, exhaustive, terminal(1-3), custom, chemical, size_branch</td>
</tr>
<tr class="even">
<td>custom_motifs</td>
<td>list</td>
<td>[]</td>
<td>Custom motifs when using ‘custom’ feature set</td>
</tr>
<tr class="odd">
<td>remove_redundant</td>
<td>bool</td>
<td>True</td>
<td>Remove redundant motifs via deduplicate_motifs</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with motif abundances (motifs as columns, samples as rows)</strong></td>
</tr>
</tbody>
</table>
<div id="fewer-franchise" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>quantify_motifs(test_df.iloc[:, <span class="dv">1</span>:], test_df.iloc[:, <span class="dv">0</span>].values.tolist(), [<span class="st">'known'</span>, <span class="st">'exhaustive'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">control_1</th>
<th data-quarto-table-cell-role="th">tumor_1</th>
<th data-quarto-table-cell-role="th">control_2</th>
<th data-quarto-table-cell-role="th">tumor_2</th>
<th data-quarto-table-cell-role="th">control_3</th>
<th data-quarto-table-cell-role="th">tumor_3</th>
<th data-quarto-table-cell-role="th">control_4</th>
<th data-quarto-table-cell-role="th">tumor_4</th>
<th data-quarto-table-cell-role="th">control_5</th>
<th data-quarto-table-cell-role="th">tumor_5</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">control_16</th>
<th data-quarto-table-cell-role="th">tumor_16</th>
<th data-quarto-table-cell-role="th">control_17</th>
<th data-quarto-table-cell-role="th">tumor_17</th>
<th data-quarto-table-cell-role="th">control_18</th>
<th data-quarto-table-cell-role="th">tumor_18</th>
<th data-quarto-table-cell-role="th">control_19</th>
<th data-quarto-table-cell-role="th">tumor_19</th>
<th data-quarto-table-cell-role="th">control_20</th>
<th data-quarto-table-cell-role="th">tumor_20</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">H_antigen_type2</th>
<td>1.347737</td>
<td>0.892651</td>
<td>2.468405</td>
<td>1.810795</td>
<td>1.589162</td>
<td>0.449339</td>
<td>2.640132</td>
<td>0.572828</td>
<td>2.763890</td>
<td>0.737076</td>
<td>...</td>
<td>1.070249</td>
<td>0.647786</td>
<td>1.440912</td>
<td>1.810304</td>
<td>1.722289</td>
<td>1.475260</td>
<td>4.847788</td>
<td>4.552496</td>
<td>0.480035</td>
<td>0.494123</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Internal_LacNAc_type2</th>
<td>8.845085</td>
<td>10.063160</td>
<td>13.435501</td>
<td>28.834006</td>
<td>5.585973</td>
<td>11.359659</td>
<td>11.672584</td>
<td>21.193308</td>
<td>12.734919</td>
<td>28.597709</td>
<td>...</td>
<td>10.883437</td>
<td>17.991155</td>
<td>21.166792</td>
<td>16.161351</td>
<td>11.909325</td>
<td>29.924308</td>
<td>12.820872</td>
<td>19.107379</td>
<td>8.802443</td>
<td>10.268911</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Terminal_LacNAc_type2</th>
<td>52.982192</td>
<td>13.183951</td>
<td>24.413523</td>
<td>12.870782</td>
<td>9.555884</td>
<td>9.822266</td>
<td>12.628910</td>
<td>13.916662</td>
<td>26.569737</td>
<td>10.733867</td>
<td>...</td>
<td>18.779972</td>
<td>12.157928</td>
<td>14.828507</td>
<td>20.879287</td>
<td>27.689619</td>
<td>10.734756</td>
<td>28.328965</td>
<td>37.870847</td>
<td>14.835019</td>
<td>8.910804</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Disialyl_T_antigen</th>
<td>20.803836</td>
<td>36.895471</td>
<td>32.803297</td>
<td>20.401157</td>
<td>33.971366</td>
<td>30.150599</td>
<td>37.703636</td>
<td>24.728411</td>
<td>31.798990</td>
<td>15.989214</td>
<td>...</td>
<td>46.337629</td>
<td>39.476930</td>
<td>39.087708</td>
<td>40.348217</td>
<td>35.791797</td>
<td>22.968160</td>
<td>11.026029</td>
<td>2.613718</td>
<td>44.676379</td>
<td>46.125360</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Oglycan_core1</th>
<td>37.329013</td>
<td>75.567842</td>
<td>59.998893</td>
<td>57.608119</td>
<td>83.293693</td>
<td>78.436161</td>
<td>73.308916</td>
<td>64.356888</td>
<td>58.197862</td>
<td>60.329536</td>
<td>...</td>
<td>68.269613</td>
<td>68.762287</td>
<td>62.541874</td>
<td>60.699726</td>
<td>58.713271</td>
<td>58.203265</td>
<td>58.826129</td>
<td>42.904325</td>
<td>74.390026</td>
<td>79.515568</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Mucin_elongated_core2</th>
<td>61.827277</td>
<td>23.247111</td>
<td>37.849024</td>
<td>41.704788</td>
<td>15.141858</td>
<td>21.181925</td>
<td>24.301494</td>
<td>35.109970</td>
<td>39.304656</td>
<td>39.331576</td>
<td>...</td>
<td>29.663409</td>
<td>30.149083</td>
<td>35.995300</td>
<td>37.040638</td>
<td>39.598944</td>
<td>40.659064</td>
<td>41.149838</td>
<td>56.978227</td>
<td>23.637462</td>
<td>19.179715</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Gal</th>
<td>163.691481</td>
<td>126.500106</td>
<td>141.895063</td>
<td>147.702533</td>
<td>115.056369</td>
<td>132.721945</td>
<td>122.804259</td>
<td>138.398297</td>
<td>141.412183</td>
<td>167.203077</td>
<td>...</td>
<td>133.838024</td>
<td>140.218313</td>
<td>142.530133</td>
<td>139.697255</td>
<td>138.848449</td>
<td>154.791018</td>
<td>142.588964</td>
<td>157.426027</td>
<td>122.916027</td>
<td>120.555251</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">GalNAc</th>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>...</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
<td>100.000000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">GalOS(b1-3)GalNAc</th>
<td>0.843710</td>
<td>1.185047</td>
<td>2.152084</td>
<td>0.687093</td>
<td>1.564450</td>
<td>0.381914</td>
<td>2.389590</td>
<td>0.533142</td>
<td>2.497482</td>
<td>0.338889</td>
<td>...</td>
<td>2.066978</td>
<td>1.088630</td>
<td>1.462826</td>
<td>2.259636</td>
<td>1.687785</td>
<td>1.137672</td>
<td>0.024033</td>
<td>0.117449</td>
<td>1.972512</td>
<td>1.304717</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">GlcNAc6S(b1-6)GalNAc</th>
<td>2.707913</td>
<td>4.438043</td>
<td>6.198123</td>
<td>6.684838</td>
<td>1.478960</td>
<td>11.921934</td>
<td>0.892356</td>
<td>3.821469</td>
<td>4.605009</td>
<td>28.210391</td>
<td>...</td>
<td>6.241593</td>
<td>11.157860</td>
<td>7.997660</td>
<td>4.916252</td>
<td>0.937290</td>
<td>15.269626</td>
<td>1.463159</td>
<td>0.565249</td>
<td>1.251077</td>
<td>2.680253</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Neu5Ac</th>
<td>80.494155</td>
<td>134.094482</td>
<td>120.708503</td>
<td>125.892731</td>
<td>128.626161</td>
<td>137.543517</td>
<td>132.135127</td>
<td>124.740497</td>
<td>118.279272</td>
<td>134.227059</td>
<td>...</td>
<td>149.089683</td>
<td>152.360772</td>
<td>145.124475</td>
<td>140.251427</td>
<td>125.331418</td>
<td>121.962226</td>
<td>91.599064</td>
<td>72.000898</td>
<td>142.956534</td>
<td>148.579697</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Gal(b1-3)GalNAc</th>
<td>99.156290</td>
<td>98.814953</td>
<td>97.847916</td>
<td>99.312907</td>
<td>98.435550</td>
<td>99.618086</td>
<td>97.610410</td>
<td>99.466858</td>
<td>97.502518</td>
<td>99.661111</td>
<td>...</td>
<td>97.933022</td>
<td>98.911370</td>
<td>98.537174</td>
<td>97.740364</td>
<td>98.312215</td>
<td>98.862328</td>
<td>99.975967</td>
<td>99.882551</td>
<td>98.027488</td>
<td>98.695283</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Neu5Ac(a2-3)Gal</th>
<td>57.345927</td>
<td>94.670033</td>
<td>83.675402</td>
<td>103.574200</td>
<td>91.775344</td>
<td>106.231617</td>
<td>90.136699</td>
<td>98.461821</td>
<td>81.110136</td>
<td>117.087919</td>
<td>...</td>
<td>97.928245</td>
<td>109.749014</td>
<td>101.760261</td>
<td>93.222423</td>
<td>86.403840</td>
<td>96.715461</td>
<td>80.029183</td>
<td>69.040921</td>
<td>95.565848</td>
<td>99.973512</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Neu5Ac(a2-6)GalNAc</th>
<td>23.063482</td>
<td>39.304399</td>
<td>36.644881</td>
<td>22.263129</td>
<td>36.571122</td>
<td>31.229766</td>
<td>41.628644</td>
<td>26.256121</td>
<td>37.088978</td>
<td>17.054227</td>
<td>...</td>
<td>50.675599</td>
<td>41.982557</td>
<td>42.829042</td>
<td>46.391984</td>
<td>38.682564</td>
<td>25.118814</td>
<td>11.540028</td>
<td>2.937334</td>
<td>47.171520</td>
<td>48.274238</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Neu5Ac(a2-8)Neu5Ac</th>
<td>0.084745</td>
<td>0.120050</td>
<td>0.388219</td>
<td>0.055402</td>
<td>0.279696</td>
<td>0.082135</td>
<td>0.369784</td>
<td>0.022555</td>
<td>0.080158</td>
<td>0.084913</td>
<td>...</td>
<td>0.485839</td>
<td>0.629202</td>
<td>0.535171</td>
<td>0.637019</td>
<td>0.245015</td>
<td>0.127952</td>
<td>0.029853</td>
<td>0.022643</td>
<td>0.219166</td>
<td>0.331947</td>
</tr>
</tbody>
</table>

<p>15 rows × 40 columns</p>
</div>
</div>
</div>
<hr>
</section>
<section id="get_k_saccharides" class="level3">
<h3 class="anchored" data-anchor-id="get_k_saccharides">get_k_saccharides</h3>
<blockquote class="blockquote">
<pre><code> get_k_saccharides (glycans:list[str]|set[str], size:int=2,
                    up_to:bool=False, just_motifs:bool=False,
                    terminal:bool=False)</code></pre>
</blockquote>
<p><em>Extracts k-saccharide fragments from glycan sequences with options for different fragment sizes and positions</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycans</td>
<td>list[str] | set[str]</td>
<td></td>
<td>List or set of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="even">
<td>size</td>
<td>int</td>
<td>2</td>
<td>Number of monosaccharides per fragment</td>
</tr>
<tr class="odd">
<td>up_to</td>
<td>bool</td>
<td>False</td>
<td>Include fragments up to size k (adds monosaccharides)</td>
</tr>
<tr class="even">
<td>just_motifs</td>
<td>bool</td>
<td>False</td>
<td>Return nested list of motifs instead of count DataFrame</td>
</tr>
<tr class="odd">
<td>terminal</td>
<td>bool</td>
<td>False</td>
<td>Only count terminal fragments</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>pandas.core.frame.DataFrame | list[list[str]]</strong></td>
<td></td>
<td><strong>DataFrame of k-saccharide counts or list of motifs per glycan</strong></td>
</tr>
</tbody>
</table>
<div id="toxic-implement" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>glycans <span class="op">=</span> [<span class="st">'Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc'</span>,</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">'GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcN4P(b1-6)GlcN4P'</span>]</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> get_k_saccharides(glycans, size <span class="op">=</span> <span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="wanted-chemical" class="cell">
<div class="cell-output cell-output-display">
<style type="text/css">
#T_7b1fd_row0_col0, #T_7b1fd_row0_col1, #T_7b1fd_row0_col2, #T_7b1fd_row0_col3, #T_7b1fd_row0_col4, #T_7b1fd_row0_col5, #T_7b1fd_row0_col6, #T_7b1fd_row0_col7, #T_7b1fd_row0_col8, #T_7b1fd_row0_col9, #T_7b1fd_row0_col10, #T_7b1fd_row0_col11, #T_7b1fd_row0_col12, #T_7b1fd_row0_col13, #T_7b1fd_row0_col14, #T_7b1fd_row0_col15, #T_7b1fd_row0_col16, #T_7b1fd_row0_col17, #T_7b1fd_row0_col18, #T_7b1fd_row0_col19, #T_7b1fd_row0_col20, #T_7b1fd_row0_col21, #T_7b1fd_row0_col22, #T_7b1fd_row0_col23, #T_7b1fd_row1_col0, #T_7b1fd_row1_col1, #T_7b1fd_row1_col2, #T_7b1fd_row1_col3, #T_7b1fd_row1_col4, #T_7b1fd_row1_col5, #T_7b1fd_row1_col6, #T_7b1fd_row1_col7, #T_7b1fd_row1_col8, #T_7b1fd_row1_col9, #T_7b1fd_row1_col10, #T_7b1fd_row1_col11, #T_7b1fd_row1_col12, #T_7b1fd_row1_col13, #T_7b1fd_row1_col14, #T_7b1fd_row1_col15, #T_7b1fd_row1_col16, #T_7b1fd_row1_col17, #T_7b1fd_row1_col18, #T_7b1fd_row1_col19, #T_7b1fd_row1_col20, #T_7b1fd_row1_col21, #T_7b1fd_row1_col22, #T_7b1fd_row1_col23, #T_7b1fd_row2_col0, #T_7b1fd_row2_col1, #T_7b1fd_row2_col2, #T_7b1fd_row2_col3, #T_7b1fd_row2_col4, #T_7b1fd_row2_col5, #T_7b1fd_row2_col6, #T_7b1fd_row2_col7, #T_7b1fd_row2_col8, #T_7b1fd_row2_col9, #T_7b1fd_row2_col10, #T_7b1fd_row2_col11, #T_7b1fd_row2_col12, #T_7b1fd_row2_col13, #T_7b1fd_row2_col14, #T_7b1fd_row2_col15, #T_7b1fd_row2_col16, #T_7b1fd_row2_col17, #T_7b1fd_row2_col18, #T_7b1fd_row2_col19, #T_7b1fd_row2_col20, #T_7b1fd_row2_col21, #T_7b1fd_row2_col22, #T_7b1fd_row2_col23 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>

<table id="T_7b1fd" class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_7b1fd_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">Fuc(a1-3)[GlcNAc(b1-4)]GlcNAc</th>
<th id="T_7b1fd_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">GalNAc(a1-4)GlcNAcA(a1-4)Kdo</th>
<th id="T_7b1fd_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">GlcN(b1-7)Kdo(a2-5)Kdo</th>
<th id="T_7b1fd_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">GlcN(b1-?)[GlcNAcA(a1-?)]Kdo</th>
<th id="T_7b1fd_level0_col4" class="col_heading level0 col4" data-quarto-table-cell-role="th">GlcNAcA(a1-4)Kdo(a2-5)Kdo</th>
<th id="T_7b1fd_level0_col5" class="col_heading level0 col5" data-quarto-table-cell-role="th">GlcNAcA(a1-4)[GlcN(b1-7)]Kdo</th>
<th id="T_7b1fd_level0_col6" class="col_heading level0 col6" data-quarto-table-cell-role="th">Kdo(a2-4)Kdo(a2-6)GlcN4P</th>
<th id="T_7b1fd_level0_col7" class="col_heading level0 col7" data-quarto-table-cell-role="th">Kdo(a2-4)[Kdo(a2-5)]Kdo</th>
<th id="T_7b1fd_level0_col8" class="col_heading level0 col8" data-quarto-table-cell-role="th">Kdo(a2-5)Kdo(a2-6)GlcN4P</th>
<th id="T_7b1fd_level0_col9" class="col_heading level0 col9" data-quarto-table-cell-role="th">Kdo(a2-6)GlcN4P(b1-6)GlcN4P</th>
<th id="T_7b1fd_level0_col10" class="col_heading level0 col10" data-quarto-table-cell-role="th">Kdo(a2-?)Kdo(a2-?)GlcN4P</th>
<th id="T_7b1fd_level0_col11" class="col_heading level0 col11" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man</th>
<th id="T_7b1fd_level0_col12" class="col_heading level0 col12" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)Man</th>
<th id="T_7b1fd_level0_col13" class="col_heading level0 col13" data-quarto-table-cell-role="th">Man(a1-3)Man(a1-6)Man</th>
<th id="T_7b1fd_level0_col14" class="col_heading level0 col14" data-quarto-table-cell-role="th">Man(a1-3)Man(b1-4)GlcNAc</th>
<th id="T_7b1fd_level0_col15" class="col_heading level0 col15" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man</th>
<th id="T_7b1fd_level0_col16" class="col_heading level0 col16" data-quarto-table-cell-role="th">Man(a1-6)Man(b1-4)GlcNAc</th>
<th id="T_7b1fd_level0_col17" class="col_heading level0 col17" data-quarto-table-cell-role="th">Man(a1-?)Man(a1-?)Man</th>
<th id="T_7b1fd_level0_col18" class="col_heading level0 col18" data-quarto-table-cell-role="th">Man(a1-?)Man(b1-?)GlcNAc</th>
<th id="T_7b1fd_level0_col19" class="col_heading level0 col19" data-quarto-table-cell-role="th">Man(a1-?)[Xyl(b1-?)]Man</th>
<th id="T_7b1fd_level0_col20" class="col_heading level0 col20" data-quarto-table-cell-role="th">Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_7b1fd_level0_col21" class="col_heading level0 col21" data-quarto-table-cell-role="th">Xyl(b1-2)Man(b1-4)GlcNAc</th>
<th id="T_7b1fd_level0_col22" class="col_heading level0 col22" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)]Man</th>
<th id="T_7b1fd_level0_col23" class="col_heading level0 col23" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-6)]Man</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th id="T_7b1fd_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</th>
<td id="T_7b1fd_row0_col0" class="data row0 col0">1</td>
<td id="T_7b1fd_row0_col1" class="data row0 col1">0</td>
<td id="T_7b1fd_row0_col2" class="data row0 col2">0</td>
<td id="T_7b1fd_row0_col3" class="data row0 col3">0</td>
<td id="T_7b1fd_row0_col4" class="data row0 col4">0</td>
<td id="T_7b1fd_row0_col5" class="data row0 col5">0</td>
<td id="T_7b1fd_row0_col6" class="data row0 col6">0</td>
<td id="T_7b1fd_row0_col7" class="data row0 col7">0</td>
<td id="T_7b1fd_row0_col8" class="data row0 col8">0</td>
<td id="T_7b1fd_row0_col9" class="data row0 col9">0</td>
<td id="T_7b1fd_row0_col10" class="data row0 col10">0</td>
<td id="T_7b1fd_row0_col11" class="data row0 col11">0</td>
<td id="T_7b1fd_row0_col12" class="data row0 col12">0</td>
<td id="T_7b1fd_row0_col13" class="data row0 col13">0</td>
<td id="T_7b1fd_row0_col14" class="data row0 col14">1</td>
<td id="T_7b1fd_row0_col15" class="data row0 col15">1</td>
<td id="T_7b1fd_row0_col16" class="data row0 col16">1</td>
<td id="T_7b1fd_row0_col17" class="data row0 col17">0</td>
<td id="T_7b1fd_row0_col18" class="data row0 col18">2</td>
<td id="T_7b1fd_row0_col19" class="data row0 col19">2</td>
<td id="T_7b1fd_row0_col20" class="data row0 col20">1</td>
<td id="T_7b1fd_row0_col21" class="data row0 col21">1</td>
<td id="T_7b1fd_row0_col22" class="data row0 col22">1</td>
<td id="T_7b1fd_row0_col23" class="data row0 col23">1</td>
</tr>
<tr class="even">
<th id="T_7b1fd_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</th>
<td id="T_7b1fd_row1_col0" class="data row1 col0">0</td>
<td id="T_7b1fd_row1_col1" class="data row1 col1">0</td>
<td id="T_7b1fd_row1_col2" class="data row1 col2">0</td>
<td id="T_7b1fd_row1_col3" class="data row1 col3">0</td>
<td id="T_7b1fd_row1_col4" class="data row1 col4">0</td>
<td id="T_7b1fd_row1_col5" class="data row1 col5">0</td>
<td id="T_7b1fd_row1_col6" class="data row1 col6">0</td>
<td id="T_7b1fd_row1_col7" class="data row1 col7">0</td>
<td id="T_7b1fd_row1_col8" class="data row1 col8">0</td>
<td id="T_7b1fd_row1_col9" class="data row1 col9">0</td>
<td id="T_7b1fd_row1_col10" class="data row1 col10">0</td>
<td id="T_7b1fd_row1_col11" class="data row1 col11">1</td>
<td id="T_7b1fd_row1_col12" class="data row1 col12">1</td>
<td id="T_7b1fd_row1_col13" class="data row1 col13">1</td>
<td id="T_7b1fd_row1_col14" class="data row1 col14">1</td>
<td id="T_7b1fd_row1_col15" class="data row1 col15">1</td>
<td id="T_7b1fd_row1_col16" class="data row1 col16">1</td>
<td id="T_7b1fd_row1_col17" class="data row1 col17">3</td>
<td id="T_7b1fd_row1_col18" class="data row1 col18">2</td>
<td id="T_7b1fd_row1_col19" class="data row1 col19">0</td>
<td id="T_7b1fd_row1_col20" class="data row1 col20">1</td>
<td id="T_7b1fd_row1_col21" class="data row1 col21">0</td>
<td id="T_7b1fd_row1_col22" class="data row1 col22">0</td>
<td id="T_7b1fd_row1_col23" class="data row1 col23">0</td>
</tr>
<tr class="odd">
<th id="T_7b1fd_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">2</th>
<td id="T_7b1fd_row2_col0" class="data row2 col0">0</td>
<td id="T_7b1fd_row2_col1" class="data row2 col1">1</td>
<td id="T_7b1fd_row2_col2" class="data row2 col2">1</td>
<td id="T_7b1fd_row2_col3" class="data row2 col3">1</td>
<td id="T_7b1fd_row2_col4" class="data row2 col4">1</td>
<td id="T_7b1fd_row2_col5" class="data row2 col5">1</td>
<td id="T_7b1fd_row2_col6" class="data row2 col6">1</td>
<td id="T_7b1fd_row2_col7" class="data row2 col7">1</td>
<td id="T_7b1fd_row2_col8" class="data row2 col8">1</td>
<td id="T_7b1fd_row2_col9" class="data row2 col9">1</td>
<td id="T_7b1fd_row2_col10" class="data row2 col10">2</td>
<td id="T_7b1fd_row2_col11" class="data row2 col11">0</td>
<td id="T_7b1fd_row2_col12" class="data row2 col12">0</td>
<td id="T_7b1fd_row2_col13" class="data row2 col13">0</td>
<td id="T_7b1fd_row2_col14" class="data row2 col14">0</td>
<td id="T_7b1fd_row2_col15" class="data row2 col15">0</td>
<td id="T_7b1fd_row2_col16" class="data row2 col16">0</td>
<td id="T_7b1fd_row2_col17" class="data row2 col17">0</td>
<td id="T_7b1fd_row2_col18" class="data row2 col18">0</td>
<td id="T_7b1fd_row2_col19" class="data row2 col19">0</td>
<td id="T_7b1fd_row2_col20" class="data row2 col20">0</td>
<td id="T_7b1fd_row2_col21" class="data row2 col21">0</td>
<td id="T_7b1fd_row2_col22" class="data row2 col22">0</td>
<td id="T_7b1fd_row2_col23" class="data row2 col23">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
<section id="get_terminal_structures" class="level3">
<h3 class="anchored" data-anchor-id="get_terminal_structures">get_terminal_structures</h3>
<blockquote class="blockquote">
<pre><code> get_terminal_structures (glycan:str|networkx.classes.digraph.DiGraph,
                          size:int=1)</code></pre>
</blockquote>
<p><em>Identifies terminal monosaccharide sequences from non-reducing ends of glycan structure</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>IUPAC-condensed glycan sequence or NetworkX graph</td>
</tr>
<tr class="even">
<td>size</td>
<td>int</td>
<td>1</td>
<td>Number of monosaccharides in terminal fragment (1 or 2)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of terminal structures with linkages</strong></td>
</tr>
</tbody>
</table>
<div id="promising-november" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>get_terminal_structures(<span class="st">"Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['Neu5Ac(a2-3)', 'Neu5Ac(a2-6)']</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_molecular_properties" class="level3">
<h3 class="anchored" data-anchor-id="get_molecular_properties">get_molecular_properties</h3>
<blockquote class="blockquote">
<pre><code> get_molecular_properties (glycan_list:list[str], verbose:bool=False,
                           placeholder:bool=False)</code></pre>
</blockquote>
<p><em>Retrieves molecular properties from PubChem for a list of glycans using their SMILES representations</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_list</td>
<td>list</td>
<td></td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>Print SMILES not found on PubChem</td>
</tr>
<tr class="odd">
<td>placeholder</td>
<td>bool</td>
<td>False</td>
<td>Return dummy values instead of dropping failed requests</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with molecular parameters from PubChem</strong></td>
</tr>
</tbody>
</table>
<div id="exposed-cisco" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> get_molecular_properties([<span class="st">"Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="relevant-census" class="cell">
<div class="cell-output cell-output-display">
<style type="text/css">
#T_0cc55_row0_col0, #T_0cc55_row0_col1, #T_0cc55_row0_col2, #T_0cc55_row0_col3, #T_0cc55_row0_col4, #T_0cc55_row0_col5, #T_0cc55_row0_col6, #T_0cc55_row0_col7, #T_0cc55_row0_col8, #T_0cc55_row0_col9, #T_0cc55_row0_col10, #T_0cc55_row0_col11, #T_0cc55_row0_col12, #T_0cc55_row0_col13, #T_0cc55_row0_col14, #T_0cc55_row0_col15, #T_0cc55_row0_col16, #T_0cc55_row0_col17, #T_0cc55_row0_col18 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>

<table id="T_0cc55" class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_0cc55_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">rotatable_bond_count</th>
<th id="T_0cc55_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">undefined_atom_stereo_count</th>
<th id="T_0cc55_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">monoisotopic_mass</th>
<th id="T_0cc55_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">isotope_atom_count</th>
<th id="T_0cc55_level0_col4" class="col_heading level0 col4" data-quarto-table-cell-role="th">atom_stereo_count</th>
<th id="T_0cc55_level0_col5" class="col_heading level0 col5" data-quarto-table-cell-role="th">molecular_weight</th>
<th id="T_0cc55_level0_col6" class="col_heading level0 col6" data-quarto-table-cell-role="th">tpsa</th>
<th id="T_0cc55_level0_col7" class="col_heading level0 col7" data-quarto-table-cell-role="th">h_bond_donor_count</th>
<th id="T_0cc55_level0_col8" class="col_heading level0 col8" data-quarto-table-cell-role="th">defined_atom_stereo_count</th>
<th id="T_0cc55_level0_col9" class="col_heading level0 col9" data-quarto-table-cell-role="th">xlogp</th>
<th id="T_0cc55_level0_col10" class="col_heading level0 col10" data-quarto-table-cell-role="th">bond_stereo_count</th>
<th id="T_0cc55_level0_col11" class="col_heading level0 col11" data-quarto-table-cell-role="th">covalent_unit_count</th>
<th id="T_0cc55_level0_col12" class="col_heading level0 col12" data-quarto-table-cell-role="th">h_bond_acceptor_count</th>
<th id="T_0cc55_level0_col13" class="col_heading level0 col13" data-quarto-table-cell-role="th">undefined_bond_stereo_count</th>
<th id="T_0cc55_level0_col14" class="col_heading level0 col14" data-quarto-table-cell-role="th">defined_bond_stereo_count</th>
<th id="T_0cc55_level0_col15" class="col_heading level0 col15" data-quarto-table-cell-role="th">heavy_atom_count</th>
<th id="T_0cc55_level0_col16" class="col_heading level0 col16" data-quarto-table-cell-role="th">charge</th>
<th id="T_0cc55_level0_col17" class="col_heading level0 col17" data-quarto-table-cell-role="th">complexity</th>
<th id="T_0cc55_level0_col18" class="col_heading level0 col18" data-quarto-table-cell-role="th">exact_mass</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th id="T_0cc55_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<td id="T_0cc55_row0_col0" class="data row0 col0">43</td>
<td id="T_0cc55_row0_col1" class="data row0 col1">1</td>
<td id="T_0cc55_row0_col2" class="data row0 col2">2222.7830048</td>
<td id="T_0cc55_row0_col3" class="data row0 col3">0</td>
<td id="T_0cc55_row0_col4" class="data row0 col4">57</td>
<td id="T_0cc55_row0_col5" class="data row0 col5">2224.0</td>
<td id="T_0cc55_row0_col6" class="data row0 col6">1070</td>
<td id="T_0cc55_row0_col7" class="data row0 col7">39</td>
<td id="T_0cc55_row0_col8" class="data row0 col8">56</td>
<td id="T_0cc55_row0_col9" class="data row0 col9">-23.600000</td>
<td id="T_0cc55_row0_col10" class="data row0 col10">0</td>
<td id="T_0cc55_row0_col11" class="data row0 col11">1</td>
<td id="T_0cc55_row0_col12" class="data row0 col12">62</td>
<td id="T_0cc55_row0_col13" class="data row0 col13">0</td>
<td id="T_0cc55_row0_col14" class="data row0 col14">0</td>
<td id="T_0cc55_row0_col15" class="data row0 col15">152</td>
<td id="T_0cc55_row0_col16" class="data row0 col16">0</td>
<td id="T_0cc55_row0_col17" class="data row0 col17">4410</td>
<td id="T_0cc55_row0_col18" class="data row0 col18">2222.7830048</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
<section id="get_glycan_similarity" class="level3">
<h3 class="anchored" data-anchor-id="get_glycan_similarity">get_glycan_similarity</h3>
<blockquote class="blockquote">
<pre><code> get_glycan_similarity (glycan1:str|networkx.classes.digraph.DiGraph,
                        glycan2:str|networkx.classes.digraph.DiGraph,
                        motifs:pandas.core.frame.DataFrame|None=None,
                        feature_set:list=['known', 'exhaustive',
                        'terminal'])</code></pre>
</blockquote>
<p><em>Calculates cosine similarity between two glycans based on their motif count fingerprints</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan1</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>IUPAC-condensed glycan sequence or NetworkX graph</td>
</tr>
<tr class="even">
<td>glycan2</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>IUPAC-condensed glycan sequence or NetworkX graph</td>
</tr>
<tr class="odd">
<td>motifs</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Motif dataframe (name + sequence); defaults to motif_list</td>
</tr>
<tr class="even">
<td>feature_set</td>
<td>list</td>
<td>[‘known’, ‘exhaustive’, ‘terminal’]</td>
<td>Feature types to analyze: known, graph, exhaustive, terminal(1-3), custom, chemical, size_branch</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>float</strong></td>
<td></td>
<td><strong>Cosine similarity between glycan1 and glycan2</strong></td>
</tr>
</tbody>
</table>
<div id="134c3212-d536-46eb-b0be-d64e2383b3e8" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>get_glycan_similarity(<span class="st">"Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>, <span class="st">"Neu5Ac(a2-3)Gal(b1-4)[Neu5Ac(a2-6)]GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>0.7276068751089989</code></pre>
</div>
</div>
</section>
</section>
<section id="graph" class="level2">
<h2 class="anchored" data-anchor-id="graph">graph</h2>
<blockquote class="blockquote">
<p>convert glycan sequences to graphs and contains helper functions to search for motifs / check whether two sequences describe the same sequence, etc.</p>
</blockquote>
<hr>
<section id="glycan_to_nxgraph" class="level3">
<h3 class="anchored" data-anchor-id="glycan_to_nxgraph">glycan_to_nxGraph</h3>
<blockquote class="blockquote">
<pre><code> glycan_to_nxGraph (glycan:str,
                    libr:glycowork.glycan_data.loader.HashableDict[str,int
                    ]|None=None, termini:str='ignore',
                    termini_list:tuple[str]|None=None)</code></pre>
</blockquote>
<p><em>Wrapper for converting glycans into networkx graphs; also works with floating substituents</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td>libr</td>
<td>glycowork.glycan_data.loader.HashableDict[str, int] | None</td>
<td>None</td>
<td>Dictionary of form glycoletter:index</td>
</tr>
<tr class="odd">
<td>termini</td>
<td>str</td>
<td>ignore</td>
<td>How to encode terminal/internal position; options: ignore, calc, provided</td>
</tr>
<tr class="even">
<td>termini_list</td>
<td>tuple[str] | None</td>
<td>None</td>
<td>List of positions from terminal/internal/flexible</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DiGraph</strong></td>
<td></td>
<td><strong>NetworkX graph object of glycan</strong></td>
</tr>
</tbody>
</table>
<div id="blank-dining" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Glycan to networkx Graph (only edges printed)'</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(glycan_to_nxGraph(<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>).edges())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Glycan to networkx Graph (only edges printed)
[(1, 0), (3, 2), (4, 1), (4, 3), (5, 4), (6, 5), (7, 6), (9, 8), (10, 7), (10, 9)]</code></pre>
</div>
</div>
<hr>
</section>
<section id="graph_to_string" class="level3">
<h3 class="anchored" data-anchor-id="graph_to_string">graph_to_string</h3>
<blockquote class="blockquote">
<pre><code> graph_to_string (graph:networkx.classes.digraph.DiGraph,
                  canonicalize:bool=True, order_by:str='length')</code></pre>
</blockquote>
<p><em>Convert glycan graph back to IUPAC-condensed format, handling disconnected components</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>graph</td>
<td>DiGraph</td>
<td></td>
<td>Glycan graph (assumes root node is the one with the highest index)</td>
</tr>
<tr class="even">
<td>canonicalize</td>
<td>bool</td>
<td>True</td>
<td>Whether to output canonicalized IUPAC-condensed</td>
</tr>
<tr class="odd">
<td>order_by</td>
<td>str</td>
<td>length</td>
<td>canonicalize by ‘length’ or ‘linkage’</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td></td>
<td><strong>IUPAC-condensed glycan string</strong></td>
</tr>
</tbody>
</table>
<div id="fancy-identification" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>graph_to_string(glycan_to_nxGraph(<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</code></pre>
</div>
</div>
<hr>
</section>
<section id="compare_glycans" class="level3">
<h3 class="anchored" data-anchor-id="compare_glycans">compare_glycans</h3>
<blockquote class="blockquote">
<pre><code> compare_glycans (glycan_a:str|networkx.classes.digraph.DiGraph,
                  glycan_b:str|networkx.classes.digraph.DiGraph,
                  return_matches:bool=False)</code></pre>
</blockquote>
<p><em>Check whether two glycans are identical</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_a</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>First glycan to compare</td>
</tr>
<tr class="even">
<td>glycan_b</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Second glycan to compare</td>
</tr>
<tr class="odd">
<td>return_matches</td>
<td>bool</td>
<td>False</td>
<td>Whether to return node mapping between glycans</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>bool</strong></td>
<td></td>
<td><strong>True if glycans are same, False if not</strong></td>
</tr>
</tbody>
</table>
<div id="lucky-mistake" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Graph Isomorphism Test"</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(compare_glycans(<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>,</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Graph Isomorphism Test
True</code></pre>
</div>
</div>
<hr>
</section>
<section id="subgraph_isomorphism" class="level3">
<h3 class="anchored" data-anchor-id="subgraph_isomorphism">subgraph_isomorphism</h3>
<blockquote class="blockquote">
<pre><code> subgraph_isomorphism (glycan:str|networkx.classes.digraph.DiGraph,
                       motif:str|networkx.classes.digraph.DiGraph,
                       termini_list:list=[], count:bool=False,
                       return_matches:bool=False)</code></pre>
</blockquote>
<p><em>Check if motif exists as subgraph in glycan</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan sequence or graph</td>
</tr>
<tr class="even">
<td>motif</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan motif sequence or graph</td>
</tr>
<tr class="odd">
<td>termini_list</td>
<td>list</td>
<td>[]</td>
<td>List of monosaccharide positions from terminal/internal/flexible</td>
</tr>
<tr class="even">
<td>count</td>
<td>bool</td>
<td>False</td>
<td>Whether to return count instead of presence/absence</td>
</tr>
<tr class="odd">
<td>return_matches</td>
<td>bool</td>
<td>False</td>
<td>Whether to return matched subgraphs as node lists</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>bool | int | tuple[int, list[list[int]]]</strong></td>
<td></td>
<td><strong>Boolean presence, count, or (count, matches)</strong></td>
</tr>
</tbody>
</table>
<div id="superb-failing" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Subgraph Isomorphism Test"</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(subgraph_isomorphism(<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>,</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>                           <span class="st">'Fuc(a1-6)GlcNAc'</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Subgraph Isomorphism Test
True</code></pre>
</div>
</div>
<hr>
</section>
<section id="generate_graph_features" class="level3">
<h3 class="anchored" data-anchor-id="generate_graph_features">generate_graph_features</h3>
<blockquote class="blockquote">
<pre><code> generate_graph_features (glycan:str|networkx.classes.digraph.DiGraph,
                          glycan_graph:bool=True, label:str='network')</code></pre>
</blockquote>
<p><em>Compute graph features of glycan or network</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan sequence or network graph</td>
</tr>
<tr class="even">
<td>glycan_graph</td>
<td>bool</td>
<td>True</td>
<td>True if input is glycan, False if network</td>
</tr>
<tr class="odd">
<td>label</td>
<td>str</td>
<td>network</td>
<td>Label for output dataframe if glycan_graph=False</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>Dataframe of graph features</strong></td>
</tr>
</tbody>
</table>
<div id="drawn-malawi" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>generate_graph_features(<span class="st">"Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">diameter</th>
<th data-quarto-table-cell-role="th">branching</th>
<th data-quarto-table-cell-role="th">nbrLeaves</th>
<th data-quarto-table-cell-role="th">avgDeg</th>
<th data-quarto-table-cell-role="th">varDeg</th>
<th data-quarto-table-cell-role="th">maxDeg</th>
<th data-quarto-table-cell-role="th">nbrDeg4</th>
<th data-quarto-table-cell-role="th">max_deg_leaves</th>
<th data-quarto-table-cell-role="th">mean_deg_leaves</th>
<th data-quarto-table-cell-role="th">deg_assort</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">flow_edgeMax</th>
<th data-quarto-table-cell-role="th">flow_edgeMin</th>
<th data-quarto-table-cell-role="th">flow_edgeAvg</th>
<th data-quarto-table-cell-role="th">flow_edgeVar</th>
<th data-quarto-table-cell-role="th">secorderMax</th>
<th data-quarto-table-cell-role="th">secorderMin</th>
<th data-quarto-table-cell-role="th">secorderAvg</th>
<th data-quarto-table-cell-role="th">secorderVar</th>
<th data-quarto-table-cell-role="th">egap</th>
<th data-quarto-table-cell-role="th">entropyStation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>
<td>8</td>
<td>1</td>
<td>3</td>
<td>1.818182</td>
<td>0.330579</td>
<td>3.0</td>
<td>0</td>
<td>3.0</td>
<td>3.0</td>
<td>-1.850372e-15</td>
<td>...</td>
<td>0.333333</td>
<td>0.111111</td>
<td>0.217778</td>
<td>0.007289</td>
<td>45.607017</td>
<td>20.736441</td>
<td>31.679285</td>
<td>62.422895</td>
<td>0.340654</td>
<td>-2.180184</td>
</tr>
</tbody>
</table>

<p>1 rows × 49 columns</p>
</div>
</div>
</div>
<hr>
</section>
<section id="largest_subgraph" class="level3">
<h3 class="anchored" data-anchor-id="largest_subgraph">largest_subgraph</h3>
<blockquote class="blockquote">
<pre><code> largest_subgraph (glycan_a:str|networkx.classes.digraph.DiGraph,
                   glycan_b:str|networkx.classes.digraph.DiGraph)</code></pre>
</blockquote>
<p><em>Find the largest common subgraph of two glycans</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_a</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td>First glycan</td>
</tr>
<tr class="even">
<td>glycan_b</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td>Second glycan</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td><strong>Largest common subgraph in IUPAC format</strong></td>
</tr>
</tbody>
</table>
<div id="oriented-anime" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>glycan1 <span class="op">=</span> <span class="st">'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>glycan2 <span class="op">=</span> <span class="st">'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>largest_subgraph(glycan1, glycan2)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Fuc(a1-6)GlcNAc'</code></pre>
</div>
</div>
<hr>
</section>
<section id="ensure_graph" class="level3">
<h3 class="anchored" data-anchor-id="ensure_graph">ensure_graph</h3>
<blockquote class="blockquote">
<pre><code> ensure_graph (glycan:str|networkx.classes.digraph.DiGraph, **kwargs)</code></pre>
</blockquote>
<p><em>Ensures function compatibility with string glycans and graph glycans</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td>Glycan in IUPAC-condensed format or as networkx graph</td>
</tr>
<tr class="even">
<td>kwargs</td>
<td>VAR_KEYWORD</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DiGraph</strong></td>
<td><strong>NetworkX graph object of glycan</strong></td>
</tr>
</tbody>
</table>
<div id="thousand-syndrome" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>ensure_graph(<span class="st">"Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>&lt;networkx.classes.digraph.DiGraph&gt;</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_possible_topologies" class="level3">
<h3 class="anchored" data-anchor-id="get_possible_topologies">get_possible_topologies</h3>
<blockquote class="blockquote">
<pre><code> get_possible_topologies (glycan:str|networkx.classes.digraph.DiGraph,
                          exhaustive:bool=False,
                          allowed_disaccharides:set[str]|None=None,
                          modification_map:dict[str,set[str]]={'6S':
                          {'Gal', 'GlcNAc'}, '3S': {'Gal'}, '4S':
                          {'GalNAc'}, 'OS': {'Gal', 'GalNAc', 'GlcNAc'}},
                          return_graphs:bool=False)</code></pre>
</blockquote>
<p><em>Create possible glycan graphs given a floating substituent</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan with floating substituent</td>
</tr>
<tr class="even">
<td>exhaustive</td>
<td>bool</td>
<td>False</td>
<td>Whether to allow additions at internal positions</td>
</tr>
<tr class="odd">
<td>allowed_disaccharides</td>
<td>set[str] | None</td>
<td>None</td>
<td>Permitted disaccharides when creating possible glycans</td>
</tr>
<tr class="even">
<td>modification_map</td>
<td>dict</td>
<td>{‘6S’: {‘Gal’, ‘GlcNAc’}, ‘3S’: {‘Gal’}, ‘4S’: {‘GalNAc’}, ‘OS’: {‘Gal’, ‘GalNAc’, ‘GlcNAc’}}</td>
<td>Maps modifications to valid attachments</td>
</tr>
<tr class="odd">
<td>return_graphs</td>
<td>bool</td>
<td>False</td>
<td>Whether to return glycan graphs (otherwise return converted strings)</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of possible topology strings or graphs</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="possible_topology_check" class="level3">
<h3 class="anchored" data-anchor-id="possible_topology_check">possible_topology_check</h3>
<blockquote class="blockquote">
<pre><code> possible_topology_check (glycan:str|networkx.classes.digraph.DiGraph,
                          glycans:list[str|networkx.classes.digraph.DiGrap
                          h], exhaustive:bool=False, **kwargs)</code></pre>
</blockquote>
<p><em>Check whether glycan with floating substituent could match glycans from a list</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan with floating substituent</td>
</tr>
<tr class="even">
<td>glycans</td>
<td>list</td>
<td></td>
<td>List of glycans to check against</td>
</tr>
<tr class="odd">
<td>exhaustive</td>
<td>bool</td>
<td>False</td>
<td>Whether to allow additions at internal positions</td>
</tr>
<tr class="even">
<td>kwargs</td>
<td>VAR_KEYWORD</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of matching glycans</strong></td>
</tr>
</tbody>
</table>
<div id="original-australia" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>possible_topology_check(<span class="st">"{Neu5Ac(a2-3)}Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc"</span>,</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>                       [<span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>, <span class="st">"Neu5Ac(a2-3)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc"</span>,</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>                       <span class="st">"Neu5Ac(a2-6)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['Neu5Ac(a2-3)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc']</code></pre>
</div>
</div>
<hr>
</section>
<section id="deduplicate_glycans" class="level3">
<h3 class="anchored" data-anchor-id="deduplicate_glycans">deduplicate_glycans</h3>
<blockquote class="blockquote">
<pre><code> deduplicate_glycans (glycans:list[str]|set[str])</code></pre>
</blockquote>
<p><em>Remove duplicate glycans from a list/set, even if they have different strings</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycans</td>
<td>list[str] | set[str]</td>
<td>List/set of glycans to deduplicate</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td><strong>Deduplicated list of glycans</strong></td>
</tr>
</tbody>
</table>
<div id="1f7f48f9-5db1-4a0b-8a0c-d1d1a65b21b2" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>deduplicate_glycans([<span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>, <span class="st">"Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-3)]GalNAc"</span>,</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]GalNAc"</span>, <span class="st">"Neu5Ac(a2-6)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-3)]GalNAc',
 'Fuc(a1-2)Gal(b1-3)GalNAc',
 'Neu5Ac(a2-6)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc']</code></pre>
</div>
</div>
</section>
</section>
<section id="processing" class="level2">
<h2 class="anchored" data-anchor-id="processing">processing</h2>
<blockquote class="blockquote">
<p>process IUPAC-condensed glycan sequences into glycoletters etc.</p>
</blockquote>
<hr>
<section id="min_process_glycans" class="level3">
<h3 class="anchored" data-anchor-id="min_process_glycans">min_process_glycans</h3>
<blockquote class="blockquote">
<pre><code> min_process_glycans (glycan_list:list[str])</code></pre>
</blockquote>
<p><em>Convert list of glycans into a nested lists of glycoletters</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_list</td>
<td>list</td>
<td>List of glycans in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td><strong>List of glycoletter lists</strong></td>
</tr>
</tbody>
</table>
<div id="initial-armor" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>min_process_glycans([<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>[['Man', 'a1-3', 'Man', 'a1-6', 'Man', 'b1-4', 'GlcNAc', 'b1-4', 'GlcNAc'],
 ['Man',
  'a1-2',
  'Man',
  'a1-3',
  'Man',
  'a1-6',
  'Man',
  'b1-4',
  'GlcNAc',
  'b1-4',
  'GlcNAc']]</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_lib" class="level3">
<h3 class="anchored" data-anchor-id="get_lib">get_lib</h3>
<blockquote class="blockquote">
<pre><code> get_lib (glycan_list:list[str])</code></pre>
</blockquote>
<p><em>Returns dictionary mapping glycoletters to indices</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_list</td>
<td>list</td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>Dictionary of glycoletter:index mappings</strong></td>
</tr>
</tbody>
</table>
<div id="checked-picnic" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>get_lib([<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>{'GlcNAc': 0, 'Man': 1, 'a1-2': 2, 'a1-3': 3, 'a1-6': 4, 'b1-4': 5}</code></pre>
</div>
</div>
<hr>
</section>
<section id="expand_lib" class="level3">
<h3 class="anchored" data-anchor-id="expand_lib">expand_lib</h3>
<blockquote class="blockquote">
<pre><code> expand_lib (libr_in:dict[str,int], glycan_list:list[str])</code></pre>
</blockquote>
<p><em>Updates libr with newly introduced glycoletters</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>libr_in</td>
<td>dict</td>
<td>Existing dictionary of glycoletter:index</td>
</tr>
<tr class="even">
<td>glycan_list</td>
<td>list</td>
<td>List of IUPAC-condensed glycan sequences</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>Updated dictionary with new glycoletters</strong></td>
</tr>
</tbody>
</table>
<div id="gentle-acquisition" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>lib1 <span class="op">=</span> get_lib([<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>,</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc'</span>])</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>lib2 <span class="op">=</span> expand_lib(lib1, [<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>])</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>lib2</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>{'GlcNAc': 0, 'Man': 1, 'a1-2': 2, 'a1-3': 3, 'a1-6': 4, 'b1-4': 5, 'Fuc': 6}</code></pre>
</div>
</div>
<hr>
</section>
<section id="presence_to_matrix" class="level3">
<h3 class="anchored" data-anchor-id="presence_to_matrix">presence_to_matrix</h3>
<blockquote class="blockquote">
<pre><code> presence_to_matrix (df:pandas.core.frame.DataFrame,
                     glycan_col_name:str='glycan',
                     label_col_name:str='Species')</code></pre>
</blockquote>
<p><em>Converts a dataframe with glycan occurrence to absence/presence matrix</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>DataFrame</td>
<td></td>
<td>DataFrame with glycan occurrence</td>
</tr>
<tr class="even">
<td>glycan_col_name</td>
<td>str</td>
<td>glycan</td>
<td>Column name for glycans</td>
</tr>
<tr class="odd">
<td>label_col_name</td>
<td>str</td>
<td>Species</td>
<td>Column name for labels</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>Matrix with labels as rows and glycan occurrences as columns</strong></td>
</tr>
</tbody>
</table>
<div id="specified-television" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> presence_to_matrix(df_species[df_species.Order <span class="op">==</span> <span class="st">'Fabales'</span>].reset_index(drop <span class="op">=</span> <span class="va">True</span>),</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>                         label_col_name <span class="op">=</span> <span class="st">'Family'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="turkish-badge" class="cell">
<div class="cell-output cell-output-display">
<style type="text/css">
#T_0aa46_row0_col0, #T_0aa46_row0_col1, #T_0aa46_row0_col2, #T_0aa46_row0_col3, #T_0aa46_row0_col4, #T_0aa46_row0_col5, #T_0aa46_row0_col6, #T_0aa46_row0_col7, #T_0aa46_row0_col8, #T_0aa46_row0_col9, #T_0aa46_row0_col10, #T_0aa46_row0_col11, #T_0aa46_row0_col12, #T_0aa46_row0_col13, #T_0aa46_row0_col14, #T_0aa46_row0_col15, #T_0aa46_row0_col16, #T_0aa46_row0_col17, #T_0aa46_row0_col18, #T_0aa46_row0_col19, #T_0aa46_row0_col20, #T_0aa46_row0_col21, #T_0aa46_row0_col22, #T_0aa46_row0_col23, #T_0aa46_row0_col24, #T_0aa46_row0_col25, #T_0aa46_row0_col26, #T_0aa46_row0_col27, #T_0aa46_row0_col28, #T_0aa46_row0_col29, #T_0aa46_row0_col30, #T_0aa46_row0_col31, #T_0aa46_row0_col32, #T_0aa46_row0_col33, #T_0aa46_row0_col34, #T_0aa46_row0_col35, #T_0aa46_row0_col36, #T_0aa46_row0_col37, #T_0aa46_row0_col38, #T_0aa46_row0_col39, #T_0aa46_row0_col40, #T_0aa46_row0_col41, #T_0aa46_row0_col42, #T_0aa46_row0_col43, #T_0aa46_row0_col44, #T_0aa46_row0_col45, #T_0aa46_row0_col46, #T_0aa46_row0_col47, #T_0aa46_row0_col48, #T_0aa46_row0_col49, #T_0aa46_row0_col50, #T_0aa46_row0_col51, #T_0aa46_row0_col52, #T_0aa46_row0_col53, #T_0aa46_row0_col54, #T_0aa46_row0_col55, #T_0aa46_row0_col56, #T_0aa46_row0_col57, #T_0aa46_row0_col58, #T_0aa46_row0_col59, #T_0aa46_row0_col60, #T_0aa46_row0_col61, #T_0aa46_row0_col62, #T_0aa46_row0_col63, #T_0aa46_row0_col64, #T_0aa46_row0_col65, #T_0aa46_row0_col66, #T_0aa46_row0_col67, #T_0aa46_row0_col68, #T_0aa46_row0_col69, #T_0aa46_row0_col70, #T_0aa46_row0_col71, #T_0aa46_row0_col72, #T_0aa46_row0_col73, #T_0aa46_row0_col74, #T_0aa46_row0_col75, #T_0aa46_row0_col76, #T_0aa46_row0_col77, #T_0aa46_row0_col78, #T_0aa46_row0_col79, #T_0aa46_row0_col80, #T_0aa46_row0_col81, #T_0aa46_row0_col82, #T_0aa46_row0_col83, #T_0aa46_row0_col84, #T_0aa46_row0_col85, #T_0aa46_row0_col86, #T_0aa46_row0_col87, #T_0aa46_row0_col88, #T_0aa46_row0_col89, #T_0aa46_row0_col90, #T_0aa46_row0_col91, #T_0aa46_row0_col92, #T_0aa46_row0_col93, #T_0aa46_row0_col94, #T_0aa46_row0_col95, #T_0aa46_row0_col96, #T_0aa46_row0_col97, #T_0aa46_row0_col98, #T_0aa46_row0_col99, #T_0aa46_row0_col100, #T_0aa46_row0_col101, #T_0aa46_row0_col102, #T_0aa46_row0_col103, #T_0aa46_row0_col104, #T_0aa46_row0_col105, #T_0aa46_row0_col106, #T_0aa46_row0_col107, #T_0aa46_row0_col108, #T_0aa46_row0_col109, #T_0aa46_row0_col110, #T_0aa46_row0_col111, #T_0aa46_row0_col112, #T_0aa46_row0_col113, #T_0aa46_row0_col114, #T_0aa46_row0_col115, #T_0aa46_row0_col116, #T_0aa46_row0_col117, #T_0aa46_row0_col118, #T_0aa46_row0_col119, #T_0aa46_row0_col120, #T_0aa46_row0_col121, #T_0aa46_row0_col122, #T_0aa46_row0_col123, #T_0aa46_row0_col124, #T_0aa46_row0_col125, #T_0aa46_row0_col126, #T_0aa46_row0_col127, #T_0aa46_row0_col128, #T_0aa46_row0_col129, #T_0aa46_row0_col130, #T_0aa46_row0_col131, #T_0aa46_row0_col132, #T_0aa46_row0_col133, #T_0aa46_row0_col134, #T_0aa46_row0_col135, #T_0aa46_row0_col136, #T_0aa46_row0_col137, #T_0aa46_row0_col138, #T_0aa46_row0_col139, #T_0aa46_row0_col140, #T_0aa46_row0_col141, #T_0aa46_row0_col142, #T_0aa46_row0_col143, #T_0aa46_row0_col144, #T_0aa46_row0_col145, #T_0aa46_row0_col146, #T_0aa46_row0_col147, #T_0aa46_row0_col148, #T_0aa46_row0_col149, #T_0aa46_row0_col150, #T_0aa46_row0_col151, #T_0aa46_row0_col152, #T_0aa46_row0_col153, #T_0aa46_row0_col154, #T_0aa46_row0_col155, #T_0aa46_row0_col156, #T_0aa46_row0_col157, #T_0aa46_row0_col158, #T_0aa46_row0_col159, #T_0aa46_row0_col160, #T_0aa46_row0_col161, #T_0aa46_row0_col162, #T_0aa46_row0_col163, #T_0aa46_row0_col164, #T_0aa46_row0_col165, #T_0aa46_row0_col166, #T_0aa46_row0_col167, #T_0aa46_row0_col168, #T_0aa46_row0_col169, #T_0aa46_row0_col170, #T_0aa46_row0_col171, #T_0aa46_row0_col172, #T_0aa46_row0_col173, #T_0aa46_row0_col174, #T_0aa46_row0_col175, #T_0aa46_row0_col176, #T_0aa46_row0_col177, #T_0aa46_row0_col178, #T_0aa46_row0_col179, #T_0aa46_row0_col180, #T_0aa46_row0_col181, #T_0aa46_row0_col182, #T_0aa46_row0_col183, #T_0aa46_row0_col184, #T_0aa46_row0_col185, #T_0aa46_row0_col186, #T_0aa46_row0_col187, #T_0aa46_row0_col188, #T_0aa46_row0_col189, #T_0aa46_row0_col190, #T_0aa46_row0_col191, #T_0aa46_row0_col192, #T_0aa46_row0_col193, #T_0aa46_row0_col194, #T_0aa46_row0_col195, #T_0aa46_row0_col196, #T_0aa46_row0_col197, #T_0aa46_row0_col198, #T_0aa46_row0_col199, #T_0aa46_row0_col200, #T_0aa46_row0_col201, #T_0aa46_row0_col202, #T_0aa46_row0_col203, #T_0aa46_row0_col204, #T_0aa46_row0_col205, #T_0aa46_row0_col206, #T_0aa46_row0_col207, #T_0aa46_row0_col208, #T_0aa46_row0_col209, #T_0aa46_row0_col210, #T_0aa46_row0_col211, #T_0aa46_row0_col212, #T_0aa46_row0_col213, #T_0aa46_row0_col214, #T_0aa46_row0_col215, #T_0aa46_row0_col216, #T_0aa46_row0_col217, #T_0aa46_row0_col218, #T_0aa46_row0_col219, #T_0aa46_row0_col220, #T_0aa46_row0_col221, #T_0aa46_row0_col222, #T_0aa46_row0_col223, #T_0aa46_row0_col224, #T_0aa46_row0_col225, #T_0aa46_row0_col226, #T_0aa46_row0_col227, #T_0aa46_row0_col228, #T_0aa46_row0_col229, #T_0aa46_row0_col230, #T_0aa46_row0_col231, #T_0aa46_row0_col232, #T_0aa46_row0_col233, #T_0aa46_row0_col234, #T_0aa46_row0_col235, #T_0aa46_row0_col236, #T_0aa46_row0_col237, #T_0aa46_row0_col238, #T_0aa46_row0_col239, #T_0aa46_row0_col240, #T_0aa46_row0_col241, #T_0aa46_row0_col242, #T_0aa46_row0_col243, #T_0aa46_row0_col244, #T_0aa46_row0_col245, #T_0aa46_row0_col246, #T_0aa46_row0_col247, #T_0aa46_row0_col248, #T_0aa46_row0_col249, #T_0aa46_row0_col250, #T_0aa46_row0_col251, #T_0aa46_row0_col252, #T_0aa46_row0_col253, #T_0aa46_row0_col254, #T_0aa46_row0_col255, #T_0aa46_row0_col256, #T_0aa46_row0_col257, #T_0aa46_row0_col258, #T_0aa46_row0_col259, #T_0aa46_row0_col260, #T_0aa46_row0_col261, #T_0aa46_row0_col262, #T_0aa46_row0_col263, #T_0aa46_row0_col264, #T_0aa46_row0_col265, #T_0aa46_row0_col266, #T_0aa46_row0_col267, #T_0aa46_row0_col268, #T_0aa46_row0_col269, #T_0aa46_row0_col270, #T_0aa46_row0_col271, #T_0aa46_row0_col272, #T_0aa46_row0_col273, #T_0aa46_row0_col274, #T_0aa46_row0_col275, #T_0aa46_row0_col276, #T_0aa46_row0_col277, #T_0aa46_row0_col278, #T_0aa46_row0_col279, #T_0aa46_row0_col280, #T_0aa46_row0_col281, #T_0aa46_row0_col282, #T_0aa46_row0_col283, #T_0aa46_row0_col284, #T_0aa46_row0_col285, #T_0aa46_row0_col286, #T_0aa46_row0_col287, #T_0aa46_row0_col288, #T_0aa46_row0_col289, #T_0aa46_row0_col290, #T_0aa46_row0_col291, #T_0aa46_row0_col292, #T_0aa46_row0_col293, #T_0aa46_row0_col294, #T_0aa46_row0_col295, #T_0aa46_row0_col296, #T_0aa46_row0_col297, #T_0aa46_row0_col298, #T_0aa46_row0_col299, #T_0aa46_row0_col300, #T_0aa46_row0_col301, #T_0aa46_row0_col302, #T_0aa46_row0_col303, #T_0aa46_row0_col304, #T_0aa46_row0_col305, #T_0aa46_row0_col306, #T_0aa46_row0_col307, #T_0aa46_row0_col308, #T_0aa46_row0_col309, #T_0aa46_row0_col310, #T_0aa46_row0_col311, #T_0aa46_row0_col312, #T_0aa46_row0_col313, #T_0aa46_row0_col314, #T_0aa46_row0_col315, #T_0aa46_row0_col316, #T_0aa46_row0_col317, #T_0aa46_row0_col318, #T_0aa46_row0_col319, #T_0aa46_row0_col320, #T_0aa46_row0_col321, #T_0aa46_row0_col322, #T_0aa46_row0_col323, #T_0aa46_row0_col324, #T_0aa46_row0_col325, #T_0aa46_row0_col326, #T_0aa46_row0_col327, #T_0aa46_row0_col328, #T_0aa46_row0_col329, #T_0aa46_row0_col330, #T_0aa46_row0_col331, #T_0aa46_row0_col332, #T_0aa46_row0_col333, #T_0aa46_row0_col334, #T_0aa46_row0_col335, #T_0aa46_row0_col336, #T_0aa46_row0_col337, #T_0aa46_row0_col338, #T_0aa46_row0_col339, #T_0aa46_row0_col340, #T_0aa46_row0_col341, #T_0aa46_row0_col342, #T_0aa46_row0_col343, #T_0aa46_row0_col344, #T_0aa46_row0_col345, #T_0aa46_row0_col346, #T_0aa46_row0_col347, #T_0aa46_row0_col348, #T_0aa46_row0_col349, #T_0aa46_row0_col350, #T_0aa46_row0_col351, #T_0aa46_row0_col352, #T_0aa46_row0_col353, #T_0aa46_row0_col354, #T_0aa46_row0_col355, #T_0aa46_row0_col356, #T_0aa46_row0_col357, #T_0aa46_row0_col358, #T_0aa46_row0_col359, #T_0aa46_row0_col360, #T_0aa46_row0_col361, #T_0aa46_row0_col362, #T_0aa46_row0_col363, #T_0aa46_row0_col364, #T_0aa46_row0_col365, #T_0aa46_row0_col366, #T_0aa46_row0_col367, #T_0aa46_row0_col368, #T_0aa46_row0_col369, #T_0aa46_row0_col370, #T_0aa46_row0_col371, #T_0aa46_row0_col372, #T_0aa46_row0_col373, #T_0aa46_row0_col374, #T_0aa46_row0_col375, #T_0aa46_row0_col376, #T_0aa46_row0_col377, #T_0aa46_row0_col378, #T_0aa46_row0_col379, #T_0aa46_row0_col380, #T_0aa46_row0_col381, #T_0aa46_row0_col382, #T_0aa46_row0_col383, #T_0aa46_row0_col384, #T_0aa46_row0_col385, #T_0aa46_row0_col386, #T_0aa46_row0_col387, #T_0aa46_row0_col388, #T_0aa46_row0_col389, #T_0aa46_row0_col390, #T_0aa46_row0_col391, #T_0aa46_row0_col392, #T_0aa46_row0_col393, #T_0aa46_row0_col394, #T_0aa46_row0_col395, #T_0aa46_row0_col396, #T_0aa46_row0_col397, #T_0aa46_row0_col398, #T_0aa46_row0_col399, #T_0aa46_row0_col400, #T_0aa46_row0_col401, #T_0aa46_row0_col402, #T_0aa46_row0_col403, #T_0aa46_row0_col404, #T_0aa46_row0_col405, #T_0aa46_row0_col406, #T_0aa46_row0_col407, #T_0aa46_row0_col408, #T_0aa46_row0_col409, #T_0aa46_row0_col410, #T_0aa46_row0_col411, #T_0aa46_row0_col412, #T_0aa46_row0_col413, #T_0aa46_row0_col414, #T_0aa46_row0_col415, #T_0aa46_row0_col416, #T_0aa46_row0_col417, #T_0aa46_row0_col418, #T_0aa46_row0_col419, #T_0aa46_row0_col420, #T_0aa46_row0_col421, #T_0aa46_row0_col422, #T_0aa46_row0_col423, #T_0aa46_row0_col424, #T_0aa46_row0_col425, #T_0aa46_row0_col426, #T_0aa46_row0_col427, #T_0aa46_row0_col428, #T_0aa46_row0_col429, #T_0aa46_row0_col430, #T_0aa46_row0_col431, #T_0aa46_row0_col432, #T_0aa46_row0_col433, #T_0aa46_row0_col434, #T_0aa46_row0_col435, #T_0aa46_row0_col436, #T_0aa46_row0_col437, #T_0aa46_row0_col438, #T_0aa46_row0_col439, #T_0aa46_row0_col440, #T_0aa46_row0_col441, #T_0aa46_row0_col442, #T_0aa46_row0_col443, #T_0aa46_row0_col444, #T_0aa46_row0_col445, #T_0aa46_row0_col446, #T_0aa46_row0_col447, #T_0aa46_row0_col448, #T_0aa46_row0_col449, #T_0aa46_row0_col450, #T_0aa46_row0_col451, #T_0aa46_row0_col452, #T_0aa46_row0_col453, #T_0aa46_row0_col454, #T_0aa46_row0_col455, #T_0aa46_row0_col456, #T_0aa46_row0_col457, #T_0aa46_row0_col458, #T_0aa46_row0_col459, #T_0aa46_row0_col460, #T_0aa46_row0_col461, #T_0aa46_row0_col462, #T_0aa46_row0_col463, #T_0aa46_row0_col464, #T_0aa46_row0_col465, #T_0aa46_row0_col466, #T_0aa46_row0_col467, #T_0aa46_row0_col468, #T_0aa46_row0_col469, #T_0aa46_row0_col470, #T_0aa46_row0_col471, #T_0aa46_row0_col472, #T_0aa46_row0_col473, #T_0aa46_row0_col474, #T_0aa46_row0_col475, #T_0aa46_row0_col476, #T_0aa46_row0_col477, #T_0aa46_row0_col478, #T_0aa46_row0_col479, #T_0aa46_row0_col480, #T_0aa46_row0_col481, #T_0aa46_row0_col482, #T_0aa46_row0_col483, #T_0aa46_row1_col0, #T_0aa46_row1_col1, #T_0aa46_row1_col2, #T_0aa46_row1_col3, #T_0aa46_row1_col4, #T_0aa46_row1_col5, #T_0aa46_row1_col6, #T_0aa46_row1_col7, #T_0aa46_row1_col8, #T_0aa46_row1_col9, #T_0aa46_row1_col10, #T_0aa46_row1_col11, #T_0aa46_row1_col12, #T_0aa46_row1_col13, #T_0aa46_row1_col14, #T_0aa46_row1_col15, #T_0aa46_row1_col16, #T_0aa46_row1_col17, #T_0aa46_row1_col18, #T_0aa46_row1_col19, #T_0aa46_row1_col20, #T_0aa46_row1_col21, #T_0aa46_row1_col22, #T_0aa46_row1_col23, #T_0aa46_row1_col24, #T_0aa46_row1_col25, #T_0aa46_row1_col26, #T_0aa46_row1_col27, #T_0aa46_row1_col28, #T_0aa46_row1_col29, #T_0aa46_row1_col30, #T_0aa46_row1_col31, #T_0aa46_row1_col32, #T_0aa46_row1_col33, #T_0aa46_row1_col34, #T_0aa46_row1_col35, #T_0aa46_row1_col36, #T_0aa46_row1_col37, #T_0aa46_row1_col38, #T_0aa46_row1_col39, #T_0aa46_row1_col40, #T_0aa46_row1_col41, #T_0aa46_row1_col42, #T_0aa46_row1_col43, #T_0aa46_row1_col44, #T_0aa46_row1_col45, #T_0aa46_row1_col46, #T_0aa46_row1_col47, #T_0aa46_row1_col48, #T_0aa46_row1_col49, #T_0aa46_row1_col50, #T_0aa46_row1_col51, #T_0aa46_row1_col52, #T_0aa46_row1_col53, #T_0aa46_row1_col54, #T_0aa46_row1_col55, #T_0aa46_row1_col56, #T_0aa46_row1_col57, #T_0aa46_row1_col58, #T_0aa46_row1_col59, #T_0aa46_row1_col60, #T_0aa46_row1_col61, #T_0aa46_row1_col62, #T_0aa46_row1_col63, #T_0aa46_row1_col64, #T_0aa46_row1_col65, #T_0aa46_row1_col66, #T_0aa46_row1_col67, #T_0aa46_row1_col68, #T_0aa46_row1_col69, #T_0aa46_row1_col70, #T_0aa46_row1_col71, #T_0aa46_row1_col72, #T_0aa46_row1_col73, #T_0aa46_row1_col74, #T_0aa46_row1_col75, #T_0aa46_row1_col76, #T_0aa46_row1_col77, #T_0aa46_row1_col78, #T_0aa46_row1_col79, #T_0aa46_row1_col80, #T_0aa46_row1_col81, #T_0aa46_row1_col82, #T_0aa46_row1_col83, #T_0aa46_row1_col84, #T_0aa46_row1_col85, #T_0aa46_row1_col86, #T_0aa46_row1_col87, #T_0aa46_row1_col88, #T_0aa46_row1_col89, #T_0aa46_row1_col90, #T_0aa46_row1_col91, #T_0aa46_row1_col92, #T_0aa46_row1_col93, #T_0aa46_row1_col94, #T_0aa46_row1_col95, #T_0aa46_row1_col96, #T_0aa46_row1_col97, #T_0aa46_row1_col98, #T_0aa46_row1_col99, #T_0aa46_row1_col100, #T_0aa46_row1_col101, #T_0aa46_row1_col102, #T_0aa46_row1_col103, #T_0aa46_row1_col104, #T_0aa46_row1_col105, #T_0aa46_row1_col106, #T_0aa46_row1_col107, #T_0aa46_row1_col108, #T_0aa46_row1_col109, #T_0aa46_row1_col110, #T_0aa46_row1_col111, #T_0aa46_row1_col112, #T_0aa46_row1_col113, #T_0aa46_row1_col114, #T_0aa46_row1_col115, #T_0aa46_row1_col116, #T_0aa46_row1_col117, #T_0aa46_row1_col118, #T_0aa46_row1_col119, #T_0aa46_row1_col120, #T_0aa46_row1_col121, #T_0aa46_row1_col122, #T_0aa46_row1_col123, #T_0aa46_row1_col124, #T_0aa46_row1_col125, #T_0aa46_row1_col126, #T_0aa46_row1_col127, #T_0aa46_row1_col128, #T_0aa46_row1_col129, #T_0aa46_row1_col130, #T_0aa46_row1_col131, #T_0aa46_row1_col132, #T_0aa46_row1_col133, #T_0aa46_row1_col134, #T_0aa46_row1_col135, #T_0aa46_row1_col136, #T_0aa46_row1_col137, #T_0aa46_row1_col138, #T_0aa46_row1_col139, #T_0aa46_row1_col140, #T_0aa46_row1_col141, #T_0aa46_row1_col142, #T_0aa46_row1_col143, #T_0aa46_row1_col144, #T_0aa46_row1_col145, #T_0aa46_row1_col146, #T_0aa46_row1_col147, #T_0aa46_row1_col148, #T_0aa46_row1_col149, #T_0aa46_row1_col150, #T_0aa46_row1_col151, #T_0aa46_row1_col152, #T_0aa46_row1_col153, #T_0aa46_row1_col154, #T_0aa46_row1_col155, #T_0aa46_row1_col156, #T_0aa46_row1_col157, #T_0aa46_row1_col158, #T_0aa46_row1_col159, #T_0aa46_row1_col160, #T_0aa46_row1_col161, #T_0aa46_row1_col162, #T_0aa46_row1_col163, #T_0aa46_row1_col164, #T_0aa46_row1_col165, #T_0aa46_row1_col166, #T_0aa46_row1_col167, #T_0aa46_row1_col168, #T_0aa46_row1_col169, #T_0aa46_row1_col170, #T_0aa46_row1_col171, #T_0aa46_row1_col172, #T_0aa46_row1_col173, #T_0aa46_row1_col174, #T_0aa46_row1_col175, #T_0aa46_row1_col176, #T_0aa46_row1_col177, #T_0aa46_row1_col178, #T_0aa46_row1_col179, #T_0aa46_row1_col180, #T_0aa46_row1_col181, #T_0aa46_row1_col182, #T_0aa46_row1_col183, #T_0aa46_row1_col184, #T_0aa46_row1_col185, #T_0aa46_row1_col186, #T_0aa46_row1_col187, #T_0aa46_row1_col188, #T_0aa46_row1_col189, #T_0aa46_row1_col190, #T_0aa46_row1_col191, #T_0aa46_row1_col192, #T_0aa46_row1_col193, #T_0aa46_row1_col194, #T_0aa46_row1_col195, #T_0aa46_row1_col196, #T_0aa46_row1_col197, #T_0aa46_row1_col198, #T_0aa46_row1_col199, #T_0aa46_row1_col200, #T_0aa46_row1_col201, #T_0aa46_row1_col202, #T_0aa46_row1_col203, #T_0aa46_row1_col204, #T_0aa46_row1_col205, #T_0aa46_row1_col206, #T_0aa46_row1_col207, #T_0aa46_row1_col208, #T_0aa46_row1_col209, #T_0aa46_row1_col210, #T_0aa46_row1_col211, #T_0aa46_row1_col212, #T_0aa46_row1_col213, #T_0aa46_row1_col214, #T_0aa46_row1_col215, #T_0aa46_row1_col216, #T_0aa46_row1_col217, #T_0aa46_row1_col218, #T_0aa46_row1_col219, #T_0aa46_row1_col220, #T_0aa46_row1_col221, #T_0aa46_row1_col222, #T_0aa46_row1_col223, #T_0aa46_row1_col224, #T_0aa46_row1_col225, #T_0aa46_row1_col226, #T_0aa46_row1_col227, #T_0aa46_row1_col228, #T_0aa46_row1_col229, #T_0aa46_row1_col230, #T_0aa46_row1_col231, #T_0aa46_row1_col232, #T_0aa46_row1_col233, #T_0aa46_row1_col234, #T_0aa46_row1_col235, #T_0aa46_row1_col236, #T_0aa46_row1_col237, #T_0aa46_row1_col238, #T_0aa46_row1_col239, #T_0aa46_row1_col240, #T_0aa46_row1_col241, #T_0aa46_row1_col242, #T_0aa46_row1_col243, #T_0aa46_row1_col244, #T_0aa46_row1_col245, #T_0aa46_row1_col246, #T_0aa46_row1_col247, #T_0aa46_row1_col248, #T_0aa46_row1_col249, #T_0aa46_row1_col250, #T_0aa46_row1_col251, #T_0aa46_row1_col252, #T_0aa46_row1_col253, #T_0aa46_row1_col254, #T_0aa46_row1_col255, #T_0aa46_row1_col256, #T_0aa46_row1_col257, #T_0aa46_row1_col258, #T_0aa46_row1_col259, #T_0aa46_row1_col260, #T_0aa46_row1_col261, #T_0aa46_row1_col262, #T_0aa46_row1_col263, #T_0aa46_row1_col264, #T_0aa46_row1_col265, #T_0aa46_row1_col266, #T_0aa46_row1_col267, #T_0aa46_row1_col268, #T_0aa46_row1_col269, #T_0aa46_row1_col270, #T_0aa46_row1_col271, #T_0aa46_row1_col272, #T_0aa46_row1_col273, #T_0aa46_row1_col274, #T_0aa46_row1_col275, #T_0aa46_row1_col276, #T_0aa46_row1_col277, #T_0aa46_row1_col278, #T_0aa46_row1_col279, #T_0aa46_row1_col280, #T_0aa46_row1_col281, #T_0aa46_row1_col282, #T_0aa46_row1_col283, #T_0aa46_row1_col284, #T_0aa46_row1_col285, #T_0aa46_row1_col286, #T_0aa46_row1_col287, #T_0aa46_row1_col288, #T_0aa46_row1_col289, #T_0aa46_row1_col290, #T_0aa46_row1_col291, #T_0aa46_row1_col292, #T_0aa46_row1_col293, #T_0aa46_row1_col294, #T_0aa46_row1_col295, #T_0aa46_row1_col296, #T_0aa46_row1_col297, #T_0aa46_row1_col298, #T_0aa46_row1_col299, #T_0aa46_row1_col300, #T_0aa46_row1_col301, #T_0aa46_row1_col302, #T_0aa46_row1_col303, #T_0aa46_row1_col304, #T_0aa46_row1_col305, #T_0aa46_row1_col306, #T_0aa46_row1_col307, #T_0aa46_row1_col308, #T_0aa46_row1_col309, #T_0aa46_row1_col310, #T_0aa46_row1_col311, #T_0aa46_row1_col312, #T_0aa46_row1_col313, #T_0aa46_row1_col314, #T_0aa46_row1_col315, #T_0aa46_row1_col316, #T_0aa46_row1_col317, #T_0aa46_row1_col318, #T_0aa46_row1_col319, #T_0aa46_row1_col320, #T_0aa46_row1_col321, #T_0aa46_row1_col322, #T_0aa46_row1_col323, #T_0aa46_row1_col324, #T_0aa46_row1_col325, #T_0aa46_row1_col326, #T_0aa46_row1_col327, #T_0aa46_row1_col328, #T_0aa46_row1_col329, #T_0aa46_row1_col330, #T_0aa46_row1_col331, #T_0aa46_row1_col332, #T_0aa46_row1_col333, #T_0aa46_row1_col334, #T_0aa46_row1_col335, #T_0aa46_row1_col336, #T_0aa46_row1_col337, #T_0aa46_row1_col338, #T_0aa46_row1_col339, #T_0aa46_row1_col340, #T_0aa46_row1_col341, #T_0aa46_row1_col342, #T_0aa46_row1_col343, #T_0aa46_row1_col344, #T_0aa46_row1_col345, #T_0aa46_row1_col346, #T_0aa46_row1_col347, #T_0aa46_row1_col348, #T_0aa46_row1_col349, #T_0aa46_row1_col350, #T_0aa46_row1_col351, #T_0aa46_row1_col352, #T_0aa46_row1_col353, #T_0aa46_row1_col354, #T_0aa46_row1_col355, #T_0aa46_row1_col356, #T_0aa46_row1_col357, #T_0aa46_row1_col358, #T_0aa46_row1_col359, #T_0aa46_row1_col360, #T_0aa46_row1_col361, #T_0aa46_row1_col362, #T_0aa46_row1_col363, #T_0aa46_row1_col364, #T_0aa46_row1_col365, #T_0aa46_row1_col366, #T_0aa46_row1_col367, #T_0aa46_row1_col368, #T_0aa46_row1_col369, #T_0aa46_row1_col370, #T_0aa46_row1_col371, #T_0aa46_row1_col372, #T_0aa46_row1_col373, #T_0aa46_row1_col374, #T_0aa46_row1_col375, #T_0aa46_row1_col376, #T_0aa46_row1_col377, #T_0aa46_row1_col378, #T_0aa46_row1_col379, #T_0aa46_row1_col380, #T_0aa46_row1_col381, #T_0aa46_row1_col382, #T_0aa46_row1_col383, #T_0aa46_row1_col384, #T_0aa46_row1_col385, #T_0aa46_row1_col386, #T_0aa46_row1_col387, #T_0aa46_row1_col388, #T_0aa46_row1_col389, #T_0aa46_row1_col390, #T_0aa46_row1_col391, #T_0aa46_row1_col392, #T_0aa46_row1_col393, #T_0aa46_row1_col394, #T_0aa46_row1_col395, #T_0aa46_row1_col396, #T_0aa46_row1_col397, #T_0aa46_row1_col398, #T_0aa46_row1_col399, #T_0aa46_row1_col400, #T_0aa46_row1_col401, #T_0aa46_row1_col402, #T_0aa46_row1_col403, #T_0aa46_row1_col404, #T_0aa46_row1_col405, #T_0aa46_row1_col406, #T_0aa46_row1_col407, #T_0aa46_row1_col408, #T_0aa46_row1_col409, #T_0aa46_row1_col410, #T_0aa46_row1_col411, #T_0aa46_row1_col412, #T_0aa46_row1_col413, #T_0aa46_row1_col414, #T_0aa46_row1_col415, #T_0aa46_row1_col416, #T_0aa46_row1_col417, #T_0aa46_row1_col418, #T_0aa46_row1_col419, #T_0aa46_row1_col420, #T_0aa46_row1_col421, #T_0aa46_row1_col422, #T_0aa46_row1_col423, #T_0aa46_row1_col424, #T_0aa46_row1_col425, #T_0aa46_row1_col426, #T_0aa46_row1_col427, #T_0aa46_row1_col428, #T_0aa46_row1_col429, #T_0aa46_row1_col430, #T_0aa46_row1_col431, #T_0aa46_row1_col432, #T_0aa46_row1_col433, #T_0aa46_row1_col434, #T_0aa46_row1_col435, #T_0aa46_row1_col436, #T_0aa46_row1_col437, #T_0aa46_row1_col438, #T_0aa46_row1_col439, #T_0aa46_row1_col440, #T_0aa46_row1_col441, #T_0aa46_row1_col442, #T_0aa46_row1_col443, #T_0aa46_row1_col444, #T_0aa46_row1_col445, #T_0aa46_row1_col446, #T_0aa46_row1_col447, #T_0aa46_row1_col448, #T_0aa46_row1_col449, #T_0aa46_row1_col450, #T_0aa46_row1_col451, #T_0aa46_row1_col452, #T_0aa46_row1_col453, #T_0aa46_row1_col454, #T_0aa46_row1_col455, #T_0aa46_row1_col456, #T_0aa46_row1_col457, #T_0aa46_row1_col458, #T_0aa46_row1_col459, #T_0aa46_row1_col460, #T_0aa46_row1_col461, #T_0aa46_row1_col462, #T_0aa46_row1_col463, #T_0aa46_row1_col464, #T_0aa46_row1_col465, #T_0aa46_row1_col466, #T_0aa46_row1_col467, #T_0aa46_row1_col468, #T_0aa46_row1_col469, #T_0aa46_row1_col470, #T_0aa46_row1_col471, #T_0aa46_row1_col472, #T_0aa46_row1_col473, #T_0aa46_row1_col474, #T_0aa46_row1_col475, #T_0aa46_row1_col476, #T_0aa46_row1_col477, #T_0aa46_row1_col478, #T_0aa46_row1_col479, #T_0aa46_row1_col480, #T_0aa46_row1_col481, #T_0aa46_row1_col482, #T_0aa46_row1_col483, #T_0aa46_row2_col0, #T_0aa46_row2_col1, #T_0aa46_row2_col2, #T_0aa46_row2_col3, #T_0aa46_row2_col4, #T_0aa46_row2_col5, #T_0aa46_row2_col6, #T_0aa46_row2_col7, #T_0aa46_row2_col8, #T_0aa46_row2_col9, #T_0aa46_row2_col10, #T_0aa46_row2_col11, #T_0aa46_row2_col12, #T_0aa46_row2_col13, #T_0aa46_row2_col14, #T_0aa46_row2_col15, #T_0aa46_row2_col16, #T_0aa46_row2_col17, #T_0aa46_row2_col18, #T_0aa46_row2_col19, #T_0aa46_row2_col20, #T_0aa46_row2_col21, #T_0aa46_row2_col22, #T_0aa46_row2_col23, #T_0aa46_row2_col24, #T_0aa46_row2_col25, #T_0aa46_row2_col26, #T_0aa46_row2_col27, #T_0aa46_row2_col28, #T_0aa46_row2_col29, #T_0aa46_row2_col30, #T_0aa46_row2_col31, #T_0aa46_row2_col32, #T_0aa46_row2_col33, #T_0aa46_row2_col34, #T_0aa46_row2_col35, #T_0aa46_row2_col36, #T_0aa46_row2_col37, #T_0aa46_row2_col38, #T_0aa46_row2_col39, #T_0aa46_row2_col40, #T_0aa46_row2_col41, #T_0aa46_row2_col42, #T_0aa46_row2_col43, #T_0aa46_row2_col44, #T_0aa46_row2_col45, #T_0aa46_row2_col46, #T_0aa46_row2_col47, #T_0aa46_row2_col48, #T_0aa46_row2_col49, #T_0aa46_row2_col50, #T_0aa46_row2_col51, #T_0aa46_row2_col52, #T_0aa46_row2_col53, #T_0aa46_row2_col54, #T_0aa46_row2_col55, #T_0aa46_row2_col56, #T_0aa46_row2_col57, #T_0aa46_row2_col58, #T_0aa46_row2_col59, #T_0aa46_row2_col60, #T_0aa46_row2_col61, #T_0aa46_row2_col62, #T_0aa46_row2_col63, #T_0aa46_row2_col64, #T_0aa46_row2_col65, #T_0aa46_row2_col66, #T_0aa46_row2_col67, #T_0aa46_row2_col68, #T_0aa46_row2_col69, #T_0aa46_row2_col70, #T_0aa46_row2_col71, #T_0aa46_row2_col72, #T_0aa46_row2_col73, #T_0aa46_row2_col74, #T_0aa46_row2_col75, #T_0aa46_row2_col76, #T_0aa46_row2_col77, #T_0aa46_row2_col78, #T_0aa46_row2_col79, #T_0aa46_row2_col80, #T_0aa46_row2_col81, #T_0aa46_row2_col82, #T_0aa46_row2_col83, #T_0aa46_row2_col84, #T_0aa46_row2_col85, #T_0aa46_row2_col86, #T_0aa46_row2_col87, #T_0aa46_row2_col88, #T_0aa46_row2_col89, #T_0aa46_row2_col90, #T_0aa46_row2_col91, #T_0aa46_row2_col92, #T_0aa46_row2_col93, #T_0aa46_row2_col94, #T_0aa46_row2_col95, #T_0aa46_row2_col96, #T_0aa46_row2_col97, #T_0aa46_row2_col98, #T_0aa46_row2_col99, #T_0aa46_row2_col100, #T_0aa46_row2_col101, #T_0aa46_row2_col102, #T_0aa46_row2_col103, #T_0aa46_row2_col104, #T_0aa46_row2_col105, #T_0aa46_row2_col106, #T_0aa46_row2_col107, #T_0aa46_row2_col108, #T_0aa46_row2_col109, #T_0aa46_row2_col110, #T_0aa46_row2_col111, #T_0aa46_row2_col112, #T_0aa46_row2_col113, #T_0aa46_row2_col114, #T_0aa46_row2_col115, #T_0aa46_row2_col116, #T_0aa46_row2_col117, #T_0aa46_row2_col118, #T_0aa46_row2_col119, #T_0aa46_row2_col120, #T_0aa46_row2_col121, #T_0aa46_row2_col122, #T_0aa46_row2_col123, #T_0aa46_row2_col124, #T_0aa46_row2_col125, #T_0aa46_row2_col126, #T_0aa46_row2_col127, #T_0aa46_row2_col128, #T_0aa46_row2_col129, #T_0aa46_row2_col130, #T_0aa46_row2_col131, #T_0aa46_row2_col132, #T_0aa46_row2_col133, #T_0aa46_row2_col134, #T_0aa46_row2_col135, #T_0aa46_row2_col136, #T_0aa46_row2_col137, #T_0aa46_row2_col138, #T_0aa46_row2_col139, #T_0aa46_row2_col140, #T_0aa46_row2_col141, #T_0aa46_row2_col142, #T_0aa46_row2_col143, #T_0aa46_row2_col144, #T_0aa46_row2_col145, #T_0aa46_row2_col146, #T_0aa46_row2_col147, #T_0aa46_row2_col148, #T_0aa46_row2_col149, #T_0aa46_row2_col150, #T_0aa46_row2_col151, #T_0aa46_row2_col152, #T_0aa46_row2_col153, #T_0aa46_row2_col154, #T_0aa46_row2_col155, #T_0aa46_row2_col156, #T_0aa46_row2_col157, #T_0aa46_row2_col158, #T_0aa46_row2_col159, #T_0aa46_row2_col160, #T_0aa46_row2_col161, #T_0aa46_row2_col162, #T_0aa46_row2_col163, #T_0aa46_row2_col164, #T_0aa46_row2_col165, #T_0aa46_row2_col166, #T_0aa46_row2_col167, #T_0aa46_row2_col168, #T_0aa46_row2_col169, #T_0aa46_row2_col170, #T_0aa46_row2_col171, #T_0aa46_row2_col172, #T_0aa46_row2_col173, #T_0aa46_row2_col174, #T_0aa46_row2_col175, #T_0aa46_row2_col176, #T_0aa46_row2_col177, #T_0aa46_row2_col178, #T_0aa46_row2_col179, #T_0aa46_row2_col180, #T_0aa46_row2_col181, #T_0aa46_row2_col182, #T_0aa46_row2_col183, #T_0aa46_row2_col184, #T_0aa46_row2_col185, #T_0aa46_row2_col186, #T_0aa46_row2_col187, #T_0aa46_row2_col188, #T_0aa46_row2_col189, #T_0aa46_row2_col190, #T_0aa46_row2_col191, #T_0aa46_row2_col192, #T_0aa46_row2_col193, #T_0aa46_row2_col194, #T_0aa46_row2_col195, #T_0aa46_row2_col196, #T_0aa46_row2_col197, #T_0aa46_row2_col198, #T_0aa46_row2_col199, #T_0aa46_row2_col200, #T_0aa46_row2_col201, #T_0aa46_row2_col202, #T_0aa46_row2_col203, #T_0aa46_row2_col204, #T_0aa46_row2_col205, #T_0aa46_row2_col206, #T_0aa46_row2_col207, #T_0aa46_row2_col208, #T_0aa46_row2_col209, #T_0aa46_row2_col210, #T_0aa46_row2_col211, #T_0aa46_row2_col212, #T_0aa46_row2_col213, #T_0aa46_row2_col214, #T_0aa46_row2_col215, #T_0aa46_row2_col216, #T_0aa46_row2_col217, #T_0aa46_row2_col218, #T_0aa46_row2_col219, #T_0aa46_row2_col220, #T_0aa46_row2_col221, #T_0aa46_row2_col222, #T_0aa46_row2_col223, #T_0aa46_row2_col224, #T_0aa46_row2_col225, #T_0aa46_row2_col226, #T_0aa46_row2_col227, #T_0aa46_row2_col228, #T_0aa46_row2_col229, #T_0aa46_row2_col230, #T_0aa46_row2_col231, #T_0aa46_row2_col232, #T_0aa46_row2_col233, #T_0aa46_row2_col234, #T_0aa46_row2_col235, #T_0aa46_row2_col236, #T_0aa46_row2_col237, #T_0aa46_row2_col238, #T_0aa46_row2_col239, #T_0aa46_row2_col240, #T_0aa46_row2_col241, #T_0aa46_row2_col242, #T_0aa46_row2_col243, #T_0aa46_row2_col244, #T_0aa46_row2_col245, #T_0aa46_row2_col246, #T_0aa46_row2_col247, #T_0aa46_row2_col248, #T_0aa46_row2_col249, #T_0aa46_row2_col250, #T_0aa46_row2_col251, #T_0aa46_row2_col252, #T_0aa46_row2_col253, #T_0aa46_row2_col254, #T_0aa46_row2_col255, #T_0aa46_row2_col256, #T_0aa46_row2_col257, #T_0aa46_row2_col258, #T_0aa46_row2_col259, #T_0aa46_row2_col260, #T_0aa46_row2_col261, #T_0aa46_row2_col262, #T_0aa46_row2_col263, #T_0aa46_row2_col264, #T_0aa46_row2_col265, #T_0aa46_row2_col266, #T_0aa46_row2_col267, #T_0aa46_row2_col268, #T_0aa46_row2_col269, #T_0aa46_row2_col270, #T_0aa46_row2_col271, #T_0aa46_row2_col272, #T_0aa46_row2_col273, #T_0aa46_row2_col274, #T_0aa46_row2_col275, #T_0aa46_row2_col276, #T_0aa46_row2_col277, #T_0aa46_row2_col278, #T_0aa46_row2_col279, #T_0aa46_row2_col280, #T_0aa46_row2_col281, #T_0aa46_row2_col282, #T_0aa46_row2_col283, #T_0aa46_row2_col284, #T_0aa46_row2_col285, #T_0aa46_row2_col286, #T_0aa46_row2_col287, #T_0aa46_row2_col288, #T_0aa46_row2_col289, #T_0aa46_row2_col290, #T_0aa46_row2_col291, #T_0aa46_row2_col292, #T_0aa46_row2_col293, #T_0aa46_row2_col294, #T_0aa46_row2_col295, #T_0aa46_row2_col296, #T_0aa46_row2_col297, #T_0aa46_row2_col298, #T_0aa46_row2_col299, #T_0aa46_row2_col300, #T_0aa46_row2_col301, #T_0aa46_row2_col302, #T_0aa46_row2_col303, #T_0aa46_row2_col304, #T_0aa46_row2_col305, #T_0aa46_row2_col306, #T_0aa46_row2_col307, #T_0aa46_row2_col308, #T_0aa46_row2_col309, #T_0aa46_row2_col310, #T_0aa46_row2_col311, #T_0aa46_row2_col312, #T_0aa46_row2_col313, #T_0aa46_row2_col314, #T_0aa46_row2_col315, #T_0aa46_row2_col316, #T_0aa46_row2_col317, #T_0aa46_row2_col318, #T_0aa46_row2_col319, #T_0aa46_row2_col320, #T_0aa46_row2_col321, #T_0aa46_row2_col322, #T_0aa46_row2_col323, #T_0aa46_row2_col324, #T_0aa46_row2_col325, #T_0aa46_row2_col326, #T_0aa46_row2_col327, #T_0aa46_row2_col328, #T_0aa46_row2_col329, #T_0aa46_row2_col330, #T_0aa46_row2_col331, #T_0aa46_row2_col332, #T_0aa46_row2_col333, #T_0aa46_row2_col334, #T_0aa46_row2_col335, #T_0aa46_row2_col336, #T_0aa46_row2_col337, #T_0aa46_row2_col338, #T_0aa46_row2_col339, #T_0aa46_row2_col340, #T_0aa46_row2_col341, #T_0aa46_row2_col342, #T_0aa46_row2_col343, #T_0aa46_row2_col344, #T_0aa46_row2_col345, #T_0aa46_row2_col346, #T_0aa46_row2_col347, #T_0aa46_row2_col348, #T_0aa46_row2_col349, #T_0aa46_row2_col350, #T_0aa46_row2_col351, #T_0aa46_row2_col352, #T_0aa46_row2_col353, #T_0aa46_row2_col354, #T_0aa46_row2_col355, #T_0aa46_row2_col356, #T_0aa46_row2_col357, #T_0aa46_row2_col358, #T_0aa46_row2_col359, #T_0aa46_row2_col360, #T_0aa46_row2_col361, #T_0aa46_row2_col362, #T_0aa46_row2_col363, #T_0aa46_row2_col364, #T_0aa46_row2_col365, #T_0aa46_row2_col366, #T_0aa46_row2_col367, #T_0aa46_row2_col368, #T_0aa46_row2_col369, #T_0aa46_row2_col370, #T_0aa46_row2_col371, #T_0aa46_row2_col372, #T_0aa46_row2_col373, #T_0aa46_row2_col374, #T_0aa46_row2_col375, #T_0aa46_row2_col376, #T_0aa46_row2_col377, #T_0aa46_row2_col378, #T_0aa46_row2_col379, #T_0aa46_row2_col380, #T_0aa46_row2_col381, #T_0aa46_row2_col382, #T_0aa46_row2_col383, #T_0aa46_row2_col384, #T_0aa46_row2_col385, #T_0aa46_row2_col386, #T_0aa46_row2_col387, #T_0aa46_row2_col388, #T_0aa46_row2_col389, #T_0aa46_row2_col390, #T_0aa46_row2_col391, #T_0aa46_row2_col392, #T_0aa46_row2_col393, #T_0aa46_row2_col394, #T_0aa46_row2_col395, #T_0aa46_row2_col396, #T_0aa46_row2_col397, #T_0aa46_row2_col398, #T_0aa46_row2_col399, #T_0aa46_row2_col400, #T_0aa46_row2_col401, #T_0aa46_row2_col402, #T_0aa46_row2_col403, #T_0aa46_row2_col404, #T_0aa46_row2_col405, #T_0aa46_row2_col406, #T_0aa46_row2_col407, #T_0aa46_row2_col408, #T_0aa46_row2_col409, #T_0aa46_row2_col410, #T_0aa46_row2_col411, #T_0aa46_row2_col412, #T_0aa46_row2_col413, #T_0aa46_row2_col414, #T_0aa46_row2_col415, #T_0aa46_row2_col416, #T_0aa46_row2_col417, #T_0aa46_row2_col418, #T_0aa46_row2_col419, #T_0aa46_row2_col420, #T_0aa46_row2_col421, #T_0aa46_row2_col422, #T_0aa46_row2_col423, #T_0aa46_row2_col424, #T_0aa46_row2_col425, #T_0aa46_row2_col426, #T_0aa46_row2_col427, #T_0aa46_row2_col428, #T_0aa46_row2_col429, #T_0aa46_row2_col430, #T_0aa46_row2_col431, #T_0aa46_row2_col432, #T_0aa46_row2_col433, #T_0aa46_row2_col434, #T_0aa46_row2_col435, #T_0aa46_row2_col436, #T_0aa46_row2_col437, #T_0aa46_row2_col438, #T_0aa46_row2_col439, #T_0aa46_row2_col440, #T_0aa46_row2_col441, #T_0aa46_row2_col442, #T_0aa46_row2_col443, #T_0aa46_row2_col444, #T_0aa46_row2_col445, #T_0aa46_row2_col446, #T_0aa46_row2_col447, #T_0aa46_row2_col448, #T_0aa46_row2_col449, #T_0aa46_row2_col450, #T_0aa46_row2_col451, #T_0aa46_row2_col452, #T_0aa46_row2_col453, #T_0aa46_row2_col454, #T_0aa46_row2_col455, #T_0aa46_row2_col456, #T_0aa46_row2_col457, #T_0aa46_row2_col458, #T_0aa46_row2_col459, #T_0aa46_row2_col460, #T_0aa46_row2_col461, #T_0aa46_row2_col462, #T_0aa46_row2_col463, #T_0aa46_row2_col464, #T_0aa46_row2_col465, #T_0aa46_row2_col466, #T_0aa46_row2_col467, #T_0aa46_row2_col468, #T_0aa46_row2_col469, #T_0aa46_row2_col470, #T_0aa46_row2_col471, #T_0aa46_row2_col472, #T_0aa46_row2_col473, #T_0aa46_row2_col474, #T_0aa46_row2_col475, #T_0aa46_row2_col476, #T_0aa46_row2_col477, #T_0aa46_row2_col478, #T_0aa46_row2_col479, #T_0aa46_row2_col480, #T_0aa46_row2_col481, #T_0aa46_row2_col482, #T_0aa46_row2_col483, #T_0aa46_row3_col0, #T_0aa46_row3_col1, #T_0aa46_row3_col2, #T_0aa46_row3_col3, #T_0aa46_row3_col4, #T_0aa46_row3_col5, #T_0aa46_row3_col6, #T_0aa46_row3_col7, #T_0aa46_row3_col8, #T_0aa46_row3_col9, #T_0aa46_row3_col10, #T_0aa46_row3_col11, #T_0aa46_row3_col12, #T_0aa46_row3_col13, #T_0aa46_row3_col14, #T_0aa46_row3_col15, #T_0aa46_row3_col16, #T_0aa46_row3_col17, #T_0aa46_row3_col18, #T_0aa46_row3_col19, #T_0aa46_row3_col20, #T_0aa46_row3_col21, #T_0aa46_row3_col22, #T_0aa46_row3_col23, #T_0aa46_row3_col24, #T_0aa46_row3_col25, #T_0aa46_row3_col26, #T_0aa46_row3_col27, #T_0aa46_row3_col28, #T_0aa46_row3_col29, #T_0aa46_row3_col30, #T_0aa46_row3_col31, #T_0aa46_row3_col32, #T_0aa46_row3_col33, #T_0aa46_row3_col34, #T_0aa46_row3_col35, #T_0aa46_row3_col36, #T_0aa46_row3_col37, #T_0aa46_row3_col38, #T_0aa46_row3_col39, #T_0aa46_row3_col40, #T_0aa46_row3_col41, #T_0aa46_row3_col42, #T_0aa46_row3_col43, #T_0aa46_row3_col44, #T_0aa46_row3_col45, #T_0aa46_row3_col46, #T_0aa46_row3_col47, #T_0aa46_row3_col48, #T_0aa46_row3_col49, #T_0aa46_row3_col50, #T_0aa46_row3_col51, #T_0aa46_row3_col52, #T_0aa46_row3_col53, #T_0aa46_row3_col54, #T_0aa46_row3_col55, #T_0aa46_row3_col56, #T_0aa46_row3_col57, #T_0aa46_row3_col58, #T_0aa46_row3_col59, #T_0aa46_row3_col60, #T_0aa46_row3_col61, #T_0aa46_row3_col62, #T_0aa46_row3_col63, #T_0aa46_row3_col64, #T_0aa46_row3_col65, #T_0aa46_row3_col66, #T_0aa46_row3_col67, #T_0aa46_row3_col68, #T_0aa46_row3_col69, #T_0aa46_row3_col70, #T_0aa46_row3_col71, #T_0aa46_row3_col72, #T_0aa46_row3_col73, #T_0aa46_row3_col74, #T_0aa46_row3_col75, #T_0aa46_row3_col76, #T_0aa46_row3_col77, #T_0aa46_row3_col78, #T_0aa46_row3_col79, #T_0aa46_row3_col80, #T_0aa46_row3_col81, #T_0aa46_row3_col82, #T_0aa46_row3_col83, #T_0aa46_row3_col84, #T_0aa46_row3_col85, #T_0aa46_row3_col86, #T_0aa46_row3_col87, #T_0aa46_row3_col88, #T_0aa46_row3_col89, #T_0aa46_row3_col90, #T_0aa46_row3_col91, #T_0aa46_row3_col92, #T_0aa46_row3_col93, #T_0aa46_row3_col94, #T_0aa46_row3_col95, #T_0aa46_row3_col96, #T_0aa46_row3_col97, #T_0aa46_row3_col98, #T_0aa46_row3_col99, #T_0aa46_row3_col100, #T_0aa46_row3_col101, #T_0aa46_row3_col102, #T_0aa46_row3_col103, #T_0aa46_row3_col104, #T_0aa46_row3_col105, #T_0aa46_row3_col106, #T_0aa46_row3_col107, #T_0aa46_row3_col108, #T_0aa46_row3_col109, #T_0aa46_row3_col110, #T_0aa46_row3_col111, #T_0aa46_row3_col112, #T_0aa46_row3_col113, #T_0aa46_row3_col114, #T_0aa46_row3_col115, #T_0aa46_row3_col116, #T_0aa46_row3_col117, #T_0aa46_row3_col118, #T_0aa46_row3_col119, #T_0aa46_row3_col120, #T_0aa46_row3_col121, #T_0aa46_row3_col122, #T_0aa46_row3_col123, #T_0aa46_row3_col124, #T_0aa46_row3_col125, #T_0aa46_row3_col126, #T_0aa46_row3_col127, #T_0aa46_row3_col128, #T_0aa46_row3_col129, #T_0aa46_row3_col130, #T_0aa46_row3_col131, #T_0aa46_row3_col132, #T_0aa46_row3_col133, #T_0aa46_row3_col134, #T_0aa46_row3_col135, #T_0aa46_row3_col136, #T_0aa46_row3_col137, #T_0aa46_row3_col138, #T_0aa46_row3_col139, #T_0aa46_row3_col140, #T_0aa46_row3_col141, #T_0aa46_row3_col142, #T_0aa46_row3_col143, #T_0aa46_row3_col144, #T_0aa46_row3_col145, #T_0aa46_row3_col146, #T_0aa46_row3_col147, #T_0aa46_row3_col148, #T_0aa46_row3_col149, #T_0aa46_row3_col150, #T_0aa46_row3_col151, #T_0aa46_row3_col152, #T_0aa46_row3_col153, #T_0aa46_row3_col154, #T_0aa46_row3_col155, #T_0aa46_row3_col156, #T_0aa46_row3_col157, #T_0aa46_row3_col158, #T_0aa46_row3_col159, #T_0aa46_row3_col160, #T_0aa46_row3_col161, #T_0aa46_row3_col162, #T_0aa46_row3_col163, #T_0aa46_row3_col164, #T_0aa46_row3_col165, #T_0aa46_row3_col166, #T_0aa46_row3_col167, #T_0aa46_row3_col168, #T_0aa46_row3_col169, #T_0aa46_row3_col170, #T_0aa46_row3_col171, #T_0aa46_row3_col172, #T_0aa46_row3_col173, #T_0aa46_row3_col174, #T_0aa46_row3_col175, #T_0aa46_row3_col176, #T_0aa46_row3_col177, #T_0aa46_row3_col178, #T_0aa46_row3_col179, #T_0aa46_row3_col180, #T_0aa46_row3_col181, #T_0aa46_row3_col182, #T_0aa46_row3_col183, #T_0aa46_row3_col184, #T_0aa46_row3_col185, #T_0aa46_row3_col186, #T_0aa46_row3_col187, #T_0aa46_row3_col188, #T_0aa46_row3_col189, #T_0aa46_row3_col190, #T_0aa46_row3_col191, #T_0aa46_row3_col192, #T_0aa46_row3_col193, #T_0aa46_row3_col194, #T_0aa46_row3_col195, #T_0aa46_row3_col196, #T_0aa46_row3_col197, #T_0aa46_row3_col198, #T_0aa46_row3_col199, #T_0aa46_row3_col200, #T_0aa46_row3_col201, #T_0aa46_row3_col202, #T_0aa46_row3_col203, #T_0aa46_row3_col204, #T_0aa46_row3_col205, #T_0aa46_row3_col206, #T_0aa46_row3_col207, #T_0aa46_row3_col208, #T_0aa46_row3_col209, #T_0aa46_row3_col210, #T_0aa46_row3_col211, #T_0aa46_row3_col212, #T_0aa46_row3_col213, #T_0aa46_row3_col214, #T_0aa46_row3_col215, #T_0aa46_row3_col216, #T_0aa46_row3_col217, #T_0aa46_row3_col218, #T_0aa46_row3_col219, #T_0aa46_row3_col220, #T_0aa46_row3_col221, #T_0aa46_row3_col222, #T_0aa46_row3_col223, #T_0aa46_row3_col224, #T_0aa46_row3_col225, #T_0aa46_row3_col226, #T_0aa46_row3_col227, #T_0aa46_row3_col228, #T_0aa46_row3_col229, #T_0aa46_row3_col230, #T_0aa46_row3_col231, #T_0aa46_row3_col232, #T_0aa46_row3_col233, #T_0aa46_row3_col234, #T_0aa46_row3_col235, #T_0aa46_row3_col236, #T_0aa46_row3_col237, #T_0aa46_row3_col238, #T_0aa46_row3_col239, #T_0aa46_row3_col240, #T_0aa46_row3_col241, #T_0aa46_row3_col242, #T_0aa46_row3_col243, #T_0aa46_row3_col244, #T_0aa46_row3_col245, #T_0aa46_row3_col246, #T_0aa46_row3_col247, #T_0aa46_row3_col248, #T_0aa46_row3_col249, #T_0aa46_row3_col250, #T_0aa46_row3_col251, #T_0aa46_row3_col252, #T_0aa46_row3_col253, #T_0aa46_row3_col254, #T_0aa46_row3_col255, #T_0aa46_row3_col256, #T_0aa46_row3_col257, #T_0aa46_row3_col258, #T_0aa46_row3_col259, #T_0aa46_row3_col260, #T_0aa46_row3_col261, #T_0aa46_row3_col262, #T_0aa46_row3_col263, #T_0aa46_row3_col264, #T_0aa46_row3_col265, #T_0aa46_row3_col266, #T_0aa46_row3_col267, #T_0aa46_row3_col268, #T_0aa46_row3_col269, #T_0aa46_row3_col270, #T_0aa46_row3_col271, #T_0aa46_row3_col272, #T_0aa46_row3_col273, #T_0aa46_row3_col274, #T_0aa46_row3_col275, #T_0aa46_row3_col276, #T_0aa46_row3_col277, #T_0aa46_row3_col278, #T_0aa46_row3_col279, #T_0aa46_row3_col280, #T_0aa46_row3_col281, #T_0aa46_row3_col282, #T_0aa46_row3_col283, #T_0aa46_row3_col284, #T_0aa46_row3_col285, #T_0aa46_row3_col286, #T_0aa46_row3_col287, #T_0aa46_row3_col288, #T_0aa46_row3_col289, #T_0aa46_row3_col290, #T_0aa46_row3_col291, #T_0aa46_row3_col292, #T_0aa46_row3_col293, #T_0aa46_row3_col294, #T_0aa46_row3_col295, #T_0aa46_row3_col296, #T_0aa46_row3_col297, #T_0aa46_row3_col298, #T_0aa46_row3_col299, #T_0aa46_row3_col300, #T_0aa46_row3_col301, #T_0aa46_row3_col302, #T_0aa46_row3_col303, #T_0aa46_row3_col304, #T_0aa46_row3_col305, #T_0aa46_row3_col306, #T_0aa46_row3_col307, #T_0aa46_row3_col308, #T_0aa46_row3_col309, #T_0aa46_row3_col310, #T_0aa46_row3_col311, #T_0aa46_row3_col312, #T_0aa46_row3_col313, #T_0aa46_row3_col314, #T_0aa46_row3_col315, #T_0aa46_row3_col316, #T_0aa46_row3_col317, #T_0aa46_row3_col318, #T_0aa46_row3_col319, #T_0aa46_row3_col320, #T_0aa46_row3_col321, #T_0aa46_row3_col322, #T_0aa46_row3_col323, #T_0aa46_row3_col324, #T_0aa46_row3_col325, #T_0aa46_row3_col326, #T_0aa46_row3_col327, #T_0aa46_row3_col328, #T_0aa46_row3_col329, #T_0aa46_row3_col330, #T_0aa46_row3_col331, #T_0aa46_row3_col332, #T_0aa46_row3_col333, #T_0aa46_row3_col334, #T_0aa46_row3_col335, #T_0aa46_row3_col336, #T_0aa46_row3_col337, #T_0aa46_row3_col338, #T_0aa46_row3_col339, #T_0aa46_row3_col340, #T_0aa46_row3_col341, #T_0aa46_row3_col342, #T_0aa46_row3_col343, #T_0aa46_row3_col344, #T_0aa46_row3_col345, #T_0aa46_row3_col346, #T_0aa46_row3_col347, #T_0aa46_row3_col348, #T_0aa46_row3_col349, #T_0aa46_row3_col350, #T_0aa46_row3_col351, #T_0aa46_row3_col352, #T_0aa46_row3_col353, #T_0aa46_row3_col354, #T_0aa46_row3_col355, #T_0aa46_row3_col356, #T_0aa46_row3_col357, #T_0aa46_row3_col358, #T_0aa46_row3_col359, #T_0aa46_row3_col360, #T_0aa46_row3_col361, #T_0aa46_row3_col362, #T_0aa46_row3_col363, #T_0aa46_row3_col364, #T_0aa46_row3_col365, #T_0aa46_row3_col366, #T_0aa46_row3_col367, #T_0aa46_row3_col368, #T_0aa46_row3_col369, #T_0aa46_row3_col370, #T_0aa46_row3_col371, #T_0aa46_row3_col372, #T_0aa46_row3_col373, #T_0aa46_row3_col374, #T_0aa46_row3_col375, #T_0aa46_row3_col376, #T_0aa46_row3_col377, #T_0aa46_row3_col378, #T_0aa46_row3_col379, #T_0aa46_row3_col380, #T_0aa46_row3_col381, #T_0aa46_row3_col382, #T_0aa46_row3_col383, #T_0aa46_row3_col384, #T_0aa46_row3_col385, #T_0aa46_row3_col386, #T_0aa46_row3_col387, #T_0aa46_row3_col388, #T_0aa46_row3_col389, #T_0aa46_row3_col390, #T_0aa46_row3_col391, #T_0aa46_row3_col392, #T_0aa46_row3_col393, #T_0aa46_row3_col394, #T_0aa46_row3_col395, #T_0aa46_row3_col396, #T_0aa46_row3_col397, #T_0aa46_row3_col398, #T_0aa46_row3_col399, #T_0aa46_row3_col400, #T_0aa46_row3_col401, #T_0aa46_row3_col402, #T_0aa46_row3_col403, #T_0aa46_row3_col404, #T_0aa46_row3_col405, #T_0aa46_row3_col406, #T_0aa46_row3_col407, #T_0aa46_row3_col408, #T_0aa46_row3_col409, #T_0aa46_row3_col410, #T_0aa46_row3_col411, #T_0aa46_row3_col412, #T_0aa46_row3_col413, #T_0aa46_row3_col414, #T_0aa46_row3_col415, #T_0aa46_row3_col416, #T_0aa46_row3_col417, #T_0aa46_row3_col418, #T_0aa46_row3_col419, #T_0aa46_row3_col420, #T_0aa46_row3_col421, #T_0aa46_row3_col422, #T_0aa46_row3_col423, #T_0aa46_row3_col424, #T_0aa46_row3_col425, #T_0aa46_row3_col426, #T_0aa46_row3_col427, #T_0aa46_row3_col428, #T_0aa46_row3_col429, #T_0aa46_row3_col430, #T_0aa46_row3_col431, #T_0aa46_row3_col432, #T_0aa46_row3_col433, #T_0aa46_row3_col434, #T_0aa46_row3_col435, #T_0aa46_row3_col436, #T_0aa46_row3_col437, #T_0aa46_row3_col438, #T_0aa46_row3_col439, #T_0aa46_row3_col440, #T_0aa46_row3_col441, #T_0aa46_row3_col442, #T_0aa46_row3_col443, #T_0aa46_row3_col444, #T_0aa46_row3_col445, #T_0aa46_row3_col446, #T_0aa46_row3_col447, #T_0aa46_row3_col448, #T_0aa46_row3_col449, #T_0aa46_row3_col450, #T_0aa46_row3_col451, #T_0aa46_row3_col452, #T_0aa46_row3_col453, #T_0aa46_row3_col454, #T_0aa46_row3_col455, #T_0aa46_row3_col456, #T_0aa46_row3_col457, #T_0aa46_row3_col458, #T_0aa46_row3_col459, #T_0aa46_row3_col460, #T_0aa46_row3_col461, #T_0aa46_row3_col462, #T_0aa46_row3_col463, #T_0aa46_row3_col464, #T_0aa46_row3_col465, #T_0aa46_row3_col466, #T_0aa46_row3_col467, #T_0aa46_row3_col468, #T_0aa46_row3_col469, #T_0aa46_row3_col470, #T_0aa46_row3_col471, #T_0aa46_row3_col472, #T_0aa46_row3_col473, #T_0aa46_row3_col474, #T_0aa46_row3_col475, #T_0aa46_row3_col476, #T_0aa46_row3_col477, #T_0aa46_row3_col478, #T_0aa46_row3_col479, #T_0aa46_row3_col480, #T_0aa46_row3_col481, #T_0aa46_row3_col482, #T_0aa46_row3_col483 {
  font-size: 11pt;
  font-family: Helvetica;
  border-collapse: collapse;
  border: 1px solid black;
}
</style>

<table id="T_0aa46" class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th class="index_name level0" data-quarto-table-cell-role="th">glycan</th>
<th id="T_0aa46_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">Apif(a1-2)Xyl(b1-2)[Glc6Ac(b1-4)]Glc</th>
<th id="T_0aa46_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">Ara(a1-2)Ara(a1-6)GlcNAc</th>
<th id="T_0aa46_level0_col2" class="col_heading level0 col2" data-quarto-table-cell-role="th">Ara(a1-2)Glc(b1-2)Ara</th>
<th id="T_0aa46_level0_col3" class="col_heading level0 col3" data-quarto-table-cell-role="th">Ara(a1-2)GlcA</th>
<th id="T_0aa46_level0_col4" class="col_heading level0 col4" data-quarto-table-cell-role="th">Ara(a1-2)[Glc(b1-6)]Glc</th>
<th id="T_0aa46_level0_col5" class="col_heading level0 col5" data-quarto-table-cell-role="th">Ara(a1-3)Gal(b1-6)Gal</th>
<th id="T_0aa46_level0_col6" class="col_heading level0 col6" data-quarto-table-cell-role="th">Ara(a1-6)Glc</th>
<th id="T_0aa46_level0_col7" class="col_heading level0 col7" data-quarto-table-cell-role="th">Araf(a1-3)Araf(a1-5)[Araf(a1-6)Gal(b1-6)Glc(b1-6)Man(a1-3)]Araf(a1-5)Araf(a1-3)Araf(a1-3)Araf</th>
<th id="T_0aa46_level0_col8" class="col_heading level0 col8" data-quarto-table-cell-role="th">Araf(a1-3)Gal(b1-6)Gal</th>
<th id="T_0aa46_level0_col9" class="col_heading level0 col9" data-quarto-table-cell-role="th">D-Apif(b1-2)Glc</th>
<th id="T_0aa46_level0_col10" class="col_heading level0 col10" data-quarto-table-cell-role="th">D-Apif(b1-2)GlcA</th>
<th id="T_0aa46_level0_col11" class="col_heading level0 col11" data-quarto-table-cell-role="th">D-Apif(b1-3)Xyl(b1-2)[Glc6Ac(b1-4)]Glc</th>
<th id="T_0aa46_level0_col12" class="col_heading level0 col12" data-quarto-table-cell-role="th">D-Apif(b1-3)Xyl(b1-4)Rha(a1-2)Ara</th>
<th id="T_0aa46_level0_col13" class="col_heading level0 col13" data-quarto-table-cell-role="th">D-Apif(b1-3)Xyl(b1-4)Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col14" class="col_heading level0 col14" data-quarto-table-cell-role="th">D-Apif(b1-3)Xyl(b1-4)[Glc(b1-3)]Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col15" class="col_heading level0 col15" data-quarto-table-cell-role="th">D-Apif(b1-3)[Gal(b1-4)Xyl(b1-4)]Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col16" class="col_heading level0 col16" data-quarto-table-cell-role="th">D-Apif(b1-3)[Gal(b1-4)Xyl(b1-4)]Rha(a1-2)[Rha(a1-3)]D-Fuc</th>
<th id="T_0aa46_level0_col17" class="col_heading level0 col17" data-quarto-table-cell-role="th">D-Apif(b1-3)[Gal(b1-4)Xyl(b1-4)]Rha(a1-3)D-Fuc</th>
<th id="T_0aa46_level0_col18" class="col_heading level0 col18" data-quarto-table-cell-role="th">D-Apif(b1-6)Glc</th>
<th id="T_0aa46_level0_col19" class="col_heading level0 col19" data-quarto-table-cell-role="th">D-ApifOMe(b1-3)XylOMe(b1-4)RhaOMe(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col20" class="col_heading level0 col20" data-quarto-table-cell-role="th">D-ApifOMe(b1-3)XylOMe(b1-4)[GlcOMe(b1-3)]RhaOMe(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col21" class="col_heading level0 col21" data-quarto-table-cell-role="th">Fruf(a2-1)[Glc(b1-2)][Glc(b1-3)Glc4Ac6Ac(b1-3)]Glc</th>
<th id="T_0aa46_level0_col22" class="col_heading level0 col22" data-quarto-table-cell-role="th">Fruf(a2-1)[Glc(b1-2)][Glc(b1-3)Glc4Ac6Ac(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col23" class="col_heading level0 col23" data-quarto-table-cell-role="th">Fruf(a2-1)[Glc(b1-2)][Glc(b1-3)Glc6Ac(b1-3)]Glc</th>
<th id="T_0aa46_level0_col24" class="col_heading level0 col24" data-quarto-table-cell-role="th">Fruf(a2-1)[Glc(b1-2)][Glc(b1-3)Glc6Ac(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col25" class="col_heading level0 col25" data-quarto-table-cell-role="th">Fruf(b2-1)Glc3Ac6Ac</th>
<th id="T_0aa46_level0_col26" class="col_heading level0 col26" data-quarto-table-cell-role="th">Fruf(b2-1)Glc4Ac6Ac</th>
<th id="T_0aa46_level0_col27" class="col_heading level0 col27" data-quarto-table-cell-role="th">Fruf(b2-1)Glc6Ac</th>
<th id="T_0aa46_level0_col28" class="col_heading level0 col28" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)]Glc</th>
<th id="T_0aa46_level0_col29" class="col_heading level0 col29" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-3)Glc(b1-3)]Glc</th>
<th id="T_0aa46_level0_col30" class="col_heading level0 col30" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col31" class="col_heading level0 col31" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-4)Glc(b1-3)]Glc</th>
<th id="T_0aa46_level0_col32" class="col_heading level0 col32" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-4)Glc(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col33" class="col_heading level0 col33" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-4)Glc6Ac(b1-3)]Glc</th>
<th id="T_0aa46_level0_col34" class="col_heading level0 col34" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc(b1-4)Glc6Ac(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col35" class="col_heading level0 col35" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc6Ac(b1-3)]Glc</th>
<th id="T_0aa46_level0_col36" class="col_heading level0 col36" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-2)][Glc6Ac(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col37" class="col_heading level0 col37" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc(b1-4)Glc6Ac(b1-3)]Glc6Ac</th>
<th id="T_0aa46_level0_col38" class="col_heading level0 col38" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc3Ac(b1-2)]Glc</th>
<th id="T_0aa46_level0_col39" class="col_heading level0 col39" data-quarto-table-cell-role="th">Fruf(b2-1)[Glc6Ac(b1-2)]Glc</th>
<th id="T_0aa46_level0_col40" class="col_heading level0 col40" data-quarto-table-cell-role="th">Fruf1Ac(b2-1)Glc2Ac4Ac6Ac</th>
<th id="T_0aa46_level0_col41" class="col_heading level0 col41" data-quarto-table-cell-role="th">Fuc(a1-2)Gal(b1-2)Xyl(a1-6)Glc</th>
<th id="T_0aa46_level0_col42" class="col_heading level0 col42" data-quarto-table-cell-role="th">Fuc(a1-2)Gal(b1-2)Xyl(a1-6)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col43" class="col_heading level0 col43" data-quarto-table-cell-role="th">Fuc(a1-2)Gal(b1-2)Xyl(a1-6)[Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col44" class="col_heading level0 col44" data-quarto-table-cell-role="th">Fuc(a1-2)Gal(b1-2)Xyl(a1-6)[Glc(b1-4)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col45" class="col_heading level0 col45" data-quarto-table-cell-role="th">Fuc(a1-2)Gal(b1-4)Xyl</th>
<th id="T_0aa46_level0_col46" class="col_heading level0 col46" data-quarto-table-cell-role="th">Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col47" class="col_heading level0 col47" data-quarto-table-cell-role="th">Fuc(a1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col48" class="col_heading level0 col48" data-quarto-table-cell-role="th">Fuc(a1-6)GlcNAc(b1-2)[Man(a1-6)]Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col49" class="col_heading level0 col49" data-quarto-table-cell-role="th">Gal(?1-?)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Man(a1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col50" class="col_heading level0 col50" data-quarto-table-cell-role="th">Gal(?1-?)[Gal(?1-?)]GlcNAc(?1-?)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(?1-?)Man(a1-3)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col51" class="col_heading level0 col51" data-quarto-table-cell-role="th">Gal(a1-4)Gal</th>
<th id="T_0aa46_level0_col52" class="col_heading level0 col52" data-quarto-table-cell-role="th">Gal(a1-6)Gal</th>
<th id="T_0aa46_level0_col53" class="col_heading level0 col53" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal</th>
<th id="T_0aa46_level0_col54" class="col_heading level0 col54" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Gal(a1-6)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col55" class="col_heading level0 col55" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Gal(a1-6)Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col56" class="col_heading level0 col56" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Gal(a1-6)[Fruf(b2-1)]Glc</th>
<th id="T_0aa46_level0_col57" class="col_heading level0 col57" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Glc</th>
<th id="T_0aa46_level0_col58" class="col_heading level0 col58" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col59" class="col_heading level0 col59" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Gal(a1-6)Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col60" class="col_heading level0 col60" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Glc</th>
<th id="T_0aa46_level0_col61" class="col_heading level0 col61" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col62" class="col_heading level0 col62" data-quarto-table-cell-role="th">Gal(a1-6)Gal(a1-6)Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col63" class="col_heading level0 col63" data-quarto-table-cell-role="th">Gal(a1-6)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col64" class="col_heading level0 col64" data-quarto-table-cell-role="th">Gal(a1-6)Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col65" class="col_heading level0 col65" data-quarto-table-cell-role="th">Gal(a1-6)Man</th>
<th id="T_0aa46_level0_col66" class="col_heading level0 col66" data-quarto-table-cell-role="th">Gal(a1-6)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col67" class="col_heading level0 col67" data-quarto-table-cell-role="th">Gal(a1-6)Man(b1-4)Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col68" class="col_heading level0 col68" data-quarto-table-cell-role="th">Gal(a1-6)Man(b1-4)Man(b1-4)Man(b1-4)[Gal(a1-6)]Man(b1-4)Man(b1-4)Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col69" class="col_heading level0 col69" data-quarto-table-cell-role="th">Gal(a1-6)Man(b1-4)Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col70" class="col_heading level0 col70" data-quarto-table-cell-role="th">Gal(a1-6)Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col71" class="col_heading level0 col71" data-quarto-table-cell-role="th">Gal(b1-2)Glc</th>
<th id="T_0aa46_level0_col72" class="col_heading level0 col72" data-quarto-table-cell-role="th">Gal(b1-2)GlcA</th>
<th id="T_0aa46_level0_col73" class="col_heading level0 col73" data-quarto-table-cell-role="th">Gal(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col74" class="col_heading level0 col74" data-quarto-table-cell-role="th">Gal(b1-2)Xyl(a1-6)Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col75" class="col_heading level0 col75" data-quarto-table-cell-role="th">Gal(b1-2)Xyl(a1-6)[Glc(b1-4)]Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col76" class="col_heading level0 col76" data-quarto-table-cell-role="th">Gal(b1-2)Xyl(a1-6)[Glc(b1-4)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col77" class="col_heading level0 col77" data-quarto-table-cell-role="th">Gal(b1-2)[Xyl(b1-3)]GlcA</th>
<th id="T_0aa46_level0_col78" class="col_heading level0 col78" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col79" class="col_heading level0 col79" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col80" class="col_heading level0 col80" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col81" class="col_heading level0 col81" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col82" class="col_heading level0 col82" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col83" class="col_heading level0 col83" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-2)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col84" class="col_heading level0 col84" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-4)Man(a1-3)[Gal(b1-3)GlcNAc(b1-4)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col85" class="col_heading level0 col85" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-4)Man(a1-3)[GlcNAc(b1-4)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col86" class="col_heading level0 col86" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-4)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col87" class="col_heading level0 col87" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-4)Man(a1-6)[GlcNAc(b1-4)Man(a1-3)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col88" class="col_heading level0 col88" data-quarto-table-cell-role="th">Gal(b1-3)GlcNAc(b1-4)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col89" class="col_heading level0 col89" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col90" class="col_heading level0 col90" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col91" class="col_heading level0 col91" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col92" class="col_heading level0 col92" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3/6)[Gal(b1-3)GlcNAc(b1-2)Man(a1-3/6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col93" class="col_heading level0 col93" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3/6)[GlcNAc(b1-2)Man(a1-3/6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col94" class="col_heading level0 col94" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3/6)[Xyl(b1-2)][Man(a1-3/6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col95" class="col_heading level0 col95" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col96" class="col_heading level0 col96" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col97" class="col_heading level0 col97" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)[Man(a1-6)]Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col98" class="col_heading level0 col98" data-quarto-table-cell-role="th">Gal(b1-3)[Fuc(a1-6)]GlcNAc(b1-2)[Man(a1-6)]Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col99" class="col_heading level0 col99" data-quarto-table-cell-role="th">Gal(b1-4)Gal(b1-4)Man</th>
<th id="T_0aa46_level0_col100" class="col_heading level0 col100" data-quarto-table-cell-role="th">Gal(b1-4)Gal(b1-4)ManOMe</th>
<th id="T_0aa46_level0_col101" class="col_heading level0 col101" data-quarto-table-cell-role="th">Gal(b1-4)GlcA</th>
<th id="T_0aa46_level0_col102" class="col_heading level0 col102" data-quarto-table-cell-role="th">Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>
<th id="T_0aa46_level0_col103" class="col_heading level0 col103" data-quarto-table-cell-role="th">Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col104" class="col_heading level0 col104" data-quarto-table-cell-role="th">Gal(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col105" class="col_heading level0 col105" data-quarto-table-cell-role="th">Gal(b1-4)Man(b1-4)Man(b1-4)Gal</th>
<th id="T_0aa46_level0_col106" class="col_heading level0 col106" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col107" class="col_heading level0 col107" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)D-Fuc1CoumOMe</th>
<th id="T_0aa46_level0_col108" class="col_heading level0 col108" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)D-Fuc1FerOMe</th>
<th id="T_0aa46_level0_col109" class="col_heading level0 col109" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)Fuc</th>
<th id="T_0aa46_level0_col110" class="col_heading level0 col110" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)Fuc4Ac</th>
<th id="T_0aa46_level0_col111" class="col_heading level0 col111" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)[Rha(a1-3)]D-Fuc</th>
<th id="T_0aa46_level0_col112" class="col_heading level0 col112" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)[Rha(a1-3)]D-Fuc1CoumOMe</th>
<th id="T_0aa46_level0_col113" class="col_heading level0 col113" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)[Rha(a1-3)]D-FucOMeOSin</th>
<th id="T_0aa46_level0_col114" class="col_heading level0 col114" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)Rha(a1-2)[Rha(a1-3)]Fuc</th>
<th id="T_0aa46_level0_col115" class="col_heading level0 col115" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)[D-Apif(b1-3)]Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col116" class="col_heading level0 col116" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)[D-Apif(b1-3)]Rha(a1-2)D-Fuc1CoumOMe</th>
<th id="T_0aa46_level0_col117" class="col_heading level0 col117" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)[D-Apif(b1-3)]Rha(a1-2)[Rha(a1-3)]D-Fuc</th>
<th id="T_0aa46_level0_col118" class="col_heading level0 col118" data-quarto-table-cell-role="th">Gal(b1-4)Xyl(b1-4)[D-Apif(b1-3)]Rha(a1-2)[Rha(a1-3)]D-Fuc1CoumOMe</th>
<th id="T_0aa46_level0_col119" class="col_heading level0 col119" data-quarto-table-cell-role="th">GalA(a1-2)[Araf(a1-5)Araf(a1-4)]Rha(b1-4)GalA</th>
<th id="T_0aa46_level0_col120" class="col_heading level0 col120" data-quarto-table-cell-role="th">GalA(a1-4)GalA(a1-4)GalA</th>
<th id="T_0aa46_level0_col121" class="col_heading level0 col121" data-quarto-table-cell-role="th">GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-2)Rha(a1-4)GalA(a1-2)Rha(a1-4)GalA(a1-2)GalA</th>
<th id="T_0aa46_level0_col122" class="col_heading level0 col122" data-quarto-table-cell-role="th">GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA(a1-4)GalA</th>
<th id="T_0aa46_level0_col123" class="col_heading level0 col123" data-quarto-table-cell-role="th">GalOMe(b1-2)[XylOMe(b1-3)]GlcAOMe</th>
<th id="T_0aa46_level0_col124" class="col_heading level0 col124" data-quarto-table-cell-role="th">GalOMe(b1-4)XylOMe(b1-4)RhaOMe(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col125" class="col_heading level0 col125" data-quarto-table-cell-role="th">GalOMe(b1-4)XylOMe(b1-4)RhaOMe(a1-2)[RhaOMe(a1-3)]D-FucOMe</th>
<th id="T_0aa46_level0_col126" class="col_heading level0 col126" data-quarto-table-cell-role="th">GalOMe(b1-4)XylOMe(b1-4)[D-ApifOMe(b1-3)]RhaOMe(a1-2)[RhaOMe(a1-3)]D-FucOMe</th>
<th id="T_0aa46_level0_col127" class="col_heading level0 col127" data-quarto-table-cell-role="th">Galf(b1-2)[Galf(b1-4)]Man</th>
<th id="T_0aa46_level0_col128" class="col_heading level0 col128" data-quarto-table-cell-role="th">Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col129" class="col_heading level0 col129" data-quarto-table-cell-role="th">Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col130" class="col_heading level0 col130" data-quarto-table-cell-role="th">Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col131" class="col_heading level0 col131" data-quarto-table-cell-role="th">Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col132" class="col_heading level0 col132" data-quarto-table-cell-role="th">Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col133" class="col_heading level0 col133" data-quarto-table-cell-role="th">Glc(a1-2)Rha(a1-6)Glc</th>
<th id="T_0aa46_level0_col134" class="col_heading level0 col134" data-quarto-table-cell-role="th">Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col135" class="col_heading level0 col135" data-quarto-table-cell-role="th">Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col136" class="col_heading level0 col136" data-quarto-table-cell-role="th">Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col137" class="col_heading level0 col137" data-quarto-table-cell-role="th">Glc(a1-4)Glc(a1-2)Rha(a1-6)Glc</th>
<th id="T_0aa46_level0_col138" class="col_heading level0 col138" data-quarto-table-cell-role="th">Glc(a1-4)Glc(a1-4)Glc(a1-6)Glc</th>
<th id="T_0aa46_level0_col139" class="col_heading level0 col139" data-quarto-table-cell-role="th">Glc(a1-4)Glc(a1-4)GlcA</th>
<th id="T_0aa46_level0_col140" class="col_heading level0 col140" data-quarto-table-cell-role="th">Glc(a1-4)GlcA(b1-2)GlcA</th>
<th id="T_0aa46_level0_col141" class="col_heading level0 col141" data-quarto-table-cell-role="th">Glc(b1-2)Ara</th>
<th id="T_0aa46_level0_col142" class="col_heading level0 col142" data-quarto-table-cell-role="th">Glc(b1-2)Ara(a1-2)GlcA</th>
<th id="T_0aa46_level0_col143" class="col_heading level0 col143" data-quarto-table-cell-role="th">Glc(b1-2)Gal(b1-2)Gal(b1-2)GlcA</th>
<th id="T_0aa46_level0_col144" class="col_heading level0 col144" data-quarto-table-cell-role="th">Glc(b1-2)Gal(b1-2)GlcA</th>
<th id="T_0aa46_level0_col145" class="col_heading level0 col145" data-quarto-table-cell-role="th">Glc(b1-2)Gal(b1-2)GlcA(b1-3)[Glc(b1-3)]Ara</th>
<th id="T_0aa46_level0_col146" class="col_heading level0 col146" data-quarto-table-cell-role="th">Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col147" class="col_heading level0 col147" data-quarto-table-cell-role="th">Glc(b1-2)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col148" class="col_heading level0 col148" data-quarto-table-cell-role="th">Glc(b1-2)Glc(a1-2)FrufOBzOCin</th>
<th id="T_0aa46_level0_col149" class="col_heading level0 col149" data-quarto-table-cell-role="th">Glc(b1-2)Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col150" class="col_heading level0 col150" data-quarto-table-cell-role="th">Glc(b1-2)GlcA</th>
<th id="T_0aa46_level0_col151" class="col_heading level0 col151" data-quarto-table-cell-role="th">Glc(b1-2)Xyl</th>
<th id="T_0aa46_level0_col152" class="col_heading level0 col152" data-quarto-table-cell-role="th">Glc(b1-2)[Ara(a1-3)]GlcA6Me</th>
<th id="T_0aa46_level0_col153" class="col_heading level0 col153" data-quarto-table-cell-role="th">Glc(b1-2)[Ara(a1-3)]GlcAOMe</th>
<th id="T_0aa46_level0_col154" class="col_heading level0 col154" data-quarto-table-cell-role="th">Glc(b1-2)[Ara(a1-6)]Glc</th>
<th id="T_0aa46_level0_col155" class="col_heading level0 col155" data-quarto-table-cell-role="th">Glc(b1-2)[Glc(b1-3)]Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col156" class="col_heading level0 col156" data-quarto-table-cell-role="th">Glc(b1-2)[Glc(b1-3)]Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col157" class="col_heading level0 col157" data-quarto-table-cell-role="th">Glc(b1-2)[Glc(b1-3)]Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col158" class="col_heading level0 col158" data-quarto-table-cell-role="th">Glc(b1-2)[Glc6Ac(b1-3)]Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col159" class="col_heading level0 col159" data-quarto-table-cell-role="th">Glc(b1-2)[Glc6Ac(b1-3)]Glc1Fer(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col160" class="col_heading level0 col160" data-quarto-table-cell-role="th">Glc(b1-2)[Glc6Ac(b1-3)]Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col161" class="col_heading level0 col161" data-quarto-table-cell-role="th">Glc(b1-2)[Rha(a1-3)]GlcA</th>
<th id="T_0aa46_level0_col162" class="col_heading level0 col162" data-quarto-table-cell-role="th">Glc(b1-2)[Xyl(b1-2)Ara(a1-6)]Glc</th>
<th id="T_0aa46_level0_col163" class="col_heading level0 col163" data-quarto-table-cell-role="th">Glc(b1-2)[Xyl(b1-2)D-Fuc(b1-6)]Glc</th>
<th id="T_0aa46_level0_col164" class="col_heading level0 col164" data-quarto-table-cell-role="th">Glc(b1-3)Ara</th>
<th id="T_0aa46_level0_col165" class="col_heading level0 col165" data-quarto-table-cell-role="th">Glc(b1-3)Glc</th>
<th id="T_0aa46_level0_col166" class="col_heading level0 col166" data-quarto-table-cell-role="th">Glc(b1-3)Glc(b1-3)[Glc(b1-2)]Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col167" class="col_heading level0 col167" data-quarto-table-cell-role="th">Glc(b1-3)Glc(b1-3)[Glc(b1-2)]Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col168" class="col_heading level0 col168" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col169" class="col_heading level0 col169" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col170" class="col_heading level0 col170" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Coum6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col171" class="col_heading level0 col171" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col172" class="col_heading level0 col172" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col173" class="col_heading level0 col173" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col174" class="col_heading level0 col174" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col175" class="col_heading level0 col175" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)]Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col176" class="col_heading level0 col176" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)][Rha(a1-4)]Glc1Coum6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col177" class="col_heading level0 col177" data-quarto-table-cell-role="th">Glc(b1-3)Glc6Ac(b1-3)[Glc(b1-2)][Rha(a1-4)]Glc1Fer6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col178" class="col_heading level0 col178" data-quarto-table-cell-role="th">Glc(b1-3)Rha1Fer(a1-4)Fruf(b2-1)GlcOBz</th>
<th id="T_0aa46_level0_col179" class="col_heading level0 col179" data-quarto-table-cell-role="th">Glc(b1-3)[Araf(a1-4)]Rha(a1-2)Glc</th>
<th id="T_0aa46_level0_col180" class="col_heading level0 col180" data-quarto-table-cell-role="th">Glc(b1-3)[Xyl(b1-4)]Rha(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col181" class="col_heading level0 col181" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col182" class="col_heading level0 col182" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc(a1-2)Fruf</th>
<th id="T_0aa46_level0_col183" class="col_heading level0 col183" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc1Coum6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col184" class="col_heading level0 col184" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col185" class="col_heading level0 col185" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col186" class="col_heading level0 col186" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col187" class="col_heading level0 col187" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-3)[Glc(b1-2)]Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col188" class="col_heading level0 col188" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col189" class="col_heading level0 col189" data-quarto-table-cell-role="th">Glc(b1-4)Glc(b1-4)Glc(b1-4)Man</th>
<th id="T_0aa46_level0_col190" class="col_heading level0 col190" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col191" class="col_heading level0 col191" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col192" class="col_heading level0 col192" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col193" class="col_heading level0 col193" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Coum6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col194" class="col_heading level0 col194" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col195" class="col_heading level0 col195" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col196" class="col_heading level0 col196" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col197" class="col_heading level0 col197" data-quarto-table-cell-role="th">Glc(b1-4)Glc6Ac(b1-3)[Glc(b1-2)]Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col198" class="col_heading level0 col198" data-quarto-table-cell-role="th">Glc(b1-4)Man(b1-4)Glc</th>
<th id="T_0aa46_level0_col199" class="col_heading level0 col199" data-quarto-table-cell-role="th">Glc(b1-4)Rha</th>
<th id="T_0aa46_level0_col200" class="col_heading level0 col200" data-quarto-table-cell-role="th">Glc(b1-4)Rha1Fer(a1-4)Fruf(b2-1)GlcOBz</th>
<th id="T_0aa46_level0_col201" class="col_heading level0 col201" data-quarto-table-cell-role="th">Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col202" class="col_heading level0 col202" data-quarto-table-cell-role="th">Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col203" class="col_heading level0 col203" data-quarto-table-cell-role="th">Glc(b1-6)Glc(b1-3)Glc</th>
<th id="T_0aa46_level0_col204" class="col_heading level0 col204" data-quarto-table-cell-role="th">Glc(b1-6)GlcNAc</th>
<th id="T_0aa46_level0_col205" class="col_heading level0 col205" data-quarto-table-cell-role="th">Glc1Cer</th>
<th id="T_0aa46_level0_col206" class="col_heading level0 col206" data-quarto-table-cell-role="th">Glc2Ac(b1-4)[D-Apif(b1-3)Xyl(b1-2)]Glc</th>
<th id="T_0aa46_level0_col207" class="col_heading level0 col207" data-quarto-table-cell-role="th">Glc2Ac3Ac4Ac6Ac(b1-3)Ara</th>
<th id="T_0aa46_level0_col208" class="col_heading level0 col208" data-quarto-table-cell-role="th">Glc3Ac(b1-2)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col209" class="col_heading level0 col209" data-quarto-table-cell-role="th">Glc6Ac(a1-2)Fru</th>
<th id="T_0aa46_level0_col210" class="col_heading level0 col210" data-quarto-table-cell-role="th">Glc6Ac(b1-2)Glc(a1-2)Fru</th>
<th id="T_0aa46_level0_col211" class="col_heading level0 col211" data-quarto-table-cell-role="th">Glc6Ac(b1-2)Glc(a1-2)FrufOBzOCin</th>
<th id="T_0aa46_level0_col212" class="col_heading level0 col212" data-quarto-table-cell-role="th">Glc6Ac(b1-3)Ara</th>
<th id="T_0aa46_level0_col213" class="col_heading level0 col213" data-quarto-table-cell-role="th">Glc6Ac(b1-3)Glc6Ac(b1-3)[Glc6Ac(b1-2)]Glc1Fer6Ac(a1-2)Fruf1CoumOAcOBz</th>
<th id="T_0aa46_level0_col214" class="col_heading level0 col214" data-quarto-table-cell-role="th">Glc6Ac(b1-3)Glc6Ac(b1-3)[Glc6Ac(b1-2)][RhaOAc(a1-4)]Glc1Fer6Ac(a1-2)Fruf1CoumOAcOBz</th>
<th id="T_0aa46_level0_col215" class="col_heading level0 col215" data-quarto-table-cell-role="th">Glc6Ac(b1-3)[Glc(b1-2)]Glc1Coum(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col216" class="col_heading level0 col216" data-quarto-table-cell-role="th">Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1CoumOBz</th>
<th id="T_0aa46_level0_col217" class="col_heading level0 col217" data-quarto-table-cell-role="th">Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col218" class="col_heading level0 col218" data-quarto-table-cell-role="th">Glc6Ac(b1-3)[Glc(b1-2)]Glc1Fer6Ac(a1-2)Fruf1FerOBz</th>
<th id="T_0aa46_level0_col219" class="col_heading level0 col219" data-quarto-table-cell-role="th">GlcA(b1-2)Glc</th>
<th id="T_0aa46_level0_col220" class="col_heading level0 col220" data-quarto-table-cell-role="th">GlcA(b1-2)GlcA</th>
<th id="T_0aa46_level0_col221" class="col_heading level0 col221" data-quarto-table-cell-role="th">GlcA(b1-2)GlcA(b1-2)Rha</th>
<th id="T_0aa46_level0_col222" class="col_heading level0 col222" data-quarto-table-cell-role="th">GlcA4Me(a1-2)[Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)]Xyl</th>
<th id="T_0aa46_level0_col223" class="col_heading level0 col223" data-quarto-table-cell-role="th">GlcA4Me(a1-2)[Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)]Xyl</th>
<th id="T_0aa46_level0_col224" class="col_heading level0 col224" data-quarto-table-cell-role="th">GlcA4Me(a1-2)[Xyl(b1-4)]Xyl</th>
<th id="T_0aa46_level0_col225" class="col_heading level0 col225" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col226" class="col_heading level0 col226" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col227" class="col_heading level0 col227" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col228" class="col_heading level0 col228" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col229" class="col_heading level0 col229" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Gal(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col230" class="col_heading level0 col230" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col231" class="col_heading level0 col231" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col232" class="col_heading level0 col232" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col233" class="col_heading level0 col233" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col234" class="col_heading level0 col234" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3/6)[Xyl(b1-2)][Man(a1-3/6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col235" class="col_heading level0 col235" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-3/6)[Xyl(b1-2)][Man(a1-3/6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col236" class="col_heading level0 col236" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col237" class="col_heading level0 col237" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col238" class="col_heading level0 col238" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col239" class="col_heading level0 col239" data-quarto-table-cell-role="th">GlcNAc(b1-2)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col240" class="col_heading level0 col240" data-quarto-table-cell-role="th">GlcNAc(b1-4)Man(a1-3)[GlcNAc(b1-4)Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col241" class="col_heading level0 col241" data-quarto-table-cell-role="th">GlcNAc(b1-4)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col242" class="col_heading level0 col242" data-quarto-table-cell-role="th">GlcNAc(b1-4)Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col243" class="col_heading level0 col243" data-quarto-table-cell-role="th">GlcNAc(b1-4)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col244" class="col_heading level0 col244" data-quarto-table-cell-role="th">GlcNAc(b1-4)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col245" class="col_heading level0 col245" data-quarto-table-cell-role="th">GlcOMe(b1-3)[XylOMe(b1-4)]RhaOMe(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col246" class="col_heading level0 col246" data-quarto-table-cell-role="th">Glcf(b1-2)Xyl(b1-4)Rha(b1-4)[Xyl(b1-3)]Xyl</th>
<th id="T_0aa46_level0_col247" class="col_heading level0 col247" data-quarto-table-cell-role="th">Hexf(?1-?)Xyl(b1-4)Rha(b1-4)[Xyl(a1-3)]Xyl</th>
<th id="T_0aa46_level0_col248" class="col_heading level0 col248" data-quarto-table-cell-role="th">L-Lyx(a1-2)Ara(a1-2)GlcA</th>
<th id="T_0aa46_level0_col249" class="col_heading level0 col249" data-quarto-table-cell-role="th">Lyx(a1-2)Ara(a1-2)GlcA</th>
<th id="T_0aa46_level0_col250" class="col_heading level0 col250" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col251" class="col_heading level0 col251" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col252" class="col_heading level0 col252" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col253" class="col_heading level0 col253" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col254" class="col_heading level0 col254" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col255" class="col_heading level0 col255" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col256" class="col_heading level0 col256" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col257" class="col_heading level0 col257" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col258" class="col_heading level0 col258" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col259" class="col_heading level0 col259" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col260" class="col_heading level0 col260" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col261" class="col_heading level0 col261" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col262" class="col_heading level0 col262" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col263" class="col_heading level0 col263" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)Man(a1-6)[Man(a1-2)Man(a1-3)]Man(a1-3)[Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col264" class="col_heading level0 col264" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-2)[Man(a1-6)]Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col265" class="col_heading level0 col265" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col266" class="col_heading level0 col266" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col267" class="col_heading level0 col267" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-3)[Man(a1-2)Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col268" class="col_heading level0 col268" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col269" class="col_heading level0 col269" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col270" class="col_heading level0 col270" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col271" class="col_heading level0 col271" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col272" class="col_heading level0 col272" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col273" class="col_heading level0 col273" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col274" class="col_heading level0 col274" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-6)[Man(a1-2)Man(a1-3)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col275" class="col_heading level0 col275" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAcN</th>
<th id="T_0aa46_level0_col276" class="col_heading level0 col276" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col277" class="col_heading level0 col277" data-quarto-table-cell-role="th">Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col278" class="col_heading level0 col278" data-quarto-table-cell-role="th">Man(a1-2)[Man(a1-3)]Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col279" class="col_heading level0 col279" data-quarto-table-cell-role="th">Man(a1-3)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col280" class="col_heading level0 col280" data-quarto-table-cell-role="th">Man(a1-3)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col281" class="col_heading level0 col281" data-quarto-table-cell-role="th">Man(a1-3)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col282" class="col_heading level0 col282" data-quarto-table-cell-role="th">Man(a1-3)Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col283" class="col_heading level0 col283" data-quarto-table-cell-role="th">Man(a1-3)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col284" class="col_heading level0 col284" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-3)[Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col285" class="col_heading level0 col285" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col286" class="col_heading level0 col286" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col287" class="col_heading level0 col287" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col288" class="col_heading level0 col288" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col289" class="col_heading level0 col289" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col290" class="col_heading level0 col290" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col291" class="col_heading level0 col291" data-quarto-table-cell-role="th">Man(a1-3)[Man(a1-6)][Xylf(a1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col292" class="col_heading level0 col292" data-quarto-table-cell-role="th">Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc-ol</th>
<th id="T_0aa46_level0_col293" class="col_heading level0 col293" data-quarto-table-cell-role="th">Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAcN</th>
<th id="T_0aa46_level0_col294" class="col_heading level0 col294" data-quarto-table-cell-role="th">Man(a1-3)[Xyl(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]Hex</th>
<th id="T_0aa46_level0_col295" class="col_heading level0 col295" data-quarto-table-cell-role="th">Man(a1-3)[Xylf(b1-2)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col296" class="col_heading level0 col296" data-quarto-table-cell-role="th">Man(a1-3/6)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col297" class="col_heading level0 col297" data-quarto-table-cell-role="th">Man(a1-3/6)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col298" class="col_heading level0 col298" data-quarto-table-cell-role="th">Man(a1-3/6)Man(a1-6)[Xyl(b1-2)][Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col299" class="col_heading level0 col299" data-quarto-table-cell-role="th">Man(a1-3/6)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col300" class="col_heading level0 col300" data-quarto-table-cell-role="th">Man(b1-2)Man</th>
<th id="T_0aa46_level0_col301" class="col_heading level0 col301" data-quarto-table-cell-role="th">Man(b1-4)Gal(b1-4)Gal(b1-4)Man</th>
<th id="T_0aa46_level0_col302" class="col_heading level0 col302" data-quarto-table-cell-role="th">Man(b1-4)Gal(b1-4)Gal(b1-4)ManOMe</th>
<th id="T_0aa46_level0_col303" class="col_heading level0 col303" data-quarto-table-cell-role="th">Man(b1-4)Man</th>
<th id="T_0aa46_level0_col304" class="col_heading level0 col304" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col305" class="col_heading level0 col305" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col306" class="col_heading level0 col306" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col307" class="col_heading level0 col307" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col308" class="col_heading level0 col308" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col309" class="col_heading level0 col309" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col310" class="col_heading level0 col310" data-quarto-table-cell-role="th">Man(b1-4)Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)Man(b1-4)Man(b1-4)[Man(b1-6)]Man(b1-4)[Man(b1-6)]Man(b1-4)Man(b1-4)Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-3)Gal(a1-3)Gal(a1-6)]Man(b1-4)Man(b1-4)Man(b1-4)[Man(b1-6)]Man(b1-4)[Man(b1-6)]Man(b1-4)Man(b1-4)Man(b1-4)[Man(b1-6)]Man(b1-4)[Man(b1-6)]Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col311" class="col_heading level0 col311" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col312" class="col_heading level0 col312" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man(b1-4)Man</th>
<th id="T_0aa46_level0_col313" class="col_heading level0 col313" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col314" class="col_heading level0 col314" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col315" class="col_heading level0 col315" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)Man</th>
<th id="T_0aa46_level0_col316" class="col_heading level0 col316" data-quarto-table-cell-role="th">Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man</th>
<th id="T_0aa46_level0_col317" class="col_heading level0 col317" data-quarto-table-cell-role="th">Man(b1-6)Glc</th>
<th id="T_0aa46_level0_col318" class="col_heading level0 col318" data-quarto-table-cell-role="th">Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>
<th id="T_0aa46_level0_col319" class="col_heading level0 col319" data-quarto-table-cell-role="th">Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col320" class="col_heading level0 col320" data-quarto-table-cell-role="th">Rha(a1-2)Ara</th>
<th id="T_0aa46_level0_col321" class="col_heading level0 col321" data-quarto-table-cell-role="th">Rha(a1-2)Ara(a1-2)GlcA</th>
<th id="T_0aa46_level0_col322" class="col_heading level0 col322" data-quarto-table-cell-role="th">Rha(a1-2)Ara(a1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col323" class="col_heading level0 col323" data-quarto-table-cell-role="th">Rha(a1-2)Ara(a1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col324" class="col_heading level0 col324" data-quarto-table-cell-role="th">Rha(a1-2)D-Ara(b1-2)GlcA</th>
<th id="T_0aa46_level0_col325" class="col_heading level0 col325" data-quarto-table-cell-role="th">Rha(a1-2)Gal(b1-2)Glc</th>
<th id="T_0aa46_level0_col326" class="col_heading level0 col326" data-quarto-table-cell-role="th">Rha(a1-2)Gal(b1-2)GlcA</th>
<th id="T_0aa46_level0_col327" class="col_heading level0 col327" data-quarto-table-cell-role="th">Rha(a1-2)Gal(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col328" class="col_heading level0 col328" data-quarto-table-cell-role="th">Rha(a1-2)Gal(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col329" class="col_heading level0 col329" data-quarto-table-cell-role="th">Rha(a1-2)Glc</th>
<th id="T_0aa46_level0_col330" class="col_heading level0 col330" data-quarto-table-cell-role="th">Rha(a1-2)Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col331" class="col_heading level0 col331" data-quarto-table-cell-role="th">Rha(a1-2)Glc(b1-2)GlcA</th>
<th id="T_0aa46_level0_col332" class="col_heading level0 col332" data-quarto-table-cell-role="th">Rha(a1-2)Glc(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col333" class="col_heading level0 col333" data-quarto-table-cell-role="th">Rha(a1-2)Glc(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col334" class="col_heading level0 col334" data-quarto-table-cell-role="th">Rha(a1-2)Glc(b1-6)Glc</th>
<th id="T_0aa46_level0_col335" class="col_heading level0 col335" data-quarto-table-cell-role="th">Rha(a1-2)GlcA(b1-2)GlcA</th>
<th id="T_0aa46_level0_col336" class="col_heading level0 col336" data-quarto-table-cell-role="th">Rha(a1-2)GlcAOMe(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col337" class="col_heading level0 col337" data-quarto-table-cell-role="th">Rha(a1-2)Rha(a1-2)Gal(b1-4)[Glc(b1-2)]GlcA</th>
<th id="T_0aa46_level0_col338" class="col_heading level0 col338" data-quarto-table-cell-role="th">Rha(a1-2)Xyl</th>
<th id="T_0aa46_level0_col339" class="col_heading level0 col339" data-quarto-table-cell-role="th">Rha(a1-2)Xyl(b1-2)Glc</th>
<th id="T_0aa46_level0_col340" class="col_heading level0 col340" data-quarto-table-cell-role="th">Rha(a1-2)Xyl(b1-2)GlcA</th>
<th id="T_0aa46_level0_col341" class="col_heading level0 col341" data-quarto-table-cell-role="th">Rha(a1-2)Xyl(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col342" class="col_heading level0 col342" data-quarto-table-cell-role="th">Rha(a1-2)Xyl(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col343" class="col_heading level0 col343" data-quarto-table-cell-role="th">Rha(a1-2)Xyl3Ac</th>
<th id="T_0aa46_level0_col344" class="col_heading level0 col344" data-quarto-table-cell-role="th">Rha(a1-2)Xyl4Ac</th>
<th id="T_0aa46_level0_col345" class="col_heading level0 col345" data-quarto-table-cell-role="th">Rha(a1-2)[Glc(b1-3)]Glc</th>
<th id="T_0aa46_level0_col346" class="col_heading level0 col346" data-quarto-table-cell-role="th">Rha(a1-2)[Glc(b1-6)]Gal(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col347" class="col_heading level0 col347" data-quarto-table-cell-role="th">Rha(a1-2)[Rha(a1-4)]Glc</th>
<th id="T_0aa46_level0_col348" class="col_heading level0 col348" data-quarto-table-cell-role="th">Rha(a1-2)[Rha(a1-6)]Gal</th>
<th id="T_0aa46_level0_col349" class="col_heading level0 col349" data-quarto-table-cell-role="th">Rha(a1-2)[Rha(a1-6)]Glc</th>
<th id="T_0aa46_level0_col350" class="col_heading level0 col350" data-quarto-table-cell-role="th">Rha(a1-2)[Xyl(b1-4)]Glc</th>
<th id="T_0aa46_level0_col351" class="col_heading level0 col351" data-quarto-table-cell-role="th">Rha(a1-2)[Xyl(b1-4)]Glc(b1-6)Glc</th>
<th id="T_0aa46_level0_col352" class="col_heading level0 col352" data-quarto-table-cell-role="th">Rha(a1-3)GlcA</th>
<th id="T_0aa46_level0_col353" class="col_heading level0 col353" data-quarto-table-cell-role="th">Rha(a1-3)[Rha(a1-4)]Gal</th>
<th id="T_0aa46_level0_col354" class="col_heading level0 col354" data-quarto-table-cell-role="th">Rha(a1-4)Gal(b1-2)GlcA</th>
<th id="T_0aa46_level0_col355" class="col_heading level0 col355" data-quarto-table-cell-role="th">Rha(a1-4)Gal(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col356" class="col_heading level0 col356" data-quarto-table-cell-role="th">Rha(a1-4)Gal(b1-2)GlcOMe</th>
<th id="T_0aa46_level0_col357" class="col_heading level0 col357" data-quarto-table-cell-role="th">Rha(a1-4)Gal(b1-4)Gal(b1-4)GalGro</th>
<th id="T_0aa46_level0_col358" class="col_heading level0 col358" data-quarto-table-cell-role="th">Rha(a1-4)Xyl(b1-2)Glc</th>
<th id="T_0aa46_level0_col359" class="col_heading level0 col359" data-quarto-table-cell-role="th">Rha(a1-4)Xyl(b1-2)GlcA</th>
<th id="T_0aa46_level0_col360" class="col_heading level0 col360" data-quarto-table-cell-role="th">Rha(a1-4)Xyl(b1-2)GlcAOMe</th>
<th id="T_0aa46_level0_col361" class="col_heading level0 col361" data-quarto-table-cell-role="th">Rha(a1-6)Glc</th>
<th id="T_0aa46_level0_col362" class="col_heading level0 col362" data-quarto-table-cell-role="th">Rha(a1-6)[Xyl(b1-3)Xyl(b1-2)]Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col363" class="col_heading level0 col363" data-quarto-table-cell-role="th">Rha(b1-2)Glc(b1-2)GlcA</th>
<th id="T_0aa46_level0_col364" class="col_heading level0 col364" data-quarto-table-cell-role="th">Rha1Fer(a1-4)Fruf(b2-1)GlcOBz</th>
<th id="T_0aa46_level0_col365" class="col_heading level0 col365" data-quarto-table-cell-role="th">RhaOMe(a1-2)[RhaOMe(a1-6)]GlcOMe-ol</th>
<th id="T_0aa46_level0_col366" class="col_heading level0 col366" data-quarto-table-cell-role="th">RhaOMe(a1-6)GlcOMe(b1-2)GlcOMe-ol</th>
<th id="T_0aa46_level0_col367" class="col_heading level0 col367" data-quarto-table-cell-role="th">Xyl(a1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col368" class="col_heading level0 col368" data-quarto-table-cell-role="th">Xyl(a1-3)Fuc(a1-4)Rha</th>
<th id="T_0aa46_level0_col369" class="col_heading level0 col369" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col370" class="col_heading level0 col370" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col371" class="col_heading level0 col371" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col372" class="col_heading level0 col372" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col373" class="col_heading level0 col373" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col374" class="col_heading level0 col374" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col375" class="col_heading level0 col375" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col376" class="col_heading level0 col376" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col377" class="col_heading level0 col377" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col378" class="col_heading level0 col378" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col379" class="col_heading level0 col379" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col380" class="col_heading level0 col380" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col381" class="col_heading level0 col381" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col382" class="col_heading level0 col382" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col383" class="col_heading level0 col383" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col384" class="col_heading level0 col384" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col385" class="col_heading level0 col385" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col386" class="col_heading level0 col386" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col387" class="col_heading level0 col387" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col388" class="col_heading level0 col388" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col389" class="col_heading level0 col389" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col390" class="col_heading level0 col390" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col391" class="col_heading level0 col391" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col392" class="col_heading level0 col392" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col393" class="col_heading level0 col393" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col394" class="col_heading level0 col394" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col395" class="col_heading level0 col395" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col396" class="col_heading level0 col396" data-quarto-table-cell-role="th">Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc-ol</th>
<th id="T_0aa46_level0_col397" class="col_heading level0 col397" data-quarto-table-cell-role="th">Xyl(b1-2)Ara(a1-6)Glc</th>
<th id="T_0aa46_level0_col398" class="col_heading level0 col398" data-quarto-table-cell-role="th">Xyl(b1-2)Ara(a1-6)GlcNAc</th>
<th id="T_0aa46_level0_col399" class="col_heading level0 col399" data-quarto-table-cell-role="th">Xyl(b1-2)Ara(a1-6)[Glc(b1-2)]Glc</th>
<th id="T_0aa46_level0_col400" class="col_heading level0 col400" data-quarto-table-cell-role="th">Xyl(b1-2)Ara(a1-6)[Glc(b1-4)]GlcNAc</th>
<th id="T_0aa46_level0_col401" class="col_heading level0 col401" data-quarto-table-cell-role="th">Xyl(b1-2)D-Fuc(b1-6)Glc</th>
<th id="T_0aa46_level0_col402" class="col_heading level0 col402" data-quarto-table-cell-role="th">Xyl(b1-2)D-Fuc(b1-6)GlcNAc</th>
<th id="T_0aa46_level0_col403" class="col_heading level0 col403" data-quarto-table-cell-role="th">Xyl(b1-2)D-Fuc(b1-6)[Glc(b1-2)]Glc</th>
<th id="T_0aa46_level0_col404" class="col_heading level0 col404" data-quarto-table-cell-role="th">Xyl(b1-2)Fuc(a1-6)Glc</th>
<th id="T_0aa46_level0_col405" class="col_heading level0 col405" data-quarto-table-cell-role="th">Xyl(b1-2)Fuc(a1-6)GlcNAc</th>
<th id="T_0aa46_level0_col406" class="col_heading level0 col406" data-quarto-table-cell-role="th">Xyl(b1-2)Fuc(b1-6)Glc</th>
<th id="T_0aa46_level0_col407" class="col_heading level0 col407" data-quarto-table-cell-role="th">Xyl(b1-2)Fuc(b1-6)GlcNAc</th>
<th id="T_0aa46_level0_col408" class="col_heading level0 col408" data-quarto-table-cell-role="th">Xyl(b1-2)Fuc(b1-6)[Glc(b1-2)]Glc</th>
<th id="T_0aa46_level0_col409" class="col_heading level0 col409" data-quarto-table-cell-role="th">Xyl(b1-2)Gal(b1-2)GlcA6Me</th>
<th id="T_0aa46_level0_col410" class="col_heading level0 col410" data-quarto-table-cell-role="th">Xyl(b1-2)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col411" class="col_heading level0 col411" data-quarto-table-cell-role="th">Xyl(b1-2)Rha(a1-2)Ara</th>
<th id="T_0aa46_level0_col412" class="col_heading level0 col412" data-quarto-table-cell-role="th">Xyl(b1-2)Xyl(b1-3)[Rha(b1-2)Rha(b1-4)]Xyl</th>
<th id="T_0aa46_level0_col413" class="col_heading level0 col413" data-quarto-table-cell-role="th">Xyl(b1-2)[Glc(b1-3)]Ara</th>
<th id="T_0aa46_level0_col414" class="col_heading level0 col414" data-quarto-table-cell-role="th">Xyl(b1-2)[Glc2Ac(b1-4)]Glc</th>
<th id="T_0aa46_level0_col415" class="col_heading level0 col415" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(a1-3)Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col416" class="col_heading level0 col416" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(a1-3)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col417" class="col_heading level0 col417" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(a1-6)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col418" class="col_heading level0 col418" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col419" class="col_heading level0 col419" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col420" class="col_heading level0 col420" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col421" class="col_heading level0 col421" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc-ol</th>
<th id="T_0aa46_level0_col422" class="col_heading level0 col422" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc</th>
<th id="T_0aa46_level0_col423" class="col_heading level0 col423" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)ManNAc</th>
<th id="T_0aa46_level0_col424" class="col_heading level0 col424" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-6)]Man(a1-3)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col425" class="col_heading level0 col425" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
<th id="T_0aa46_level0_col426" class="col_heading level0 col426" data-quarto-table-cell-role="th">Xyl(b1-2)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
<th id="T_0aa46_level0_col427" class="col_heading level0 col427" data-quarto-table-cell-role="th">Xyl(b1-2)[Rha(a1-3)]GlcA</th>
<th id="T_0aa46_level0_col428" class="col_heading level0 col428" data-quarto-table-cell-role="th">Xyl(b1-3)Ara</th>
<th id="T_0aa46_level0_col429" class="col_heading level0 col429" data-quarto-table-cell-role="th">Xyl(b1-3)Xyl(b1-2)[Rha(a1-6)]Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col430" class="col_heading level0 col430" data-quarto-table-cell-role="th">Xyl(b1-3)Xyl(b1-4)Rha(a1-2)[Rha(a1-6)]Glc</th>
<th id="T_0aa46_level0_col431" class="col_heading level0 col431" data-quarto-table-cell-role="th">Xyl(b1-3)Xyl(b1-4)Rha(a1-2)[Rha(a1-6)]Glc(b1-2)Glc</th>
<th id="T_0aa46_level0_col432" class="col_heading level0 col432" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)Ara</th>
<th id="T_0aa46_level0_col433" class="col_heading level0 col433" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)D-Fuc</th>
<th id="T_0aa46_level0_col434" class="col_heading level0 col434" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col435" class="col_heading level0 col435" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)Fuc</th>
<th id="T_0aa46_level0_col436" class="col_heading level0 col436" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)Fuc3Ac</th>
<th id="T_0aa46_level0_col437" class="col_heading level0 col437" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)Fuc4Ac</th>
<th id="T_0aa46_level0_col438" class="col_heading level0 col438" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)Glc</th>
<th id="T_0aa46_level0_col439" class="col_heading level0 col439" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)[Rha(a1-3)]Fuc4Ac</th>
<th id="T_0aa46_level0_col440" class="col_heading level0 col440" data-quarto-table-cell-role="th">Xyl(b1-4)Rha(a1-2)[Rha(a1-6)]Glc</th>
<th id="T_0aa46_level0_col441" class="col_heading level0 col441" data-quarto-table-cell-role="th">Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl3Ac(b1-4)Xyl(b1-4)Xyl(b1-4)[GlcA(a1-2)]Xyl(b1-4)Xyl</th>
<th id="T_0aa46_level0_col442" class="col_heading level0 col442" data-quarto-table-cell-role="th">Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl3Ac(b1-4)Xyl(b1-4)Xyl(b1-4)[GlcA(a1-2)]Xyl3Ac(b1-4)Xyl</th>
<th id="T_0aa46_level0_col443" class="col_heading level0 col443" data-quarto-table-cell-role="th">Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl3Ac(b1-4)Xyl(b1-4)Xyl(b1-4)[GlcA4Me(a1-2)]Xyl(b1-4)Xyl</th>
<th id="T_0aa46_level0_col444" class="col_heading level0 col444" data-quarto-table-cell-role="th">Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl3Ac(b1-4)Xyl(b1-4)Xyl(b1-4)[GlcA4Me(a1-2)]Xyl3Ac(b1-4)Xyl</th>
<th id="T_0aa46_level0_col445" class="col_heading level0 col445" data-quarto-table-cell-role="th">Xyl(b1-4)Xyl(b1-4)[GlcA(a1-2)]Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl</th>
<th id="T_0aa46_level0_col446" class="col_heading level0 col446" data-quarto-table-cell-role="th">Xyl(b1-4)[GlcAOMe(a1-2)]Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl</th>
<th id="T_0aa46_level0_col447" class="col_heading level0 col447" data-quarto-table-cell-role="th">Xyl2Ac3Ac4Ac(b1-3)Ara</th>
<th id="T_0aa46_level0_col448" class="col_heading level0 col448" data-quarto-table-cell-role="th">Xyl4Ac(b1-3)Ara</th>
<th id="T_0aa46_level0_col449" class="col_heading level0 col449" data-quarto-table-cell-role="th">XylOMe(b1-2)[RhaOMe(a1-6)]GlcOMe(b1-2)GlcOMe-ol</th>
<th id="T_0aa46_level0_col450" class="col_heading level0 col450" data-quarto-table-cell-role="th">XylOMe(b1-3)XylOMe(b1-2)[RhaOMe(a1-6)]GlcOMe(b1-2)GlcOMe-ol</th>
<th id="T_0aa46_level0_col451" class="col_heading level0 col451" data-quarto-table-cell-role="th">XylOMe(b1-4)RhaOMe(a1-2)D-FucOMe</th>
<th id="T_0aa46_level0_col452" class="col_heading level0 col452" data-quarto-table-cell-role="th">XylOMe(b1-4)RhaOMe(a1-2)[RhaOMe(a1-6)]GlcOMe</th>
<th id="T_0aa46_level0_col453" class="col_heading level0 col453" data-quarto-table-cell-role="th">XylOMe(b1-4)RhaOMe(a1-2)[RhaOMe(a1-6)]GlcOMe-ol</th>
<th id="T_0aa46_level0_col454" class="col_heading level0 col454" data-quarto-table-cell-role="th">Xylf(b1-2)Xyl(b1-3)[Rha(b1-2)Rha(b1-4)]Xyl</th>
<th id="T_0aa46_level0_col455" class="col_heading level0 col455" data-quarto-table-cell-role="th">[Araf(a1-3)Gal(b1-3)Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col456" class="col_heading level0 col456" data-quarto-table-cell-role="th">[Araf(a1-3)Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col457" class="col_heading level0 col457" data-quarto-table-cell-role="th">[Gal(a1-4)Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)[Man(b1-4)Man(b1-4)Man(b1-4)Gal(a1-6)]Man(b1-2)[Gal(a1-6)]Man(b1-2)[Gal(a1-4)Gal(a1-6)]Man(b1-4)Man</th>
<th id="T_0aa46_level0_col458" class="col_heading level0 col458" data-quarto-table-cell-role="th">[Gal(a1-6)]Man(b1-4)Man</th>
<th id="T_0aa46_level0_col459" class="col_heading level0 col459" data-quarto-table-cell-role="th">[Gal(a1-6)]Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col460" class="col_heading level0 col460" data-quarto-table-cell-role="th">[Gal(a1-6)]Man(b1-4)Man(b1-4)Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col461" class="col_heading level0 col461" data-quarto-table-cell-role="th">[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)Man(b1-4)Man</th>
<th id="T_0aa46_level0_col462" class="col_heading level0 col462" data-quarto-table-cell-role="th">[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)[Gal(a1-6)]Man(b1-4)Man</th>
<th id="T_0aa46_level0_col463" class="col_heading level0 col463" data-quarto-table-cell-role="th">[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col464" class="col_heading level0 col464" data-quarto-table-cell-role="th">[Gal(b1-3)Gal(b1-6)[Araf(a1-3)]Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col465" class="col_heading level0 col465" data-quarto-table-cell-role="th">[Gal(b1-3)Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col466" class="col_heading level0 col466" data-quarto-table-cell-role="th">[Gal(b1-6)Gal(b1-6)Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col467" class="col_heading level0 col467" data-quarto-table-cell-role="th">[Gal(b1-6)Gal(b1-6)]Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col468" class="col_heading level0 col468" data-quarto-table-cell-role="th">[Gal(b1-6)]Gal(b1-3)Gal(b1-3)Gal(b1-3)Gal(b1-3)Gal(b1-3)Gal(b1-3)Gal</th>
<th id="T_0aa46_level0_col469" class="col_heading level0 col469" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Araf(a1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col470" class="col_heading level0 col470" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Araf(a1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col471" class="col_heading level0 col471" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Araf(a1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col472" class="col_heading level0 col472" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Araf(a1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col473" class="col_heading level0 col473" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-5)Araf(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col474" class="col_heading level0 col474" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col475" class="col_heading level0 col475" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-5)Araf(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col476" class="col_heading level0 col476" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col477" class="col_heading level0 col477" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Fuc(a1-2)Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col478" class="col_heading level0 col478" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-5)Araf(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col479" class="col_heading level0 col479" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col480" class="col_heading level0 col480" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-5)Araf(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col481" class="col_heading level0 col481" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col482" class="col_heading level0 col482" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Gal(b1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
<th id="T_0aa46_level0_col483" class="col_heading level0 col483" data-quarto-table-cell-role="th">[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)[Gal(b1-5)Araf(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc</th>
</tr>
<tr class="even">
<th class="index_name level0" data-quarto-table-cell-role="th">Family</th>
<th class="blank col0" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col1" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col2" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col3" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col4" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col5" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col6" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col7" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col8" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col9" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col10" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col11" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col12" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col13" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col14" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col15" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col16" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col17" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col18" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col19" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col20" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col21" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col22" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col23" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col24" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col25" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col26" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col27" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col28" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col29" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col30" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col31" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col32" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col33" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col34" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col35" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col36" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col37" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col38" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col39" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col40" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col41" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col42" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col43" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col44" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col45" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col46" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col47" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col48" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col49" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col50" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col51" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col52" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col53" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col54" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col55" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col56" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col57" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col58" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col59" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col60" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col61" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col62" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col63" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col64" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col65" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col66" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col67" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col68" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col69" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col70" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col71" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col72" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col73" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col74" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col75" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col76" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col77" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col78" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col79" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col80" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col81" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col82" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col83" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col84" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col85" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col86" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col87" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col88" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col89" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col90" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col91" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col92" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col93" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col94" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col95" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col96" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col97" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col98" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col99" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col100" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col101" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col102" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col103" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col104" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col105" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col106" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col107" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col108" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col109" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col110" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col111" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col112" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col113" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col114" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col115" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col116" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col117" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col118" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col119" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col120" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col121" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col122" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col123" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col124" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col125" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col126" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col127" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col128" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col129" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col130" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col131" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col132" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col133" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col134" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col135" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col136" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col137" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col138" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col139" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col140" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col141" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col142" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col143" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col144" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col145" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col146" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col147" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col148" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col149" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col150" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col151" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col152" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col153" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col154" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col155" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col156" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col157" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col158" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col159" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col160" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col161" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col162" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col163" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col164" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col165" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col166" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col167" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col168" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col169" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col170" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col171" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col172" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col173" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col174" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col175" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col176" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col177" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col178" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col179" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col180" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col181" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col182" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col183" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col184" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col185" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col186" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col187" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col188" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col189" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col190" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col191" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col192" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col193" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col194" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col195" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col196" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col197" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col198" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col199" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col200" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col201" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col202" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col203" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col204" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col205" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col206" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col207" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col208" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col209" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col210" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col211" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col212" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col213" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col214" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col215" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col216" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col217" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col218" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col219" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col220" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col221" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col222" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col223" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col224" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col225" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col226" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col227" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col228" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col229" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col230" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col231" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col232" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col233" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col234" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col235" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col236" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col237" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col238" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col239" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col240" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col241" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col242" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col243" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col244" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col245" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col246" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col247" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col248" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col249" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col250" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col251" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col252" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col253" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col254" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col255" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col256" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col257" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col258" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col259" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col260" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col261" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col262" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col263" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col264" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col265" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col266" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col267" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col268" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col269" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col270" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col271" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col272" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col273" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col274" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col275" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col276" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col277" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col278" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col279" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col280" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col281" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col282" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col283" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col284" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col285" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col286" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col287" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col288" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col289" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col290" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col291" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col292" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col293" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col294" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col295" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col296" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col297" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col298" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col299" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col300" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col301" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col302" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col303" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col304" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col305" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col306" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col307" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col308" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col309" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col310" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col311" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col312" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col313" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col314" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col315" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col316" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col317" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col318" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col319" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col320" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col321" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col322" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col323" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col324" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col325" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col326" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col327" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col328" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col329" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col330" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col331" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col332" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col333" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col334" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col335" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col336" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col337" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col338" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col339" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col340" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col341" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col342" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col343" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col344" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col345" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col346" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col347" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col348" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col349" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col350" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col351" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col352" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col353" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col354" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col355" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col356" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col357" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col358" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col359" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col360" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col361" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col362" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col363" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col364" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col365" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col366" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col367" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col368" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col369" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col370" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col371" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col372" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col373" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col374" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col375" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col376" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col377" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col378" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col379" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col380" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col381" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col382" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col383" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col384" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col385" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col386" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col387" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col388" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col389" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col390" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col391" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col392" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col393" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col394" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col395" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col396" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col397" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col398" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col399" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col400" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col401" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col402" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col403" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col404" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col405" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col406" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col407" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col408" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col409" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col410" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col411" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col412" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col413" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col414" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col415" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col416" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col417" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col418" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col419" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col420" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col421" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col422" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col423" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col424" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col425" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col426" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col427" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col428" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col429" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col430" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col431" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col432" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col433" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col434" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col435" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col436" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col437" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col438" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col439" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col440" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col441" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col442" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col443" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col444" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col445" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col446" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col447" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col448" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col449" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col450" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col451" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col452" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col453" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col454" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col455" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col456" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col457" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col458" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col459" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col460" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col461" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col462" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col463" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col464" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col465" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col466" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col467" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col468" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col469" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col470" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col471" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col472" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col473" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col474" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col475" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col476" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col477" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col478" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col479" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col480" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col481" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col482" data-quarto-table-cell-role="th">&nbsp;</th>
<th class="blank col483" data-quarto-table-cell-role="th">&nbsp;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th id="T_0aa46_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">Fabaceae</th>
<td id="T_0aa46_row0_col0" class="data row0 col0">1</td>
<td id="T_0aa46_row0_col1" class="data row0 col1">4</td>
<td id="T_0aa46_row0_col2" class="data row0 col2">1</td>
<td id="T_0aa46_row0_col3" class="data row0 col3">3</td>
<td id="T_0aa46_row0_col4" class="data row0 col4">1</td>
<td id="T_0aa46_row0_col5" class="data row0 col5">1</td>
<td id="T_0aa46_row0_col6" class="data row0 col6">1</td>
<td id="T_0aa46_row0_col7" class="data row0 col7">0</td>
<td id="T_0aa46_row0_col8" class="data row0 col8">1</td>
<td id="T_0aa46_row0_col9" class="data row0 col9">3</td>
<td id="T_0aa46_row0_col10" class="data row0 col10">1</td>
<td id="T_0aa46_row0_col11" class="data row0 col11">1</td>
<td id="T_0aa46_row0_col12" class="data row0 col12">1</td>
<td id="T_0aa46_row0_col13" class="data row0 col13">0</td>
<td id="T_0aa46_row0_col14" class="data row0 col14">0</td>
<td id="T_0aa46_row0_col15" class="data row0 col15">0</td>
<td id="T_0aa46_row0_col16" class="data row0 col16">0</td>
<td id="T_0aa46_row0_col17" class="data row0 col17">0</td>
<td id="T_0aa46_row0_col18" class="data row0 col18">1</td>
<td id="T_0aa46_row0_col19" class="data row0 col19">0</td>
<td id="T_0aa46_row0_col20" class="data row0 col20">0</td>
<td id="T_0aa46_row0_col21" class="data row0 col21">0</td>
<td id="T_0aa46_row0_col22" class="data row0 col22">0</td>
<td id="T_0aa46_row0_col23" class="data row0 col23">0</td>
<td id="T_0aa46_row0_col24" class="data row0 col24">0</td>
<td id="T_0aa46_row0_col25" class="data row0 col25">0</td>
<td id="T_0aa46_row0_col26" class="data row0 col26">0</td>
<td id="T_0aa46_row0_col27" class="data row0 col27">0</td>
<td id="T_0aa46_row0_col28" class="data row0 col28">0</td>
<td id="T_0aa46_row0_col29" class="data row0 col29">0</td>
<td id="T_0aa46_row0_col30" class="data row0 col30">0</td>
<td id="T_0aa46_row0_col31" class="data row0 col31">0</td>
<td id="T_0aa46_row0_col32" class="data row0 col32">0</td>
<td id="T_0aa46_row0_col33" class="data row0 col33">0</td>
<td id="T_0aa46_row0_col34" class="data row0 col34">0</td>
<td id="T_0aa46_row0_col35" class="data row0 col35">0</td>
<td id="T_0aa46_row0_col36" class="data row0 col36">0</td>
<td id="T_0aa46_row0_col37" class="data row0 col37">0</td>
<td id="T_0aa46_row0_col38" class="data row0 col38">0</td>
<td id="T_0aa46_row0_col39" class="data row0 col39">0</td>
<td id="T_0aa46_row0_col40" class="data row0 col40">0</td>
<td id="T_0aa46_row0_col41" class="data row0 col41">1</td>
<td id="T_0aa46_row0_col42" class="data row0 col42">2</td>
<td id="T_0aa46_row0_col43" class="data row0 col43">1</td>
<td id="T_0aa46_row0_col44" class="data row0 col44">1</td>
<td id="T_0aa46_row0_col45" class="data row0 col45">1</td>
<td id="T_0aa46_row0_col46" class="data row0 col46">1</td>
<td id="T_0aa46_row0_col47" class="data row0 col47">2</td>
<td id="T_0aa46_row0_col48" class="data row0 col48">1</td>
<td id="T_0aa46_row0_col49" class="data row0 col49">1</td>
<td id="T_0aa46_row0_col50" class="data row0 col50">1</td>
<td id="T_0aa46_row0_col51" class="data row0 col51">1</td>
<td id="T_0aa46_row0_col52" class="data row0 col52">4</td>
<td id="T_0aa46_row0_col53" class="data row0 col53">2</td>
<td id="T_0aa46_row0_col54" class="data row0 col54">1</td>
<td id="T_0aa46_row0_col55" class="data row0 col55">1</td>
<td id="T_0aa46_row0_col56" class="data row0 col56">2</td>
<td id="T_0aa46_row0_col57" class="data row0 col57">2</td>
<td id="T_0aa46_row0_col58" class="data row0 col58">7</td>
<td id="T_0aa46_row0_col59" class="data row0 col59">7</td>
<td id="T_0aa46_row0_col60" class="data row0 col60">4</td>
<td id="T_0aa46_row0_col61" class="data row0 col61">4</td>
<td id="T_0aa46_row0_col62" class="data row0 col62">4</td>
<td id="T_0aa46_row0_col63" class="data row0 col63">4</td>
<td id="T_0aa46_row0_col64" class="data row0 col64">4</td>
<td id="T_0aa46_row0_col65" class="data row0 col65">2</td>
<td id="T_0aa46_row0_col66" class="data row0 col66">8</td>
<td id="T_0aa46_row0_col67" class="data row0 col67">4</td>
<td id="T_0aa46_row0_col68" class="data row0 col68">2</td>
<td id="T_0aa46_row0_col69" class="data row0 col69">5</td>
<td id="T_0aa46_row0_col70" class="data row0 col70">4</td>
<td id="T_0aa46_row0_col71" class="data row0 col71">2</td>
<td id="T_0aa46_row0_col72" class="data row0 col72">2</td>
<td id="T_0aa46_row0_col73" class="data row0 col73">1</td>
<td id="T_0aa46_row0_col74" class="data row0 col74">1</td>
<td id="T_0aa46_row0_col75" class="data row0 col75">1</td>
<td id="T_0aa46_row0_col76" class="data row0 col76">1</td>
<td id="T_0aa46_row0_col77" class="data row0 col77">0</td>
<td id="T_0aa46_row0_col78" class="data row0 col78">1</td>
<td id="T_0aa46_row0_col79" class="data row0 col79">1</td>
<td id="T_0aa46_row0_col80" class="data row0 col80">3</td>
<td id="T_0aa46_row0_col81" class="data row0 col81">1</td>
<td id="T_0aa46_row0_col82" class="data row0 col82">1</td>
<td id="T_0aa46_row0_col83" class="data row0 col83">2</td>
<td id="T_0aa46_row0_col84" class="data row0 col84">1</td>
<td id="T_0aa46_row0_col85" class="data row0 col85">1</td>
<td id="T_0aa46_row0_col86" class="data row0 col86">1</td>
<td id="T_0aa46_row0_col87" class="data row0 col87">1</td>
<td id="T_0aa46_row0_col88" class="data row0 col88">2</td>
<td id="T_0aa46_row0_col89" class="data row0 col89">5</td>
<td id="T_0aa46_row0_col90" class="data row0 col90">1</td>
<td id="T_0aa46_row0_col91" class="data row0 col91">1</td>
<td id="T_0aa46_row0_col92" class="data row0 col92">2</td>
<td id="T_0aa46_row0_col93" class="data row0 col93">2</td>
<td id="T_0aa46_row0_col94" class="data row0 col94">1</td>
<td id="T_0aa46_row0_col95" class="data row0 col95">1</td>
<td id="T_0aa46_row0_col96" class="data row0 col96">1</td>
<td id="T_0aa46_row0_col97" class="data row0 col97">1</td>
<td id="T_0aa46_row0_col98" class="data row0 col98">2</td>
<td id="T_0aa46_row0_col99" class="data row0 col99">1</td>
<td id="T_0aa46_row0_col100" class="data row0 col100">1</td>
<td id="T_0aa46_row0_col101" class="data row0 col101">1</td>
<td id="T_0aa46_row0_col102" class="data row0 col102">1</td>
<td id="T_0aa46_row0_col103" class="data row0 col103">1</td>
<td id="T_0aa46_row0_col104" class="data row0 col104">3</td>
<td id="T_0aa46_row0_col105" class="data row0 col105">2</td>
<td id="T_0aa46_row0_col106" class="data row0 col106">0</td>
<td id="T_0aa46_row0_col107" class="data row0 col107">0</td>
<td id="T_0aa46_row0_col108" class="data row0 col108">0</td>
<td id="T_0aa46_row0_col109" class="data row0 col109">0</td>
<td id="T_0aa46_row0_col110" class="data row0 col110">0</td>
<td id="T_0aa46_row0_col111" class="data row0 col111">0</td>
<td id="T_0aa46_row0_col112" class="data row0 col112">0</td>
<td id="T_0aa46_row0_col113" class="data row0 col113">0</td>
<td id="T_0aa46_row0_col114" class="data row0 col114">0</td>
<td id="T_0aa46_row0_col115" class="data row0 col115">0</td>
<td id="T_0aa46_row0_col116" class="data row0 col116">0</td>
<td id="T_0aa46_row0_col117" class="data row0 col117">0</td>
<td id="T_0aa46_row0_col118" class="data row0 col118">0</td>
<td id="T_0aa46_row0_col119" class="data row0 col119">1</td>
<td id="T_0aa46_row0_col120" class="data row0 col120">1</td>
<td id="T_0aa46_row0_col121" class="data row0 col121">1</td>
<td id="T_0aa46_row0_col122" class="data row0 col122">1</td>
<td id="T_0aa46_row0_col123" class="data row0 col123">0</td>
<td id="T_0aa46_row0_col124" class="data row0 col124">0</td>
<td id="T_0aa46_row0_col125" class="data row0 col125">0</td>
<td id="T_0aa46_row0_col126" class="data row0 col126">0</td>
<td id="T_0aa46_row0_col127" class="data row0 col127">1</td>
<td id="T_0aa46_row0_col128" class="data row0 col128">0</td>
<td id="T_0aa46_row0_col129" class="data row0 col129">1</td>
<td id="T_0aa46_row0_col130" class="data row0 col130">1</td>
<td id="T_0aa46_row0_col131" class="data row0 col131">1</td>
<td id="T_0aa46_row0_col132" class="data row0 col132">1</td>
<td id="T_0aa46_row0_col133" class="data row0 col133">1</td>
<td id="T_0aa46_row0_col134" class="data row0 col134">1</td>
<td id="T_0aa46_row0_col135" class="data row0 col135">3</td>
<td id="T_0aa46_row0_col136" class="data row0 col136">1</td>
<td id="T_0aa46_row0_col137" class="data row0 col137">1</td>
<td id="T_0aa46_row0_col138" class="data row0 col138">1</td>
<td id="T_0aa46_row0_col139" class="data row0 col139">1</td>
<td id="T_0aa46_row0_col140" class="data row0 col140">1</td>
<td id="T_0aa46_row0_col141" class="data row0 col141">2</td>
<td id="T_0aa46_row0_col142" class="data row0 col142">2</td>
<td id="T_0aa46_row0_col143" class="data row0 col143">1</td>
<td id="T_0aa46_row0_col144" class="data row0 col144">3</td>
<td id="T_0aa46_row0_col145" class="data row0 col145">1</td>
<td id="T_0aa46_row0_col146" class="data row0 col146">5</td>
<td id="T_0aa46_row0_col147" class="data row0 col147">0</td>
<td id="T_0aa46_row0_col148" class="data row0 col148">0</td>
<td id="T_0aa46_row0_col149" class="data row0 col149">1</td>
<td id="T_0aa46_row0_col150" class="data row0 col150">3</td>
<td id="T_0aa46_row0_col151" class="data row0 col151">1</td>
<td id="T_0aa46_row0_col152" class="data row0 col152">1</td>
<td id="T_0aa46_row0_col153" class="data row0 col153">1</td>
<td id="T_0aa46_row0_col154" class="data row0 col154">2</td>
<td id="T_0aa46_row0_col155" class="data row0 col155">0</td>
<td id="T_0aa46_row0_col156" class="data row0 col156">0</td>
<td id="T_0aa46_row0_col157" class="data row0 col157">0</td>
<td id="T_0aa46_row0_col158" class="data row0 col158">0</td>
<td id="T_0aa46_row0_col159" class="data row0 col159">0</td>
<td id="T_0aa46_row0_col160" class="data row0 col160">0</td>
<td id="T_0aa46_row0_col161" class="data row0 col161">2</td>
<td id="T_0aa46_row0_col162" class="data row0 col162">1</td>
<td id="T_0aa46_row0_col163" class="data row0 col163">1</td>
<td id="T_0aa46_row0_col164" class="data row0 col164">4</td>
<td id="T_0aa46_row0_col165" class="data row0 col165">1</td>
<td id="T_0aa46_row0_col166" class="data row0 col166">0</td>
<td id="T_0aa46_row0_col167" class="data row0 col167">0</td>
<td id="T_0aa46_row0_col168" class="data row0 col168">0</td>
<td id="T_0aa46_row0_col169" class="data row0 col169">0</td>
<td id="T_0aa46_row0_col170" class="data row0 col170">0</td>
<td id="T_0aa46_row0_col171" class="data row0 col171">0</td>
<td id="T_0aa46_row0_col172" class="data row0 col172">0</td>
<td id="T_0aa46_row0_col173" class="data row0 col173">0</td>
<td id="T_0aa46_row0_col174" class="data row0 col174">0</td>
<td id="T_0aa46_row0_col175" class="data row0 col175">0</td>
<td id="T_0aa46_row0_col176" class="data row0 col176">0</td>
<td id="T_0aa46_row0_col177" class="data row0 col177">0</td>
<td id="T_0aa46_row0_col178" class="data row0 col178">0</td>
<td id="T_0aa46_row0_col179" class="data row0 col179">2</td>
<td id="T_0aa46_row0_col180" class="data row0 col180">0</td>
<td id="T_0aa46_row0_col181" class="data row0 col181">0</td>
<td id="T_0aa46_row0_col182" class="data row0 col182">0</td>
<td id="T_0aa46_row0_col183" class="data row0 col183">0</td>
<td id="T_0aa46_row0_col184" class="data row0 col184">0</td>
<td id="T_0aa46_row0_col185" class="data row0 col185">0</td>
<td id="T_0aa46_row0_col186" class="data row0 col186">0</td>
<td id="T_0aa46_row0_col187" class="data row0 col187">0</td>
<td id="T_0aa46_row0_col188" class="data row0 col188">1</td>
<td id="T_0aa46_row0_col189" class="data row0 col189">1</td>
<td id="T_0aa46_row0_col190" class="data row0 col190">0</td>
<td id="T_0aa46_row0_col191" class="data row0 col191">0</td>
<td id="T_0aa46_row0_col192" class="data row0 col192">0</td>
<td id="T_0aa46_row0_col193" class="data row0 col193">0</td>
<td id="T_0aa46_row0_col194" class="data row0 col194">0</td>
<td id="T_0aa46_row0_col195" class="data row0 col195">0</td>
<td id="T_0aa46_row0_col196" class="data row0 col196">0</td>
<td id="T_0aa46_row0_col197" class="data row0 col197">0</td>
<td id="T_0aa46_row0_col198" class="data row0 col198">1</td>
<td id="T_0aa46_row0_col199" class="data row0 col199">2</td>
<td id="T_0aa46_row0_col200" class="data row0 col200">0</td>
<td id="T_0aa46_row0_col201" class="data row0 col201">1</td>
<td id="T_0aa46_row0_col202" class="data row0 col202">1</td>
<td id="T_0aa46_row0_col203" class="data row0 col203">1</td>
<td id="T_0aa46_row0_col204" class="data row0 col204">1</td>
<td id="T_0aa46_row0_col205" class="data row0 col205">5</td>
<td id="T_0aa46_row0_col206" class="data row0 col206">1</td>
<td id="T_0aa46_row0_col207" class="data row0 col207">1</td>
<td id="T_0aa46_row0_col208" class="data row0 col208">0</td>
<td id="T_0aa46_row0_col209" class="data row0 col209">0</td>
<td id="T_0aa46_row0_col210" class="data row0 col210">0</td>
<td id="T_0aa46_row0_col211" class="data row0 col211">0</td>
<td id="T_0aa46_row0_col212" class="data row0 col212">1</td>
<td id="T_0aa46_row0_col213" class="data row0 col213">0</td>
<td id="T_0aa46_row0_col214" class="data row0 col214">0</td>
<td id="T_0aa46_row0_col215" class="data row0 col215">0</td>
<td id="T_0aa46_row0_col216" class="data row0 col216">0</td>
<td id="T_0aa46_row0_col217" class="data row0 col217">0</td>
<td id="T_0aa46_row0_col218" class="data row0 col218">0</td>
<td id="T_0aa46_row0_col219" class="data row0 col219">1</td>
<td id="T_0aa46_row0_col220" class="data row0 col220">3</td>
<td id="T_0aa46_row0_col221" class="data row0 col221">2</td>
<td id="T_0aa46_row0_col222" class="data row0 col222">0</td>
<td id="T_0aa46_row0_col223" class="data row0 col223">0</td>
<td id="T_0aa46_row0_col224" class="data row0 col224">0</td>
<td id="T_0aa46_row0_col225" class="data row0 col225">1</td>
<td id="T_0aa46_row0_col226" class="data row0 col226">1</td>
<td id="T_0aa46_row0_col227" class="data row0 col227">4</td>
<td id="T_0aa46_row0_col228" class="data row0 col228">6</td>
<td id="T_0aa46_row0_col229" class="data row0 col229">1</td>
<td id="T_0aa46_row0_col230" class="data row0 col230">1</td>
<td id="T_0aa46_row0_col231" class="data row0 col231">1</td>
<td id="T_0aa46_row0_col232" class="data row0 col232">1</td>
<td id="T_0aa46_row0_col233" class="data row0 col233">3</td>
<td id="T_0aa46_row0_col234" class="data row0 col234">4</td>
<td id="T_0aa46_row0_col235" class="data row0 col235">2</td>
<td id="T_0aa46_row0_col236" class="data row0 col236">1</td>
<td id="T_0aa46_row0_col237" class="data row0 col237">1</td>
<td id="T_0aa46_row0_col238" class="data row0 col238">1</td>
<td id="T_0aa46_row0_col239" class="data row0 col239">4</td>
<td id="T_0aa46_row0_col240" class="data row0 col240">1</td>
<td id="T_0aa46_row0_col241" class="data row0 col241">1</td>
<td id="T_0aa46_row0_col242" class="data row0 col242">1</td>
<td id="T_0aa46_row0_col243" class="data row0 col243">1</td>
<td id="T_0aa46_row0_col244" class="data row0 col244">1</td>
<td id="T_0aa46_row0_col245" class="data row0 col245">0</td>
<td id="T_0aa46_row0_col246" class="data row0 col246">0</td>
<td id="T_0aa46_row0_col247" class="data row0 col247">0</td>
<td id="T_0aa46_row0_col248" class="data row0 col248">1</td>
<td id="T_0aa46_row0_col249" class="data row0 col249">1</td>
<td id="T_0aa46_row0_col250" class="data row0 col250">1</td>
<td id="T_0aa46_row0_col251" class="data row0 col251">1</td>
<td id="T_0aa46_row0_col252" class="data row0 col252">1</td>
<td id="T_0aa46_row0_col253" class="data row0 col253">1</td>
<td id="T_0aa46_row0_col254" class="data row0 col254">1</td>
<td id="T_0aa46_row0_col255" class="data row0 col255">1</td>
<td id="T_0aa46_row0_col256" class="data row0 col256">7</td>
<td id="T_0aa46_row0_col257" class="data row0 col257">2</td>
<td id="T_0aa46_row0_col258" class="data row0 col258">5</td>
<td id="T_0aa46_row0_col259" class="data row0 col259">1</td>
<td id="T_0aa46_row0_col260" class="data row0 col260">2</td>
<td id="T_0aa46_row0_col261" class="data row0 col261">1</td>
<td id="T_0aa46_row0_col262" class="data row0 col262">1</td>
<td id="T_0aa46_row0_col263" class="data row0 col263">1</td>
<td id="T_0aa46_row0_col264" class="data row0 col264">1</td>
<td id="T_0aa46_row0_col265" class="data row0 col265">1</td>
<td id="T_0aa46_row0_col266" class="data row0 col266">1</td>
<td id="T_0aa46_row0_col267" class="data row0 col267">1</td>
<td id="T_0aa46_row0_col268" class="data row0 col268">2</td>
<td id="T_0aa46_row0_col269" class="data row0 col269">1</td>
<td id="T_0aa46_row0_col270" class="data row0 col270">5</td>
<td id="T_0aa46_row0_col271" class="data row0 col271">1</td>
<td id="T_0aa46_row0_col272" class="data row0 col272">1</td>
<td id="T_0aa46_row0_col273" class="data row0 col273">1</td>
<td id="T_0aa46_row0_col274" class="data row0 col274">1</td>
<td id="T_0aa46_row0_col275" class="data row0 col275">1</td>
<td id="T_0aa46_row0_col276" class="data row0 col276">3</td>
<td id="T_0aa46_row0_col277" class="data row0 col277">1</td>
<td id="T_0aa46_row0_col278" class="data row0 col278">1</td>
<td id="T_0aa46_row0_col279" class="data row0 col279">1</td>
<td id="T_0aa46_row0_col280" class="data row0 col280">1</td>
<td id="T_0aa46_row0_col281" class="data row0 col281">4</td>
<td id="T_0aa46_row0_col282" class="data row0 col282">1</td>
<td id="T_0aa46_row0_col283" class="data row0 col283">1</td>
<td id="T_0aa46_row0_col284" class="data row0 col284">1</td>
<td id="T_0aa46_row0_col285" class="data row0 col285">1</td>
<td id="T_0aa46_row0_col286" class="data row0 col286">1</td>
<td id="T_0aa46_row0_col287" class="data row0 col287">5</td>
<td id="T_0aa46_row0_col288" class="data row0 col288">1</td>
<td id="T_0aa46_row0_col289" class="data row0 col289">11</td>
<td id="T_0aa46_row0_col290" class="data row0 col290">2</td>
<td id="T_0aa46_row0_col291" class="data row0 col291">1</td>
<td id="T_0aa46_row0_col292" class="data row0 col292">1</td>
<td id="T_0aa46_row0_col293" class="data row0 col293">1</td>
<td id="T_0aa46_row0_col294" class="data row0 col294">1</td>
<td id="T_0aa46_row0_col295" class="data row0 col295">1</td>
<td id="T_0aa46_row0_col296" class="data row0 col296">1</td>
<td id="T_0aa46_row0_col297" class="data row0 col297">1</td>
<td id="T_0aa46_row0_col298" class="data row0 col298">1</td>
<td id="T_0aa46_row0_col299" class="data row0 col299">1</td>
<td id="T_0aa46_row0_col300" class="data row0 col300">2</td>
<td id="T_0aa46_row0_col301" class="data row0 col301">1</td>
<td id="T_0aa46_row0_col302" class="data row0 col302">1</td>
<td id="T_0aa46_row0_col303" class="data row0 col303">1</td>
<td id="T_0aa46_row0_col304" class="data row0 col304">4</td>
<td id="T_0aa46_row0_col305" class="data row0 col305">6</td>
<td id="T_0aa46_row0_col306" class="data row0 col306">4</td>
<td id="T_0aa46_row0_col307" class="data row0 col307">4</td>
<td id="T_0aa46_row0_col308" class="data row0 col308">4</td>
<td id="T_0aa46_row0_col309" class="data row0 col309">1</td>
<td id="T_0aa46_row0_col310" class="data row0 col310">1</td>
<td id="T_0aa46_row0_col311" class="data row0 col311">5</td>
<td id="T_0aa46_row0_col312" class="data row0 col312">4</td>
<td id="T_0aa46_row0_col313" class="data row0 col313">1</td>
<td id="T_0aa46_row0_col314" class="data row0 col314">4</td>
<td id="T_0aa46_row0_col315" class="data row0 col315">1</td>
<td id="T_0aa46_row0_col316" class="data row0 col316">1</td>
<td id="T_0aa46_row0_col317" class="data row0 col317">0</td>
<td id="T_0aa46_row0_col318" class="data row0 col318">1</td>
<td id="T_0aa46_row0_col319" class="data row0 col319">1</td>
<td id="T_0aa46_row0_col320" class="data row0 col320">1</td>
<td id="T_0aa46_row0_col321" class="data row0 col321">7</td>
<td id="T_0aa46_row0_col322" class="data row0 col322">1</td>
<td id="T_0aa46_row0_col323" class="data row0 col323">1</td>
<td id="T_0aa46_row0_col324" class="data row0 col324">2</td>
<td id="T_0aa46_row0_col325" class="data row0 col325">3</td>
<td id="T_0aa46_row0_col326" class="data row0 col326">23</td>
<td id="T_0aa46_row0_col327" class="data row0 col327">6</td>
<td id="T_0aa46_row0_col328" class="data row0 col328">7</td>
<td id="T_0aa46_row0_col329" class="data row0 col329">0</td>
<td id="T_0aa46_row0_col330" class="data row0 col330">1</td>
<td id="T_0aa46_row0_col331" class="data row0 col331">9</td>
<td id="T_0aa46_row0_col332" class="data row0 col332">3</td>
<td id="T_0aa46_row0_col333" class="data row0 col333">4</td>
<td id="T_0aa46_row0_col334" class="data row0 col334">1</td>
<td id="T_0aa46_row0_col335" class="data row0 col335">3</td>
<td id="T_0aa46_row0_col336" class="data row0 col336">1</td>
<td id="T_0aa46_row0_col337" class="data row0 col337">1</td>
<td id="T_0aa46_row0_col338" class="data row0 col338">1</td>
<td id="T_0aa46_row0_col339" class="data row0 col339">1</td>
<td id="T_0aa46_row0_col340" class="data row0 col340">3</td>
<td id="T_0aa46_row0_col341" class="data row0 col341">3</td>
<td id="T_0aa46_row0_col342" class="data row0 col342">2</td>
<td id="T_0aa46_row0_col343" class="data row0 col343">1</td>
<td id="T_0aa46_row0_col344" class="data row0 col344">1</td>
<td id="T_0aa46_row0_col345" class="data row0 col345">1</td>
<td id="T_0aa46_row0_col346" class="data row0 col346">1</td>
<td id="T_0aa46_row0_col347" class="data row0 col347">1</td>
<td id="T_0aa46_row0_col348" class="data row0 col348">0</td>
<td id="T_0aa46_row0_col349" class="data row0 col349">2</td>
<td id="T_0aa46_row0_col350" class="data row0 col350">1</td>
<td id="T_0aa46_row0_col351" class="data row0 col351">1</td>
<td id="T_0aa46_row0_col352" class="data row0 col352">1</td>
<td id="T_0aa46_row0_col353" class="data row0 col353">1</td>
<td id="T_0aa46_row0_col354" class="data row0 col354">1</td>
<td id="T_0aa46_row0_col355" class="data row0 col355">1</td>
<td id="T_0aa46_row0_col356" class="data row0 col356">1</td>
<td id="T_0aa46_row0_col357" class="data row0 col357">1</td>
<td id="T_0aa46_row0_col358" class="data row0 col358">1</td>
<td id="T_0aa46_row0_col359" class="data row0 col359">1</td>
<td id="T_0aa46_row0_col360" class="data row0 col360">1</td>
<td id="T_0aa46_row0_col361" class="data row0 col361">0</td>
<td id="T_0aa46_row0_col362" class="data row0 col362">1</td>
<td id="T_0aa46_row0_col363" class="data row0 col363">2</td>
<td id="T_0aa46_row0_col364" class="data row0 col364">0</td>
<td id="T_0aa46_row0_col365" class="data row0 col365">1</td>
<td id="T_0aa46_row0_col366" class="data row0 col366">1</td>
<td id="T_0aa46_row0_col367" class="data row0 col367">1</td>
<td id="T_0aa46_row0_col368" class="data row0 col368">1</td>
<td id="T_0aa46_row0_col369" class="data row0 col369">2</td>
<td id="T_0aa46_row0_col370" class="data row0 col370">1</td>
<td id="T_0aa46_row0_col371" class="data row0 col371">1</td>
<td id="T_0aa46_row0_col372" class="data row0 col372">2</td>
<td id="T_0aa46_row0_col373" class="data row0 col373">1</td>
<td id="T_0aa46_row0_col374" class="data row0 col374">2</td>
<td id="T_0aa46_row0_col375" class="data row0 col375">2</td>
<td id="T_0aa46_row0_col376" class="data row0 col376">1</td>
<td id="T_0aa46_row0_col377" class="data row0 col377">1</td>
<td id="T_0aa46_row0_col378" class="data row0 col378">1</td>
<td id="T_0aa46_row0_col379" class="data row0 col379">1</td>
<td id="T_0aa46_row0_col380" class="data row0 col380">2</td>
<td id="T_0aa46_row0_col381" class="data row0 col381">1</td>
<td id="T_0aa46_row0_col382" class="data row0 col382">1</td>
<td id="T_0aa46_row0_col383" class="data row0 col383">1</td>
<td id="T_0aa46_row0_col384" class="data row0 col384">1</td>
<td id="T_0aa46_row0_col385" class="data row0 col385">1</td>
<td id="T_0aa46_row0_col386" class="data row0 col386">2</td>
<td id="T_0aa46_row0_col387" class="data row0 col387">1</td>
<td id="T_0aa46_row0_col388" class="data row0 col388">1</td>
<td id="T_0aa46_row0_col389" class="data row0 col389">1</td>
<td id="T_0aa46_row0_col390" class="data row0 col390">1</td>
<td id="T_0aa46_row0_col391" class="data row0 col391">2</td>
<td id="T_0aa46_row0_col392" class="data row0 col392">1</td>
<td id="T_0aa46_row0_col393" class="data row0 col393">1</td>
<td id="T_0aa46_row0_col394" class="data row0 col394">1</td>
<td id="T_0aa46_row0_col395" class="data row0 col395">1</td>
<td id="T_0aa46_row0_col396" class="data row0 col396">1</td>
<td id="T_0aa46_row0_col397" class="data row0 col397">7</td>
<td id="T_0aa46_row0_col398" class="data row0 col398">1</td>
<td id="T_0aa46_row0_col399" class="data row0 col399">1</td>
<td id="T_0aa46_row0_col400" class="data row0 col400">1</td>
<td id="T_0aa46_row0_col401" class="data row0 col401">2</td>
<td id="T_0aa46_row0_col402" class="data row0 col402">3</td>
<td id="T_0aa46_row0_col403" class="data row0 col403">1</td>
<td id="T_0aa46_row0_col404" class="data row0 col404">1</td>
<td id="T_0aa46_row0_col405" class="data row0 col405">1</td>
<td id="T_0aa46_row0_col406" class="data row0 col406">1</td>
<td id="T_0aa46_row0_col407" class="data row0 col407">1</td>
<td id="T_0aa46_row0_col408" class="data row0 col408">1</td>
<td id="T_0aa46_row0_col409" class="data row0 col409">1</td>
<td id="T_0aa46_row0_col410" class="data row0 col410">1</td>
<td id="T_0aa46_row0_col411" class="data row0 col411">1</td>
<td id="T_0aa46_row0_col412" class="data row0 col412">0</td>
<td id="T_0aa46_row0_col413" class="data row0 col413">1</td>
<td id="T_0aa46_row0_col414" class="data row0 col414">1</td>
<td id="T_0aa46_row0_col415" class="data row0 col415">1</td>
<td id="T_0aa46_row0_col416" class="data row0 col416">1</td>
<td id="T_0aa46_row0_col417" class="data row0 col417">1</td>
<td id="T_0aa46_row0_col418" class="data row0 col418">3</td>
<td id="T_0aa46_row0_col419" class="data row0 col419">16</td>
<td id="T_0aa46_row0_col420" class="data row0 col420">18</td>
<td id="T_0aa46_row0_col421" class="data row0 col421">2</td>
<td id="T_0aa46_row0_col422" class="data row0 col422">1</td>
<td id="T_0aa46_row0_col423" class="data row0 col423">2</td>
<td id="T_0aa46_row0_col424" class="data row0 col424">1</td>
<td id="T_0aa46_row0_col425" class="data row0 col425">9</td>
<td id="T_0aa46_row0_col426" class="data row0 col426">13</td>
<td id="T_0aa46_row0_col427" class="data row0 col427">2</td>
<td id="T_0aa46_row0_col428" class="data row0 col428">1</td>
<td id="T_0aa46_row0_col429" class="data row0 col429">1</td>
<td id="T_0aa46_row0_col430" class="data row0 col430">3</td>
<td id="T_0aa46_row0_col431" class="data row0 col431">2</td>
<td id="T_0aa46_row0_col432" class="data row0 col432">1</td>
<td id="T_0aa46_row0_col433" class="data row0 col433">0</td>
<td id="T_0aa46_row0_col434" class="data row0 col434">0</td>
<td id="T_0aa46_row0_col435" class="data row0 col435">0</td>
<td id="T_0aa46_row0_col436" class="data row0 col436">0</td>
<td id="T_0aa46_row0_col437" class="data row0 col437">0</td>
<td id="T_0aa46_row0_col438" class="data row0 col438">0</td>
<td id="T_0aa46_row0_col439" class="data row0 col439">0</td>
<td id="T_0aa46_row0_col440" class="data row0 col440">2</td>
<td id="T_0aa46_row0_col441" class="data row0 col441">1</td>
<td id="T_0aa46_row0_col442" class="data row0 col442">1</td>
<td id="T_0aa46_row0_col443" class="data row0 col443">1</td>
<td id="T_0aa46_row0_col444" class="data row0 col444">1</td>
<td id="T_0aa46_row0_col445" class="data row0 col445">1</td>
<td id="T_0aa46_row0_col446" class="data row0 col446">1</td>
<td id="T_0aa46_row0_col447" class="data row0 col447">1</td>
<td id="T_0aa46_row0_col448" class="data row0 col448">1</td>
<td id="T_0aa46_row0_col449" class="data row0 col449">1</td>
<td id="T_0aa46_row0_col450" class="data row0 col450">1</td>
<td id="T_0aa46_row0_col451" class="data row0 col451">0</td>
<td id="T_0aa46_row0_col452" class="data row0 col452">1</td>
<td id="T_0aa46_row0_col453" class="data row0 col453">1</td>
<td id="T_0aa46_row0_col454" class="data row0 col454">0</td>
<td id="T_0aa46_row0_col455" class="data row0 col455">1</td>
<td id="T_0aa46_row0_col456" class="data row0 col456">1</td>
<td id="T_0aa46_row0_col457" class="data row0 col457">1</td>
<td id="T_0aa46_row0_col458" class="data row0 col458">4</td>
<td id="T_0aa46_row0_col459" class="data row0 col459">1</td>
<td id="T_0aa46_row0_col460" class="data row0 col460">2</td>
<td id="T_0aa46_row0_col461" class="data row0 col461">1</td>
<td id="T_0aa46_row0_col462" class="data row0 col462">1</td>
<td id="T_0aa46_row0_col463" class="data row0 col463">1</td>
<td id="T_0aa46_row0_col464" class="data row0 col464">1</td>
<td id="T_0aa46_row0_col465" class="data row0 col465">1</td>
<td id="T_0aa46_row0_col466" class="data row0 col466">1</td>
<td id="T_0aa46_row0_col467" class="data row0 col467">1</td>
<td id="T_0aa46_row0_col468" class="data row0 col468">1</td>
<td id="T_0aa46_row0_col469" class="data row0 col469">1</td>
<td id="T_0aa46_row0_col470" class="data row0 col470">1</td>
<td id="T_0aa46_row0_col471" class="data row0 col471">1</td>
<td id="T_0aa46_row0_col472" class="data row0 col472">1</td>
<td id="T_0aa46_row0_col473" class="data row0 col473">1</td>
<td id="T_0aa46_row0_col474" class="data row0 col474">1</td>
<td id="T_0aa46_row0_col475" class="data row0 col475">1</td>
<td id="T_0aa46_row0_col476" class="data row0 col476">1</td>
<td id="T_0aa46_row0_col477" class="data row0 col477">1</td>
<td id="T_0aa46_row0_col478" class="data row0 col478">1</td>
<td id="T_0aa46_row0_col479" class="data row0 col479">1</td>
<td id="T_0aa46_row0_col480" class="data row0 col480">1</td>
<td id="T_0aa46_row0_col481" class="data row0 col481">1</td>
<td id="T_0aa46_row0_col482" class="data row0 col482">1</td>
<td id="T_0aa46_row0_col483" class="data row0 col483">1</td>
</tr>
<tr class="even">
<th id="T_0aa46_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">Fagaceae</th>
<td id="T_0aa46_row1_col0" class="data row1 col0">0</td>
<td id="T_0aa46_row1_col1" class="data row1 col1">0</td>
<td id="T_0aa46_row1_col2" class="data row1 col2">0</td>
<td id="T_0aa46_row1_col3" class="data row1 col3">0</td>
<td id="T_0aa46_row1_col4" class="data row1 col4">0</td>
<td id="T_0aa46_row1_col5" class="data row1 col5">0</td>
<td id="T_0aa46_row1_col6" class="data row1 col6">0</td>
<td id="T_0aa46_row1_col7" class="data row1 col7">0</td>
<td id="T_0aa46_row1_col8" class="data row1 col8">0</td>
<td id="T_0aa46_row1_col9" class="data row1 col9">0</td>
<td id="T_0aa46_row1_col10" class="data row1 col10">0</td>
<td id="T_0aa46_row1_col11" class="data row1 col11">0</td>
<td id="T_0aa46_row1_col12" class="data row1 col12">0</td>
<td id="T_0aa46_row1_col13" class="data row1 col13">0</td>
<td id="T_0aa46_row1_col14" class="data row1 col14">0</td>
<td id="T_0aa46_row1_col15" class="data row1 col15">0</td>
<td id="T_0aa46_row1_col16" class="data row1 col16">0</td>
<td id="T_0aa46_row1_col17" class="data row1 col17">0</td>
<td id="T_0aa46_row1_col18" class="data row1 col18">0</td>
<td id="T_0aa46_row1_col19" class="data row1 col19">0</td>
<td id="T_0aa46_row1_col20" class="data row1 col20">0</td>
<td id="T_0aa46_row1_col21" class="data row1 col21">0</td>
<td id="T_0aa46_row1_col22" class="data row1 col22">0</td>
<td id="T_0aa46_row1_col23" class="data row1 col23">0</td>
<td id="T_0aa46_row1_col24" class="data row1 col24">0</td>
<td id="T_0aa46_row1_col25" class="data row1 col25">0</td>
<td id="T_0aa46_row1_col26" class="data row1 col26">0</td>
<td id="T_0aa46_row1_col27" class="data row1 col27">0</td>
<td id="T_0aa46_row1_col28" class="data row1 col28">0</td>
<td id="T_0aa46_row1_col29" class="data row1 col29">0</td>
<td id="T_0aa46_row1_col30" class="data row1 col30">0</td>
<td id="T_0aa46_row1_col31" class="data row1 col31">0</td>
<td id="T_0aa46_row1_col32" class="data row1 col32">0</td>
<td id="T_0aa46_row1_col33" class="data row1 col33">0</td>
<td id="T_0aa46_row1_col34" class="data row1 col34">0</td>
<td id="T_0aa46_row1_col35" class="data row1 col35">0</td>
<td id="T_0aa46_row1_col36" class="data row1 col36">0</td>
<td id="T_0aa46_row1_col37" class="data row1 col37">0</td>
<td id="T_0aa46_row1_col38" class="data row1 col38">0</td>
<td id="T_0aa46_row1_col39" class="data row1 col39">0</td>
<td id="T_0aa46_row1_col40" class="data row1 col40">0</td>
<td id="T_0aa46_row1_col41" class="data row1 col41">0</td>
<td id="T_0aa46_row1_col42" class="data row1 col42">0</td>
<td id="T_0aa46_row1_col43" class="data row1 col43">0</td>
<td id="T_0aa46_row1_col44" class="data row1 col44">0</td>
<td id="T_0aa46_row1_col45" class="data row1 col45">0</td>
<td id="T_0aa46_row1_col46" class="data row1 col46">0</td>
<td id="T_0aa46_row1_col47" class="data row1 col47">0</td>
<td id="T_0aa46_row1_col48" class="data row1 col48">0</td>
<td id="T_0aa46_row1_col49" class="data row1 col49">0</td>
<td id="T_0aa46_row1_col50" class="data row1 col50">0</td>
<td id="T_0aa46_row1_col51" class="data row1 col51">0</td>
<td id="T_0aa46_row1_col52" class="data row1 col52">0</td>
<td id="T_0aa46_row1_col53" class="data row1 col53">0</td>
<td id="T_0aa46_row1_col54" class="data row1 col54">0</td>
<td id="T_0aa46_row1_col55" class="data row1 col55">0</td>
<td id="T_0aa46_row1_col56" class="data row1 col56">0</td>
<td id="T_0aa46_row1_col57" class="data row1 col57">0</td>
<td id="T_0aa46_row1_col58" class="data row1 col58">0</td>
<td id="T_0aa46_row1_col59" class="data row1 col59">0</td>
<td id="T_0aa46_row1_col60" class="data row1 col60">0</td>
<td id="T_0aa46_row1_col61" class="data row1 col61">0</td>
<td id="T_0aa46_row1_col62" class="data row1 col62">0</td>
<td id="T_0aa46_row1_col63" class="data row1 col63">0</td>
<td id="T_0aa46_row1_col64" class="data row1 col64">0</td>
<td id="T_0aa46_row1_col65" class="data row1 col65">0</td>
<td id="T_0aa46_row1_col66" class="data row1 col66">0</td>
<td id="T_0aa46_row1_col67" class="data row1 col67">0</td>
<td id="T_0aa46_row1_col68" class="data row1 col68">0</td>
<td id="T_0aa46_row1_col69" class="data row1 col69">0</td>
<td id="T_0aa46_row1_col70" class="data row1 col70">0</td>
<td id="T_0aa46_row1_col71" class="data row1 col71">0</td>
<td id="T_0aa46_row1_col72" class="data row1 col72">0</td>
<td id="T_0aa46_row1_col73" class="data row1 col73">0</td>
<td id="T_0aa46_row1_col74" class="data row1 col74">0</td>
<td id="T_0aa46_row1_col75" class="data row1 col75">0</td>
<td id="T_0aa46_row1_col76" class="data row1 col76">0</td>
<td id="T_0aa46_row1_col77" class="data row1 col77">0</td>
<td id="T_0aa46_row1_col78" class="data row1 col78">0</td>
<td id="T_0aa46_row1_col79" class="data row1 col79">0</td>
<td id="T_0aa46_row1_col80" class="data row1 col80">0</td>
<td id="T_0aa46_row1_col81" class="data row1 col81">0</td>
<td id="T_0aa46_row1_col82" class="data row1 col82">0</td>
<td id="T_0aa46_row1_col83" class="data row1 col83">0</td>
<td id="T_0aa46_row1_col84" class="data row1 col84">0</td>
<td id="T_0aa46_row1_col85" class="data row1 col85">0</td>
<td id="T_0aa46_row1_col86" class="data row1 col86">0</td>
<td id="T_0aa46_row1_col87" class="data row1 col87">0</td>
<td id="T_0aa46_row1_col88" class="data row1 col88">0</td>
<td id="T_0aa46_row1_col89" class="data row1 col89">0</td>
<td id="T_0aa46_row1_col90" class="data row1 col90">0</td>
<td id="T_0aa46_row1_col91" class="data row1 col91">0</td>
<td id="T_0aa46_row1_col92" class="data row1 col92">0</td>
<td id="T_0aa46_row1_col93" class="data row1 col93">0</td>
<td id="T_0aa46_row1_col94" class="data row1 col94">0</td>
<td id="T_0aa46_row1_col95" class="data row1 col95">0</td>
<td id="T_0aa46_row1_col96" class="data row1 col96">0</td>
<td id="T_0aa46_row1_col97" class="data row1 col97">0</td>
<td id="T_0aa46_row1_col98" class="data row1 col98">0</td>
<td id="T_0aa46_row1_col99" class="data row1 col99">0</td>
<td id="T_0aa46_row1_col100" class="data row1 col100">0</td>
<td id="T_0aa46_row1_col101" class="data row1 col101">0</td>
<td id="T_0aa46_row1_col102" class="data row1 col102">0</td>
<td id="T_0aa46_row1_col103" class="data row1 col103">0</td>
<td id="T_0aa46_row1_col104" class="data row1 col104">0</td>
<td id="T_0aa46_row1_col105" class="data row1 col105">0</td>
<td id="T_0aa46_row1_col106" class="data row1 col106">0</td>
<td id="T_0aa46_row1_col107" class="data row1 col107">0</td>
<td id="T_0aa46_row1_col108" class="data row1 col108">0</td>
<td id="T_0aa46_row1_col109" class="data row1 col109">0</td>
<td id="T_0aa46_row1_col110" class="data row1 col110">0</td>
<td id="T_0aa46_row1_col111" class="data row1 col111">0</td>
<td id="T_0aa46_row1_col112" class="data row1 col112">0</td>
<td id="T_0aa46_row1_col113" class="data row1 col113">0</td>
<td id="T_0aa46_row1_col114" class="data row1 col114">0</td>
<td id="T_0aa46_row1_col115" class="data row1 col115">0</td>
<td id="T_0aa46_row1_col116" class="data row1 col116">0</td>
<td id="T_0aa46_row1_col117" class="data row1 col117">0</td>
<td id="T_0aa46_row1_col118" class="data row1 col118">0</td>
<td id="T_0aa46_row1_col119" class="data row1 col119">0</td>
<td id="T_0aa46_row1_col120" class="data row1 col120">0</td>
<td id="T_0aa46_row1_col121" class="data row1 col121">0</td>
<td id="T_0aa46_row1_col122" class="data row1 col122">0</td>
<td id="T_0aa46_row1_col123" class="data row1 col123">0</td>
<td id="T_0aa46_row1_col124" class="data row1 col124">0</td>
<td id="T_0aa46_row1_col125" class="data row1 col125">0</td>
<td id="T_0aa46_row1_col126" class="data row1 col126">0</td>
<td id="T_0aa46_row1_col127" class="data row1 col127">0</td>
<td id="T_0aa46_row1_col128" class="data row1 col128">0</td>
<td id="T_0aa46_row1_col129" class="data row1 col129">0</td>
<td id="T_0aa46_row1_col130" class="data row1 col130">0</td>
<td id="T_0aa46_row1_col131" class="data row1 col131">0</td>
<td id="T_0aa46_row1_col132" class="data row1 col132">0</td>
<td id="T_0aa46_row1_col133" class="data row1 col133">0</td>
<td id="T_0aa46_row1_col134" class="data row1 col134">0</td>
<td id="T_0aa46_row1_col135" class="data row1 col135">0</td>
<td id="T_0aa46_row1_col136" class="data row1 col136">0</td>
<td id="T_0aa46_row1_col137" class="data row1 col137">0</td>
<td id="T_0aa46_row1_col138" class="data row1 col138">0</td>
<td id="T_0aa46_row1_col139" class="data row1 col139">0</td>
<td id="T_0aa46_row1_col140" class="data row1 col140">0</td>
<td id="T_0aa46_row1_col141" class="data row1 col141">0</td>
<td id="T_0aa46_row1_col142" class="data row1 col142">0</td>
<td id="T_0aa46_row1_col143" class="data row1 col143">0</td>
<td id="T_0aa46_row1_col144" class="data row1 col144">0</td>
<td id="T_0aa46_row1_col145" class="data row1 col145">0</td>
<td id="T_0aa46_row1_col146" class="data row1 col146">0</td>
<td id="T_0aa46_row1_col147" class="data row1 col147">0</td>
<td id="T_0aa46_row1_col148" class="data row1 col148">0</td>
<td id="T_0aa46_row1_col149" class="data row1 col149">0</td>
<td id="T_0aa46_row1_col150" class="data row1 col150">0</td>
<td id="T_0aa46_row1_col151" class="data row1 col151">0</td>
<td id="T_0aa46_row1_col152" class="data row1 col152">0</td>
<td id="T_0aa46_row1_col153" class="data row1 col153">0</td>
<td id="T_0aa46_row1_col154" class="data row1 col154">0</td>
<td id="T_0aa46_row1_col155" class="data row1 col155">0</td>
<td id="T_0aa46_row1_col156" class="data row1 col156">0</td>
<td id="T_0aa46_row1_col157" class="data row1 col157">0</td>
<td id="T_0aa46_row1_col158" class="data row1 col158">0</td>
<td id="T_0aa46_row1_col159" class="data row1 col159">0</td>
<td id="T_0aa46_row1_col160" class="data row1 col160">0</td>
<td id="T_0aa46_row1_col161" class="data row1 col161">0</td>
<td id="T_0aa46_row1_col162" class="data row1 col162">0</td>
<td id="T_0aa46_row1_col163" class="data row1 col163">0</td>
<td id="T_0aa46_row1_col164" class="data row1 col164">0</td>
<td id="T_0aa46_row1_col165" class="data row1 col165">0</td>
<td id="T_0aa46_row1_col166" class="data row1 col166">0</td>
<td id="T_0aa46_row1_col167" class="data row1 col167">0</td>
<td id="T_0aa46_row1_col168" class="data row1 col168">0</td>
<td id="T_0aa46_row1_col169" class="data row1 col169">0</td>
<td id="T_0aa46_row1_col170" class="data row1 col170">0</td>
<td id="T_0aa46_row1_col171" class="data row1 col171">0</td>
<td id="T_0aa46_row1_col172" class="data row1 col172">0</td>
<td id="T_0aa46_row1_col173" class="data row1 col173">0</td>
<td id="T_0aa46_row1_col174" class="data row1 col174">0</td>
<td id="T_0aa46_row1_col175" class="data row1 col175">0</td>
<td id="T_0aa46_row1_col176" class="data row1 col176">0</td>
<td id="T_0aa46_row1_col177" class="data row1 col177">0</td>
<td id="T_0aa46_row1_col178" class="data row1 col178">0</td>
<td id="T_0aa46_row1_col179" class="data row1 col179">0</td>
<td id="T_0aa46_row1_col180" class="data row1 col180">0</td>
<td id="T_0aa46_row1_col181" class="data row1 col181">0</td>
<td id="T_0aa46_row1_col182" class="data row1 col182">0</td>
<td id="T_0aa46_row1_col183" class="data row1 col183">0</td>
<td id="T_0aa46_row1_col184" class="data row1 col184">0</td>
<td id="T_0aa46_row1_col185" class="data row1 col185">0</td>
<td id="T_0aa46_row1_col186" class="data row1 col186">0</td>
<td id="T_0aa46_row1_col187" class="data row1 col187">0</td>
<td id="T_0aa46_row1_col188" class="data row1 col188">0</td>
<td id="T_0aa46_row1_col189" class="data row1 col189">0</td>
<td id="T_0aa46_row1_col190" class="data row1 col190">0</td>
<td id="T_0aa46_row1_col191" class="data row1 col191">0</td>
<td id="T_0aa46_row1_col192" class="data row1 col192">0</td>
<td id="T_0aa46_row1_col193" class="data row1 col193">0</td>
<td id="T_0aa46_row1_col194" class="data row1 col194">0</td>
<td id="T_0aa46_row1_col195" class="data row1 col195">0</td>
<td id="T_0aa46_row1_col196" class="data row1 col196">0</td>
<td id="T_0aa46_row1_col197" class="data row1 col197">0</td>
<td id="T_0aa46_row1_col198" class="data row1 col198">0</td>
<td id="T_0aa46_row1_col199" class="data row1 col199">0</td>
<td id="T_0aa46_row1_col200" class="data row1 col200">0</td>
<td id="T_0aa46_row1_col201" class="data row1 col201">0</td>
<td id="T_0aa46_row1_col202" class="data row1 col202">0</td>
<td id="T_0aa46_row1_col203" class="data row1 col203">0</td>
<td id="T_0aa46_row1_col204" class="data row1 col204">0</td>
<td id="T_0aa46_row1_col205" class="data row1 col205">0</td>
<td id="T_0aa46_row1_col206" class="data row1 col206">0</td>
<td id="T_0aa46_row1_col207" class="data row1 col207">0</td>
<td id="T_0aa46_row1_col208" class="data row1 col208">0</td>
<td id="T_0aa46_row1_col209" class="data row1 col209">0</td>
<td id="T_0aa46_row1_col210" class="data row1 col210">0</td>
<td id="T_0aa46_row1_col211" class="data row1 col211">0</td>
<td id="T_0aa46_row1_col212" class="data row1 col212">0</td>
<td id="T_0aa46_row1_col213" class="data row1 col213">0</td>
<td id="T_0aa46_row1_col214" class="data row1 col214">0</td>
<td id="T_0aa46_row1_col215" class="data row1 col215">0</td>
<td id="T_0aa46_row1_col216" class="data row1 col216">0</td>
<td id="T_0aa46_row1_col217" class="data row1 col217">0</td>
<td id="T_0aa46_row1_col218" class="data row1 col218">0</td>
<td id="T_0aa46_row1_col219" class="data row1 col219">0</td>
<td id="T_0aa46_row1_col220" class="data row1 col220">0</td>
<td id="T_0aa46_row1_col221" class="data row1 col221">0</td>
<td id="T_0aa46_row1_col222" class="data row1 col222">1</td>
<td id="T_0aa46_row1_col223" class="data row1 col223">1</td>
<td id="T_0aa46_row1_col224" class="data row1 col224">1</td>
<td id="T_0aa46_row1_col225" class="data row1 col225">0</td>
<td id="T_0aa46_row1_col226" class="data row1 col226">0</td>
<td id="T_0aa46_row1_col227" class="data row1 col227">0</td>
<td id="T_0aa46_row1_col228" class="data row1 col228">0</td>
<td id="T_0aa46_row1_col229" class="data row1 col229">0</td>
<td id="T_0aa46_row1_col230" class="data row1 col230">0</td>
<td id="T_0aa46_row1_col231" class="data row1 col231">0</td>
<td id="T_0aa46_row1_col232" class="data row1 col232">0</td>
<td id="T_0aa46_row1_col233" class="data row1 col233">0</td>
<td id="T_0aa46_row1_col234" class="data row1 col234">0</td>
<td id="T_0aa46_row1_col235" class="data row1 col235">0</td>
<td id="T_0aa46_row1_col236" class="data row1 col236">0</td>
<td id="T_0aa46_row1_col237" class="data row1 col237">0</td>
<td id="T_0aa46_row1_col238" class="data row1 col238">0</td>
<td id="T_0aa46_row1_col239" class="data row1 col239">0</td>
<td id="T_0aa46_row1_col240" class="data row1 col240">0</td>
<td id="T_0aa46_row1_col241" class="data row1 col241">0</td>
<td id="T_0aa46_row1_col242" class="data row1 col242">0</td>
<td id="T_0aa46_row1_col243" class="data row1 col243">0</td>
<td id="T_0aa46_row1_col244" class="data row1 col244">0</td>
<td id="T_0aa46_row1_col245" class="data row1 col245">0</td>
<td id="T_0aa46_row1_col246" class="data row1 col246">0</td>
<td id="T_0aa46_row1_col247" class="data row1 col247">0</td>
<td id="T_0aa46_row1_col248" class="data row1 col248">0</td>
<td id="T_0aa46_row1_col249" class="data row1 col249">0</td>
<td id="T_0aa46_row1_col250" class="data row1 col250">0</td>
<td id="T_0aa46_row1_col251" class="data row1 col251">0</td>
<td id="T_0aa46_row1_col252" class="data row1 col252">0</td>
<td id="T_0aa46_row1_col253" class="data row1 col253">0</td>
<td id="T_0aa46_row1_col254" class="data row1 col254">0</td>
<td id="T_0aa46_row1_col255" class="data row1 col255">0</td>
<td id="T_0aa46_row1_col256" class="data row1 col256">0</td>
<td id="T_0aa46_row1_col257" class="data row1 col257">0</td>
<td id="T_0aa46_row1_col258" class="data row1 col258">0</td>
<td id="T_0aa46_row1_col259" class="data row1 col259">0</td>
<td id="T_0aa46_row1_col260" class="data row1 col260">0</td>
<td id="T_0aa46_row1_col261" class="data row1 col261">0</td>
<td id="T_0aa46_row1_col262" class="data row1 col262">0</td>
<td id="T_0aa46_row1_col263" class="data row1 col263">0</td>
<td id="T_0aa46_row1_col264" class="data row1 col264">0</td>
<td id="T_0aa46_row1_col265" class="data row1 col265">0</td>
<td id="T_0aa46_row1_col266" class="data row1 col266">0</td>
<td id="T_0aa46_row1_col267" class="data row1 col267">0</td>
<td id="T_0aa46_row1_col268" class="data row1 col268">0</td>
<td id="T_0aa46_row1_col269" class="data row1 col269">0</td>
<td id="T_0aa46_row1_col270" class="data row1 col270">0</td>
<td id="T_0aa46_row1_col271" class="data row1 col271">0</td>
<td id="T_0aa46_row1_col272" class="data row1 col272">0</td>
<td id="T_0aa46_row1_col273" class="data row1 col273">0</td>
<td id="T_0aa46_row1_col274" class="data row1 col274">0</td>
<td id="T_0aa46_row1_col275" class="data row1 col275">0</td>
<td id="T_0aa46_row1_col276" class="data row1 col276">0</td>
<td id="T_0aa46_row1_col277" class="data row1 col277">0</td>
<td id="T_0aa46_row1_col278" class="data row1 col278">0</td>
<td id="T_0aa46_row1_col279" class="data row1 col279">0</td>
<td id="T_0aa46_row1_col280" class="data row1 col280">0</td>
<td id="T_0aa46_row1_col281" class="data row1 col281">0</td>
<td id="T_0aa46_row1_col282" class="data row1 col282">0</td>
<td id="T_0aa46_row1_col283" class="data row1 col283">0</td>
<td id="T_0aa46_row1_col284" class="data row1 col284">0</td>
<td id="T_0aa46_row1_col285" class="data row1 col285">0</td>
<td id="T_0aa46_row1_col286" class="data row1 col286">0</td>
<td id="T_0aa46_row1_col287" class="data row1 col287">0</td>
<td id="T_0aa46_row1_col288" class="data row1 col288">0</td>
<td id="T_0aa46_row1_col289" class="data row1 col289">0</td>
<td id="T_0aa46_row1_col290" class="data row1 col290">0</td>
<td id="T_0aa46_row1_col291" class="data row1 col291">0</td>
<td id="T_0aa46_row1_col292" class="data row1 col292">0</td>
<td id="T_0aa46_row1_col293" class="data row1 col293">0</td>
<td id="T_0aa46_row1_col294" class="data row1 col294">0</td>
<td id="T_0aa46_row1_col295" class="data row1 col295">0</td>
<td id="T_0aa46_row1_col296" class="data row1 col296">0</td>
<td id="T_0aa46_row1_col297" class="data row1 col297">0</td>
<td id="T_0aa46_row1_col298" class="data row1 col298">0</td>
<td id="T_0aa46_row1_col299" class="data row1 col299">0</td>
<td id="T_0aa46_row1_col300" class="data row1 col300">0</td>
<td id="T_0aa46_row1_col301" class="data row1 col301">0</td>
<td id="T_0aa46_row1_col302" class="data row1 col302">0</td>
<td id="T_0aa46_row1_col303" class="data row1 col303">0</td>
<td id="T_0aa46_row1_col304" class="data row1 col304">0</td>
<td id="T_0aa46_row1_col305" class="data row1 col305">0</td>
<td id="T_0aa46_row1_col306" class="data row1 col306">0</td>
<td id="T_0aa46_row1_col307" class="data row1 col307">0</td>
<td id="T_0aa46_row1_col308" class="data row1 col308">0</td>
<td id="T_0aa46_row1_col309" class="data row1 col309">0</td>
<td id="T_0aa46_row1_col310" class="data row1 col310">0</td>
<td id="T_0aa46_row1_col311" class="data row1 col311">0</td>
<td id="T_0aa46_row1_col312" class="data row1 col312">0</td>
<td id="T_0aa46_row1_col313" class="data row1 col313">0</td>
<td id="T_0aa46_row1_col314" class="data row1 col314">0</td>
<td id="T_0aa46_row1_col315" class="data row1 col315">0</td>
<td id="T_0aa46_row1_col316" class="data row1 col316">0</td>
<td id="T_0aa46_row1_col317" class="data row1 col317">0</td>
<td id="T_0aa46_row1_col318" class="data row1 col318">0</td>
<td id="T_0aa46_row1_col319" class="data row1 col319">0</td>
<td id="T_0aa46_row1_col320" class="data row1 col320">0</td>
<td id="T_0aa46_row1_col321" class="data row1 col321">0</td>
<td id="T_0aa46_row1_col322" class="data row1 col322">0</td>
<td id="T_0aa46_row1_col323" class="data row1 col323">0</td>
<td id="T_0aa46_row1_col324" class="data row1 col324">0</td>
<td id="T_0aa46_row1_col325" class="data row1 col325">0</td>
<td id="T_0aa46_row1_col326" class="data row1 col326">0</td>
<td id="T_0aa46_row1_col327" class="data row1 col327">0</td>
<td id="T_0aa46_row1_col328" class="data row1 col328">0</td>
<td id="T_0aa46_row1_col329" class="data row1 col329">0</td>
<td id="T_0aa46_row1_col330" class="data row1 col330">0</td>
<td id="T_0aa46_row1_col331" class="data row1 col331">0</td>
<td id="T_0aa46_row1_col332" class="data row1 col332">0</td>
<td id="T_0aa46_row1_col333" class="data row1 col333">0</td>
<td id="T_0aa46_row1_col334" class="data row1 col334">0</td>
<td id="T_0aa46_row1_col335" class="data row1 col335">0</td>
<td id="T_0aa46_row1_col336" class="data row1 col336">0</td>
<td id="T_0aa46_row1_col337" class="data row1 col337">0</td>
<td id="T_0aa46_row1_col338" class="data row1 col338">0</td>
<td id="T_0aa46_row1_col339" class="data row1 col339">0</td>
<td id="T_0aa46_row1_col340" class="data row1 col340">0</td>
<td id="T_0aa46_row1_col341" class="data row1 col341">0</td>
<td id="T_0aa46_row1_col342" class="data row1 col342">0</td>
<td id="T_0aa46_row1_col343" class="data row1 col343">0</td>
<td id="T_0aa46_row1_col344" class="data row1 col344">0</td>
<td id="T_0aa46_row1_col345" class="data row1 col345">0</td>
<td id="T_0aa46_row1_col346" class="data row1 col346">0</td>
<td id="T_0aa46_row1_col347" class="data row1 col347">0</td>
<td id="T_0aa46_row1_col348" class="data row1 col348">0</td>
<td id="T_0aa46_row1_col349" class="data row1 col349">0</td>
<td id="T_0aa46_row1_col350" class="data row1 col350">0</td>
<td id="T_0aa46_row1_col351" class="data row1 col351">0</td>
<td id="T_0aa46_row1_col352" class="data row1 col352">0</td>
<td id="T_0aa46_row1_col353" class="data row1 col353">0</td>
<td id="T_0aa46_row1_col354" class="data row1 col354">0</td>
<td id="T_0aa46_row1_col355" class="data row1 col355">0</td>
<td id="T_0aa46_row1_col356" class="data row1 col356">0</td>
<td id="T_0aa46_row1_col357" class="data row1 col357">0</td>
<td id="T_0aa46_row1_col358" class="data row1 col358">0</td>
<td id="T_0aa46_row1_col359" class="data row1 col359">0</td>
<td id="T_0aa46_row1_col360" class="data row1 col360">0</td>
<td id="T_0aa46_row1_col361" class="data row1 col361">0</td>
<td id="T_0aa46_row1_col362" class="data row1 col362">0</td>
<td id="T_0aa46_row1_col363" class="data row1 col363">0</td>
<td id="T_0aa46_row1_col364" class="data row1 col364">0</td>
<td id="T_0aa46_row1_col365" class="data row1 col365">0</td>
<td id="T_0aa46_row1_col366" class="data row1 col366">0</td>
<td id="T_0aa46_row1_col367" class="data row1 col367">0</td>
<td id="T_0aa46_row1_col368" class="data row1 col368">0</td>
<td id="T_0aa46_row1_col369" class="data row1 col369">0</td>
<td id="T_0aa46_row1_col370" class="data row1 col370">0</td>
<td id="T_0aa46_row1_col371" class="data row1 col371">0</td>
<td id="T_0aa46_row1_col372" class="data row1 col372">0</td>
<td id="T_0aa46_row1_col373" class="data row1 col373">0</td>
<td id="T_0aa46_row1_col374" class="data row1 col374">0</td>
<td id="T_0aa46_row1_col375" class="data row1 col375">0</td>
<td id="T_0aa46_row1_col376" class="data row1 col376">0</td>
<td id="T_0aa46_row1_col377" class="data row1 col377">0</td>
<td id="T_0aa46_row1_col378" class="data row1 col378">0</td>
<td id="T_0aa46_row1_col379" class="data row1 col379">0</td>
<td id="T_0aa46_row1_col380" class="data row1 col380">0</td>
<td id="T_0aa46_row1_col381" class="data row1 col381">0</td>
<td id="T_0aa46_row1_col382" class="data row1 col382">0</td>
<td id="T_0aa46_row1_col383" class="data row1 col383">0</td>
<td id="T_0aa46_row1_col384" class="data row1 col384">0</td>
<td id="T_0aa46_row1_col385" class="data row1 col385">0</td>
<td id="T_0aa46_row1_col386" class="data row1 col386">0</td>
<td id="T_0aa46_row1_col387" class="data row1 col387">0</td>
<td id="T_0aa46_row1_col388" class="data row1 col388">0</td>
<td id="T_0aa46_row1_col389" class="data row1 col389">0</td>
<td id="T_0aa46_row1_col390" class="data row1 col390">0</td>
<td id="T_0aa46_row1_col391" class="data row1 col391">0</td>
<td id="T_0aa46_row1_col392" class="data row1 col392">0</td>
<td id="T_0aa46_row1_col393" class="data row1 col393">0</td>
<td id="T_0aa46_row1_col394" class="data row1 col394">0</td>
<td id="T_0aa46_row1_col395" class="data row1 col395">0</td>
<td id="T_0aa46_row1_col396" class="data row1 col396">0</td>
<td id="T_0aa46_row1_col397" class="data row1 col397">0</td>
<td id="T_0aa46_row1_col398" class="data row1 col398">0</td>
<td id="T_0aa46_row1_col399" class="data row1 col399">0</td>
<td id="T_0aa46_row1_col400" class="data row1 col400">0</td>
<td id="T_0aa46_row1_col401" class="data row1 col401">0</td>
<td id="T_0aa46_row1_col402" class="data row1 col402">0</td>
<td id="T_0aa46_row1_col403" class="data row1 col403">0</td>
<td id="T_0aa46_row1_col404" class="data row1 col404">0</td>
<td id="T_0aa46_row1_col405" class="data row1 col405">0</td>
<td id="T_0aa46_row1_col406" class="data row1 col406">0</td>
<td id="T_0aa46_row1_col407" class="data row1 col407">0</td>
<td id="T_0aa46_row1_col408" class="data row1 col408">0</td>
<td id="T_0aa46_row1_col409" class="data row1 col409">0</td>
<td id="T_0aa46_row1_col410" class="data row1 col410">0</td>
<td id="T_0aa46_row1_col411" class="data row1 col411">0</td>
<td id="T_0aa46_row1_col412" class="data row1 col412">0</td>
<td id="T_0aa46_row1_col413" class="data row1 col413">0</td>
<td id="T_0aa46_row1_col414" class="data row1 col414">0</td>
<td id="T_0aa46_row1_col415" class="data row1 col415">0</td>
<td id="T_0aa46_row1_col416" class="data row1 col416">0</td>
<td id="T_0aa46_row1_col417" class="data row1 col417">0</td>
<td id="T_0aa46_row1_col418" class="data row1 col418">0</td>
<td id="T_0aa46_row1_col419" class="data row1 col419">0</td>
<td id="T_0aa46_row1_col420" class="data row1 col420">0</td>
<td id="T_0aa46_row1_col421" class="data row1 col421">0</td>
<td id="T_0aa46_row1_col422" class="data row1 col422">0</td>
<td id="T_0aa46_row1_col423" class="data row1 col423">0</td>
<td id="T_0aa46_row1_col424" class="data row1 col424">0</td>
<td id="T_0aa46_row1_col425" class="data row1 col425">0</td>
<td id="T_0aa46_row1_col426" class="data row1 col426">0</td>
<td id="T_0aa46_row1_col427" class="data row1 col427">0</td>
<td id="T_0aa46_row1_col428" class="data row1 col428">0</td>
<td id="T_0aa46_row1_col429" class="data row1 col429">0</td>
<td id="T_0aa46_row1_col430" class="data row1 col430">0</td>
<td id="T_0aa46_row1_col431" class="data row1 col431">0</td>
<td id="T_0aa46_row1_col432" class="data row1 col432">0</td>
<td id="T_0aa46_row1_col433" class="data row1 col433">0</td>
<td id="T_0aa46_row1_col434" class="data row1 col434">0</td>
<td id="T_0aa46_row1_col435" class="data row1 col435">0</td>
<td id="T_0aa46_row1_col436" class="data row1 col436">0</td>
<td id="T_0aa46_row1_col437" class="data row1 col437">0</td>
<td id="T_0aa46_row1_col438" class="data row1 col438">0</td>
<td id="T_0aa46_row1_col439" class="data row1 col439">0</td>
<td id="T_0aa46_row1_col440" class="data row1 col440">0</td>
<td id="T_0aa46_row1_col441" class="data row1 col441">0</td>
<td id="T_0aa46_row1_col442" class="data row1 col442">0</td>
<td id="T_0aa46_row1_col443" class="data row1 col443">0</td>
<td id="T_0aa46_row1_col444" class="data row1 col444">0</td>
<td id="T_0aa46_row1_col445" class="data row1 col445">0</td>
<td id="T_0aa46_row1_col446" class="data row1 col446">0</td>
<td id="T_0aa46_row1_col447" class="data row1 col447">0</td>
<td id="T_0aa46_row1_col448" class="data row1 col448">0</td>
<td id="T_0aa46_row1_col449" class="data row1 col449">0</td>
<td id="T_0aa46_row1_col450" class="data row1 col450">0</td>
<td id="T_0aa46_row1_col451" class="data row1 col451">0</td>
<td id="T_0aa46_row1_col452" class="data row1 col452">0</td>
<td id="T_0aa46_row1_col453" class="data row1 col453">0</td>
<td id="T_0aa46_row1_col454" class="data row1 col454">0</td>
<td id="T_0aa46_row1_col455" class="data row1 col455">0</td>
<td id="T_0aa46_row1_col456" class="data row1 col456">0</td>
<td id="T_0aa46_row1_col457" class="data row1 col457">0</td>
<td id="T_0aa46_row1_col458" class="data row1 col458">0</td>
<td id="T_0aa46_row1_col459" class="data row1 col459">0</td>
<td id="T_0aa46_row1_col460" class="data row1 col460">0</td>
<td id="T_0aa46_row1_col461" class="data row1 col461">0</td>
<td id="T_0aa46_row1_col462" class="data row1 col462">0</td>
<td id="T_0aa46_row1_col463" class="data row1 col463">0</td>
<td id="T_0aa46_row1_col464" class="data row1 col464">0</td>
<td id="T_0aa46_row1_col465" class="data row1 col465">0</td>
<td id="T_0aa46_row1_col466" class="data row1 col466">0</td>
<td id="T_0aa46_row1_col467" class="data row1 col467">0</td>
<td id="T_0aa46_row1_col468" class="data row1 col468">0</td>
<td id="T_0aa46_row1_col469" class="data row1 col469">0</td>
<td id="T_0aa46_row1_col470" class="data row1 col470">0</td>
<td id="T_0aa46_row1_col471" class="data row1 col471">0</td>
<td id="T_0aa46_row1_col472" class="data row1 col472">0</td>
<td id="T_0aa46_row1_col473" class="data row1 col473">0</td>
<td id="T_0aa46_row1_col474" class="data row1 col474">0</td>
<td id="T_0aa46_row1_col475" class="data row1 col475">0</td>
<td id="T_0aa46_row1_col476" class="data row1 col476">0</td>
<td id="T_0aa46_row1_col477" class="data row1 col477">0</td>
<td id="T_0aa46_row1_col478" class="data row1 col478">0</td>
<td id="T_0aa46_row1_col479" class="data row1 col479">0</td>
<td id="T_0aa46_row1_col480" class="data row1 col480">0</td>
<td id="T_0aa46_row1_col481" class="data row1 col481">0</td>
<td id="T_0aa46_row1_col482" class="data row1 col482">0</td>
<td id="T_0aa46_row1_col483" class="data row1 col483">0</td>
</tr>
<tr class="odd">
<th id="T_0aa46_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">Polygalaceae</th>
<td id="T_0aa46_row2_col0" class="data row2 col0">0</td>
<td id="T_0aa46_row2_col1" class="data row2 col1">0</td>
<td id="T_0aa46_row2_col2" class="data row2 col2">0</td>
<td id="T_0aa46_row2_col3" class="data row2 col3">0</td>
<td id="T_0aa46_row2_col4" class="data row2 col4">0</td>
<td id="T_0aa46_row2_col5" class="data row2 col5">0</td>
<td id="T_0aa46_row2_col6" class="data row2 col6">0</td>
<td id="T_0aa46_row2_col7" class="data row2 col7">1</td>
<td id="T_0aa46_row2_col8" class="data row2 col8">0</td>
<td id="T_0aa46_row2_col9" class="data row2 col9">0</td>
<td id="T_0aa46_row2_col10" class="data row2 col10">0</td>
<td id="T_0aa46_row2_col11" class="data row2 col11">0</td>
<td id="T_0aa46_row2_col12" class="data row2 col12">0</td>
<td id="T_0aa46_row2_col13" class="data row2 col13">0</td>
<td id="T_0aa46_row2_col14" class="data row2 col14">0</td>
<td id="T_0aa46_row2_col15" class="data row2 col15">1</td>
<td id="T_0aa46_row2_col16" class="data row2 col16">1</td>
<td id="T_0aa46_row2_col17" class="data row2 col17">1</td>
<td id="T_0aa46_row2_col18" class="data row2 col18">0</td>
<td id="T_0aa46_row2_col19" class="data row2 col19">0</td>
<td id="T_0aa46_row2_col20" class="data row2 col20">0</td>
<td id="T_0aa46_row2_col21" class="data row2 col21">1</td>
<td id="T_0aa46_row2_col22" class="data row2 col22">1</td>
<td id="T_0aa46_row2_col23" class="data row2 col23">1</td>
<td id="T_0aa46_row2_col24" class="data row2 col24">1</td>
<td id="T_0aa46_row2_col25" class="data row2 col25">1</td>
<td id="T_0aa46_row2_col26" class="data row2 col26">1</td>
<td id="T_0aa46_row2_col27" class="data row2 col27">1</td>
<td id="T_0aa46_row2_col28" class="data row2 col28">1</td>
<td id="T_0aa46_row2_col29" class="data row2 col29">1</td>
<td id="T_0aa46_row2_col30" class="data row2 col30">1</td>
<td id="T_0aa46_row2_col31" class="data row2 col31">1</td>
<td id="T_0aa46_row2_col32" class="data row2 col32">1</td>
<td id="T_0aa46_row2_col33" class="data row2 col33">1</td>
<td id="T_0aa46_row2_col34" class="data row2 col34">1</td>
<td id="T_0aa46_row2_col35" class="data row2 col35">1</td>
<td id="T_0aa46_row2_col36" class="data row2 col36">1</td>
<td id="T_0aa46_row2_col37" class="data row2 col37">1</td>
<td id="T_0aa46_row2_col38" class="data row2 col38">1</td>
<td id="T_0aa46_row2_col39" class="data row2 col39">1</td>
<td id="T_0aa46_row2_col40" class="data row2 col40">1</td>
<td id="T_0aa46_row2_col41" class="data row2 col41">0</td>
<td id="T_0aa46_row2_col42" class="data row2 col42">0</td>
<td id="T_0aa46_row2_col43" class="data row2 col43">0</td>
<td id="T_0aa46_row2_col44" class="data row2 col44">0</td>
<td id="T_0aa46_row2_col45" class="data row2 col45">0</td>
<td id="T_0aa46_row2_col46" class="data row2 col46">0</td>
<td id="T_0aa46_row2_col47" class="data row2 col47">0</td>
<td id="T_0aa46_row2_col48" class="data row2 col48">0</td>
<td id="T_0aa46_row2_col49" class="data row2 col49">0</td>
<td id="T_0aa46_row2_col50" class="data row2 col50">0</td>
<td id="T_0aa46_row2_col51" class="data row2 col51">0</td>
<td id="T_0aa46_row2_col52" class="data row2 col52">0</td>
<td id="T_0aa46_row2_col53" class="data row2 col53">0</td>
<td id="T_0aa46_row2_col54" class="data row2 col54">0</td>
<td id="T_0aa46_row2_col55" class="data row2 col55">0</td>
<td id="T_0aa46_row2_col56" class="data row2 col56">0</td>
<td id="T_0aa46_row2_col57" class="data row2 col57">0</td>
<td id="T_0aa46_row2_col58" class="data row2 col58">0</td>
<td id="T_0aa46_row2_col59" class="data row2 col59">0</td>
<td id="T_0aa46_row2_col60" class="data row2 col60">0</td>
<td id="T_0aa46_row2_col61" class="data row2 col61">0</td>
<td id="T_0aa46_row2_col62" class="data row2 col62">0</td>
<td id="T_0aa46_row2_col63" class="data row2 col63">0</td>
<td id="T_0aa46_row2_col64" class="data row2 col64">0</td>
<td id="T_0aa46_row2_col65" class="data row2 col65">0</td>
<td id="T_0aa46_row2_col66" class="data row2 col66">0</td>
<td id="T_0aa46_row2_col67" class="data row2 col67">0</td>
<td id="T_0aa46_row2_col68" class="data row2 col68">0</td>
<td id="T_0aa46_row2_col69" class="data row2 col69">0</td>
<td id="T_0aa46_row2_col70" class="data row2 col70">0</td>
<td id="T_0aa46_row2_col71" class="data row2 col71">0</td>
<td id="T_0aa46_row2_col72" class="data row2 col72">0</td>
<td id="T_0aa46_row2_col73" class="data row2 col73">0</td>
<td id="T_0aa46_row2_col74" class="data row2 col74">0</td>
<td id="T_0aa46_row2_col75" class="data row2 col75">0</td>
<td id="T_0aa46_row2_col76" class="data row2 col76">0</td>
<td id="T_0aa46_row2_col77" class="data row2 col77">0</td>
<td id="T_0aa46_row2_col78" class="data row2 col78">0</td>
<td id="T_0aa46_row2_col79" class="data row2 col79">0</td>
<td id="T_0aa46_row2_col80" class="data row2 col80">0</td>
<td id="T_0aa46_row2_col81" class="data row2 col81">0</td>
<td id="T_0aa46_row2_col82" class="data row2 col82">0</td>
<td id="T_0aa46_row2_col83" class="data row2 col83">0</td>
<td id="T_0aa46_row2_col84" class="data row2 col84">0</td>
<td id="T_0aa46_row2_col85" class="data row2 col85">0</td>
<td id="T_0aa46_row2_col86" class="data row2 col86">0</td>
<td id="T_0aa46_row2_col87" class="data row2 col87">0</td>
<td id="T_0aa46_row2_col88" class="data row2 col88">0</td>
<td id="T_0aa46_row2_col89" class="data row2 col89">0</td>
<td id="T_0aa46_row2_col90" class="data row2 col90">0</td>
<td id="T_0aa46_row2_col91" class="data row2 col91">0</td>
<td id="T_0aa46_row2_col92" class="data row2 col92">0</td>
<td id="T_0aa46_row2_col93" class="data row2 col93">0</td>
<td id="T_0aa46_row2_col94" class="data row2 col94">0</td>
<td id="T_0aa46_row2_col95" class="data row2 col95">0</td>
<td id="T_0aa46_row2_col96" class="data row2 col96">0</td>
<td id="T_0aa46_row2_col97" class="data row2 col97">0</td>
<td id="T_0aa46_row2_col98" class="data row2 col98">0</td>
<td id="T_0aa46_row2_col99" class="data row2 col99">0</td>
<td id="T_0aa46_row2_col100" class="data row2 col100">0</td>
<td id="T_0aa46_row2_col101" class="data row2 col101">0</td>
<td id="T_0aa46_row2_col102" class="data row2 col102">0</td>
<td id="T_0aa46_row2_col103" class="data row2 col103">0</td>
<td id="T_0aa46_row2_col104" class="data row2 col104">0</td>
<td id="T_0aa46_row2_col105" class="data row2 col105">0</td>
<td id="T_0aa46_row2_col106" class="data row2 col106">2</td>
<td id="T_0aa46_row2_col107" class="data row2 col107">1</td>
<td id="T_0aa46_row2_col108" class="data row2 col108">1</td>
<td id="T_0aa46_row2_col109" class="data row2 col109">2</td>
<td id="T_0aa46_row2_col110" class="data row2 col110">1</td>
<td id="T_0aa46_row2_col111" class="data row2 col111">2</td>
<td id="T_0aa46_row2_col112" class="data row2 col112">2</td>
<td id="T_0aa46_row2_col113" class="data row2 col113">1</td>
<td id="T_0aa46_row2_col114" class="data row2 col114">2</td>
<td id="T_0aa46_row2_col115" class="data row2 col115">1</td>
<td id="T_0aa46_row2_col116" class="data row2 col116">1</td>
<td id="T_0aa46_row2_col117" class="data row2 col117">1</td>
<td id="T_0aa46_row2_col118" class="data row2 col118">2</td>
<td id="T_0aa46_row2_col119" class="data row2 col119">0</td>
<td id="T_0aa46_row2_col120" class="data row2 col120">0</td>
<td id="T_0aa46_row2_col121" class="data row2 col121">0</td>
<td id="T_0aa46_row2_col122" class="data row2 col122">0</td>
<td id="T_0aa46_row2_col123" class="data row2 col123">0</td>
<td id="T_0aa46_row2_col124" class="data row2 col124">1</td>
<td id="T_0aa46_row2_col125" class="data row2 col125">1</td>
<td id="T_0aa46_row2_col126" class="data row2 col126">1</td>
<td id="T_0aa46_row2_col127" class="data row2 col127">0</td>
<td id="T_0aa46_row2_col128" class="data row2 col128">1</td>
<td id="T_0aa46_row2_col129" class="data row2 col129">0</td>
<td id="T_0aa46_row2_col130" class="data row2 col130">0</td>
<td id="T_0aa46_row2_col131" class="data row2 col131">0</td>
<td id="T_0aa46_row2_col132" class="data row2 col132">0</td>
<td id="T_0aa46_row2_col133" class="data row2 col133">0</td>
<td id="T_0aa46_row2_col134" class="data row2 col134">0</td>
<td id="T_0aa46_row2_col135" class="data row2 col135">0</td>
<td id="T_0aa46_row2_col136" class="data row2 col136">0</td>
<td id="T_0aa46_row2_col137" class="data row2 col137">0</td>
<td id="T_0aa46_row2_col138" class="data row2 col138">0</td>
<td id="T_0aa46_row2_col139" class="data row2 col139">0</td>
<td id="T_0aa46_row2_col140" class="data row2 col140">0</td>
<td id="T_0aa46_row2_col141" class="data row2 col141">0</td>
<td id="T_0aa46_row2_col142" class="data row2 col142">0</td>
<td id="T_0aa46_row2_col143" class="data row2 col143">0</td>
<td id="T_0aa46_row2_col144" class="data row2 col144">0</td>
<td id="T_0aa46_row2_col145" class="data row2 col145">0</td>
<td id="T_0aa46_row2_col146" class="data row2 col146">0</td>
<td id="T_0aa46_row2_col147" class="data row2 col147">1</td>
<td id="T_0aa46_row2_col148" class="data row2 col148">1</td>
<td id="T_0aa46_row2_col149" class="data row2 col149">0</td>
<td id="T_0aa46_row2_col150" class="data row2 col150">0</td>
<td id="T_0aa46_row2_col151" class="data row2 col151">0</td>
<td id="T_0aa46_row2_col152" class="data row2 col152">0</td>
<td id="T_0aa46_row2_col153" class="data row2 col153">0</td>
<td id="T_0aa46_row2_col154" class="data row2 col154">0</td>
<td id="T_0aa46_row2_col155" class="data row2 col155">1</td>
<td id="T_0aa46_row2_col156" class="data row2 col156">1</td>
<td id="T_0aa46_row2_col157" class="data row2 col157">1</td>
<td id="T_0aa46_row2_col158" class="data row2 col158">1</td>
<td id="T_0aa46_row2_col159" class="data row2 col159">1</td>
<td id="T_0aa46_row2_col160" class="data row2 col160">1</td>
<td id="T_0aa46_row2_col161" class="data row2 col161">0</td>
<td id="T_0aa46_row2_col162" class="data row2 col162">0</td>
<td id="T_0aa46_row2_col163" class="data row2 col163">0</td>
<td id="T_0aa46_row2_col164" class="data row2 col164">0</td>
<td id="T_0aa46_row2_col165" class="data row2 col165">0</td>
<td id="T_0aa46_row2_col166" class="data row2 col166">1</td>
<td id="T_0aa46_row2_col167" class="data row2 col167">1</td>
<td id="T_0aa46_row2_col168" class="data row2 col168">1</td>
<td id="T_0aa46_row2_col169" class="data row2 col169">1</td>
<td id="T_0aa46_row2_col170" class="data row2 col170">1</td>
<td id="T_0aa46_row2_col171" class="data row2 col171">2</td>
<td id="T_0aa46_row2_col172" class="data row2 col172">2</td>
<td id="T_0aa46_row2_col173" class="data row2 col173">1</td>
<td id="T_0aa46_row2_col174" class="data row2 col174">1</td>
<td id="T_0aa46_row2_col175" class="data row2 col175">1</td>
<td id="T_0aa46_row2_col176" class="data row2 col176">1</td>
<td id="T_0aa46_row2_col177" class="data row2 col177">1</td>
<td id="T_0aa46_row2_col178" class="data row2 col178">1</td>
<td id="T_0aa46_row2_col179" class="data row2 col179">0</td>
<td id="T_0aa46_row2_col180" class="data row2 col180">0</td>
<td id="T_0aa46_row2_col181" class="data row2 col181">1</td>
<td id="T_0aa46_row2_col182" class="data row2 col182">1</td>
<td id="T_0aa46_row2_col183" class="data row2 col183">1</td>
<td id="T_0aa46_row2_col184" class="data row2 col184">1</td>
<td id="T_0aa46_row2_col185" class="data row2 col185">1</td>
<td id="T_0aa46_row2_col186" class="data row2 col186">1</td>
<td id="T_0aa46_row2_col187" class="data row2 col187">1</td>
<td id="T_0aa46_row2_col188" class="data row2 col188">0</td>
<td id="T_0aa46_row2_col189" class="data row2 col189">0</td>
<td id="T_0aa46_row2_col190" class="data row2 col190">1</td>
<td id="T_0aa46_row2_col191" class="data row2 col191">1</td>
<td id="T_0aa46_row2_col192" class="data row2 col192">1</td>
<td id="T_0aa46_row2_col193" class="data row2 col193">1</td>
<td id="T_0aa46_row2_col194" class="data row2 col194">1</td>
<td id="T_0aa46_row2_col195" class="data row2 col195">1</td>
<td id="T_0aa46_row2_col196" class="data row2 col196">1</td>
<td id="T_0aa46_row2_col197" class="data row2 col197">1</td>
<td id="T_0aa46_row2_col198" class="data row2 col198">0</td>
<td id="T_0aa46_row2_col199" class="data row2 col199">0</td>
<td id="T_0aa46_row2_col200" class="data row2 col200">1</td>
<td id="T_0aa46_row2_col201" class="data row2 col201">0</td>
<td id="T_0aa46_row2_col202" class="data row2 col202">0</td>
<td id="T_0aa46_row2_col203" class="data row2 col203">0</td>
<td id="T_0aa46_row2_col204" class="data row2 col204">0</td>
<td id="T_0aa46_row2_col205" class="data row2 col205">1</td>
<td id="T_0aa46_row2_col206" class="data row2 col206">0</td>
<td id="T_0aa46_row2_col207" class="data row2 col207">0</td>
<td id="T_0aa46_row2_col208" class="data row2 col208">1</td>
<td id="T_0aa46_row2_col209" class="data row2 col209">1</td>
<td id="T_0aa46_row2_col210" class="data row2 col210">1</td>
<td id="T_0aa46_row2_col211" class="data row2 col211">1</td>
<td id="T_0aa46_row2_col212" class="data row2 col212">0</td>
<td id="T_0aa46_row2_col213" class="data row2 col213">1</td>
<td id="T_0aa46_row2_col214" class="data row2 col214">1</td>
<td id="T_0aa46_row2_col215" class="data row2 col215">1</td>
<td id="T_0aa46_row2_col216" class="data row2 col216">1</td>
<td id="T_0aa46_row2_col217" class="data row2 col217">1</td>
<td id="T_0aa46_row2_col218" class="data row2 col218">1</td>
<td id="T_0aa46_row2_col219" class="data row2 col219">0</td>
<td id="T_0aa46_row2_col220" class="data row2 col220">0</td>
<td id="T_0aa46_row2_col221" class="data row2 col221">0</td>
<td id="T_0aa46_row2_col222" class="data row2 col222">0</td>
<td id="T_0aa46_row2_col223" class="data row2 col223">0</td>
<td id="T_0aa46_row2_col224" class="data row2 col224">0</td>
<td id="T_0aa46_row2_col225" class="data row2 col225">0</td>
<td id="T_0aa46_row2_col226" class="data row2 col226">0</td>
<td id="T_0aa46_row2_col227" class="data row2 col227">0</td>
<td id="T_0aa46_row2_col228" class="data row2 col228">0</td>
<td id="T_0aa46_row2_col229" class="data row2 col229">0</td>
<td id="T_0aa46_row2_col230" class="data row2 col230">0</td>
<td id="T_0aa46_row2_col231" class="data row2 col231">0</td>
<td id="T_0aa46_row2_col232" class="data row2 col232">0</td>
<td id="T_0aa46_row2_col233" class="data row2 col233">0</td>
<td id="T_0aa46_row2_col234" class="data row2 col234">0</td>
<td id="T_0aa46_row2_col235" class="data row2 col235">0</td>
<td id="T_0aa46_row2_col236" class="data row2 col236">0</td>
<td id="T_0aa46_row2_col237" class="data row2 col237">0</td>
<td id="T_0aa46_row2_col238" class="data row2 col238">0</td>
<td id="T_0aa46_row2_col239" class="data row2 col239">0</td>
<td id="T_0aa46_row2_col240" class="data row2 col240">0</td>
<td id="T_0aa46_row2_col241" class="data row2 col241">0</td>
<td id="T_0aa46_row2_col242" class="data row2 col242">0</td>
<td id="T_0aa46_row2_col243" class="data row2 col243">0</td>
<td id="T_0aa46_row2_col244" class="data row2 col244">0</td>
<td id="T_0aa46_row2_col245" class="data row2 col245">0</td>
<td id="T_0aa46_row2_col246" class="data row2 col246">1</td>
<td id="T_0aa46_row2_col247" class="data row2 col247">1</td>
<td id="T_0aa46_row2_col248" class="data row2 col248">0</td>
<td id="T_0aa46_row2_col249" class="data row2 col249">0</td>
<td id="T_0aa46_row2_col250" class="data row2 col250">0</td>
<td id="T_0aa46_row2_col251" class="data row2 col251">0</td>
<td id="T_0aa46_row2_col252" class="data row2 col252">0</td>
<td id="T_0aa46_row2_col253" class="data row2 col253">0</td>
<td id="T_0aa46_row2_col254" class="data row2 col254">0</td>
<td id="T_0aa46_row2_col255" class="data row2 col255">0</td>
<td id="T_0aa46_row2_col256" class="data row2 col256">0</td>
<td id="T_0aa46_row2_col257" class="data row2 col257">0</td>
<td id="T_0aa46_row2_col258" class="data row2 col258">0</td>
<td id="T_0aa46_row2_col259" class="data row2 col259">0</td>
<td id="T_0aa46_row2_col260" class="data row2 col260">0</td>
<td id="T_0aa46_row2_col261" class="data row2 col261">0</td>
<td id="T_0aa46_row2_col262" class="data row2 col262">0</td>
<td id="T_0aa46_row2_col263" class="data row2 col263">0</td>
<td id="T_0aa46_row2_col264" class="data row2 col264">0</td>
<td id="T_0aa46_row2_col265" class="data row2 col265">0</td>
<td id="T_0aa46_row2_col266" class="data row2 col266">0</td>
<td id="T_0aa46_row2_col267" class="data row2 col267">0</td>
<td id="T_0aa46_row2_col268" class="data row2 col268">0</td>
<td id="T_0aa46_row2_col269" class="data row2 col269">0</td>
<td id="T_0aa46_row2_col270" class="data row2 col270">0</td>
<td id="T_0aa46_row2_col271" class="data row2 col271">0</td>
<td id="T_0aa46_row2_col272" class="data row2 col272">0</td>
<td id="T_0aa46_row2_col273" class="data row2 col273">0</td>
<td id="T_0aa46_row2_col274" class="data row2 col274">0</td>
<td id="T_0aa46_row2_col275" class="data row2 col275">0</td>
<td id="T_0aa46_row2_col276" class="data row2 col276">0</td>
<td id="T_0aa46_row2_col277" class="data row2 col277">0</td>
<td id="T_0aa46_row2_col278" class="data row2 col278">0</td>
<td id="T_0aa46_row2_col279" class="data row2 col279">0</td>
<td id="T_0aa46_row2_col280" class="data row2 col280">0</td>
<td id="T_0aa46_row2_col281" class="data row2 col281">0</td>
<td id="T_0aa46_row2_col282" class="data row2 col282">0</td>
<td id="T_0aa46_row2_col283" class="data row2 col283">0</td>
<td id="T_0aa46_row2_col284" class="data row2 col284">0</td>
<td id="T_0aa46_row2_col285" class="data row2 col285">0</td>
<td id="T_0aa46_row2_col286" class="data row2 col286">0</td>
<td id="T_0aa46_row2_col287" class="data row2 col287">0</td>
<td id="T_0aa46_row2_col288" class="data row2 col288">0</td>
<td id="T_0aa46_row2_col289" class="data row2 col289">0</td>
<td id="T_0aa46_row2_col290" class="data row2 col290">0</td>
<td id="T_0aa46_row2_col291" class="data row2 col291">0</td>
<td id="T_0aa46_row2_col292" class="data row2 col292">0</td>
<td id="T_0aa46_row2_col293" class="data row2 col293">0</td>
<td id="T_0aa46_row2_col294" class="data row2 col294">0</td>
<td id="T_0aa46_row2_col295" class="data row2 col295">0</td>
<td id="T_0aa46_row2_col296" class="data row2 col296">0</td>
<td id="T_0aa46_row2_col297" class="data row2 col297">0</td>
<td id="T_0aa46_row2_col298" class="data row2 col298">0</td>
<td id="T_0aa46_row2_col299" class="data row2 col299">0</td>
<td id="T_0aa46_row2_col300" class="data row2 col300">0</td>
<td id="T_0aa46_row2_col301" class="data row2 col301">0</td>
<td id="T_0aa46_row2_col302" class="data row2 col302">0</td>
<td id="T_0aa46_row2_col303" class="data row2 col303">0</td>
<td id="T_0aa46_row2_col304" class="data row2 col304">0</td>
<td id="T_0aa46_row2_col305" class="data row2 col305">0</td>
<td id="T_0aa46_row2_col306" class="data row2 col306">0</td>
<td id="T_0aa46_row2_col307" class="data row2 col307">0</td>
<td id="T_0aa46_row2_col308" class="data row2 col308">0</td>
<td id="T_0aa46_row2_col309" class="data row2 col309">0</td>
<td id="T_0aa46_row2_col310" class="data row2 col310">0</td>
<td id="T_0aa46_row2_col311" class="data row2 col311">0</td>
<td id="T_0aa46_row2_col312" class="data row2 col312">0</td>
<td id="T_0aa46_row2_col313" class="data row2 col313">0</td>
<td id="T_0aa46_row2_col314" class="data row2 col314">0</td>
<td id="T_0aa46_row2_col315" class="data row2 col315">0</td>
<td id="T_0aa46_row2_col316" class="data row2 col316">0</td>
<td id="T_0aa46_row2_col317" class="data row2 col317">1</td>
<td id="T_0aa46_row2_col318" class="data row2 col318">0</td>
<td id="T_0aa46_row2_col319" class="data row2 col319">0</td>
<td id="T_0aa46_row2_col320" class="data row2 col320">0</td>
<td id="T_0aa46_row2_col321" class="data row2 col321">0</td>
<td id="T_0aa46_row2_col322" class="data row2 col322">0</td>
<td id="T_0aa46_row2_col323" class="data row2 col323">0</td>
<td id="T_0aa46_row2_col324" class="data row2 col324">0</td>
<td id="T_0aa46_row2_col325" class="data row2 col325">0</td>
<td id="T_0aa46_row2_col326" class="data row2 col326">0</td>
<td id="T_0aa46_row2_col327" class="data row2 col327">0</td>
<td id="T_0aa46_row2_col328" class="data row2 col328">0</td>
<td id="T_0aa46_row2_col329" class="data row2 col329">1</td>
<td id="T_0aa46_row2_col330" class="data row2 col330">0</td>
<td id="T_0aa46_row2_col331" class="data row2 col331">0</td>
<td id="T_0aa46_row2_col332" class="data row2 col332">0</td>
<td id="T_0aa46_row2_col333" class="data row2 col333">0</td>
<td id="T_0aa46_row2_col334" class="data row2 col334">0</td>
<td id="T_0aa46_row2_col335" class="data row2 col335">0</td>
<td id="T_0aa46_row2_col336" class="data row2 col336">0</td>
<td id="T_0aa46_row2_col337" class="data row2 col337">0</td>
<td id="T_0aa46_row2_col338" class="data row2 col338">0</td>
<td id="T_0aa46_row2_col339" class="data row2 col339">0</td>
<td id="T_0aa46_row2_col340" class="data row2 col340">0</td>
<td id="T_0aa46_row2_col341" class="data row2 col341">0</td>
<td id="T_0aa46_row2_col342" class="data row2 col342">0</td>
<td id="T_0aa46_row2_col343" class="data row2 col343">0</td>
<td id="T_0aa46_row2_col344" class="data row2 col344">0</td>
<td id="T_0aa46_row2_col345" class="data row2 col345">0</td>
<td id="T_0aa46_row2_col346" class="data row2 col346">0</td>
<td id="T_0aa46_row2_col347" class="data row2 col347">0</td>
<td id="T_0aa46_row2_col348" class="data row2 col348">1</td>
<td id="T_0aa46_row2_col349" class="data row2 col349">0</td>
<td id="T_0aa46_row2_col350" class="data row2 col350">0</td>
<td id="T_0aa46_row2_col351" class="data row2 col351">0</td>
<td id="T_0aa46_row2_col352" class="data row2 col352">0</td>
<td id="T_0aa46_row2_col353" class="data row2 col353">0</td>
<td id="T_0aa46_row2_col354" class="data row2 col354">0</td>
<td id="T_0aa46_row2_col355" class="data row2 col355">0</td>
<td id="T_0aa46_row2_col356" class="data row2 col356">0</td>
<td id="T_0aa46_row2_col357" class="data row2 col357">0</td>
<td id="T_0aa46_row2_col358" class="data row2 col358">0</td>
<td id="T_0aa46_row2_col359" class="data row2 col359">0</td>
<td id="T_0aa46_row2_col360" class="data row2 col360">0</td>
<td id="T_0aa46_row2_col361" class="data row2 col361">1</td>
<td id="T_0aa46_row2_col362" class="data row2 col362">0</td>
<td id="T_0aa46_row2_col363" class="data row2 col363">0</td>
<td id="T_0aa46_row2_col364" class="data row2 col364">1</td>
<td id="T_0aa46_row2_col365" class="data row2 col365">0</td>
<td id="T_0aa46_row2_col366" class="data row2 col366">0</td>
<td id="T_0aa46_row2_col367" class="data row2 col367">0</td>
<td id="T_0aa46_row2_col368" class="data row2 col368">0</td>
<td id="T_0aa46_row2_col369" class="data row2 col369">0</td>
<td id="T_0aa46_row2_col370" class="data row2 col370">0</td>
<td id="T_0aa46_row2_col371" class="data row2 col371">0</td>
<td id="T_0aa46_row2_col372" class="data row2 col372">0</td>
<td id="T_0aa46_row2_col373" class="data row2 col373">0</td>
<td id="T_0aa46_row2_col374" class="data row2 col374">0</td>
<td id="T_0aa46_row2_col375" class="data row2 col375">0</td>
<td id="T_0aa46_row2_col376" class="data row2 col376">0</td>
<td id="T_0aa46_row2_col377" class="data row2 col377">0</td>
<td id="T_0aa46_row2_col378" class="data row2 col378">0</td>
<td id="T_0aa46_row2_col379" class="data row2 col379">0</td>
<td id="T_0aa46_row2_col380" class="data row2 col380">0</td>
<td id="T_0aa46_row2_col381" class="data row2 col381">0</td>
<td id="T_0aa46_row2_col382" class="data row2 col382">0</td>
<td id="T_0aa46_row2_col383" class="data row2 col383">0</td>
<td id="T_0aa46_row2_col384" class="data row2 col384">0</td>
<td id="T_0aa46_row2_col385" class="data row2 col385">0</td>
<td id="T_0aa46_row2_col386" class="data row2 col386">0</td>
<td id="T_0aa46_row2_col387" class="data row2 col387">0</td>
<td id="T_0aa46_row2_col388" class="data row2 col388">0</td>
<td id="T_0aa46_row2_col389" class="data row2 col389">0</td>
<td id="T_0aa46_row2_col390" class="data row2 col390">0</td>
<td id="T_0aa46_row2_col391" class="data row2 col391">0</td>
<td id="T_0aa46_row2_col392" class="data row2 col392">0</td>
<td id="T_0aa46_row2_col393" class="data row2 col393">0</td>
<td id="T_0aa46_row2_col394" class="data row2 col394">0</td>
<td id="T_0aa46_row2_col395" class="data row2 col395">0</td>
<td id="T_0aa46_row2_col396" class="data row2 col396">0</td>
<td id="T_0aa46_row2_col397" class="data row2 col397">0</td>
<td id="T_0aa46_row2_col398" class="data row2 col398">0</td>
<td id="T_0aa46_row2_col399" class="data row2 col399">0</td>
<td id="T_0aa46_row2_col400" class="data row2 col400">0</td>
<td id="T_0aa46_row2_col401" class="data row2 col401">0</td>
<td id="T_0aa46_row2_col402" class="data row2 col402">0</td>
<td id="T_0aa46_row2_col403" class="data row2 col403">0</td>
<td id="T_0aa46_row2_col404" class="data row2 col404">0</td>
<td id="T_0aa46_row2_col405" class="data row2 col405">0</td>
<td id="T_0aa46_row2_col406" class="data row2 col406">0</td>
<td id="T_0aa46_row2_col407" class="data row2 col407">0</td>
<td id="T_0aa46_row2_col408" class="data row2 col408">0</td>
<td id="T_0aa46_row2_col409" class="data row2 col409">0</td>
<td id="T_0aa46_row2_col410" class="data row2 col410">0</td>
<td id="T_0aa46_row2_col411" class="data row2 col411">0</td>
<td id="T_0aa46_row2_col412" class="data row2 col412">1</td>
<td id="T_0aa46_row2_col413" class="data row2 col413">0</td>
<td id="T_0aa46_row2_col414" class="data row2 col414">0</td>
<td id="T_0aa46_row2_col415" class="data row2 col415">0</td>
<td id="T_0aa46_row2_col416" class="data row2 col416">0</td>
<td id="T_0aa46_row2_col417" class="data row2 col417">0</td>
<td id="T_0aa46_row2_col418" class="data row2 col418">0</td>
<td id="T_0aa46_row2_col419" class="data row2 col419">0</td>
<td id="T_0aa46_row2_col420" class="data row2 col420">0</td>
<td id="T_0aa46_row2_col421" class="data row2 col421">0</td>
<td id="T_0aa46_row2_col422" class="data row2 col422">0</td>
<td id="T_0aa46_row2_col423" class="data row2 col423">0</td>
<td id="T_0aa46_row2_col424" class="data row2 col424">0</td>
<td id="T_0aa46_row2_col425" class="data row2 col425">0</td>
<td id="T_0aa46_row2_col426" class="data row2 col426">0</td>
<td id="T_0aa46_row2_col427" class="data row2 col427">0</td>
<td id="T_0aa46_row2_col428" class="data row2 col428">0</td>
<td id="T_0aa46_row2_col429" class="data row2 col429">0</td>
<td id="T_0aa46_row2_col430" class="data row2 col430">0</td>
<td id="T_0aa46_row2_col431" class="data row2 col431">0</td>
<td id="T_0aa46_row2_col432" class="data row2 col432">0</td>
<td id="T_0aa46_row2_col433" class="data row2 col433">1</td>
<td id="T_0aa46_row2_col434" class="data row2 col434">0</td>
<td id="T_0aa46_row2_col435" class="data row2 col435">0</td>
<td id="T_0aa46_row2_col436" class="data row2 col436">1</td>
<td id="T_0aa46_row2_col437" class="data row2 col437">1</td>
<td id="T_0aa46_row2_col438" class="data row2 col438">1</td>
<td id="T_0aa46_row2_col439" class="data row2 col439">1</td>
<td id="T_0aa46_row2_col440" class="data row2 col440">0</td>
<td id="T_0aa46_row2_col441" class="data row2 col441">0</td>
<td id="T_0aa46_row2_col442" class="data row2 col442">0</td>
<td id="T_0aa46_row2_col443" class="data row2 col443">0</td>
<td id="T_0aa46_row2_col444" class="data row2 col444">0</td>
<td id="T_0aa46_row2_col445" class="data row2 col445">0</td>
<td id="T_0aa46_row2_col446" class="data row2 col446">0</td>
<td id="T_0aa46_row2_col447" class="data row2 col447">0</td>
<td id="T_0aa46_row2_col448" class="data row2 col448">0</td>
<td id="T_0aa46_row2_col449" class="data row2 col449">0</td>
<td id="T_0aa46_row2_col450" class="data row2 col450">0</td>
<td id="T_0aa46_row2_col451" class="data row2 col451">0</td>
<td id="T_0aa46_row2_col452" class="data row2 col452">0</td>
<td id="T_0aa46_row2_col453" class="data row2 col453">0</td>
<td id="T_0aa46_row2_col454" class="data row2 col454">1</td>
<td id="T_0aa46_row2_col455" class="data row2 col455">0</td>
<td id="T_0aa46_row2_col456" class="data row2 col456">0</td>
<td id="T_0aa46_row2_col457" class="data row2 col457">0</td>
<td id="T_0aa46_row2_col458" class="data row2 col458">0</td>
<td id="T_0aa46_row2_col459" class="data row2 col459">0</td>
<td id="T_0aa46_row2_col460" class="data row2 col460">0</td>
<td id="T_0aa46_row2_col461" class="data row2 col461">0</td>
<td id="T_0aa46_row2_col462" class="data row2 col462">0</td>
<td id="T_0aa46_row2_col463" class="data row2 col463">0</td>
<td id="T_0aa46_row2_col464" class="data row2 col464">0</td>
<td id="T_0aa46_row2_col465" class="data row2 col465">0</td>
<td id="T_0aa46_row2_col466" class="data row2 col466">0</td>
<td id="T_0aa46_row2_col467" class="data row2 col467">0</td>
<td id="T_0aa46_row2_col468" class="data row2 col468">0</td>
<td id="T_0aa46_row2_col469" class="data row2 col469">0</td>
<td id="T_0aa46_row2_col470" class="data row2 col470">0</td>
<td id="T_0aa46_row2_col471" class="data row2 col471">0</td>
<td id="T_0aa46_row2_col472" class="data row2 col472">0</td>
<td id="T_0aa46_row2_col473" class="data row2 col473">0</td>
<td id="T_0aa46_row2_col474" class="data row2 col474">0</td>
<td id="T_0aa46_row2_col475" class="data row2 col475">0</td>
<td id="T_0aa46_row2_col476" class="data row2 col476">0</td>
<td id="T_0aa46_row2_col477" class="data row2 col477">0</td>
<td id="T_0aa46_row2_col478" class="data row2 col478">0</td>
<td id="T_0aa46_row2_col479" class="data row2 col479">0</td>
<td id="T_0aa46_row2_col480" class="data row2 col480">0</td>
<td id="T_0aa46_row2_col481" class="data row2 col481">0</td>
<td id="T_0aa46_row2_col482" class="data row2 col482">0</td>
<td id="T_0aa46_row2_col483" class="data row2 col483">0</td>
</tr>
<tr class="even">
<th id="T_0aa46_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">Quillajaceae</th>
<td id="T_0aa46_row3_col0" class="data row3 col0">0</td>
<td id="T_0aa46_row3_col1" class="data row3 col1">0</td>
<td id="T_0aa46_row3_col2" class="data row3 col2">0</td>
<td id="T_0aa46_row3_col3" class="data row3 col3">0</td>
<td id="T_0aa46_row3_col4" class="data row3 col4">0</td>
<td id="T_0aa46_row3_col5" class="data row3 col5">0</td>
<td id="T_0aa46_row3_col6" class="data row3 col6">0</td>
<td id="T_0aa46_row3_col7" class="data row3 col7">0</td>
<td id="T_0aa46_row3_col8" class="data row3 col8">0</td>
<td id="T_0aa46_row3_col9" class="data row3 col9">0</td>
<td id="T_0aa46_row3_col10" class="data row3 col10">0</td>
<td id="T_0aa46_row3_col11" class="data row3 col11">0</td>
<td id="T_0aa46_row3_col12" class="data row3 col12">0</td>
<td id="T_0aa46_row3_col13" class="data row3 col13">1</td>
<td id="T_0aa46_row3_col14" class="data row3 col14">1</td>
<td id="T_0aa46_row3_col15" class="data row3 col15">0</td>
<td id="T_0aa46_row3_col16" class="data row3 col16">0</td>
<td id="T_0aa46_row3_col17" class="data row3 col17">0</td>
<td id="T_0aa46_row3_col18" class="data row3 col18">0</td>
<td id="T_0aa46_row3_col19" class="data row3 col19">1</td>
<td id="T_0aa46_row3_col20" class="data row3 col20">1</td>
<td id="T_0aa46_row3_col21" class="data row3 col21">0</td>
<td id="T_0aa46_row3_col22" class="data row3 col22">0</td>
<td id="T_0aa46_row3_col23" class="data row3 col23">0</td>
<td id="T_0aa46_row3_col24" class="data row3 col24">0</td>
<td id="T_0aa46_row3_col25" class="data row3 col25">0</td>
<td id="T_0aa46_row3_col26" class="data row3 col26">0</td>
<td id="T_0aa46_row3_col27" class="data row3 col27">0</td>
<td id="T_0aa46_row3_col28" class="data row3 col28">0</td>
<td id="T_0aa46_row3_col29" class="data row3 col29">0</td>
<td id="T_0aa46_row3_col30" class="data row3 col30">0</td>
<td id="T_0aa46_row3_col31" class="data row3 col31">0</td>
<td id="T_0aa46_row3_col32" class="data row3 col32">0</td>
<td id="T_0aa46_row3_col33" class="data row3 col33">0</td>
<td id="T_0aa46_row3_col34" class="data row3 col34">0</td>
<td id="T_0aa46_row3_col35" class="data row3 col35">0</td>
<td id="T_0aa46_row3_col36" class="data row3 col36">0</td>
<td id="T_0aa46_row3_col37" class="data row3 col37">0</td>
<td id="T_0aa46_row3_col38" class="data row3 col38">0</td>
<td id="T_0aa46_row3_col39" class="data row3 col39">0</td>
<td id="T_0aa46_row3_col40" class="data row3 col40">0</td>
<td id="T_0aa46_row3_col41" class="data row3 col41">0</td>
<td id="T_0aa46_row3_col42" class="data row3 col42">0</td>
<td id="T_0aa46_row3_col43" class="data row3 col43">0</td>
<td id="T_0aa46_row3_col44" class="data row3 col44">0</td>
<td id="T_0aa46_row3_col45" class="data row3 col45">0</td>
<td id="T_0aa46_row3_col46" class="data row3 col46">0</td>
<td id="T_0aa46_row3_col47" class="data row3 col47">0</td>
<td id="T_0aa46_row3_col48" class="data row3 col48">0</td>
<td id="T_0aa46_row3_col49" class="data row3 col49">0</td>
<td id="T_0aa46_row3_col50" class="data row3 col50">0</td>
<td id="T_0aa46_row3_col51" class="data row3 col51">0</td>
<td id="T_0aa46_row3_col52" class="data row3 col52">0</td>
<td id="T_0aa46_row3_col53" class="data row3 col53">0</td>
<td id="T_0aa46_row3_col54" class="data row3 col54">0</td>
<td id="T_0aa46_row3_col55" class="data row3 col55">0</td>
<td id="T_0aa46_row3_col56" class="data row3 col56">0</td>
<td id="T_0aa46_row3_col57" class="data row3 col57">0</td>
<td id="T_0aa46_row3_col58" class="data row3 col58">0</td>
<td id="T_0aa46_row3_col59" class="data row3 col59">0</td>
<td id="T_0aa46_row3_col60" class="data row3 col60">0</td>
<td id="T_0aa46_row3_col61" class="data row3 col61">0</td>
<td id="T_0aa46_row3_col62" class="data row3 col62">0</td>
<td id="T_0aa46_row3_col63" class="data row3 col63">0</td>
<td id="T_0aa46_row3_col64" class="data row3 col64">0</td>
<td id="T_0aa46_row3_col65" class="data row3 col65">0</td>
<td id="T_0aa46_row3_col66" class="data row3 col66">0</td>
<td id="T_0aa46_row3_col67" class="data row3 col67">0</td>
<td id="T_0aa46_row3_col68" class="data row3 col68">0</td>
<td id="T_0aa46_row3_col69" class="data row3 col69">0</td>
<td id="T_0aa46_row3_col70" class="data row3 col70">0</td>
<td id="T_0aa46_row3_col71" class="data row3 col71">0</td>
<td id="T_0aa46_row3_col72" class="data row3 col72">0</td>
<td id="T_0aa46_row3_col73" class="data row3 col73">0</td>
<td id="T_0aa46_row3_col74" class="data row3 col74">0</td>
<td id="T_0aa46_row3_col75" class="data row3 col75">0</td>
<td id="T_0aa46_row3_col76" class="data row3 col76">0</td>
<td id="T_0aa46_row3_col77" class="data row3 col77">1</td>
<td id="T_0aa46_row3_col78" class="data row3 col78">0</td>
<td id="T_0aa46_row3_col79" class="data row3 col79">0</td>
<td id="T_0aa46_row3_col80" class="data row3 col80">0</td>
<td id="T_0aa46_row3_col81" class="data row3 col81">0</td>
<td id="T_0aa46_row3_col82" class="data row3 col82">0</td>
<td id="T_0aa46_row3_col83" class="data row3 col83">0</td>
<td id="T_0aa46_row3_col84" class="data row3 col84">0</td>
<td id="T_0aa46_row3_col85" class="data row3 col85">0</td>
<td id="T_0aa46_row3_col86" class="data row3 col86">0</td>
<td id="T_0aa46_row3_col87" class="data row3 col87">0</td>
<td id="T_0aa46_row3_col88" class="data row3 col88">0</td>
<td id="T_0aa46_row3_col89" class="data row3 col89">0</td>
<td id="T_0aa46_row3_col90" class="data row3 col90">0</td>
<td id="T_0aa46_row3_col91" class="data row3 col91">0</td>
<td id="T_0aa46_row3_col92" class="data row3 col92">0</td>
<td id="T_0aa46_row3_col93" class="data row3 col93">0</td>
<td id="T_0aa46_row3_col94" class="data row3 col94">0</td>
<td id="T_0aa46_row3_col95" class="data row3 col95">0</td>
<td id="T_0aa46_row3_col96" class="data row3 col96">0</td>
<td id="T_0aa46_row3_col97" class="data row3 col97">0</td>
<td id="T_0aa46_row3_col98" class="data row3 col98">0</td>
<td id="T_0aa46_row3_col99" class="data row3 col99">0</td>
<td id="T_0aa46_row3_col100" class="data row3 col100">0</td>
<td id="T_0aa46_row3_col101" class="data row3 col101">0</td>
<td id="T_0aa46_row3_col102" class="data row3 col102">0</td>
<td id="T_0aa46_row3_col103" class="data row3 col103">0</td>
<td id="T_0aa46_row3_col104" class="data row3 col104">0</td>
<td id="T_0aa46_row3_col105" class="data row3 col105">0</td>
<td id="T_0aa46_row3_col106" class="data row3 col106">0</td>
<td id="T_0aa46_row3_col107" class="data row3 col107">0</td>
<td id="T_0aa46_row3_col108" class="data row3 col108">0</td>
<td id="T_0aa46_row3_col109" class="data row3 col109">0</td>
<td id="T_0aa46_row3_col110" class="data row3 col110">0</td>
<td id="T_0aa46_row3_col111" class="data row3 col111">0</td>
<td id="T_0aa46_row3_col112" class="data row3 col112">0</td>
<td id="T_0aa46_row3_col113" class="data row3 col113">0</td>
<td id="T_0aa46_row3_col114" class="data row3 col114">0</td>
<td id="T_0aa46_row3_col115" class="data row3 col115">0</td>
<td id="T_0aa46_row3_col116" class="data row3 col116">0</td>
<td id="T_0aa46_row3_col117" class="data row3 col117">0</td>
<td id="T_0aa46_row3_col118" class="data row3 col118">0</td>
<td id="T_0aa46_row3_col119" class="data row3 col119">0</td>
<td id="T_0aa46_row3_col120" class="data row3 col120">0</td>
<td id="T_0aa46_row3_col121" class="data row3 col121">0</td>
<td id="T_0aa46_row3_col122" class="data row3 col122">0</td>
<td id="T_0aa46_row3_col123" class="data row3 col123">1</td>
<td id="T_0aa46_row3_col124" class="data row3 col124">0</td>
<td id="T_0aa46_row3_col125" class="data row3 col125">0</td>
<td id="T_0aa46_row3_col126" class="data row3 col126">0</td>
<td id="T_0aa46_row3_col127" class="data row3 col127">0</td>
<td id="T_0aa46_row3_col128" class="data row3 col128">0</td>
<td id="T_0aa46_row3_col129" class="data row3 col129">0</td>
<td id="T_0aa46_row3_col130" class="data row3 col130">0</td>
<td id="T_0aa46_row3_col131" class="data row3 col131">0</td>
<td id="T_0aa46_row3_col132" class="data row3 col132">0</td>
<td id="T_0aa46_row3_col133" class="data row3 col133">0</td>
<td id="T_0aa46_row3_col134" class="data row3 col134">0</td>
<td id="T_0aa46_row3_col135" class="data row3 col135">0</td>
<td id="T_0aa46_row3_col136" class="data row3 col136">0</td>
<td id="T_0aa46_row3_col137" class="data row3 col137">0</td>
<td id="T_0aa46_row3_col138" class="data row3 col138">0</td>
<td id="T_0aa46_row3_col139" class="data row3 col139">0</td>
<td id="T_0aa46_row3_col140" class="data row3 col140">0</td>
<td id="T_0aa46_row3_col141" class="data row3 col141">0</td>
<td id="T_0aa46_row3_col142" class="data row3 col142">0</td>
<td id="T_0aa46_row3_col143" class="data row3 col143">0</td>
<td id="T_0aa46_row3_col144" class="data row3 col144">0</td>
<td id="T_0aa46_row3_col145" class="data row3 col145">0</td>
<td id="T_0aa46_row3_col146" class="data row3 col146">0</td>
<td id="T_0aa46_row3_col147" class="data row3 col147">0</td>
<td id="T_0aa46_row3_col148" class="data row3 col148">0</td>
<td id="T_0aa46_row3_col149" class="data row3 col149">0</td>
<td id="T_0aa46_row3_col150" class="data row3 col150">0</td>
<td id="T_0aa46_row3_col151" class="data row3 col151">0</td>
<td id="T_0aa46_row3_col152" class="data row3 col152">0</td>
<td id="T_0aa46_row3_col153" class="data row3 col153">0</td>
<td id="T_0aa46_row3_col154" class="data row3 col154">0</td>
<td id="T_0aa46_row3_col155" class="data row3 col155">0</td>
<td id="T_0aa46_row3_col156" class="data row3 col156">0</td>
<td id="T_0aa46_row3_col157" class="data row3 col157">0</td>
<td id="T_0aa46_row3_col158" class="data row3 col158">0</td>
<td id="T_0aa46_row3_col159" class="data row3 col159">0</td>
<td id="T_0aa46_row3_col160" class="data row3 col160">0</td>
<td id="T_0aa46_row3_col161" class="data row3 col161">0</td>
<td id="T_0aa46_row3_col162" class="data row3 col162">0</td>
<td id="T_0aa46_row3_col163" class="data row3 col163">0</td>
<td id="T_0aa46_row3_col164" class="data row3 col164">0</td>
<td id="T_0aa46_row3_col165" class="data row3 col165">0</td>
<td id="T_0aa46_row3_col166" class="data row3 col166">0</td>
<td id="T_0aa46_row3_col167" class="data row3 col167">0</td>
<td id="T_0aa46_row3_col168" class="data row3 col168">0</td>
<td id="T_0aa46_row3_col169" class="data row3 col169">0</td>
<td id="T_0aa46_row3_col170" class="data row3 col170">0</td>
<td id="T_0aa46_row3_col171" class="data row3 col171">0</td>
<td id="T_0aa46_row3_col172" class="data row3 col172">0</td>
<td id="T_0aa46_row3_col173" class="data row3 col173">0</td>
<td id="T_0aa46_row3_col174" class="data row3 col174">0</td>
<td id="T_0aa46_row3_col175" class="data row3 col175">0</td>
<td id="T_0aa46_row3_col176" class="data row3 col176">0</td>
<td id="T_0aa46_row3_col177" class="data row3 col177">0</td>
<td id="T_0aa46_row3_col178" class="data row3 col178">0</td>
<td id="T_0aa46_row3_col179" class="data row3 col179">0</td>
<td id="T_0aa46_row3_col180" class="data row3 col180">1</td>
<td id="T_0aa46_row3_col181" class="data row3 col181">0</td>
<td id="T_0aa46_row3_col182" class="data row3 col182">0</td>
<td id="T_0aa46_row3_col183" class="data row3 col183">0</td>
<td id="T_0aa46_row3_col184" class="data row3 col184">0</td>
<td id="T_0aa46_row3_col185" class="data row3 col185">0</td>
<td id="T_0aa46_row3_col186" class="data row3 col186">0</td>
<td id="T_0aa46_row3_col187" class="data row3 col187">0</td>
<td id="T_0aa46_row3_col188" class="data row3 col188">0</td>
<td id="T_0aa46_row3_col189" class="data row3 col189">0</td>
<td id="T_0aa46_row3_col190" class="data row3 col190">0</td>
<td id="T_0aa46_row3_col191" class="data row3 col191">0</td>
<td id="T_0aa46_row3_col192" class="data row3 col192">0</td>
<td id="T_0aa46_row3_col193" class="data row3 col193">0</td>
<td id="T_0aa46_row3_col194" class="data row3 col194">0</td>
<td id="T_0aa46_row3_col195" class="data row3 col195">0</td>
<td id="T_0aa46_row3_col196" class="data row3 col196">0</td>
<td id="T_0aa46_row3_col197" class="data row3 col197">0</td>
<td id="T_0aa46_row3_col198" class="data row3 col198">0</td>
<td id="T_0aa46_row3_col199" class="data row3 col199">0</td>
<td id="T_0aa46_row3_col200" class="data row3 col200">0</td>
<td id="T_0aa46_row3_col201" class="data row3 col201">0</td>
<td id="T_0aa46_row3_col202" class="data row3 col202">0</td>
<td id="T_0aa46_row3_col203" class="data row3 col203">0</td>
<td id="T_0aa46_row3_col204" class="data row3 col204">0</td>
<td id="T_0aa46_row3_col205" class="data row3 col205">0</td>
<td id="T_0aa46_row3_col206" class="data row3 col206">0</td>
<td id="T_0aa46_row3_col207" class="data row3 col207">0</td>
<td id="T_0aa46_row3_col208" class="data row3 col208">0</td>
<td id="T_0aa46_row3_col209" class="data row3 col209">0</td>
<td id="T_0aa46_row3_col210" class="data row3 col210">0</td>
<td id="T_0aa46_row3_col211" class="data row3 col211">0</td>
<td id="T_0aa46_row3_col212" class="data row3 col212">0</td>
<td id="T_0aa46_row3_col213" class="data row3 col213">0</td>
<td id="T_0aa46_row3_col214" class="data row3 col214">0</td>
<td id="T_0aa46_row3_col215" class="data row3 col215">0</td>
<td id="T_0aa46_row3_col216" class="data row3 col216">0</td>
<td id="T_0aa46_row3_col217" class="data row3 col217">0</td>
<td id="T_0aa46_row3_col218" class="data row3 col218">0</td>
<td id="T_0aa46_row3_col219" class="data row3 col219">0</td>
<td id="T_0aa46_row3_col220" class="data row3 col220">0</td>
<td id="T_0aa46_row3_col221" class="data row3 col221">0</td>
<td id="T_0aa46_row3_col222" class="data row3 col222">0</td>
<td id="T_0aa46_row3_col223" class="data row3 col223">0</td>
<td id="T_0aa46_row3_col224" class="data row3 col224">0</td>
<td id="T_0aa46_row3_col225" class="data row3 col225">0</td>
<td id="T_0aa46_row3_col226" class="data row3 col226">0</td>
<td id="T_0aa46_row3_col227" class="data row3 col227">0</td>
<td id="T_0aa46_row3_col228" class="data row3 col228">0</td>
<td id="T_0aa46_row3_col229" class="data row3 col229">0</td>
<td id="T_0aa46_row3_col230" class="data row3 col230">0</td>
<td id="T_0aa46_row3_col231" class="data row3 col231">0</td>
<td id="T_0aa46_row3_col232" class="data row3 col232">0</td>
<td id="T_0aa46_row3_col233" class="data row3 col233">0</td>
<td id="T_0aa46_row3_col234" class="data row3 col234">0</td>
<td id="T_0aa46_row3_col235" class="data row3 col235">0</td>
<td id="T_0aa46_row3_col236" class="data row3 col236">0</td>
<td id="T_0aa46_row3_col237" class="data row3 col237">0</td>
<td id="T_0aa46_row3_col238" class="data row3 col238">0</td>
<td id="T_0aa46_row3_col239" class="data row3 col239">0</td>
<td id="T_0aa46_row3_col240" class="data row3 col240">0</td>
<td id="T_0aa46_row3_col241" class="data row3 col241">0</td>
<td id="T_0aa46_row3_col242" class="data row3 col242">0</td>
<td id="T_0aa46_row3_col243" class="data row3 col243">0</td>
<td id="T_0aa46_row3_col244" class="data row3 col244">0</td>
<td id="T_0aa46_row3_col245" class="data row3 col245">1</td>
<td id="T_0aa46_row3_col246" class="data row3 col246">0</td>
<td id="T_0aa46_row3_col247" class="data row3 col247">0</td>
<td id="T_0aa46_row3_col248" class="data row3 col248">0</td>
<td id="T_0aa46_row3_col249" class="data row3 col249">0</td>
<td id="T_0aa46_row3_col250" class="data row3 col250">0</td>
<td id="T_0aa46_row3_col251" class="data row3 col251">0</td>
<td id="T_0aa46_row3_col252" class="data row3 col252">0</td>
<td id="T_0aa46_row3_col253" class="data row3 col253">0</td>
<td id="T_0aa46_row3_col254" class="data row3 col254">0</td>
<td id="T_0aa46_row3_col255" class="data row3 col255">0</td>
<td id="T_0aa46_row3_col256" class="data row3 col256">0</td>
<td id="T_0aa46_row3_col257" class="data row3 col257">0</td>
<td id="T_0aa46_row3_col258" class="data row3 col258">0</td>
<td id="T_0aa46_row3_col259" class="data row3 col259">0</td>
<td id="T_0aa46_row3_col260" class="data row3 col260">0</td>
<td id="T_0aa46_row3_col261" class="data row3 col261">0</td>
<td id="T_0aa46_row3_col262" class="data row3 col262">0</td>
<td id="T_0aa46_row3_col263" class="data row3 col263">0</td>
<td id="T_0aa46_row3_col264" class="data row3 col264">0</td>
<td id="T_0aa46_row3_col265" class="data row3 col265">0</td>
<td id="T_0aa46_row3_col266" class="data row3 col266">0</td>
<td id="T_0aa46_row3_col267" class="data row3 col267">0</td>
<td id="T_0aa46_row3_col268" class="data row3 col268">0</td>
<td id="T_0aa46_row3_col269" class="data row3 col269">0</td>
<td id="T_0aa46_row3_col270" class="data row3 col270">0</td>
<td id="T_0aa46_row3_col271" class="data row3 col271">0</td>
<td id="T_0aa46_row3_col272" class="data row3 col272">0</td>
<td id="T_0aa46_row3_col273" class="data row3 col273">0</td>
<td id="T_0aa46_row3_col274" class="data row3 col274">0</td>
<td id="T_0aa46_row3_col275" class="data row3 col275">0</td>
<td id="T_0aa46_row3_col276" class="data row3 col276">0</td>
<td id="T_0aa46_row3_col277" class="data row3 col277">0</td>
<td id="T_0aa46_row3_col278" class="data row3 col278">0</td>
<td id="T_0aa46_row3_col279" class="data row3 col279">0</td>
<td id="T_0aa46_row3_col280" class="data row3 col280">0</td>
<td id="T_0aa46_row3_col281" class="data row3 col281">0</td>
<td id="T_0aa46_row3_col282" class="data row3 col282">0</td>
<td id="T_0aa46_row3_col283" class="data row3 col283">0</td>
<td id="T_0aa46_row3_col284" class="data row3 col284">0</td>
<td id="T_0aa46_row3_col285" class="data row3 col285">0</td>
<td id="T_0aa46_row3_col286" class="data row3 col286">0</td>
<td id="T_0aa46_row3_col287" class="data row3 col287">0</td>
<td id="T_0aa46_row3_col288" class="data row3 col288">0</td>
<td id="T_0aa46_row3_col289" class="data row3 col289">0</td>
<td id="T_0aa46_row3_col290" class="data row3 col290">0</td>
<td id="T_0aa46_row3_col291" class="data row3 col291">0</td>
<td id="T_0aa46_row3_col292" class="data row3 col292">0</td>
<td id="T_0aa46_row3_col293" class="data row3 col293">0</td>
<td id="T_0aa46_row3_col294" class="data row3 col294">0</td>
<td id="T_0aa46_row3_col295" class="data row3 col295">0</td>
<td id="T_0aa46_row3_col296" class="data row3 col296">0</td>
<td id="T_0aa46_row3_col297" class="data row3 col297">0</td>
<td id="T_0aa46_row3_col298" class="data row3 col298">0</td>
<td id="T_0aa46_row3_col299" class="data row3 col299">0</td>
<td id="T_0aa46_row3_col300" class="data row3 col300">0</td>
<td id="T_0aa46_row3_col301" class="data row3 col301">0</td>
<td id="T_0aa46_row3_col302" class="data row3 col302">0</td>
<td id="T_0aa46_row3_col303" class="data row3 col303">0</td>
<td id="T_0aa46_row3_col304" class="data row3 col304">0</td>
<td id="T_0aa46_row3_col305" class="data row3 col305">0</td>
<td id="T_0aa46_row3_col306" class="data row3 col306">0</td>
<td id="T_0aa46_row3_col307" class="data row3 col307">0</td>
<td id="T_0aa46_row3_col308" class="data row3 col308">0</td>
<td id="T_0aa46_row3_col309" class="data row3 col309">0</td>
<td id="T_0aa46_row3_col310" class="data row3 col310">0</td>
<td id="T_0aa46_row3_col311" class="data row3 col311">0</td>
<td id="T_0aa46_row3_col312" class="data row3 col312">0</td>
<td id="T_0aa46_row3_col313" class="data row3 col313">0</td>
<td id="T_0aa46_row3_col314" class="data row3 col314">0</td>
<td id="T_0aa46_row3_col315" class="data row3 col315">0</td>
<td id="T_0aa46_row3_col316" class="data row3 col316">0</td>
<td id="T_0aa46_row3_col317" class="data row3 col317">0</td>
<td id="T_0aa46_row3_col318" class="data row3 col318">0</td>
<td id="T_0aa46_row3_col319" class="data row3 col319">0</td>
<td id="T_0aa46_row3_col320" class="data row3 col320">0</td>
<td id="T_0aa46_row3_col321" class="data row3 col321">0</td>
<td id="T_0aa46_row3_col322" class="data row3 col322">0</td>
<td id="T_0aa46_row3_col323" class="data row3 col323">0</td>
<td id="T_0aa46_row3_col324" class="data row3 col324">0</td>
<td id="T_0aa46_row3_col325" class="data row3 col325">0</td>
<td id="T_0aa46_row3_col326" class="data row3 col326">0</td>
<td id="T_0aa46_row3_col327" class="data row3 col327">0</td>
<td id="T_0aa46_row3_col328" class="data row3 col328">0</td>
<td id="T_0aa46_row3_col329" class="data row3 col329">0</td>
<td id="T_0aa46_row3_col330" class="data row3 col330">0</td>
<td id="T_0aa46_row3_col331" class="data row3 col331">0</td>
<td id="T_0aa46_row3_col332" class="data row3 col332">0</td>
<td id="T_0aa46_row3_col333" class="data row3 col333">0</td>
<td id="T_0aa46_row3_col334" class="data row3 col334">0</td>
<td id="T_0aa46_row3_col335" class="data row3 col335">0</td>
<td id="T_0aa46_row3_col336" class="data row3 col336">0</td>
<td id="T_0aa46_row3_col337" class="data row3 col337">0</td>
<td id="T_0aa46_row3_col338" class="data row3 col338">0</td>
<td id="T_0aa46_row3_col339" class="data row3 col339">0</td>
<td id="T_0aa46_row3_col340" class="data row3 col340">0</td>
<td id="T_0aa46_row3_col341" class="data row3 col341">0</td>
<td id="T_0aa46_row3_col342" class="data row3 col342">0</td>
<td id="T_0aa46_row3_col343" class="data row3 col343">0</td>
<td id="T_0aa46_row3_col344" class="data row3 col344">0</td>
<td id="T_0aa46_row3_col345" class="data row3 col345">0</td>
<td id="T_0aa46_row3_col346" class="data row3 col346">0</td>
<td id="T_0aa46_row3_col347" class="data row3 col347">0</td>
<td id="T_0aa46_row3_col348" class="data row3 col348">0</td>
<td id="T_0aa46_row3_col349" class="data row3 col349">0</td>
<td id="T_0aa46_row3_col350" class="data row3 col350">0</td>
<td id="T_0aa46_row3_col351" class="data row3 col351">0</td>
<td id="T_0aa46_row3_col352" class="data row3 col352">0</td>
<td id="T_0aa46_row3_col353" class="data row3 col353">0</td>
<td id="T_0aa46_row3_col354" class="data row3 col354">0</td>
<td id="T_0aa46_row3_col355" class="data row3 col355">0</td>
<td id="T_0aa46_row3_col356" class="data row3 col356">0</td>
<td id="T_0aa46_row3_col357" class="data row3 col357">0</td>
<td id="T_0aa46_row3_col358" class="data row3 col358">0</td>
<td id="T_0aa46_row3_col359" class="data row3 col359">0</td>
<td id="T_0aa46_row3_col360" class="data row3 col360">0</td>
<td id="T_0aa46_row3_col361" class="data row3 col361">0</td>
<td id="T_0aa46_row3_col362" class="data row3 col362">0</td>
<td id="T_0aa46_row3_col363" class="data row3 col363">0</td>
<td id="T_0aa46_row3_col364" class="data row3 col364">0</td>
<td id="T_0aa46_row3_col365" class="data row3 col365">0</td>
<td id="T_0aa46_row3_col366" class="data row3 col366">0</td>
<td id="T_0aa46_row3_col367" class="data row3 col367">0</td>
<td id="T_0aa46_row3_col368" class="data row3 col368">0</td>
<td id="T_0aa46_row3_col369" class="data row3 col369">0</td>
<td id="T_0aa46_row3_col370" class="data row3 col370">0</td>
<td id="T_0aa46_row3_col371" class="data row3 col371">0</td>
<td id="T_0aa46_row3_col372" class="data row3 col372">0</td>
<td id="T_0aa46_row3_col373" class="data row3 col373">0</td>
<td id="T_0aa46_row3_col374" class="data row3 col374">0</td>
<td id="T_0aa46_row3_col375" class="data row3 col375">0</td>
<td id="T_0aa46_row3_col376" class="data row3 col376">0</td>
<td id="T_0aa46_row3_col377" class="data row3 col377">0</td>
<td id="T_0aa46_row3_col378" class="data row3 col378">0</td>
<td id="T_0aa46_row3_col379" class="data row3 col379">0</td>
<td id="T_0aa46_row3_col380" class="data row3 col380">0</td>
<td id="T_0aa46_row3_col381" class="data row3 col381">0</td>
<td id="T_0aa46_row3_col382" class="data row3 col382">0</td>
<td id="T_0aa46_row3_col383" class="data row3 col383">0</td>
<td id="T_0aa46_row3_col384" class="data row3 col384">0</td>
<td id="T_0aa46_row3_col385" class="data row3 col385">0</td>
<td id="T_0aa46_row3_col386" class="data row3 col386">0</td>
<td id="T_0aa46_row3_col387" class="data row3 col387">0</td>
<td id="T_0aa46_row3_col388" class="data row3 col388">0</td>
<td id="T_0aa46_row3_col389" class="data row3 col389">0</td>
<td id="T_0aa46_row3_col390" class="data row3 col390">0</td>
<td id="T_0aa46_row3_col391" class="data row3 col391">0</td>
<td id="T_0aa46_row3_col392" class="data row3 col392">0</td>
<td id="T_0aa46_row3_col393" class="data row3 col393">0</td>
<td id="T_0aa46_row3_col394" class="data row3 col394">0</td>
<td id="T_0aa46_row3_col395" class="data row3 col395">0</td>
<td id="T_0aa46_row3_col396" class="data row3 col396">0</td>
<td id="T_0aa46_row3_col397" class="data row3 col397">0</td>
<td id="T_0aa46_row3_col398" class="data row3 col398">0</td>
<td id="T_0aa46_row3_col399" class="data row3 col399">0</td>
<td id="T_0aa46_row3_col400" class="data row3 col400">0</td>
<td id="T_0aa46_row3_col401" class="data row3 col401">0</td>
<td id="T_0aa46_row3_col402" class="data row3 col402">0</td>
<td id="T_0aa46_row3_col403" class="data row3 col403">0</td>
<td id="T_0aa46_row3_col404" class="data row3 col404">0</td>
<td id="T_0aa46_row3_col405" class="data row3 col405">0</td>
<td id="T_0aa46_row3_col406" class="data row3 col406">0</td>
<td id="T_0aa46_row3_col407" class="data row3 col407">0</td>
<td id="T_0aa46_row3_col408" class="data row3 col408">0</td>
<td id="T_0aa46_row3_col409" class="data row3 col409">0</td>
<td id="T_0aa46_row3_col410" class="data row3 col410">0</td>
<td id="T_0aa46_row3_col411" class="data row3 col411">0</td>
<td id="T_0aa46_row3_col412" class="data row3 col412">0</td>
<td id="T_0aa46_row3_col413" class="data row3 col413">0</td>
<td id="T_0aa46_row3_col414" class="data row3 col414">0</td>
<td id="T_0aa46_row3_col415" class="data row3 col415">0</td>
<td id="T_0aa46_row3_col416" class="data row3 col416">0</td>
<td id="T_0aa46_row3_col417" class="data row3 col417">0</td>
<td id="T_0aa46_row3_col418" class="data row3 col418">0</td>
<td id="T_0aa46_row3_col419" class="data row3 col419">0</td>
<td id="T_0aa46_row3_col420" class="data row3 col420">0</td>
<td id="T_0aa46_row3_col421" class="data row3 col421">0</td>
<td id="T_0aa46_row3_col422" class="data row3 col422">0</td>
<td id="T_0aa46_row3_col423" class="data row3 col423">0</td>
<td id="T_0aa46_row3_col424" class="data row3 col424">0</td>
<td id="T_0aa46_row3_col425" class="data row3 col425">0</td>
<td id="T_0aa46_row3_col426" class="data row3 col426">0</td>
<td id="T_0aa46_row3_col427" class="data row3 col427">0</td>
<td id="T_0aa46_row3_col428" class="data row3 col428">0</td>
<td id="T_0aa46_row3_col429" class="data row3 col429">0</td>
<td id="T_0aa46_row3_col430" class="data row3 col430">0</td>
<td id="T_0aa46_row3_col431" class="data row3 col431">0</td>
<td id="T_0aa46_row3_col432" class="data row3 col432">0</td>
<td id="T_0aa46_row3_col433" class="data row3 col433">1</td>
<td id="T_0aa46_row3_col434" class="data row3 col434">1</td>
<td id="T_0aa46_row3_col435" class="data row3 col435">1</td>
<td id="T_0aa46_row3_col436" class="data row3 col436">0</td>
<td id="T_0aa46_row3_col437" class="data row3 col437">0</td>
<td id="T_0aa46_row3_col438" class="data row3 col438">0</td>
<td id="T_0aa46_row3_col439" class="data row3 col439">0</td>
<td id="T_0aa46_row3_col440" class="data row3 col440">0</td>
<td id="T_0aa46_row3_col441" class="data row3 col441">0</td>
<td id="T_0aa46_row3_col442" class="data row3 col442">0</td>
<td id="T_0aa46_row3_col443" class="data row3 col443">0</td>
<td id="T_0aa46_row3_col444" class="data row3 col444">0</td>
<td id="T_0aa46_row3_col445" class="data row3 col445">0</td>
<td id="T_0aa46_row3_col446" class="data row3 col446">0</td>
<td id="T_0aa46_row3_col447" class="data row3 col447">0</td>
<td id="T_0aa46_row3_col448" class="data row3 col448">0</td>
<td id="T_0aa46_row3_col449" class="data row3 col449">0</td>
<td id="T_0aa46_row3_col450" class="data row3 col450">0</td>
<td id="T_0aa46_row3_col451" class="data row3 col451">1</td>
<td id="T_0aa46_row3_col452" class="data row3 col452">0</td>
<td id="T_0aa46_row3_col453" class="data row3 col453">0</td>
<td id="T_0aa46_row3_col454" class="data row3 col454">0</td>
<td id="T_0aa46_row3_col455" class="data row3 col455">0</td>
<td id="T_0aa46_row3_col456" class="data row3 col456">0</td>
<td id="T_0aa46_row3_col457" class="data row3 col457">0</td>
<td id="T_0aa46_row3_col458" class="data row3 col458">0</td>
<td id="T_0aa46_row3_col459" class="data row3 col459">0</td>
<td id="T_0aa46_row3_col460" class="data row3 col460">0</td>
<td id="T_0aa46_row3_col461" class="data row3 col461">0</td>
<td id="T_0aa46_row3_col462" class="data row3 col462">0</td>
<td id="T_0aa46_row3_col463" class="data row3 col463">0</td>
<td id="T_0aa46_row3_col464" class="data row3 col464">0</td>
<td id="T_0aa46_row3_col465" class="data row3 col465">0</td>
<td id="T_0aa46_row3_col466" class="data row3 col466">0</td>
<td id="T_0aa46_row3_col467" class="data row3 col467">0</td>
<td id="T_0aa46_row3_col468" class="data row3 col468">0</td>
<td id="T_0aa46_row3_col469" class="data row3 col469">0</td>
<td id="T_0aa46_row3_col470" class="data row3 col470">0</td>
<td id="T_0aa46_row3_col471" class="data row3 col471">0</td>
<td id="T_0aa46_row3_col472" class="data row3 col472">0</td>
<td id="T_0aa46_row3_col473" class="data row3 col473">0</td>
<td id="T_0aa46_row3_col474" class="data row3 col474">0</td>
<td id="T_0aa46_row3_col475" class="data row3 col475">0</td>
<td id="T_0aa46_row3_col476" class="data row3 col476">0</td>
<td id="T_0aa46_row3_col477" class="data row3 col477">0</td>
<td id="T_0aa46_row3_col478" class="data row3 col478">0</td>
<td id="T_0aa46_row3_col479" class="data row3 col479">0</td>
<td id="T_0aa46_row3_col480" class="data row3 col480">0</td>
<td id="T_0aa46_row3_col481" class="data row3 col481">0</td>
<td id="T_0aa46_row3_col482" class="data row3 col482">0</td>
<td id="T_0aa46_row3_col483" class="data row3 col483">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
<section id="enforce_class" class="level3">
<h3 class="anchored" data-anchor-id="enforce_class">enforce_class</h3>
<blockquote class="blockquote">
<pre><code> enforce_class (glycan:str, glycan_class:str, conf:float|None=None,
                extra_thresh:float=0.3)</code></pre>
</blockquote>
<p><em>Determines whether glycan belongs to a specified class</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed nomenclature</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td></td>
<td>Glycan class (O, N, free, or lipid)</td>
</tr>
<tr class="odd">
<td>conf</td>
<td>float | None</td>
<td>None</td>
<td>Prediction confidence to override class</td>
</tr>
<tr class="even">
<td>extra_thresh</td>
<td>float</td>
<td>0.3</td>
<td>Threshold to override class</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>bool</strong></td>
<td></td>
<td><strong>True if glycan is in glycan class</strong></td>
</tr>
</tbody>
</table>
<div id="respected-horse" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>enforce_class(<span class="st">"Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>, <span class="st">"O"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>False</code></pre>
</div>
</div>
<hr>
</section>
<section id="iupac_to_smiles" class="level3">
<h3 class="anchored" data-anchor-id="iupac_to_smiles">IUPAC_to_SMILES</h3>
<blockquote class="blockquote">
<pre><code> IUPAC_to_SMILES (glycan_list:str|list[str])</code></pre>
</blockquote>
<p><em>Convert list of IUPAC-condensed glycans to isomeric SMILES using GlyLES</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan_list</td>
<td>str | list[str]</td>
<td>List of IUPAC-condensed glycans or single glycan</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td><strong>List of corresponding SMILES strings</strong></td>
</tr>
</tbody>
</table>
<div id="national-updating" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb118"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>IUPAC_to_SMILES([<span class="st">'Neu5Ac(a2-3)Gal(b1-4)Glc'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['O1C(O)[C@H](O)[C@@H](O)[C@H](O[C@@H]2O[C@H](CO)[C@H](O)[C@H](O[C@]3(C(=O)O)C[C@H](O)[C@@H](NC(C)=O)[C@H]([C@H](O)[C@H](O)CO)O3)[C@H]2O)[C@H]1CO']</code></pre>
</div>
</div>
<hr>
</section>
<section id="canonicalize_composition" class="level3">
<h3 class="anchored" data-anchor-id="canonicalize_composition">canonicalize_composition</h3>
<blockquote class="blockquote">
<pre><code> canonicalize_composition (comp:str)</code></pre>
</blockquote>
<p><em>Converts composition from any common format to standardized dictionary</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>comp</td>
<td>str</td>
<td>Composition in Hex5HexNAc4Fuc1Neu5Ac2 or H5N4F1A2 format</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>Dictionary of monosaccharide:count</strong></td>
</tr>
</tbody>
</table>
<div id="60e989f5-bf78-4b8e-b58a-5a5ed45b1631" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_composition(<span class="st">"HexNAc2Hex1Fuc3Neu5Ac1"</span>))</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_composition(<span class="st">"N2H1F3A1"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'HexNAc': 2, 'Hex': 1, 'dHex': 3, 'Neu5Ac': 1}
{'HexNAc': 2, 'Hex': 1, 'dHex': 3, 'Neu5Ac': 1}</code></pre>
</div>
</div>
<hr>
</section>
<section id="canonicalize_iupac" class="level3">
<h3 class="anchored" data-anchor-id="canonicalize_iupac">canonicalize_iupac</h3>
<blockquote class="blockquote">
<pre><code> canonicalize_iupac (glycan:str)</code></pre>
</blockquote>
<p><em>Convert glycan from IUPAC-extended, LinearCode, GlycoCT, WURCS, Oxford, GLYCAM, GlycoWorkBench, CSDB-linear, KCF, GlyConnect IDs, and GlyTouCanIDs to standardized IUPAC-condensed format</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td>Glycan sequence in any supported format</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td><strong>Standardized IUPAC-condensed format</strong></td>
</tr>
</tbody>
</table>
<div id="outstanding-treasury" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_iupac(<span class="st">"NeuAc?1-36SGalb1-4GlcNACb1-6(Fuc?1-2Galb1-4GlcNacb1-3Galb1-3)GalNAc-sp3"</span>))</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_iupac(<span class="st">"WURCS=2.0/5,11,10/[a2122h-1b_1-5_2*NCC/3=O][a1122h-1b_1-5][a1122h-1a_1-5][a2112h-1b_1-5][a1221m-1a_1-5]/1-1-2-3-1-4-3-1-4-5-5/a4-b1_a6-k1_b4-c1_c3-d1_c6-g1_d2-e1_e4-f1_g2-h1_h4-i1_i2-j1"</span>))</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_iupac(<span class="st">"Ma3(Ma6)Mb4GNb4GN;N"</span>))</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_iupac(<span class="st">"α-D-Manp-(1→3)[α-D-Manp-(1→6)]-β-D-Manp-(1→4)-β-D-GlcpNAc-(1→4)-β-D-GlcpNAc-(1→"</span>))</span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(canonicalize_iupac(<span class="st">"""RES</span></span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a><span class="st">1b:b-dgal-HEX-1:5</span></span>
<span id="cb124-7"><a href="#cb124-7" aria-hidden="true" tabindex="-1"></a><span class="st">2s:n-acetyl</span></span>
<span id="cb124-8"><a href="#cb124-8" aria-hidden="true" tabindex="-1"></a><span class="st">3b:b-dgal-HEX-1:5</span></span>
<span id="cb124-9"><a href="#cb124-9" aria-hidden="true" tabindex="-1"></a><span class="st">4b:b-dglc-HEX-1:5</span></span>
<span id="cb124-10"><a href="#cb124-10" aria-hidden="true" tabindex="-1"></a><span class="st">5b:b-dgal-HEX-1:5</span></span>
<span id="cb124-11"><a href="#cb124-11" aria-hidden="true" tabindex="-1"></a><span class="st">6b:a-dglc-HEX-1:5</span></span>
<span id="cb124-12"><a href="#cb124-12" aria-hidden="true" tabindex="-1"></a><span class="st">7b:b-dgal-HEX-1:5</span></span>
<span id="cb124-13"><a href="#cb124-13" aria-hidden="true" tabindex="-1"></a><span class="st">8b:a-lgal-HEX-1:5|6:d</span></span>
<span id="cb124-14"><a href="#cb124-14" aria-hidden="true" tabindex="-1"></a><span class="st">9b:a-dgal-HEX-1:5</span></span>
<span id="cb124-15"><a href="#cb124-15" aria-hidden="true" tabindex="-1"></a><span class="st">10s:n-acetyl</span></span>
<span id="cb124-16"><a href="#cb124-16" aria-hidden="true" tabindex="-1"></a><span class="st">11s:n-acetyl</span></span>
<span id="cb124-17"><a href="#cb124-17" aria-hidden="true" tabindex="-1"></a><span class="st">12b:b-dglc-HEX-1:5</span></span>
<span id="cb124-18"><a href="#cb124-18" aria-hidden="true" tabindex="-1"></a><span class="st">13b:b-dgal-HEX-1:5</span></span>
<span id="cb124-19"><a href="#cb124-19" aria-hidden="true" tabindex="-1"></a><span class="st">14b:a-lgal-HEX-1:5|6:d</span></span>
<span id="cb124-20"><a href="#cb124-20" aria-hidden="true" tabindex="-1"></a><span class="st">15b:a-lgal-HEX-1:5|6:d</span></span>
<span id="cb124-21"><a href="#cb124-21" aria-hidden="true" tabindex="-1"></a><span class="st">16s:n-acetyl</span></span>
<span id="cb124-22"><a href="#cb124-22" aria-hidden="true" tabindex="-1"></a><span class="st">17s:n-acetyl</span></span>
<span id="cb124-23"><a href="#cb124-23" aria-hidden="true" tabindex="-1"></a><span class="st">18b:b-dgal-HEX-1:5</span></span>
<span id="cb124-24"><a href="#cb124-24" aria-hidden="true" tabindex="-1"></a><span class="st">LIN</span></span>
<span id="cb124-25"><a href="#cb124-25" aria-hidden="true" tabindex="-1"></a><span class="st">1:1d(2+1)2n</span></span>
<span id="cb124-26"><a href="#cb124-26" aria-hidden="true" tabindex="-1"></a><span class="st">2:1o(3+1)3d</span></span>
<span id="cb124-27"><a href="#cb124-27" aria-hidden="true" tabindex="-1"></a><span class="st">3:3o(3+1)4d</span></span>
<span id="cb124-28"><a href="#cb124-28" aria-hidden="true" tabindex="-1"></a><span class="st">4:4o(-1+1)5d</span></span>
<span id="cb124-29"><a href="#cb124-29" aria-hidden="true" tabindex="-1"></a><span class="st">5:5o(-1+1)6d</span></span>
<span id="cb124-30"><a href="#cb124-30" aria-hidden="true" tabindex="-1"></a><span class="st">6:6o(-1+1)7d</span></span>
<span id="cb124-31"><a href="#cb124-31" aria-hidden="true" tabindex="-1"></a><span class="st">7:7o(2+1)8d</span></span>
<span id="cb124-32"><a href="#cb124-32" aria-hidden="true" tabindex="-1"></a><span class="st">8:7o(3+1)9d</span></span>
<span id="cb124-33"><a href="#cb124-33" aria-hidden="true" tabindex="-1"></a><span class="st">9:9d(2+1)10n</span></span>
<span id="cb124-34"><a href="#cb124-34" aria-hidden="true" tabindex="-1"></a><span class="st">10:6d(2+1)11n</span></span>
<span id="cb124-35"><a href="#cb124-35" aria-hidden="true" tabindex="-1"></a><span class="st">11:5o(-1+1)12d</span></span>
<span id="cb124-36"><a href="#cb124-36" aria-hidden="true" tabindex="-1"></a><span class="st">12:12o(-1+1)13d</span></span>
<span id="cb124-37"><a href="#cb124-37" aria-hidden="true" tabindex="-1"></a><span class="st">13:13o(2+1)14d</span></span>
<span id="cb124-38"><a href="#cb124-38" aria-hidden="true" tabindex="-1"></a><span class="st">14:12o(-1+1)15d</span></span>
<span id="cb124-39"><a href="#cb124-39" aria-hidden="true" tabindex="-1"></a><span class="st">15:12d(2+1)16n</span></span>
<span id="cb124-40"><a href="#cb124-40" aria-hidden="true" tabindex="-1"></a><span class="st">16:4d(2+1)17n</span></span>
<span id="cb124-41"><a href="#cb124-41" aria-hidden="true" tabindex="-1"></a><span class="st">17:1o(6+1)18d</span></span>
<span id="cb124-42"><a href="#cb124-42" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)Gal(b1-3)[Neu5Ac(a2-3)Gal6S(b1-4)GlcNAc(b1-6)]GalNAc
Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc
Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc
Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc
Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-?)GlcNAc(a1-?)[Fuc(a1-2)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-?)]Gal(b1-?)GlcNAc(b1-3)Gal(b1-3)[Gal(b1-6)]GalNAc</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_possible_linkages" class="level3">
<h3 class="anchored" data-anchor-id="get_possible_linkages">get_possible_linkages</h3>
<blockquote class="blockquote">
<pre><code> get_possible_linkages (wildcard:str, linkage_list:list[str]={'a2-2',
                        'a1-3', '?1-2', '?1-?', '?2-8', 'a1-5', 'a2-6',
                        'a2-5', '?1-4', 'a1-1', 'a1-6', 'a2-11', 'a1-?',
                        'b2-6', '1-6', 'b2-4', '?2-6', 'b2-8', 'b1-?',
                        'b1-5', 'b1-3', 'a1-11', 'a2-7', 'b1-4', 'b1-1',
                        'b2-2', 'a2-3', '?1-6', 'b2-7', 'a1-8', '?1-3',
                        'a1-9', 'b1-9', '?2-3', 'b1-8', '?2-?', 'a1-4',
                        'a2-8', 'b1-2', 'b1-7', 'b1-6', 'b2-5', '1-4',
                        'a2-1', 'a2-9', 'a1-7', 'a2-?', 'a2-4', 'a1-2',
                        'b2-1', 'b2-3'})</code></pre>
</blockquote>
<p><em>Retrieves all linkages that match a given wildcard pattern</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>wildcard</td>
<td>str</td>
<td></td>
<td>Pattern to match, ? can be wildcard</td>
</tr>
<tr class="even">
<td>linkage_list</td>
<td>list</td>
<td>{‘a2-2’, ‘a1-3’, ‘?1-2’, ‘?1-?’, ‘?2-8’, ‘a1-5’, ‘a2-6’, ‘a2-5’, ‘?1-4’, ‘a1-1’, ‘a1-6’, ‘a2-11’, ‘a1-?’, ‘b2-6’, ‘1-6’, ‘b2-4’, ‘?2-6’, ‘b2-8’, ‘b1-?’, ‘b1-5’, ‘b1-3’, ‘a1-11’, ‘a2-7’, ‘b1-4’, ‘b1-1’, ‘b2-2’, ‘a2-3’, ‘?1-6’, ‘b2-7’, ‘a1-8’, ‘?1-3’, ‘a1-9’, ‘b1-9’, ‘?2-3’, ‘b1-8’, ‘?2-?’, ‘a1-4’, ‘a2-8’, ‘b1-2’, ‘b1-7’, ‘b1-6’, ‘b2-5’, ‘1-4’, ‘a2-1’, ‘a2-9’, ‘a1-7’, ‘a2-?’, ‘a2-4’, ‘a1-2’, ‘b2-1’, ‘b2-3’}</td>
<td>List of linkages to search</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>set</strong></td>
<td></td>
<td><strong>Matching linkages</strong></td>
</tr>
</tbody>
</table>
<div id="8ee0431e-8d4f-40f6-adfb-7dd962a33c69" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>get_possible_linkages(<span class="st">"a1-?"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>{'a1-1',
 'a1-2',
 'a1-3',
 'a1-4',
 'a1-5',
 'a1-6',
 'a1-7',
 'a1-8',
 'a1-9',
 'a1-?'}</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_possible_monosaccharides" class="level3">
<h3 class="anchored" data-anchor-id="get_possible_monosaccharides">get_possible_monosaccharides</h3>
<blockquote class="blockquote">
<pre><code> get_possible_monosaccharides (wildcard:str)</code></pre>
</blockquote>
<p><em>Retrieves all matching common monosaccharides of a type</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>wildcard</td>
<td>str</td>
<td>Monosaccharide type; options: Hex, HexNAc, dHex, Sia, HexA, Pen, HexOS, HexNAcOS</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>set</strong></td>
<td><strong>Matching monosaccharides</strong></td>
</tr>
</tbody>
</table>
<div id="7dfafbf3-4602-40a2-aca0-f043aa60186b" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb130"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>get_possible_monosaccharides(<span class="st">"HexNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>{'GalNAc', 'GlcNAc', 'HexNAc', 'ManNAc'}</code></pre>
</div>
</div>
<hr>
</section>
<section id="equal_repeats" class="level3">
<h3 class="anchored" data-anchor-id="equal_repeats">equal_repeats</h3>
<blockquote class="blockquote">
<pre><code> equal_repeats (r1:str, r2:str)</code></pre>
</blockquote>
<p><em>Check whether two repeat units could stem from the same repeating structure</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>r1</td>
<td>str</td>
<td>First glycan sequence</td>
</tr>
<tr class="even">
<td>r2</td>
<td>str</td>
<td>Second glycan sequence</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>bool</strong></td>
<td><strong>True if repeats are shifted versions</strong></td>
</tr>
</tbody>
</table>
<div id="bc5ca7d2-cc19-4456-85d7-02794c7b652e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>equal_repeats(<span class="st">"Fuc2S3S(a1-3)Fuc2S(a1-4)Fuc2S3S"</span>, <span class="st">"Fuc2S(a1-4)Fuc2S3S(a1-3)Fuc2S"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>True</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_class" class="level3">
<h3 class="anchored" data-anchor-id="get_class">get_class</h3>
<blockquote class="blockquote">
<pre><code> get_class (glycan:str)</code></pre>
</blockquote>
<p><em>Determines glycan class</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td>Glycan in IUPAC-condensed nomenclature</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td><strong>Glycan class (repeat, O, N, free, lipid, lipid/free, or empty)</strong></td>
</tr>
</tbody>
</table>
<div id="b1384260-2bda-41b0-82a8-2be45dd104f5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>get_class(<span class="st">"Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'N'</code></pre>
</div>
</div>
</section>
</section>
<section id="query" class="level2">
<h2 class="anchored" data-anchor-id="query">query</h2>
<blockquote class="blockquote">
<p>for interacting with the databases contained in glycowork, delivering insights for sequences of interest</p>
</blockquote>
<hr>
<section id="get_insight" class="level3">
<h3 class="anchored" data-anchor-id="get_insight">get_insight</h3>
<blockquote class="blockquote">
<pre><code> get_insight (glycan:str, motifs:pandas.core.frame.DataFrame|None=None)</code></pre>
</blockquote>
<p><em>Print meta-information about a glycan</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td>motifs</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>DataFrame of glycan motifs; default:motif_list</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td><strong>Prints glycan meta-information</strong></td>
</tr>
</tbody>
</table>
<div id="still-offset" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Test get_insight with 'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'"</span>)</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>get_insight(<span class="st">'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Test get_insight with 'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'
Let's get rolling! Give us a few moments to crunch some numbers.

This glycan occurs in the following species: ['Acanthocheilonema_viteae', 'Adeno-associated_dependoparvovirusA', 'Aedes_aegypti', 'Angiostrongylus_cantonensis', 'Anopheles_gambiae', 'Antheraea_pernyi', 'Apis_mellifera', 'Ascaris_suum', 'Autographa_californica_nucleopolyhedrovirus', 'AvianInfluenzaA_Virus', 'Bombus_ignitus', 'Bombyx_mori', 'Bos_taurus', 'Brugia_malayi', 'Caenorhabditis_elegans', 'Cardicola_forsteri', 'Cooperia_onchophora', 'Cornu_aspersum', 'Crassostrea_gigas', 'Crassostrea_virginica', 'Cricetulus_griseus', 'Danio_rerio', 'Dictyocaulus_viviparus', 'Dirofilaria_immitis', 'Drosophila_melanogaster', 'Fasciola_hepatica', 'Gallus_gallus', 'Glossina_morsitans', 'Haemonchus_contortus', 'Haliotis_tuberculata', 'Heligmosomoides_polygyrus', 'Helix_lucorum', 'Homo_sapiens', 'HumanImmunoDeficiency_Virus', 'Hylesia_metabus', 'Hypsibius_exemplaris', 'Lutzomyia_longipalpis', 'Lymantria_dispar', 'Macaca_mulatta', 'Mamestra_brassicae', 'Megathura_crenulata', 'Mus_musculus', 'Nilaparvata_lugens', 'Oesophagostomum_dentatum', 'Onchocerca_volvulus', 'Onchocerca_volvulus', 'Ophiactis_savignyi', 'Opisthorchis_viverrini', 'Ostrea_edulis', 'Ovis_aries', 'Pan_troglodytes', 'Pristionchus_pacificus', 'Ramazzottius_varieornatus', 'Rattus_norvegicus', 'Schistosoma_mansoni', 'SemlikiForest_Virus', 'Spodoptera_frugiperda', 'Sus_scrofa', 'Tick_borne_encephalitis_virus', 'Tribolium_castaneum', 'Trichinella_spiralis', 'Trichoplusia_ni', 'Trichuris_suis', 'Tropidolaemus_subannulatus', 'Volvarina_rubella', 'undetermined', 'unidentified_influenza_virus']

Puh, that's quite a lot! Here are the phyla of those species: ['Arthropoda', 'Artverviricota', 'Chordata', 'Cossaviricota', 'Echinodermata', 'Kitrinoviricota', 'Mollusca', 'Negarnaviricota', 'Nematoda', 'Platyhelminthes', 'Tardigrada', 'Virus']

This glycan contains the following motifs: ['Chitobiose', 'Trimannosylcore', 'core_fucose']

This is the GlyTouCan ID for this glycan: G63041RA

This glycan has been reported to be expressed in: ['2A3_cell_line', 'A549_cell_line', 'AML_193_cell_line', 'C10_cell_line', 'CHOK1_cell_line', 'CHOS_cell_line', 'COLO_205_cell_line', 'COLO_320_cell_line', 'CRL_1620_cell_line', 'Caco_2_cell_line', 'Cal-27_cell_line', 'Cervicovaginal_Secretion', 'Co_115_cell_line', 'EOL_1_cell_line', 'FaDu_cell_line', 'HCT_15_cell_line', 'HCT_8_cell_line', 'HEK293_cell_line', 'HEL92_1_7_cell_line', 'HEL_cell_line', 'HL_60_cell_line', 'HT_29_cell_line', 'KG_1_cell_line', 'KG_1a_cell_line', 'KM12_cell_line', 'Kasumi_1_cell_line', 'LS174T_cell_line', 'LS180_cell_line', 'LS411N_cell_line', 'LoVo_cell_line', 'MDA_MB_231BR_cell_line', 'ME_1_cell_line', 'ML_1_cell_line', 'MOLM_13_cell_line', 'MOLM_14_cell_line', 'MV4_11_cell_line', 'M_07e_cell_line', 'NB_4_cell_line', 'NS0_cell_line', 'OCI_AML2_cell_line', 'OCI_AML3_cell_line', 'PLB_985_cell_line', 'RKO_cell_line', 'SCC-9_cell_line', 'SCC_25_cell_line', 'SW1116_cell_line', 'SW1398_cell_line', 'SW1463_cell_line', 'SW480_cell_line', 'SW48_cell_line', 'SW620_cell_line', 'SW948_cell_line', 'T84_cell_line', 'TF_1_cell_line', 'THP_1_cell_line', 'U_937_cell_line', 'VU-147T_cell_line', 'WiDr_cell_line', 'alveolus_of_lung', 'brain', 'brain', 'cerebellar_cortex', 'cerebellar_cortex', 'cerebellar_cortex', 'cerebellar_cortex', 'cerebellum', 'colon', 'cortex', 'digestive_tract', 'digestive_tract', 'forebrain', 'gills', 'gills', 'heart', 'heart', 'heart', 'hindbrain', 'hippocampal_formation', 'hippocampus', 'hippocampus', 'hippocampus', 'hippocampus', 'iPS1A_cell_line', 'iPS2A_cell_line', 'kidney', 'liver', 'liver', 'liver', 'lung', 'mantle', 'mantle', 'metastatic_pancreatic_ductal_adenocarcinoma', 'milk', 'mucus', 'muscle_of_leg', 'nerve_ending', 'ovary', 'pancreas', 'placenta', 'prefrontal_cortex', 'prefrontal_cortex', 'prefrontal_cortex', 'prefrontal_cortex', 'primary_pancreatic_ductal_adenocarcinoma', 'prostate_gland', 'seminal_fluid', 'striatum', 'striatum', 'striatum', 'striatum', 'testicle', 'testis', 'trachea', 'urine', 'urothelium']

This glycan has been reported to be dysregulated in (disease, direction, sample): [('REM_sleep_behavior_disorder', 'down', 'serum'), ('benign_breast_tumor_tissues_vs_para_carcinoma_tissues', 'up', 'breast'), ('cystic_fibrosis', 'up', 'sputum'), ('female_breast_cancer', 'up', 'breast'), ('female_breast_cancer', 'up', 'cell_line'), ('prostate_cancer', 'up', 'prostate_cancer_biopsy'), ('thyroid_gland_papillary_carcinoma', 'up', 'serum'), ('urinary_bladder_cancer', 'down', 'urine')]

That's all we can do for you at this point!</code></pre>
</div>
</div>
<hr>
</section>
<section id="glytoucan_to_glycan" class="level3">
<h3 class="anchored" data-anchor-id="glytoucan_to_glycan">glytoucan_to_glycan</h3>
<blockquote class="blockquote">
<pre><code> glytoucan_to_glycan (ids:list[str], revert:bool=False, verbose:bool=True)</code></pre>
</blockquote>
<p><em>Convert between GlyTouCan IDs and IUPAC-condensed glycans</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ids</td>
<td>list</td>
<td></td>
<td>List of GlyTouCan IDs or glycans</td>
</tr>
<tr class="even">
<td>revert</td>
<td>bool</td>
<td>False</td>
<td>Whether to map glycans to IDs; default:False</td>
</tr>
<tr class="odd">
<td>verbose</td>
<td>bool</td>
<td>True</td>
<td>Whether to print missing entries; default:True</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of glycans or IDs</strong></td>
</tr>
</tbody>
</table>
<div id="blocked-niger" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>glytoucan_to_glycan([<span class="st">'G63041RA'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc']</code></pre>
</div>
</div>
</section>
</section>
<section id="regex" class="level2">
<h2 class="anchored" data-anchor-id="regex">regex</h2>
<blockquote class="blockquote">
<p>for performing regular expression-like searches in glycans, very powerful to find complicated motifs</p>
</blockquote>
<hr>
<section id="get_match" class="level3">
<h3 class="anchored" data-anchor-id="get_match">get_match</h3>
<blockquote class="blockquote">
<pre><code> get_match (pattern:str|list[str],
            glycan:str|networkx.classes.digraph.DiGraph,
            return_matches:bool=True)</code></pre>
</blockquote>
<p><em>Find matches for glyco-regular expression in glycan</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | list[str]</td>
<td></td>
<td>Expression or pre-compiled pattern; e.g., “Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc”</td>
</tr>
<tr class="even">
<td>glycan</td>
<td>str | networkx.classes.digraph.DiGraph</td>
<td></td>
<td>Glycan string or graph</td>
</tr>
<tr class="odd">
<td>return_matches</td>
<td>bool</td>
<td>True</td>
<td>Whether to return matches vs boolean</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>bool | list[str]</strong></td>
<td></td>
<td><strong>Match results</strong></td>
</tr>
</tbody>
</table>
<div id="a3b2cb52-dc10-4a39-b8c9-2a1b94c53f7d" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># {} = between min and max occurrences, e.g., "Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc"</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a><span class="co"># * = zero or more occurrences, e.g., "Hex-HexNAc-([Hex|Fuc])*-HexNAc"</span></span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a><span class="co"># + = one or more occurrences, e.g., "Hex-HexNAc-([Hex|Fuc])+-HexNAc"</span></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ? = zero or one occurrence, e.g., "Hex-HexNAc-([Hex|Fuc])?-HexNAc"</span></span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a><span class="co"># {1,} = at minimum one occurrence, e.g., "Hex-HexNAc-([Hex|Fuc]){1,}-HexNAc"</span></span>
<span id="cb145-6"><a href="#cb145-6" aria-hidden="true" tabindex="-1"></a><span class="co"># {,1} = at maximum one occurrence, e.g., "Hex-HexNAc-([Hex|Fuc]){,1}-HexNAc"</span></span>
<span id="cb145-7"><a href="#cb145-7" aria-hidden="true" tabindex="-1"></a><span class="co"># {2} = exactly two occurrences, e.g., "Hex-HexNAc-([Hex|Fuc]){2}-HexNAc"</span></span>
<span id="cb145-8"><a href="#cb145-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ^ = start of sequence, e.g., "^Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc"</span></span>
<span id="cb145-9"><a href="#cb145-9" aria-hidden="true" tabindex="-1"></a><span class="co"># % = middle of sequence (i.e., neither start nor end)</span></span>
<span id="cb145-10"><a href="#cb145-10" aria-hidden="true" tabindex="-1"></a><span class="co"># $ = end of sequence, e.g., "Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc$"</span></span>
<span id="cb145-11"><a href="#cb145-11" aria-hidden="true" tabindex="-1"></a><span class="co"># ?&lt;= = lookbehind (i.e., provided pattern must be present before rest of pattern but is not included in match), e.g., "(?&lt;=Xyl-)Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc"</span></span>
<span id="cb145-12"><a href="#cb145-12" aria-hidden="true" tabindex="-1"></a><span class="co"># ?&lt;! = negative lookbehind (i.e., provided pattern is not present before rest of pattern and is also not included in match), e.g., "(?&lt;!Xyl-)Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc"</span></span>
<span id="cb145-13"><a href="#cb145-13" aria-hidden="true" tabindex="-1"></a><span class="co"># ?= = lookahead (i.e., provided pattern must be present after rest of pattern but is not included in match), e.g., "Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc(?=-HexNAc)"</span></span>
<span id="cb145-14"><a href="#cb145-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ?! = negative lookahead (i.e., provided pattern is not present after rest of pattern and is not included in match), e.g., "Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc(?!-HexNAc)"</span></span>
<span id="cb145-15"><a href="#cb145-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-16"><a href="#cb145-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: extracting the sequence from the a1-6 branch of N-glycans</span></span>
<span id="cb145-17"><a href="#cb145-17" aria-hidden="true" tabindex="-1"></a>pattern <span class="op">=</span> <span class="st">"r[Sia]{,1}-Monosaccharide-([dHex]){,1}-Monosaccharide(?=-Mana6-Monosaccharide)"</span></span>
<span id="cb145-18"><a href="#cb145-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(get_match(pattern, <span class="st">"GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>))</span>
<span id="cb145-19"><a href="#cb145-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(get_match(pattern, <span class="st">"GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[GalNAc(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>))</span>
<span id="cb145-20"><a href="#cb145-20" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(get_match(pattern, <span class="st">"GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)GalNAc(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>))</span>
<span id="cb145-21"><a href="#cb145-21" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(get_match(pattern, <span class="st">"GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-6)GalNAc(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>['Gal(b1-4)GlcNAc']
['GalNAc(b1-4)GlcNAc']
['Neu5Ac(a2-6)GalNAc(b1-4)GlcNAc']
['Neu5Gc(a2-6)GalNAc(b1-4)[Fuc(a1-3)]GlcNAc']</code></pre>
</div>
</div>
<p>For interested users, we here compile a selection of regular expression patterns that we find useful in our own work:</p>
<ul>
<li>Lewis or sialyl-Lewis structures:<br>
pattern = “r[Sia]{,1}-[Gal|GalOS]{1}-([Fuc]){1}-[GlcNAc|GlcNAc6S]{1}”</li>
<li>Blood groups:<br>
pattern = “rFuc-([Gal|GalNAc])?-Gal-GlcNAc”</li>
<li>a1-6 branch in N-glycans:<br>
pattern = “r[Sia]{,1}-[Hex|HexNAc]{,1}-([dHex]){,1}-[Man|GlcNAc]{1}-([.-.|.]){,1}-Mana6(?=-Manb4-GlcNAc)”</li>
<li>b1-6 branch in O-glycans (from core 2/4/6):<br>
pattern = “r[Sia|dHex]{,1}-[Hex|HexNAc]{,1}-([dHex]){,1}-.b6(?=-GalNAc)”</li>
<li>b1-3 branch in O-glycans (from core 1/2):<br>
pattern = “r[Sia]{,1}-[.]{,1}-([dHex]){,1}-.b3(?=-GalNAc)”</li>
</ul>
<hr>
</section>
<section id="get_match_batch" class="level3">
<h3 class="anchored" data-anchor-id="get_match_batch">get_match_batch</h3>
<blockquote class="blockquote">
<pre><code> get_match_batch (pattern:str,
                  glycan_list:list[str|networkx.classes.digraph.DiGraph],
                  return_matches:bool=True)</code></pre>
</blockquote>
<p><em>Find glyco-regular expression matches in list of glycans</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str</td>
<td></td>
<td>Glyco-regular expression; e.g., “Hex-HexNAc-([Hex|Fuc]){1,2}-HexNAc”</td>
</tr>
<tr class="even">
<td>glycan_list</td>
<td>list</td>
<td></td>
<td>List of glycans</td>
</tr>
<tr class="odd">
<td>return_matches</td>
<td>bool</td>
<td>True</td>
<td>Whether to return matches vs boolean</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list[bool] | list[list[str]]</strong></td>
<td></td>
<td><strong>Match results for each glycan</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="motif_to_regex" class="level3">
<h3 class="anchored" data-anchor-id="motif_to_regex">motif_to_regex</h3>
<blockquote class="blockquote">
<pre><code> motif_to_regex (motif:str)</code></pre>
</blockquote>
<p><em>Convert glycan motif to regular expression pattern</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>motif</td>
<td>str</td>
<td>Glycan in IUPAC-condensed</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td><strong>Regular expression</strong></td>
</tr>
</tbody>
</table>
<div id="2b00f450-dc7b-47ad-9c6e-73cda1101b94" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb149"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>motif_to_regex(<span class="st">"Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-?)"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Fuca3-([Galb4]){1}-GlcNAcb?'</code></pre>
</div>
</div>
</section>
</section>
<section id="tokenization" class="level2">
<h2 class="anchored" data-anchor-id="tokenization">tokenization</h2>
<blockquote class="blockquote">
<p>helper functions to map m/z–&gt;composition, composition–&gt;structure, structure–&gt;motif, and more</p>
</blockquote>
<hr>
<section id="string_to_labels" class="level3">
<h3 class="anchored" data-anchor-id="string_to_labels">string_to_labels</h3>
<blockquote class="blockquote">
<pre><code> string_to_labels (character_string:str, libr:dict[str,int]|None=None)</code></pre>
</blockquote>
<p><em>Tokenize word by indexing characters in library</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>character_string</td>
<td>str</td>
<td></td>
<td>String to tokenize</td>
</tr>
<tr class="even">
<td>libr</td>
<td>dict[str, int] | None</td>
<td>None</td>
<td>Dictionary mapping characters to indices</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of character indices</strong></td>
</tr>
</tbody>
</table>
<div id="unique-welcome" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb152"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>string_to_labels([<span class="st">'Man'</span>,<span class="st">'a1-3'</span>,<span class="st">'Man'</span>,<span class="st">'a1-6'</span>,<span class="st">'Man'</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>[None, None, None, None, None]</code></pre>
</div>
</div>
<hr>
</section>
<section id="pad_sequence" class="level3">
<h3 class="anchored" data-anchor-id="pad_sequence">pad_sequence</h3>
<blockquote class="blockquote">
<pre><code> pad_sequence (seq:list[int], max_length:int, pad_label:int|None=None,
               libr:dict[str,int]|None=None)</code></pre>
</blockquote>
<p><em>Pad sequences to same length using padding token</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>seq</td>
<td>list</td>
<td></td>
<td>Sequence to pad</td>
</tr>
<tr class="even">
<td>max_length</td>
<td>int</td>
<td></td>
<td>Target length</td>
</tr>
<tr class="odd">
<td>pad_label</td>
<td>int | None</td>
<td>None</td>
<td>Padding token value</td>
</tr>
<tr class="even">
<td>libr</td>
<td>dict[str, int] | None</td>
<td>None</td>
<td>Character library</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>Padded sequence</strong></td>
</tr>
</tbody>
</table>
<div id="recreational-persian" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb155"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>pad_sequence(string_to_labels([<span class="st">'Man'</span>,<span class="st">'a1-3'</span>,<span class="st">'Man'</span>,<span class="st">'a1-6'</span>,<span class="st">'Man'</span>]), <span class="dv">7</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>[None, None, None, None, None, 25, 25]</code></pre>
</div>
</div>
<hr>
</section>
<section id="stemify_glycan" class="level3">
<h3 class="anchored" data-anchor-id="stemify_glycan">stemify_glycan</h3>
<blockquote class="blockquote">
<pre><code> stemify_glycan (glycan:str, stem_lib:dict[str,str]|None=None,
                 libr:dict[str,int]|None=None)</code></pre>
</blockquote>
<p><em>Remove modifications from all monosaccharides in glycan</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td>stem_lib</td>
<td>dict[str, str] | None</td>
<td>None</td>
<td>Modified to core monosaccharide mapping; default:created from lib</td>
</tr>
<tr class="odd">
<td>libr</td>
<td>dict[str, int] | None</td>
<td>None</td>
<td>Glycoletter to index mapping</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td></td>
<td><strong>Stemmed glycan string</strong></td>
</tr>
</tbody>
</table>
<div id="extra-isaac" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a>stemify_glycan(<span class="st">"Neu5Ac9Ac(a2-3)Gal6S(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc'</code></pre>
</div>
</div>
<hr>
</section>
<section id="stemify_dataset" class="level3">
<h3 class="anchored" data-anchor-id="stemify_dataset">stemify_dataset</h3>
<blockquote class="blockquote">
<pre><code> stemify_dataset (df:pandas.core.frame.DataFrame,
                  stem_lib:dict[str,str]|None=None,
                  libr:dict[str,int]|None=None,
                  glycan_col_name:str='glycan', rarity_filter:int=1)</code></pre>
</blockquote>
<p><em>Remove monosaccharide modifications from all glycans in dataset</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>DataFrame</td>
<td></td>
<td>DataFrame with glycan column</td>
</tr>
<tr class="even">
<td>stem_lib</td>
<td>dict[str, str] | None</td>
<td>None</td>
<td>Modified to core monosaccharide mapping; default:created from lib</td>
</tr>
<tr class="odd">
<td>libr</td>
<td>dict[str, int] | None</td>
<td>None</td>
<td>Glycoletter to index mapping</td>
</tr>
<tr class="even">
<td>glycan_col_name</td>
<td>str</td>
<td>glycan</td>
<td>Column name for glycans</td>
</tr>
<tr class="odd">
<td>rarity_filter</td>
<td>int</td>
<td>1</td>
<td>Minimum occurrences to keep modification</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame with stemified glycans</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="mask_rare_glycoletters" class="level3">
<h3 class="anchored" data-anchor-id="mask_rare_glycoletters">mask_rare_glycoletters</h3>
<blockquote class="blockquote">
<pre><code> mask_rare_glycoletters (glycans:list[str],
                         thresh_monosaccharides:int|None=None,
                         thresh_linkages:int|None=None)</code></pre>
</blockquote>
<p><em>Mask rare monosaccharides and linkages in glycans</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycans</td>
<td>list</td>
<td></td>
<td>List of IUPAC-condensed glycans</td>
</tr>
<tr class="even">
<td>thresh_monosaccharides</td>
<td>int | None</td>
<td>None</td>
<td>Threshold for rare monosaccharides (default: 0.001*len(glycans))</td>
</tr>
<tr class="odd">
<td>thresh_linkages</td>
<td>int | None</td>
<td>None</td>
<td>Threshold for rare linkages (default: 0.03*len(glycans))</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of glycans with masked rare elements</strong></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="mz_to_composition" class="level3">
<h3 class="anchored" data-anchor-id="mz_to_composition">mz_to_composition</h3>
<blockquote class="blockquote">
<pre><code> mz_to_composition (mz_value:float, mode:str='negative',
                    mass_value:str='monoisotopic', reduced:bool=False,
                    sample_prep:str='underivatized',
                    mass_tolerance:float=0.5, kingdom:str='Animalia',
                    glycan_class:str='all',
                    df_use:pandas.core.frame.DataFrame|None=None,
                    filter_out:set[str]|None=None,
                    extras:list[str]=['doubly_charged'],
                    adduct:str|None=None)</code></pre>
</blockquote>
<p><em>Map m/z value to matching monosaccharide composition</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>mz_value</td>
<td>float</td>
<td></td>
<td>m/z value from mass spec</td>
</tr>
<tr class="even">
<td>mode</td>
<td>str</td>
<td>negative</td>
<td>MS mode: positive/negative</td>
</tr>
<tr class="odd">
<td>mass_value</td>
<td>str</td>
<td>monoisotopic</td>
<td>Mass type: monoisotopic/average</td>
</tr>
<tr class="even">
<td>reduced</td>
<td>bool</td>
<td>False</td>
<td>Whether glycans are reduced</td>
</tr>
<tr class="odd">
<td>sample_prep</td>
<td>str</td>
<td>underivatized</td>
<td>Sample preparation method: underivatized/permethylated/peracetylated</td>
</tr>
<tr class="even">
<td>mass_tolerance</td>
<td>float</td>
<td>0.5</td>
<td>Mass tolerance for matching</td>
</tr>
<tr class="odd">
<td>kingdom</td>
<td>str</td>
<td>Animalia</td>
<td>Taxonomic kingdom filter for choosing a subset of glycans to consider</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td>all</td>
<td>Glycan class: N/O/lipid/free/all</td>
</tr>
<tr class="odd">
<td>df_use</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Custom glycan database</td>
</tr>
<tr class="even">
<td>filter_out</td>
<td>set[str] | None</td>
<td>None</td>
<td>Monosaccharides to ignore during composition finding</td>
</tr>
<tr class="odd">
<td>extras</td>
<td>list</td>
<td>[‘doubly_charged’]</td>
<td>Additional operations: adduct/doubly_charged</td>
</tr>
<tr class="even">
<td>adduct</td>
<td>str | None</td>
<td>None</td>
<td>Chemical formula of adduct that contributes to m/z, e.g., “C2H4O2”</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of matching compositions</strong></td>
</tr>
</tbody>
</table>
<div id="civil-renaissance" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb163"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>mz_to_composition(<span class="fl">665.4</span>, glycan_class<span class="op">=</span><span class="st">'O'</span>, filter_out<span class="op">=</span>{<span class="st">'Kdn'</span>, <span class="st">'P'</span>, <span class="st">'HexA'</span>, <span class="st">'Pen'</span>, <span class="st">'HexN'</span>, <span class="st">'Me'</span>, <span class="st">'PCho'</span>, <span class="st">'PEtN'</span>},</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>                    reduced <span class="op">=</span> <span class="va">True</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>[{'Hex': 2, 'HexNAc': 2, 'Neu5Ac': 2}]</code></pre>
</div>
</div>
<hr>
</section>
<section id="match_composition_relaxed" class="level3">
<h3 class="anchored" data-anchor-id="match_composition_relaxed">match_composition_relaxed</h3>
<blockquote class="blockquote">
<pre><code> match_composition_relaxed (composition:dict[str,int],
                            glycan_class:str='N', kingdom:str='Animalia',
                            df_use:pandas.core.frame.DataFrame|None=None,
                            reducing_end:str|None=None)</code></pre>
</blockquote>
<p><em>Map coarse-grained composition to matching glycans</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>composition</td>
<td>dict</td>
<td></td>
<td>Dictionary indicating composition (e.g.&nbsp;{“dHex”: 1, “Hex”: 1, “HexNAc”: 1})</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td>N</td>
<td>Glycan class: N/O/lipid/free</td>
</tr>
<tr class="odd">
<td>kingdom</td>
<td>str</td>
<td>Animalia</td>
<td>Taxonomic kingdom filter for choosing a subset of glycans to consider</td>
</tr>
<tr class="even">
<td>df_use</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Custom glycan database</td>
</tr>
<tr class="odd">
<td>reducing_end</td>
<td>str | None</td>
<td>None</td>
<td>Reducing end specification</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of matching glycans</strong></td>
</tr>
</tbody>
</table>
<div id="grave-acoustic" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb166"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>match_composition_relaxed({<span class="st">"Hex"</span>:<span class="dv">3</span>, <span class="st">"HexNAc"</span>:<span class="dv">2</span>, <span class="st">"dHex"</span>:<span class="dv">1</span>}, glycan_class <span class="op">=</span> <span class="st">'O'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',
 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc',
 'Gal(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',
 'Gal(?1-3/4)Gal(b1-3/4)[Fuc(a1-3/4)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',
 'Gal(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-3)GalNAc',
 'Fuc(a1-2)Gal(b1-3/4)GlcNAc(b1-3)Gal(b1-3)[Gal(b1-6)]GalNAc',
 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)[Gal(?1-?)Gal(b1-3)]GalNAc',
 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-3)Gal(b1-3)GalNAc',
 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(?1-3/4)Gal(b1-3)GalNAc',
 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal',
 'Gal(?1-?)Gal(b1-4)GlcNAc(b1-6)[Fuc(a1-2)Gal(b1-3)]GalNAc',
 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)[Gal(a1-3)Gal(b1-3)]GalNAc',
 'Gal(a1-3)Gal(b1-4)GlcNAc(b1-6)[Fuc(a1-2)Gal(b1-3)]GalNAc',
 'Fuc(a1-2)Gal(b1-3)Gal(b1-3)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',
 'Fuc(a1-2)Gal(b1-3)Gal(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc',
 'Gal(b1-4)Gal(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]GalNAc',
 'Fuc(a1-2)Gal(?1-?)Gal(b1-3/4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',
 'Gal(a1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-3)Gal(b1-3)GalNAc',
 'Man(a1-6)Glc(a1-4)GlcNAc(b1-4)[Fuc(a1-2)]Gal(b1-3)GalNAc',
 'Man(a1-6)Glc(b1-4)GlcNAc(b1-4)[Fuc(a1-2)]Gal(b1-3)GalNAc',
 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-?)Man',
 'Gal(b1-2)Gal(a1-3)[Fuc(a1-2)]Gal(b1-3)[GlcNAc(b1-6)]GalNAc',
 'Fuc(a1-2)Gal(a1-3)Gal(a1-4)Gal(b1-3)[GlcNAc(b1-6)]GalNAc',
 'Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)]Gal(b1-3)GalNAc',
 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-?)Gal(b1-6)[Gal(b1-3)]GalNAc']</code></pre>
</div>
</div>
<hr>
</section>
<section id="condense_composition_matching" class="level3">
<h3 class="anchored" data-anchor-id="condense_composition_matching">condense_composition_matching</h3>
<blockquote class="blockquote">
<pre><code> condense_composition_matching (matched_composition:list[str])</code></pre>
</blockquote>
<p><em>Find minimum set of glycans characterizing matched composition</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>matched_composition</td>
<td>list</td>
<td>List of matching glycans</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td><strong>Minimal list of representative glycans</strong></td>
</tr>
</tbody>
</table>
<div id="cheap-fifty" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb169"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>match_comp <span class="op">=</span> match_composition_relaxed({<span class="st">'Hex'</span>:<span class="dv">1</span>, <span class="st">'HexNAc'</span>:<span class="dv">1</span>, <span class="st">'Neu5Ac'</span>:<span class="dv">1</span>}, glycan_class <span class="op">=</span> <span class="st">'O'</span>)</span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(match_comp)</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>condense_composition_matching(match_comp)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>['Neu5Ac(a2-3)Gal(b1-3)GalNAc', 'Gal(b1-3)[Neu5Ac(a2-6)]GalNAc', '{Neu5Ac(a2-3/6)}Gal(b1-3)GalNAc', 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal', 'Gal(a1-3)[Neu5Ac(a2-6)]GalNAc', 'Neu5Ac(a2-3/6)Gal(b1-3)GalNAc', 'Neu5Ac(a2-6)Gal(b1-3)GalNAc', 'Gal(?1-3)[Neu5Ac(a2-6)]GalNAc', 'Neu5Ac(a2-3/6)Gal(?1-3)GalNAc', 'Neu5Ac(a2-?)Hex(?1-?)GalNAc', 'Neu5Ac(a2-3)Gal(?1-?)GalNAc', 'Neu5Ac(a2-3/6)GalNAc(a1-6)Gal', 'Neu5Ac(a2-6)Gal(a1-3)GalNAc', 'Gal(b1-4)[Neu5Ac(a2-6)]GalNAc', 'Neu5Ac(a2-3)GalNAc(b1-3)Gal']</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>['Neu5Ac(a2-3)Gal(b1-3)GalNAc',
 'Neu5Ac(a2-3/6)Gal(b1-3)GalNAc',
 'Gal(b1-3)[Neu5Ac(a2-6)]GalNAc',
 'Gal(a1-3)[Neu5Ac(a2-6)]GalNAc',
 '{Neu5Ac(a2-3/6)}Gal(b1-3)GalNAc',
 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal',
 'Neu5Ac(a2-6)Gal(b1-3)GalNAc',
 'Neu5Ac(a2-3/6)GalNAc(a1-6)Gal',
 'Neu5Ac(a2-6)Gal(a1-3)GalNAc',
 'Gal(b1-4)[Neu5Ac(a2-6)]GalNAc',
 'Neu5Ac(a2-3)GalNAc(b1-3)Gal']</code></pre>
</div>
</div>
<hr>
</section>
<section id="mz_to_structures" class="level3">
<h3 class="anchored" data-anchor-id="mz_to_structures">mz_to_structures</h3>
<blockquote class="blockquote">
<pre><code> mz_to_structures (mz_list:list[float], glycan_class:str,
                   kingdom:str='Animalia',
                   abundances:pandas.core.frame.DataFrame|None=None,
                   mode:str='negative', mass_value:str='monoisotopic',
                   sample_prep:str='underivatized',
                   mass_tolerance:float=0.5, reduced:bool=False,
                   df_use:pandas.core.frame.DataFrame|None=None,
                   filter_out:set[str]|None=None, verbose:bool=False)</code></pre>
</blockquote>
<p><em>Map precursor masses to structures, supporting accompanying relative intensities</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>mz_list</td>
<td>list</td>
<td></td>
<td>List of precursor masses</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td></td>
<td>Glycan class: N/O/lipid/free</td>
</tr>
<tr class="odd">
<td>kingdom</td>
<td>str</td>
<td>Animalia</td>
<td>Taxonomic kingdom filter for choosing a subset of glycans to consider</td>
</tr>
<tr class="even">
<td>abundances</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Sample abundances matrix</td>
</tr>
<tr class="odd">
<td>mode</td>
<td>str</td>
<td>negative</td>
<td>MS mode: positive/negative</td>
</tr>
<tr class="even">
<td>mass_value</td>
<td>str</td>
<td>monoisotopic</td>
<td>Mass type: monoisotopic/average</td>
</tr>
<tr class="odd">
<td>sample_prep</td>
<td>str</td>
<td>underivatized</td>
<td>Sample prep: underivatized/permethylated/peracetylated</td>
</tr>
<tr class="even">
<td>mass_tolerance</td>
<td>float</td>
<td>0.5</td>
<td>Mass tolerance for matching</td>
</tr>
<tr class="odd">
<td>reduced</td>
<td>bool</td>
<td>False</td>
<td>Whether glycans are reduced</td>
</tr>
<tr class="even">
<td>df_use</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Custom glycan database</td>
</tr>
<tr class="odd">
<td>filter_out</td>
<td>set[str] | None</td>
<td>None</td>
<td>Monosaccharides to ignore</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>Whether to print non-matching compositions</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>pandas.core.frame.DataFrame | list</strong></td>
<td></td>
<td><strong>DataFrame of structures x intensities or empty list</strong></td>
</tr>
</tbody>
</table>
<div id="african-bulgarian" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>mz_to_structures([<span class="fl">674.29</span>], glycan_class <span class="op">=</span> <span class="st">'O'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 compositions could not be matched. Run with verbose = True to see which compositions.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">glycan</th>
<th data-quarto-table-cell-role="th">abundance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Neu5Ac(a2-3)Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Neu5Ac(a2-3/6)Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Gal(a1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>{Neu5Ac(a2-3/6)}Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">5</th>
<td>Neu5Ac(a2-3)[GalNAc(b1-4)]Gal</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">6</th>
<td>Neu5Ac(a2-6)Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">7</th>
<td>Neu5Ac(a2-3/6)GalNAc(a1-6)Gal</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">8</th>
<td>Neu5Ac(a2-6)Gal(a1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">9</th>
<td>Gal(b1-4)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">10</th>
<td>Neu5Ac(a2-3)GalNAc(b1-3)Gal</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="compositions_to_structures" class="level3">
<h3 class="anchored" data-anchor-id="compositions_to_structures">compositions_to_structures</h3>
<blockquote class="blockquote">
<pre><code> compositions_to_structures (composition_list:list[dict[str,int]],
                             glycan_class:str='N', kingdom:str='Animalia',
                             abundances:pandas.core.frame.DataFrame|None=N
                             one,
                             df_use:pandas.core.frame.DataFrame|None=None,
                             verbose:bool=False)</code></pre>
</blockquote>
<p><em>Map compositions to structures, supporting accompanying relative intensities</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>composition_list</td>
<td>list</td>
<td></td>
<td>List of compositions like {‘Hex’: 1, ‘HexNAc’: 1}</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td>N</td>
<td>Glycan class: N/O/lipid/free</td>
</tr>
<tr class="odd">
<td>kingdom</td>
<td>str</td>
<td>Animalia</td>
<td>Taxonomic kingdom filter for choosing a subset of glycans to consider</td>
</tr>
<tr class="even">
<td>abundances</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Sample abundances matrix</td>
</tr>
<tr class="odd">
<td>df_use</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Custom glycan database</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>Whether to print non-matching compositions</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>DataFrame of structures x intensities</strong></td>
</tr>
</tbody>
</table>
<div id="frequent-bacteria" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb176"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>compositions_to_structures([{<span class="st">'Neu5Ac'</span>: <span class="dv">2</span>, <span class="st">'Hex'</span>: <span class="dv">1</span>, <span class="st">'HexNAc'</span>: <span class="dv">1</span>}], glycan_class <span class="op">=</span> <span class="st">'O'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 compositions could not be matched. Run with verbose = True to see which compositions.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">glycan</th>
<th data-quarto-table-cell-role="th">abundance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Neu5Ac(a2-8)Neu5Ac(a2-6)[Gal(b1-3)]GalNAc</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Neu5Ac(a2-3)[Neu5Ac(a2-6)]Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Neu5Ac(a2-3)Gal(b1-4)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="40e1f90d-209f-449d-be0f-12d265b5f6d3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb178"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>compositions_to_structures([<span class="st">"H1N1A2"</span>], glycan_class <span class="op">=</span> <span class="st">'O'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 compositions could not be matched. Run with verbose = True to see which compositions.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">glycan</th>
<th data-quarto-table-cell-role="th">abundance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Neu5Ac(a2-8)Neu5Ac(a2-6)[Gal(b1-3)]GalNAc</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Neu5Ac(a2-3)[Neu5Ac(a2-6)]Gal(b1-3)GalNAc</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Neu5Ac(a2-3)Gal(b1-4)[Neu5Ac(a2-6)]GalNAc</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="structure_to_basic" class="level3">
<h3 class="anchored" data-anchor-id="structure_to_basic">structure_to_basic</h3>
<blockquote class="blockquote">
<pre><code> structure_to_basic (glycan:str)</code></pre>
</blockquote>
<p><em>Convert glycan structure to base topology</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td><strong>Base topology string</strong></td>
</tr>
</tbody>
</table>
<div id="neither-vegetation" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb181"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a>structure_to_basic(<span class="st">"Neu5Ac(a2-3)Gal6S(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Neu5Ac(?1-?)HexOS(?1-?)[Neu5Ac(?1-?)]HexNAc'</code></pre>
</div>
</div>
<hr>
</section>
<section id="glycan_to_composition" class="level3">
<h3 class="anchored" data-anchor-id="glycan_to_composition">glycan_to_composition</h3>
<blockquote class="blockquote">
<pre><code> glycan_to_composition (glycan:str, stem_libr:dict[str,str]|None=None)</code></pre>
</blockquote>
<p><em>Map glycan to its composition</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td>stem_libr</td>
<td>dict[str, str] | None</td>
<td>None</td>
<td>Modified to core monosaccharide mapping; default: created from lib</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>Dictionary of monosaccharide counts</strong></td>
</tr>
</tbody>
</table>
<div id="designing-comparative" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb184"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a>glycan_to_composition(<span class="st">"Neu5Ac(a2-3)Gal6S(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>{'Hex': 1, 'HexNAc': 1, 'Neu5Ac': 2, 'S': 1}</code></pre>
</div>
</div>
<hr>
</section>
<section id="glycan_to_mass" class="level3">
<h3 class="anchored" data-anchor-id="glycan_to_mass">glycan_to_mass</h3>
<blockquote class="blockquote">
<pre><code> glycan_to_mass (glycan:str, mass_value:str='monoisotopic',
                 sample_prep:str='underivatized',
                 stem_libr:dict[str,str]|None=None,
                 adduct:str|float|None=None)</code></pre>
</blockquote>
<p><em>Calculate theoretical mass from glycan</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>glycan</td>
<td>str</td>
<td></td>
<td>Glycan in IUPAC-condensed format</td>
</tr>
<tr class="even">
<td>mass_value</td>
<td>str</td>
<td>monoisotopic</td>
<td>Mass type: monoisotopic/average</td>
</tr>
<tr class="odd">
<td>sample_prep</td>
<td>str</td>
<td>underivatized</td>
<td>Sample prep: underivatized/permethylated/peracetylated</td>
</tr>
<tr class="even">
<td>stem_libr</td>
<td>dict[str, str] | None</td>
<td>None</td>
<td>Modified to core monosaccharide mapping</td>
</tr>
<tr class="odd">
<td>adduct</td>
<td>str | float | None</td>
<td>None</td>
<td>Chemical formula of adduct (e.g., “C2H4O2”) OR its exact mass in Da</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>float</strong></td>
<td></td>
<td><strong>Theoretical mass</strong></td>
</tr>
</tbody>
</table>
<div id="headed-approval" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb187"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a>glycan_to_mass(<span class="st">"Neu5Ac(a2-3)Gal6S(b1-3)[Neu5Ac(a2-6)]GalNAc"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>1045.2903546</code></pre>
</div>
</div>
<hr>
</section>
<section id="composition_to_mass" class="level3">
<h3 class="anchored" data-anchor-id="composition_to_mass">composition_to_mass</h3>
<blockquote class="blockquote">
<pre><code> composition_to_mass (dict_comp_in:dict[str,int],
                      mass_value:str='monoisotopic',
                      sample_prep:str='underivatized',
                      adduct:str|float|None=None)</code></pre>
</blockquote>
<p><em>Calculate theoretical mass from composition</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dict_comp_in</td>
<td>dict</td>
<td></td>
<td>Composition dictionary of monosaccharide:count</td>
</tr>
<tr class="even">
<td>mass_value</td>
<td>str</td>
<td>monoisotopic</td>
<td>Mass type: monoisotopic/average</td>
</tr>
<tr class="odd">
<td>sample_prep</td>
<td>str</td>
<td>underivatized</td>
<td>Sample prep: underivatized/permethylated/peracetylated</td>
</tr>
<tr class="even">
<td>adduct</td>
<td>str | float | None</td>
<td>None</td>
<td>Chemical formula of adduct (e.g., “C2H4O2”) OR its exact mass in Da</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>float</strong></td>
<td></td>
<td><strong>Theoretical mass</strong></td>
</tr>
</tbody>
</table>
<div id="second-consensus" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb190"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>composition_to_mass({<span class="st">'Neu5Ac'</span>: <span class="dv">2</span>, <span class="st">'Hex'</span>: <span class="dv">1</span>, <span class="st">'HexNAc'</span>: <span class="dv">1</span>, <span class="st">'S'</span>: <span class="dv">1</span>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>1045.2903546</code></pre>
</div>
</div>
<hr>
</section>
<section id="calculate_adduct_mass" class="level3">
<h3 class="anchored" data-anchor-id="calculate_adduct_mass">calculate_adduct_mass</h3>
<blockquote class="blockquote">
<pre><code> calculate_adduct_mass (formula:str, mass_value:str='monoisotopic',
                        enforce_sign:bool=False)</code></pre>
</blockquote>
<p><em>Calculate mass of adduct from chemical formula, including signed formulas</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>formula</td>
<td>str</td>
<td></td>
<td>Chemical formula of adduct (e.g., “C2H4O2”, “-H2O”, “+Na”)</td>
</tr>
<tr class="even">
<td>mass_value</td>
<td>str</td>
<td>monoisotopic</td>
<td>Mass type: monoisotopic/average</td>
</tr>
<tr class="odd">
<td>enforce_sign</td>
<td>bool</td>
<td>False</td>
<td>If True, returns 0 for unsigned formulas</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>float</strong></td>
<td></td>
<td><strong>Formula mass</strong></td>
</tr>
</tbody>
</table>
<div id="b3f751be-8345-4a74-86b5-a4b9df4fd1a0" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb193"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a>calculate_adduct_mass(<span class="st">"C2H4O2"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>60.021</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_unique_topologies" class="level3">
<h3 class="anchored" data-anchor-id="get_unique_topologies">get_unique_topologies</h3>
<blockquote class="blockquote">
<pre><code> get_unique_topologies (composition:dict[str,int], glycan_type:str,
                        df_use:pandas.core.frame.DataFrame|None=None,
                        universal_replacers:dict[str,str]|None=None,
                        taxonomy_rank:str='Kingdom',
                        taxonomy_value:str='Animalia')</code></pre>
</blockquote>
<p><em>Get all observed unique base topologies for composition</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>composition</td>
<td>dict</td>
<td></td>
<td>Composition dictionary of monosaccharide:count</td>
</tr>
<tr class="even">
<td>glycan_type</td>
<td>str</td>
<td></td>
<td>Glycan class: N/O/lipid/free/repeat</td>
</tr>
<tr class="odd">
<td>df_use</td>
<td>pandas.core.frame.DataFrame | None</td>
<td>None</td>
<td>Custom glycan database to use for mapping</td>
</tr>
<tr class="even">
<td>universal_replacers</td>
<td>dict[str, str] | None</td>
<td>None</td>
<td>Base-to-specific monosaccharide mapping</td>
</tr>
<tr class="odd">
<td>taxonomy_rank</td>
<td>str</td>
<td>Kingdom</td>
<td>Taxonomic rank for filtering</td>
</tr>
<tr class="even">
<td>taxonomy_value</td>
<td>str</td>
<td>Animalia</td>
<td>Value at taxonomy rank</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>list</strong></td>
<td></td>
<td><strong>List of unique base topologies</strong></td>
</tr>
</tbody>
</table>
<div id="fceba897-6089-406d-9826-d4a0696edc94" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb196"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a>get_unique_topologies({<span class="st">'HexNAc'</span>:<span class="dv">2</span>, <span class="st">'Hex'</span>:<span class="dv">1</span>}, <span class="st">'O'</span>, universal_replacers <span class="op">=</span> {<span class="st">'dHex'</span>:<span class="st">'Fuc'</span>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>['HexNAc(?1-?)[HexNAc(?1-?)]Hex',
 'HexNAc(?1-?)HexNAc(?1-?)Hex',
 'HexNAc(?1-?)Hex(?1-?)HexNAc',
 'Hex(?1-?)[HexNAc(?1-?)]HexNAc',
 'Hex(?1-?)HexNAc(?1-?)HexNAc']</code></pre>
</div>
</div>
<hr>
</section>
<section id="get_random_glycan" class="level3">
<h3 class="anchored" data-anchor-id="get_random_glycan">get_random_glycan</h3>
<blockquote class="blockquote">
<pre><code> get_random_glycan (n:int=1, glycan_class:str='all',
                    kingdom:str='Animalia')</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>n</td>
<td>int</td>
<td>1</td>
<td>How many random glycans to sample</td>
</tr>
<tr class="even">
<td>glycan_class</td>
<td>str</td>
<td>all</td>
<td>Glycan class: N/O/lipid/free/repeat/all</td>
</tr>
<tr class="odd">
<td>kingdom</td>
<td>str</td>
<td>Animalia</td>
<td>Taxonomic kingdom filter for choosing a subset of glycans to consider</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>str | list[str]</strong></td>
<td></td>
<td><strong>Returns a random glycan or list of glycans if n &gt; 1</strong></td>
</tr>
</tbody>
</table>
<div id="6a4f1703-5e0e-4642-acf0-da80671c9b3e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb199"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a>get_random_glycan()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>'Gal(b1-4)GlcNAc(b1-2)Man(a1-3/6)[GlcNAc(b1-4)Man(a1-3/6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc'</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BojarLab\.github\.io\/glycowork\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BojarLab/glycowork/tree/master/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>