<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>glycowork - ml</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="glycowork - ml">
<meta property="og:description" content="ml contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan representations.">
<meta property="og:site-name" content="glycowork">
<meta name="twitter:title" content="glycowork - ml">
<meta name="twitter:description" content="ml contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan representations.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">glycowork</span>
    </a>
  </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">ml</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Glycowork</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">Overview</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./glycan_data.html" class="sidebar-item-text sidebar-link">glycan_data</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link active">ml</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motif.html" class="sidebar-item-text sidebar-link">motif</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./network.html" class="sidebar-item-text sidebar-link">network</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./examples.html" class="sidebar-item-text sidebar-link">Examples</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#model_training" id="toc-model_training" class="nav-link active" data-scroll-target="#model_training">model_training</a>
  <ul class="collapse">
  <li><a href="#earlystopping" id="toc-earlystopping" class="nav-link" data-scroll-target="#earlystopping">EarlyStopping</a></li>
  <li><a href="#train_model" id="toc-train_model" class="nav-link" data-scroll-target="#train_model">train_model</a></li>
  <li><a href="#training_setup" id="toc-training_setup" class="nav-link" data-scroll-target="#training_setup">training_setup</a></li>
  <li><a href="#train_ml_model" id="toc-train_ml_model" class="nav-link" data-scroll-target="#train_ml_model">train_ml_model</a></li>
  <li><a href="#analyze_ml_model" id="toc-analyze_ml_model" class="nav-link" data-scroll-target="#analyze_ml_model">analyze_ml_model</a></li>
  <li><a href="#get_mismatch" id="toc-get_mismatch" class="nav-link" data-scroll-target="#get_mismatch">get_mismatch</a></li>
  </ul></li>
  <li><a href="#models" id="toc-models" class="nav-link" data-scroll-target="#models">models</a></li>
  <li><a href="#processing" id="toc-processing" class="nav-link" data-scroll-target="#processing">processing</a></li>
  <li><a href="#inference" id="toc-inference" class="nav-link" data-scroll-target="#inference">inference</a></li>
  <li><a href="#train_test_split" id="toc-train_test_split" class="nav-link" data-scroll-target="#train_test_split">train_test_split</a>
  <ul class="collapse">
  <li><a href="#hierarchy_filter" id="toc-hierarchy_filter" class="nav-link" data-scroll-target="#hierarchy_filter">hierarchy_filter</a></li>
  <li><a href="#general_split" id="toc-general_split" class="nav-link" data-scroll-target="#general_split">general_split</a></li>
  <li><a href="#prepare_multilabel" id="toc-prepare_multilabel" class="nav-link" data-scroll-target="#prepare_multilabel">prepare_multilabel</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/BojarLab/glycowork/tree/master/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">ml</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p><code>ml</code> contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan representations. It currently contains the following modules:</p>
<ul>
<li><code>model_training</code> contains functions for training machine learning models</li>
<li><code>models</code> describes some examples for machine learning architectures applicable to glycans</li>
<li><code>processing</code> contains helper functions to prepare glycan data for model training</li>
<li><code>inference</code> can be used to analyze trained models, make predictions, or obtain glycan representations</li>
<li><code>train_test_split</code> contains various data split functions to get appropriate training and test sets</li>
</ul>
<section id="model_training" class="level2">
<h2 class="anchored" data-anchor-id="model_training">model_training</h2>
<blockquote class="blockquote">
<p>contains functions for training machine learning models</p>
</blockquote>
<hr>
<section id="earlystopping" class="level3">
<h3 class="anchored" data-anchor-id="earlystopping">EarlyStopping</h3>
<blockquote class="blockquote">
<pre><code> EarlyStopping (patience=7, verbose=False)</code></pre>
</blockquote>
<p>Early stops the training if validation loss doesn’t improve after a given patience.</p>
<hr>
</section>
<section id="train_model" class="level3">
<h3 class="anchored" data-anchor-id="train_model">train_model</h3>
<blockquote class="blockquote">
<pre><code> train_model (model, dataloaders, criterion, optimizer, scheduler,
              num_epochs=25, patience=50, mode='classification',
              mode2='multi')</code></pre>
</blockquote>
<p>trains a deep learning model on predicting glycan properties</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="even">
<td style="text-align: left;">dataloaders (PyTorch object): dictionary of dataloader objects with keys ‘train’ and ‘val’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">criterion (PyTorch object): PyTorch loss function</td>
</tr>
<tr class="even">
<td style="text-align: left;">optimizer (PyTorch object): PyTorch optimizer</td>
</tr>
<tr class="odd">
<td style="text-align: left;">scheduler (PyTorch object): PyTorch learning rate decay</td>
</tr>
<tr class="even">
<td style="text-align: left;">num_epochs (int): number of epochs for training; default:25</td>
</tr>
<tr class="odd">
<td style="text-align: left;">patience (int): number of epochs without improvement until early stop; default:50</td>
</tr>
<tr class="even">
<td style="text-align: left;">mode (string): ‘classification’, ‘multilabel’, or ‘regression’; default:classification</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mode2 (string): further specifying classification into ‘multi’ or ‘binary’ classification;default:multi</td>
</tr>
</tbody>
</table>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns the best model seen during training</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="training_setup" class="level3">
<h3 class="anchored" data-anchor-id="training_setup">training_setup</h3>
<blockquote class="blockquote">
<pre><code> training_setup (model, lr, lr_patience=4, factor=0.2,
                 weight_decay=0.0001, mode='multiclass', gsam_alpha=0.0)</code></pre>
</blockquote>
<p>prepares optimizer, learning rate scheduler, and loss criterion for model training</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="even">
<td style="text-align: left;">lr (float): learning rate</td>
</tr>
<tr class="odd">
<td style="text-align: left;">lr_patience (int): number of epochs without validation loss improvement before reducing the learning rate;default:4</td>
</tr>
<tr class="even">
<td style="text-align: left;">factor (float): factor by which learning rate is multiplied upon reduction</td>
</tr>
<tr class="odd">
<td style="text-align: left;">weight_decay (float): regularization parameter for the optimizer; default:0.001</td>
</tr>
<tr class="even">
<td style="text-align: left;">mode (string): ‘multiclass’: classification with multiple classes, ‘multilabel’: predicting several labels at the same time, ‘binary’:binary classification, ‘regression’: regression; default:‘multiclass’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gsam_alpha (float): if higher than zero, uses GSAM instead of SAM for the optimizer</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns optimizer, learning rate scheduler, and loss criterion objects</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="train_ml_model" class="level3">
<h3 class="anchored" data-anchor-id="train_ml_model">train_ml_model</h3>
<blockquote class="blockquote">
<pre><code> train_ml_model (X_train, X_test, y_train, y_test, mode='classification',
                 feature_calc=False, libr=None, return_features=False,
                 feature_set=['known', 'exhaustive'],
                 additional_features_train=None,
                 additional_features_test=None)</code></pre>
</blockquote>
<p>wrapper function to train standard machine learning models on glycans</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">X_train, X_test (list or dataframe): either lists of glycans (needs feature_calc = True) or motif dataframes such as from annotate_dataset</td>
</tr>
<tr class="even">
<td style="text-align: left;">y_train, y_test (list): lists of labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mode (string): ‘classification’ or ‘regression’; default:‘classification’</td>
</tr>
<tr class="even">
<td style="text-align: left;">feature_calc (bool): set to True for calculating motifs from glycans; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (list): sorted list of unique glycoletters observed in the glycans of our data; default:lib</td>
</tr>
<tr class="even">
<td style="text-align: left;">return_features (bool): whether to return calculated features; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">feature_set (list): which feature set to use for annotations, add more to list to expand; default:[‘known’,‘exhaustive’]; options are: ‘known’ (hand-crafted glycan features), ‘graph’ (structural graph features of glycans), and ‘exhaustive’ (all mono- and disaccharide features)</td>
</tr>
<tr class="even">
<td style="text-align: left;">additional_features_train (dataframe): additional features (apart from glycans) to be used for training. Has to be of the same length as X_train; default:None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">additional_features_test (dataframe): additional features (apart from glycans) to be used for evaluation. Has to be of the same length as X_test; default:None</td>
</tr>
</tbody>
</table>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns trained model</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>vertebrate <span class="op">=</span> [<span class="dv">1</span> <span class="cf">if</span> k <span class="op">==</span> <span class="st">'Chordata'</span> <span class="cf">else</span> <span class="dv">0</span> <span class="cf">for</span> k <span class="kw">in</span> df_species[df_species.Kingdom<span class="op">==</span><span class="st">'Animalia'</span>].Phylum.values.tolist()]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>X_train, X_test, y_train, y_test <span class="op">=</span> general_split(df_species[df_species.Kingdom<span class="op">==</span><span class="st">'Animalia'</span>].target.values.tolist(), vertebrate)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>model_ft, _, X_test <span class="op">=</span> train_ml_model(X_train, X_test, y_train, y_test, feature_calc <span class="op">=</span> <span class="va">True</span>, feature_set <span class="op">=</span> [<span class="st">'exhaustive'</span>],</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                         return_features <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Calculating Glycan Features...

Training model...
[07:14:11] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.

Evaluating model...
Accuracy of trained model on separate validation set: 0.927536231884058</code></pre>
</div>
</div>
<hr>
</section>
<section id="analyze_ml_model" class="level3">
<h3 class="anchored" data-anchor-id="analyze_ml_model">analyze_ml_model</h3>
<blockquote class="blockquote">
<pre><code> analyze_ml_model (model)</code></pre>
</blockquote>
<p>plots relevant features for model prediction</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (model object): trained machine learning model from train_ml_model</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>analyze_ml_model(model_ft)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="02_ml_files/figure-html/cell-8-output-1.png" class="img-fluid"></p>
</div>
</div>
<hr>
</section>
<section id="get_mismatch" class="level3">
<h3 class="anchored" data-anchor-id="get_mismatch">get_mismatch</h3>
<blockquote class="blockquote">
<pre><code> get_mismatch (model, X_test, y_test, n=10)</code></pre>
</blockquote>
<p>analyzes misclassifications of trained machine learning model</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (model object): trained machine learning model from train_ml_model</td>
</tr>
<tr class="even">
<td style="text-align: left;">X_test (dataframe): motif dataframe used for validating model</td>
</tr>
<tr class="odd">
<td style="text-align: left;">y_test (list): list of labels</td>
</tr>
<tr class="even">
<td style="text-align: left;">n (int): number of returned misclassifications; default:10</td>
</tr>
</tbody>
</table>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns tuples of misclassifications and their predicted probability</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>get_mismatch(model_ft, X_test, y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('Man(a1-?)Man(a1-?)[Man(a1-?)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',
  0.7171406745910645),
 ('Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.919142484664917),
 ('Man(a1-?)Man(a1-?)[Man(a1-?)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.8076043128967285),
 ('Man(a1-?)Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc', 0.49124422669410706),
 ('GlcNAc(b1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.46067529916763306),
 ('Gal(a1-4)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc', 0.9600384831428528),
 ('Gal6Aep(a1-2)[Gal3Me(a1-3)]Gal(b1-4)Glc1Cer', 0.5451778769493103),
 ('Man(a1-2)Man(a1-6)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.4874219596385956),
 ('GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]Man',
  0.4899427592754364),
 ('Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.591519832611084)]</code></pre>
</div>
</div>
</section>
</section>
<section id="models" class="level2">
<h2 class="anchored" data-anchor-id="models">models</h2>
<blockquote class="blockquote">
<p>describes some examples for machine learning architectures applicable to glycans. The main portal is prep_models which allows users to setup (trained) models by their string names</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(SweetNet)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(LectinOracle)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(LectinOracle_flex)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(NSequonPred)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(init_weights)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(prep_model)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="processing" class="level2">
<h2 class="anchored" data-anchor-id="processing">processing</h2>
<blockquote class="blockquote">
<p>contains helper functions to prepare glycan data for model training</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(dataset_to_graphs)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(dataset_to_dataloader)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(split_data_to_train)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="inference" class="level2">
<h2 class="anchored" data-anchor-id="inference">inference</h2>
<blockquote class="blockquote">
<p>can be used to analyze trained models, make predictions, or obtain glycan representations</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#requires torch_geometric</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(glycans_to_emb)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(get_lectin_preds)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(get_Nsequon_preds)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#show_doc(get_esm1b_representations)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order to run <code>get_esm1b_representations</code>, you first have to run this snippet:</p>
<p><code>!pip install fair-esm import esm model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()</code></p>
</section>
<section id="train_test_split" class="level2">
<h2 class="anchored" data-anchor-id="train_test_split">train_test_split</h2>
<blockquote class="blockquote">
<p>contains various data split functions to get appropriate training and test sets</p>
</blockquote>
<hr>
<section id="hierarchy_filter" class="level3">
<h3 class="anchored" data-anchor-id="hierarchy_filter">hierarchy_filter</h3>
<blockquote class="blockquote">
<pre><code> hierarchy_filter (df_in, rank='Domain', min_seq=5, wildcard_seed=False,
                   wildcard_list=None, wildcard_name=None, r=0.1,
                   col='target')</code></pre>
</blockquote>
<p>stratified data split in train/test at the taxonomic level, removing duplicate glycans and infrequent classes</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">df_in (dataframe): dataframe of glycan sequences and taxonomic labels</td>
</tr>
<tr class="even">
<td style="text-align: left;">rank (string): which rank should be filtered; default:‘domain’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">min_seq (int): how many glycans need to be present in class to keep it; default:5</td>
</tr>
<tr class="even">
<td style="text-align: left;">wildcard_seed (bool): set to True if you want to seed wildcard glycoletters; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">wildcard_list (list): list which glycoletters a wildcard encompasses</td>
</tr>
<tr class="even">
<td style="text-align: left;">wildcard_name (string): how the wildcard should be named in the IUPAC-condensed nomenclature</td>
</tr>
<tr class="odd">
<td style="text-align: left;">r (float): rate of replacement, default:0.1 or 10%</td>
</tr>
<tr class="even">
<td style="text-align: left;">col (string): column name for glycan sequences; default:target</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns train_x, val_x (lists of glycans (strings) after stratified shuffle split)</td>
</tr>
<tr class="even">
<td style="text-align: left;">train_y, val_y (lists of taxonomic labels (mapped integers))</td>
</tr>
<tr class="odd">
<td style="text-align: left;">id_val (taxonomic labels in text form (strings))</td>
</tr>
<tr class="even">
<td style="text-align: left;">class_list (list of unique taxonomic classes (strings))</td>
</tr>
<tr class="odd">
<td style="text-align: left;">class_converter (dictionary to map mapped integers back to text labels)</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>train_x, val_x, train_y, val_y, id_val, class_list, class_converter <span class="op">=</span> hierarchy_filter(df_species,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                                                                                       rank <span class="op">=</span> <span class="st">'Kingdom'</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(train_x[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['GlcNAc(b1-3)GlcNAc(b1-3)GlcNAcOP(a1-3)GlcNAc', 'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc-ol', 'Fruf1Ac(b2-1)Glc3Ac6Ac', 'Man(a1-3)[Gal(b1-4)]Man(?1-?)Ins', 'Gal(b1-4)Gal(b1-4)Gal(b1-4)Glc', 'Gal(b1-4)Xyl', 'Rha(a1-2)[Ara(a1-3)]Glc', 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc', 'Ara2Ac3Ac4Ac(a1-6)Glc', 'Ara(b1-2)Gal(b1-3)Gal(b1-3)Man(a1-3)Man(a1-4)GlcN(a1-6)Ins']</code></pre>
</div>
</div>
<hr>
</section>
<section id="general_split" class="level3">
<h3 class="anchored" data-anchor-id="general_split">general_split</h3>
<blockquote class="blockquote">
<pre><code> general_split (glycans, labels, test_size=0.2)</code></pre>
</blockquote>
<p>splits glycans and labels into train / test sets</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">labels (list): list of labels used for prediction</td>
</tr>
<tr class="odd">
<td style="text-align: left;">test_size (float): % size of test set; default:0.2 / 20%</td>
</tr>
</tbody>
</table>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns X_train, X_test, y_train, y_test</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>train_x, val_x, train_y, val_y <span class="op">=</span> general_split(df_species.target.values.tolist(),</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                                              df_species.Species.values.tolist())</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(train_x[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['Galf(b1-3)Gal(a1-4)Neu5Ac(a2-3)Gal(b1-4)Glc', 'Neu5Gc(a2-?)Gal(b1-?)GalNAc(b1-4)[Neu5Ac(a2-8)Neu5Ac(a2-?)]Gal(b1-4)Glc1Cer', 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-8)Neu5Ac(a2-6)]GalNAc(b1-4)Gal(b1-4)Glc1Cer', 'Man(a1-2)Man(a1-2)Man', 'Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]GalNAc', 'Gal(b1-?)GlcNAc(b1-3)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-3)Gal(b1-4)Glc-ol', 'Man(a1-2)Man6PEtN(a1-2)Man(a1-6)[GalNAc(b1-4)]Man2PEtN(a1-4)GlcN', 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc', 'GlcN(a1-6)GlcNAc', 'Man(a1-2)Man(a1-3)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc']</code></pre>
</div>
</div>
<hr>
</section>
<section id="prepare_multilabel" class="level3">
<h3 class="anchored" data-anchor-id="prepare_multilabel">prepare_multilabel</h3>
<blockquote class="blockquote">
<pre><code> prepare_multilabel (df, rank='Species', glycan_col='target')</code></pre>
</blockquote>
<p>converts a one row per glycan-species/tissue/disease association file to a format of one glycan - all associations</p>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">df (dataframe): dataframe where each row is one glycan - species association</td>
</tr>
<tr class="even">
<td style="text-align: left;">rank (string): which label column should be used; default:Species</td>
</tr>
<tr class="odd">
<td style="text-align: left;">glycan_col (string): column name of where the glycan sequences are stored; default:target</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(1) list of unique glycans in df</td>
</tr>
<tr class="even">
<td style="text-align: left;">(2) list of lists, where each inner list are all the labels of a glycan</td>
</tr>
</tbody>
</table>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>