---

title: Glycowork


keywords: fastai
sidebar: home_sidebar

summary: "Glycans are a fundamental biological sequence, similar to DNA, RNA, or proteins. Glycans are complex carbohydrates that can form branched structures comprising monosaccharides and linkages as constituents. Despite being conspicuously absent from most research, glycans are ubiquitous in biology. They decorate most proteins and lipids and direct the stability and functions of biomolecules, cells, and organisms. This also makes glycans relevant to every human disease."
description: "Glycans are a fundamental biological sequence, similar to DNA, RNA, or proteins. Glycans are complex carbohydrates that can form branched structures comprising monosaccharides and linkages as constituents. Despite being conspicuously absent from most research, glycans are ubiquitous in biology. They decorate most proteins and lipids and direct the stability and functions of biomolecules, cells, and organisms. This also makes glycans relevant to every human disease."
nb_path: "index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Install">Install<a class="anchor-link" href="#Install"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#`pip install glycowork`</span>

<span class="c1">#if you don&#39;t have a security token</span>
<span class="c1">#`pip install git+https://YOURUSERNAME:YOURPASSWORD@github.com/BojarLab/glycowork.git</span>

<span class="err">`</span><span class="n">pip</span> <span class="n">install</span> <span class="n">git</span><span class="o">+</span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">BojarLab</span><span class="o">/</span><span class="n">glycowork</span><span class="o">.</span><span class="n">git</span>
<span class="kn">import</span> <span class="nn">glycowork</span><span class="err">`</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="How-to-use">How to use<a class="anchor-link" href="#How-to-use"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Glycowork currently contains four main modules:</p>
<ul>
<li><strong>alignment</strong><ul>
<li>can be used to find similar glycan sequences by alignment according to a glycan-specific substitution matrix</li>
</ul>
</li>
<li><strong>glycan_data</strong><ul>
<li>stores several glycan datasets and contains helper functions</li>
</ul>
</li>
<li><strong>ml</strong><ul>
<li>here are all the functions for training and using machine learning models, including train-test-split, getting glycan representations, etc.</li>
</ul>
</li>
<li><strong>motif</strong><ul>
<li>contains functions for processing glycan sequences, identifying motifs and featues, and analyzing them</li>
</ul>
</li>
</ul>
<p>Below are some examples of what you can do with glycowork, be sure to check out the full documentation for everything that's there.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">glycowork.motif.graph</span> <span class="kn">import</span> <span class="n">glycan_to_graph</span>
<span class="n">glycan_to_graph</span><span class="p">(</span><span class="s1">&#39;Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>([794, 1045, 450, 1045, 450, 281, 1015],
 [(0, 1, 2, 3, 5, 6), (1, 2, 3, 4, 6, 4)])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">glycowork.motif.graph</span> <span class="kn">import</span> <span class="n">compare_glycans</span>
<span class="nb">print</span><span class="p">(</span><span class="n">compare_glycans</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">compare_glycans</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>True
False
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">glycowork.motif.query</span> <span class="kn">import</span> <span class="n">get_insight</span>
<span class="n">get_insight</span><span class="p">(</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Let&#39;s get rolling! Give us a few moments to crunch some numbers.

This glycan occurs in the following species: [&#39;Antheraea_pernyi&#39;, &#39;Apis_mellifera&#39;, &#39;Autographa_californica_nucleopolyhedrovirus&#39;, &#39;AvianInfluenzaA_Virus&#39;, &#39;Bombyx_mori&#39;, &#39;Bos_taurus&#39;, &#39;Caenorhabditis_elegans&#39;, &#39;Drosophila_melanogaster&#39;, &#39;Homo_sapiens&#39;, &#39;HumanImmunoDeficiency_Virus&#39;, &#39;Mamestra_brassicae&#39;, &#39;Megathura_crenulata&#39;, &#39;Mus_musculus&#39;, &#39;Rattus_norvegicus&#39;, &#39;Spodoptera_frugiperda&#39;, &#39;Sus_scrofa&#39;, &#39;Trichinella_spiralis&#39;]

This glycan contains the following motifs: [&#39;Chitobiose&#39;, &#39;Trimannosylcore&#39;, &#39;core_fucose&#39;]

That&#39;s all we can do for you at this point!
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcOPN(b1-6)GlcOPN&#39;</span><span class="p">]</span>
<span class="kn">from</span> <span class="nn">glycowork.motif.annotate</span> <span class="kn">import</span> <span class="n">annotate_dataset</span>
<span class="n">annotate_dataset</span><span class="p">(</span><span class="n">glycans</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;known&#39;</span><span class="p">,</span> <span class="s1">&#39;graph&#39;</span><span class="p">,</span> <span class="s1">&#39;exhaustive&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>LewisX</th>
      <th>LewisY</th>
      <th>SialylLewisX</th>
      <th>SulfoSialylLewisX</th>
      <th>LewisA</th>
      <th>LewisB</th>
      <th>SialylLewisA</th>
      <th>SulfoLewisA</th>
      <th>H_type2</th>
      <th>H_type1</th>
      <th>...</th>
      <th>GlcNAcA*a1-4*Kdo</th>
      <th>GlcOPN*b1-6*GlcOPN</th>
      <th>Kdo*a2-4*Kdo</th>
      <th>Kdo*a2-5*Kdo</th>
      <th>Kdo*a2-6*GlcOPN</th>
      <th>Man*a1-2*Man</th>
      <th>Man*a1-3*Man</th>
      <th>Man*a1-6*Man</th>
      <th>Man*b1-4*GlcNAc</th>
      <th>Xyl*b1-2*Man</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcOPN(b1-6)GlcOPN</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 1236 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">glycowork.motif.analysis</span> <span class="kn">import</span> <span class="n">get_pvals_motifs</span>
<span class="n">glycans</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)[Kdo(a2-4)]Kdo(a2-6)GlcOPN(b1-6)GlcOPN&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;</span><span class="p">,</span>
           <span class="s1">&#39;Glc(b1-3)Glc(b1-3)Glc&#39;</span><span class="p">]</span>
<span class="n">label</span> <span class="o">=</span> <span class="p">[</span><span class="mf">3.234</span><span class="p">,</span> <span class="mf">2.423</span><span class="p">,</span> <span class="mf">0.733</span><span class="p">,</span> <span class="mf">3.102</span><span class="p">,</span> <span class="mf">0.108</span><span class="p">]</span>
<span class="n">test_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;glycan&#39;</span><span class="p">:</span><span class="n">glycans</span><span class="p">,</span> <span class="s1">&#39;binding&#39;</span><span class="p">:</span><span class="n">label</span><span class="p">})</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">get_pvals_motifs</span><span class="p">(</span><span class="n">test_df</span><span class="p">,</span> <span class="n">glycan_col_name</span> <span class="o">=</span> <span class="s1">&#39;glycan&#39;</span><span class="p">,</span> <span class="n">label_col_name</span> <span class="o">=</span> <span class="s1">&#39;binding&#39;</span><span class="p">)</span>
<span class="n">out</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;corr_pval&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>motif</th>
      <th>pval</th>
      <th>corr_pval</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1075</th>
      <td>Man*b1-4*GlcNAc</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1074</th>
      <td>Man*a1-6*Man</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1073</th>
      <td>Man*a1-3*Man</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>450</th>
      <td>GlcNAc</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1015</th>
      <td>a1-6</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>364</th>
      <td>GalNFoAN</td>
      <td>1.000000</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>365</th>
      <td>GalNOPCho</td>
      <td>1.000000</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>366</th>
      <td>GalNSuc</td>
      <td>1.000000</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>352</th>
      <td>GalNAcOAcOMeA</td>
      <td>1.000000</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1076</th>
      <td>Xyl*b1-2*Man</td>
      <td>0.211325</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
<p>1077 rows × 3 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

