---

title: ml


keywords: fastai
sidebar: home_sidebar

summary: "data processing and model training + analysis using machine learning"
description: "data processing and model training + analysis using machine learning"
nb_path: "02_ml.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_ml.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>c:\users\danie\appdata\local\programs\python\python39\lib\site-packages\xgboost-1.4.0rc1-py3.9-win-amd64.egg\xgboost\compat.py:31: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="/glycowork/ml.html"><code>ml</code></a> contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan representations. It currently contains the following modules:</p>
<ul>
<li><code>model_training</code> contains functions for training machine learning models</li>
<li><code>models</code> describes some examples for machine learning architectures applicable to glycans</li>
<li><code>processing</code> contains helper functions to prepare glycan data for model training</li>
<li><code>inference</code> can be used to analyze trained models, make predictions, or obtain glycan representations</li>
<li><code>train_test_split</code> contains various data split functions to get appropriate training and test sets</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="model_training">model_training<a class="anchor-link" href="#model_training"> </a></h2><blockquote><p>contains functions for training machine learning models</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="EarlyStopping" class="doc_header"><code>class</code> <code>EarlyStopping</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L19" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>EarlyStopping</code>(<strong><code>patience</code></strong>=<em><code>7</code></em>, <strong><code>verbose</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Early stops the training if validation loss doesn't improve after a given patience.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="train_model" class="doc_header"><code>train_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L78" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>train_model</code>(<strong><code>model</code></strong>, <strong><code>dataloaders</code></strong>, <strong><code>criterion</code></strong>, <strong><code>optimizer</code></strong>, <strong><code>scheduler</code></strong>, <strong><code>num_epochs</code></strong>=<em><code>25</code></em>, <strong><code>patience</code></strong>=<em><code>50</code></em>, <strong><code>mode</code></strong>=<em><code>'classification'</code></em>, <strong><code>mode2</code></strong>=<em><code>'multi'</code></em>)</p>
</blockquote>
<p>trains a deep learning model on predicting glycan properties</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr>
<td style="text-align:left">dataloaders (PyTorch object): dictionary of dataloader objects with keys 'train' and 'val'</td>
</tr>
<tr>
<td style="text-align:left">criterion (PyTorch object): PyTorch loss function</td>
</tr>
<tr>
<td style="text-align:left">optimizer (PyTorch object): PyTorch optimizer</td>
</tr>
<tr>
<td style="text-align:left">scheduler (PyTorch object): PyTorch learning rate decay</td>
</tr>
<tr>
<td style="text-align:left">num_epochs (int): number of epochs for training; default:25</td>
</tr>
<tr>
<td style="text-align:left">patience (int): number of epochs without improvement until early stop; default:50</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'classification', 'multilabel', or 'regression'; default:classification</td>
</tr>
<tr>
<td style="text-align:left">mode2 (string): further specifying classification into 'multi' or 'binary' classification;default:multi</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns the best model seen during training</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="training_setup" class="doc_header"><code>training_setup</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L318" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>training_setup</code>(<strong><code>model</code></strong>, <strong><code>lr</code></strong>, <strong><code>lr_patience</code></strong>=<em><code>4</code></em>, <strong><code>factor</code></strong>=<em><code>0.2</code></em>, <strong><code>weight_decay</code></strong>=<em><code>0.0001</code></em>, <strong><code>mode</code></strong>=<em><code>'multiclass'</code></em>)</p>
</blockquote>
<p>prepares optimizer, learning rate scheduler, and loss criterion for model training</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr>
<td style="text-align:left">lr (float): learning rate</td>
</tr>
<tr>
<td style="text-align:left">lr_patience (int): number of epochs without validation loss improvement before reducing the learning rate;default:4</td>
</tr>
<tr>
<td style="text-align:left">factor (float): factor by which learning rate is multiplied upon reduction</td>
</tr>
<tr>
<td style="text-align:left">weight_decay (float): regularization parameter for the optimizer; default:0.001</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'multiclass': classification with multiple classes, 'multilabel': predicting several labels at the same time, 'binary':binary classification, 'regression': regression; default:'multiclass'</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns optimizer, learning rate scheduler, and loss criterion objects</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="train_ml_model" class="doc_header"><code>train_ml_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L357" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>train_ml_model</code>(<strong><code>X_train</code></strong>, <strong><code>X_test</code></strong>, <strong><code>y_train</code></strong>, <strong><code>y_test</code></strong>, <strong><code>mode</code></strong>=<em><code>'classification'</code></em>, <strong><code>feature_calc</code></strong>=<em><code>False</code></em>, <strong><code>libr</code></strong>=<em><code>None</code></em>, <strong><code>return_features</code></strong>=<em><code>False</code></em>, <strong><code>feature_set</code></strong>=<em><code>['known', 'exhaustive']</code></em>, <strong><code>additional_features_train</code></strong>=<em><code>None</code></em>, <strong><code>additional_features_test</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>wrapper function to train standard machine learning models on glycans</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">X_train, X_test (list or dataframe): either lists of glycans (needs feature_calc = True) or motif dataframes such as from annotate_dataset</td>
</tr>
<tr>
<td style="text-align:left">y_train, y_test (list): lists of labels</td>
</tr>
<tr>
<td style="text-align:left">mode (string): 'classification' or 'regression'; default:'classification'</td>
</tr>
<tr>
<td style="text-align:left">feature_calc (bool): set to True for calculating motifs from glycans; default:False</td>
</tr>
<tr>
<td style="text-align:left">libr (list): sorted list of unique glycoletters observed in the glycans of our data; default:lib</td>
</tr>
<tr>
<td style="text-align:left">return_features (bool): whether to return calculated features; default:False</td>
</tr>
<tr>
<td style="text-align:left">feature_set (list): which feature set to use for annotations, add more to list to expand; default:['known','exhaustive']; options are: 'known' (hand-crafted glycan features), 'graph' (structural graph features of glycans), and 'exhaustive' (all mono- and disaccharide features)</td>
</tr>
<tr>
<td style="text-align:left">additional_features_train (dataframe): additional features (apart from glycans) to be used for training. Has to be of the same length as X_train; default:None</td>
</tr>
<tr>
<td style="text-align:left">additional_features_test (dataframe): additional features (apart from glycans) to be used for evaluation. Has to be of the same length as X_test; default:None</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns trained model                           </td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">vertebrate</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s1">&#39;Chordata&#39;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span><span class="o">==</span><span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">Phylum</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
<span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">general_split</span><span class="p">(</span><span class="n">df_species</span><span class="p">[</span><span class="n">df_species</span><span class="o">.</span><span class="n">Kingdom</span><span class="o">==</span><span class="s1">&#39;Animalia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">vertebrate</span><span class="p">)</span>
<span class="n">model_ft</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">X_test</span> <span class="o">=</span> <span class="n">train_ml_model</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span> <span class="n">feature_calc</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">feature_set</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;exhaustive&#39;</span><span class="p">],</span>
                         <span class="n">return_features</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>
Calculating Glycan Features...

Training model...
[22:17:59] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective &#39;binary:logistic&#39; was changed from &#39;error&#39; to &#39;logloss&#39;. Explicitly set eval_metric if you&#39;d like to restore the old behavior.

Evaluating model...
Accuracy of trained model on separate validation set: 0.9302481556002683
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="analyze_ml_model" class="doc_header"><code>analyze_ml_model</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L423" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>analyze_ml_model</code>(<strong><code>model</code></strong>)</p>
</blockquote>
<p>plots relevant features for model prediction</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (model object): trained machine learning model from train_ml_model</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">analyze_ml_model</span><span class="p">(</span><span class="n">model_ft</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_png output_subarea ">
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAzX0lEQVR4nO3dd7xcVb3+8c+T0EsIKCDSQYpKr1GKgKJXRWmKgD8EBASJCioqeAUBucoFxYugaOgoIk0BgSvkUqJIDb1L7x1poYfn98dak2yGc+ZMSGavfc75vl+veWX23jNnP2cnmTV7VdkmhBBCaJoRpQOEEEIIfYkCKoQQQiNFARVCCKGRooAKIYTQSFFAhRBCaKQooEIIITRSFFAhhBAaKQqoMChJul/SK5JeqjzePwN+5idmVMYuzre/pD/Udb5OJO0g6bLSOUKoigIqDGafsz1X5fFoyTCSZip5/ndrsOYOQ18UUGFIkTSPpGMlPSbpEUkHSRqZjy0t6WJJz0h6WtLJkkbnY78HFgP+mu/Gvi9pA0kPt/38KXdZ+Q7oDEl/kPQCsEOn83eR3ZJ2l3SXpBcl/SRnvlzSC5JOkzRLfu0Gkh6W9MP8u9wv6ctt1+EkSU9JekDSjySNyMd2kPRPSb+U9AxwKvBb4CP5d38uv+6zkq7P535I0v6Vn79Ezru9pAdzhv+sHB+Zs92Tf5drJS2ajy0vabykZyXdKWmrafpLDsNGFFBhqDkBeBP4ALAq8Elg53xMwM+A9wMfBBYF9gewvR3wIFPvyg7p8nybAmcAo4GTBzh/Nz4FrA6MAb4PjAP+X866ArBN5bXvA94LLAxsD4yTtFw+dgQwD7AU8DHgK8COlfeuDdwLLJh//m7AFfl3H51fMym/bzTwWeDrkjZry7susBzwcWA/SR/M+7+Ts34GGAV8FXhZ0pzAeOCPwALA1sBvJH2o+0sUhosooMJgdpak5/LjLEkLkj4Q97Q9yfaTwC9JH4LYvtv2eNuv2X4KOIz04T09rrB9lu23SB/E/Z6/S4fYfsH2rcAtwIW277X9PPC/pEKvat/8+0wAzgO2yndsWwP72H7R9v3AL4DtKu971PYRtt+0/UpfQWxfavtm22/Zvgk4hXderwNsv2L7RuBGYOW8f2fgR7bvdHKj7WeATYD7bR+fz309cCbwxWm4RmGYiLrnMJhtZvv/WhuS1gJmBh6T1No9AngoH18QOBxYD5g7H/v3dGZ4qPJ88U7n79ITleev9LH9vsr2v21Pqmw/QLo7fG/O8UDbsYX7yd0nSWsDB5Pu3GYBZgVOb3vZ45XnLwNz5eeLAvf08WMXB9ZuVSNmMwG/HyhPGH7iDioMJQ8BrwHvtT06P0bZ/nA+/lPAwIq2R5GqtlR5f/vU/pOAOVob+c5k/rbXVN8z0PlntHlzlVnLYsCjwNPAG6TCoHrskX5y97UNqRruHGBR2/OQ2qnUx+v68hCwdD/7J1Suz+hcrfj1Ln9uGEaigApDhu3HgAuBX0gaJWlE7mTQqpaaG3gJeF7SwsD32n7EE6Q2m5Z/AbPlzgIzAz8i3UW82/P3wgGSZpG0Hqn67HTbk4HTgP+SNLekxUltQp26tD8BLNLqhJHNDTxr+9V8d7rtNOQ6BviJpGWUrCTpPcC5wLKStpM0c36sWWm7CmGKKKDCUPMVUnXUbaTquzOAhfKxA4DVgOdJ7TV/bnvvz4Af5TatvXK7z+6kD9tHSHdUD9NZp/PPaI/nczxK6qCxm+078rFvkvLeC1xGuhs6rsPPuhi4FXhc0tN53+7AgZJeBPYjFXrdOiy//kLgBeBYYHbbL5I6jmydcz8O/DcdCv4wfCkWLAxh8JG0AfAH24sUjhJCz8QdVAghhEaKAiqEEEIjRRVfCCGERhpwHJSkBYB1SOMrXiENHpyYByaGEEIIPdHvHZSkDYG9gfmA64EngdmAZUnjG84AfmH7hXqi9ilu/0IIYXDrd2xdpwLqUOAI2w/2cWwm0piLkbbPnFEp34UooEIIYXCb9gJqkBjU4UMIIfRfQHXTBjUrsCWwRPX1tg+cEclCCCGEvnQzWezZpJH315LmGQshhBB6bsAqPkm32F6hpjzTKqr4QghhcHv3VXzA5ZJWtH3zDAzUE98/uq/Z/XvrkF36mrA5hBDC9OqmgFqXtJT1faQqPgG2vVJPk4UQQhjWuimgPt3zFCGEEEKbfgsoSaPyINwXa8wTQgghAJ3voP5IGox7LakzQvvKo0v19aYQQghhRui3gLK9Sf5zyfrihBBCCEk3bVBImhdYhjQXHwC2/96rUCGEEMKA60FJ2hn4O3ABacnsC4D9u3jfbJKulnSjpFslHZD3LynpKkl3SzpV0ix5/6x5++58fInp+L1CCCEMct0sWLgHsCbwgO0NgVWB57p432vARrZXBlYB/kPSGOC/gV/a/gDwb2Cn/PqdgH/n/b/MrwshhDBMdVNAvWr7VUh3ObbvAJYb6E1OXsqbM+eHgY1IS3UAnAhslp9vmrfJxz8uqd8RxiGEEIa2bgqohyWNBs4Cxks6G3igmx8uaaSkG0hrSY0H7gGes/1m62cDC+fnCwMPAeTjzwPv6eq3CCGEMOQMWEDZ3tz2c7b3B/YFjmXqXc9A751sexVgEWAtYPl3nTST9DVJEyVNHDdu3PT+uBBCCA3VVS8+AEkLA/flzWla7t32c5IuAT4CjJY0U75LWgR4JL/sEWBR0h3bTMA8wDN9/KxxQKtkisliQwhhiOr3DkrSPpL2q+y6AjgXuBD43kA/WNL8uWoQSbMDGwO3A5cAX8gv2560nAfAOXmbfPxiD/LVFEMIIbx7ne6gvgisV9l+xvaqkkYCE4CfDfCzFwJOzK8fAZxm+1xJtwF/knQQcD2pypD85+8l3Q08C2w97b9OCCGEoaJjFZ/tSZXNw/O+yfmOqCPbN5G6pLfvv5fUHtW+/1VSoRhCCCF07CQxl6SZWxu2T4ApS8CP6nGuEEIIw1ynAuoM4HeS5mjtkDQn8FumjmMKIYQQeqJTAbUvafzSg5KulXQtcD/wRD4WQggh9Eyn2cwnA3vnOfQ+kHffbfuVWpKFEEIY1jp1M18XwPYrtm/Oj1cqx0dJWqGOkCGEEIafTr34tpR0CPA30qKFT5GW2/gAsCGwOPDdnicMIYQwLHWq4vu2pPmALUndvxcCXiENtv2d7cvqiRhCCGE4Gmgc1LPA0fkRQggh1Kab2cxDCCGE2kUBFUIIoZGigAohhNBIAxZQkuaQtK+ko/P2MpI26X20EEIIw1k3d1DHA6+R1nKCtG7TQT1LFEIIIdBdAbW07UOANwBsvwyop6lCCCEMe90UUK/n5TUMIGlp0h1VCCGE0DPdLPn+Y9JsEotKOhlYB9ihl6FCCCGEAQso2+MlXQeMIVXt7WH76Z4nCyGEMKx104tvc+BN2+fZPhd4U9JmPU8WQghhWOumDerHtp9vbdh+jlTtF0IIIfRMNwVUX6/ppu0qhBBCeNe6KaAmSjpM0tL5cRhp+Y0QQgihZ7opoL4JvA6cmh+vAWN7GSqEEELophffJGDvGrKEEEIIUwxYQElaFtgLWKL6etsb9S5WCCGE4a6bzg6nA78FjgEm9zZOCCGEkHRTQL1p+6ieJwkhhBAquukk8VdJu0taSNJ8rUfPk4UQQhjWurmD2j7/+b3KPgNLzfg4IYQQQjLgHZTtJft4DFg4SVpU0iWSbpN0q6Q98v75JI2XdFf+c968X5J+JeluSTdJWm36f70QQgiDVVczQkhaAfgQMFtrn+2TBnjbm8B3bV8naW7gWknjSTOhX2T7YEl7k7qw/wD4NLBMfqwNHJX/DCGEMAx1M1nsj4Ej8mND4BDg8wO9z/Zjtq/Lz18EbgcWBjYFTswvOxHYLD/fFDjJyZXAaEkLTdNvE0IIYcjoppPEF4CPA4/b3hFYGZhnWk4iaQlgVeAqYEHbj+VDjwML5ucLAw9V3vZw3hdCCGEY6qaAesX2W6RlNkYBTwKLdnsCSXMBZwJ72n6hesy2ySv1TsPP+5qkiZImjhs3blreGkIIYRDppg1qoqTRwNGkSWJfAq7o5odLmplUOJ1s+8959xOSFrL9WK7CezLvf4S3F3yL5H1vY3sc0CqZpqlwCyGEMHh004tvd9vP2f4tsDGwfa7q60iSgGOB220fVjl0DlO7rm8PnF3Z/5Xcm28M8HylKjCEEMIw000niYtaz23fb/um6r4O1gG2AzaSdEN+fAY4GNhY0l3AJ/I2wPnAvcDdpLu13aftVwkhhDCU9FvFJ2k2YA7gvXmskvKhUXTRecH2ZZX3tPt4H683sYxHCCGErFMb1K7AnsD7SW1PrcLmBeDI3sYKIYQw3PVbQNk+XNKRwA9t/6TGTCGEEELnNijbk4EtasoSQgghTNHNOKiLJG2Ze+WFEEIIteimgNqVtGjh65JekPSipBcGelMIIYQwPQYcqGt77jqChBBCCFXdzmb+eWD9vHmp7XN7FymEEELobqDuwcAewG35sYekn/U6WAghhOGtmzuozwCr5AljkXQicD2wTy+DhRBCGN666SQBMLryfJqW2gghhBDejW7uoH4GXC/pEtJsEuuTVsENIYQQeqabXnynSLoUWJO0vMUPbD/e62AhhBCGt6568QEfAdYlFVAzAX/pWaIh5vtH31P7OQ/ZZenazxlCCDNaN734fgPsBtwM3ALsKunXvQ4WQghheOvmDmoj4IN5OYxWL75be5oqhBDCsNdNL767gcUq24vmfSGEEELPdHMHNTdwu6Sr8/aawERJ5wDY/nyvwoUQQhi+uimg9ut5ihBCCKFNN93MJwBIGlV9ve1ne5grhBDCMDdgASXpa8CBwKvAW6TBugaW6m20EEIIw1k3VXzfA1aw/XSvw4QQQggt3fTiuwd4uddBQgghhKpu7qD2AS6XdBXwWmun7W/1LFUIIYRhr5sC6nfAxaSZJN7qbZwQQggh6aaAmtn2d3qeJNQm5gcMIQwG3bRB/a+kr0laSNJ8rUfPk4UQQhjWurmD2ib/WV1BN7qZhxBC6KluBuouWUeQEEIIoarfAkrSFp3eaPvPMz5OCCGEkHS6g/pch2MGOhZQko4DNgGetL1C3jcfcCqwBHA/sJXtf0sScDjwGdKYqx1sX9fl7xBCCGEI6reAsr3jdP7sE4AjgZMq+/YGLrJ9sKS98/YPgE8Dy+TH2sBR+c8QQgjDVLdLvk8z23+XtETb7k2BDfLzE4FLSQXUpsBJeVHEKyWNlrSQ7cd6lS80S4mu7xDd30Nosm66mc9IC1YKnceBBfPzhYGHKq97OO97h9zlfaKkiePGjetd0hBCCEX17A5qILYtye/ifeOAVsk0ze8PIYQwOAx4ByVpDkn7Sjo6by8jaZN3eb4nJC2Uf85CwJN5/yOkpeRbFsn7QgghDFPdVPEdT5ok9iN5+xHgoHd5vnOA7fPz7YGzK/u/omQM8Hy0P4UQwvDWTQG1tO1DgDcAbL9MWrSwI0mnAFcAy0l6WNJOwMHAxpLuAj6RtwHOB+4F7gaOBnaf1l8khBDC0NJNG9TrkmYnt/dIWprKshv9sb1NP4c+3sdrDYztIksIIYRhopsCan/gb8Cikk4G1gF26GGmEEIIoau5+C6UdC0whlS1t0cs/x5CCKHXBiygJP0V+CNwju1JvY8UQgghdNdJ4ufAesBtks6Q9AVJs/U4VwghhGGumyq+CcAESSOBjYBdgOOAUT3OFkIIYRjraiaJ3Ivvc8CXgNVI8+iFEEIIPdNNG9RpwFqknnxHAhNsv9XrYCGUFhPYhlBWN3dQxwLb2J7c6zAhhM6i0AzDSacVdTeyfTEwJ7BpWlNwqlhRN4QQQi91uoP6GHAxfa+sO+CKuiGEEML06LSi7o/z0wNt31c9JmnJnqYKIYQw7HXTBnUmqede1RnA6jM+TghhsIl2sdArndqglgc+DMwjaYvKoVFADNQNIYTQU53uoJYDNgFG8/Z2qBdJg3VDCCGEnunUBnU2cLakj9i+osZMIYQQQldtUNdLGkuq7ptStWf7qz1LFUII06GJ7WKRaapu2w+7mSz298D7gE8BE4BFSNV8IYQQQs90U0B9wPa+wCTbJwKfBdbubawQQgjDXTcF1Bv5z+ckrQDMAyzQu0ghhBBCd21Q4yTNC+wLnAPMBezX01QhhBCGvW7WgzomP50ALNXbOCGEEELSaaDudzq90fZhMz5OCCGEkHS6g5q7thQhhBBCm04DdQ+oM0gIIYRQNWAvPknLSrpI0i15eyVJP+p9tBBCCMNZN93Mjwb2IXc3t30TsHUvQ4UQQgjdFFBz2L66bd+bvQgTQgghtHRTQD0taWnSKrpI+gLwWE9ThRBCGPa6KaDGAr8Dlpf0CLAnsFsvwkj6D0l3Srpb0t69OEcIIYTBoZuBuvcCn5A0J6lAe5nUBvXAjAwiaSTwa2Bj4GHgGknn2L5tRp4nhBDC4NDvHZSkUZL2kXSkpI1JBdP2wN3AVj3IshZwt+17bb8O/AnYtAfnCSGEMBjY7vMBnA2cAOwKnAZcSpruaJX+3jM9D+ALwDGV7e2AI/t43deAifnxtRl4/hn2syJTZIpMkSkyTf9D+UTvIOlm2yvm5yNJHSMWs/3q9BeLfZ7vC8B/2N45b28HrG37G704Xx/nn2h7jTrO1a3I1J3I1J3I1J3I1J06MnXqJNFaZgPbk4GHe1U4ZY8Ai1a2F8n7QgghDEOdOkmsLOmF/FzA7HlbgG2PmsFZrgGWkbQkqWDaGth2Bp8jhBDCINFpLr6RdQax/aakbwAXACOB42zfWmOEcTWeq1uRqTuRqTuRqTuRqTs9z9RvG1QIIYRQUjcDdUMIIYTaRQEVQgihkaKACiGE0EhRQDWMpCUlzVbZnl3SEgUjhRBCEcOugJJ0s6Sb+njcLOmm0vmA04G3KtuT877iJM0paTtJ5zUgy3hJoyvb80q6oGCkVo6FJX1U0vqtR+E8YyTNXdkeJWntwpk2lzRPZXu0pM0KRkLS2D7+Pe1eMM9ISXeUOn9/6r5Ow64Xn6TFOx23PUMnwZ1Wkm6wvUrbvhttr1wozyzAZ0lj0j4FnAn82fZfS+Sp5Lre9qoD7as5038DXwJuI32xgDRm8PMFM10PrOb8H13SCGCi7dUKZurr33jpv7smZjob+KbtB0tlaFf3dRpwNvOhpnQB1IWnJH3e9jkAkjYFnq47hKRPAtsAnwQuAU4C1rS9Y91Z+vGWpMVa/3nzF4/S37Y2A5az/VrhHFVy5Vuo7bcklf5/31fNTelMIyWpUpCPBGYpnGle4FZJVwOTWjtLfuGh5utU+h9FMZLGAEcAHyRd4JHApB7MkDGtdgNOlnQkadaOh0gT59btb8A/gHVt3wcg6fACOfrzn8BlkiaQrtN6pImNS7oXmBloUgF1r6RvAUfl7d1JOUuaKOkw0vI6AN8Ari2YB9K/91Ml/S5v75r3lbRv4fP3pdbrNOyq+FokTSRNp3Q6sAbwFWBZ2/sUDZZJmgvA9kuS1rR9Tc3nX4V0fb5I+kD7E7Cf7Y5VpHWS9F5gTN68Enje9hsd3tLrPGcCKwMXUSmkbH+rYKYFgF8BG5HuMC8C9rD9VMFMc5I+fD+Rd10IHGT75YKZRpBWSmhlGg8cbfut/t/Ve7lmYBnb/ydpDmCk7RcL5qn1Og3rAsr2GpJusr1S3le0zrlK0odIVWxbkz54i81kLOmjOcuWwI3AX2w3YuoVSSJ9+G4LbGJ7wYJZtu9rv+0T687SH0mzk65TIzreAEhaDPiS7UNLZ2mRtB6wte2xBTPsQioM5rO9tKRlgN/a/nipTO16fZ2GbRUf8HLuAHCjpENIy4kU7dWYu5Nvkx9vAIsDa9i+v2AsbF8OXC5pD9I3py9ReG6wXEW7LandZz5gLLBXyUxNKoiqcjvBp0j/rjYGLqNwz1BJ85PuzrcB3g/8pWQeAEmrkvJsBdwH/LlsIsaSFnK9CsD2XfmOuKg6r9NwLqC2IxVIY4Fvk5b32LJUGElXAKNIVWlb5n+M95UunCRtDlxs+/ncwH41MEfBPD8lfbA9CJwCHEDqlVa8cMjfcH8GfAiYMpbN9lKF8nyMVIh/BrgaWAdYqlRVWu7uvkXOtCzpg21J24uUyJMzLcvUL4VPA6eSapY2LJWp4jXbr6dKAsidW4pUeZW6TsNxHNSmksbafiCvbzUe2AHYHFilYLQngLmBBYH5874m1L/+2PbzrQ3bzwE/LheHnUnX6ijg97afoRnXCeB4Uq43gQ1JPR//UCKIpIdJheVlwIdsbwm8UrKdB3gS+CpwEKmg/C7wesE8AHeQqog3sb2u7SOYOkSgtAmSfkha6mhj0l1vqeEdRa7TsCuggO8D51S2ZwVWBzYAvl4iEIDtzYAVSb2Z9pd0HzCvpLVKZcqa1iV4IdIH3OeAeyT9nvQfuAm1AbPbvoj0zfIB2/uTxpCVcAap6uxLwOdyx4TSBfk+pP9vvwH2kbR04TyQ7ugeAy6RdLSkj5N6hTbB3sBTwM2k3nLnAz8qlKXIdRp2nSQkXWN7zcr2ka1l5SVdaXtM/++uT65r/hKpk8Rithcd4C29ynEc8BxTuwSPJTXa7lAiT5WkWYFNSNUO6wEX2S62yKWky4F1SYXDxaSFNw+2vVyhPCJ98dqGVM03D7ATcL7tl0pkyrmWIv273gZYhnRH/hfb/yqYaU5g05xpI9Ld719sX1ggy/zA/LZva9v/YeDJBvTArO06DccC6m7bH+jn2D22m/Ct7m0kLV5qgHEfXYLHk7oET+r/XfXL7Rub2z6pYIY1gduB0cBPSAXCIbavLJWpRdLMwH+QCoZP2X5v4UgASFqB1Ca1VX//L+smaV5SO+fWtjcqcP4/Ab+x/fe2/esBXy/5Jayqjus0HAuok4FLbR/dtn9XYAPb2xTKdTz9V8HY9k515hksJK1L6ul0s+3xpfMMBpJmt/1K6Ryhb60hMP0cu8X2CnVnKqUJ9fZ1+zZwlqRtgevyvtVJdeOblQoFnNvHvkVJeUfWnAVJf6VDm0Wp6VYkXW17rfx8F1KV419I7Xar2z64QKZzOh0vca0kXUKHLzxA7WNpcrtqNZMq2y5ReyHpxUqGVpuKSZ+Ns9gu8Rk5d4djM9eWoqLUdRp2BZTtJ4GPStoI+HDefZ7tiwvGwvaZree5jv6HwPrAwcCxBSL9vI997f9AS6j+B/0asLHtpyT9nDSbRO0FFPAR0pRUp5DGrDShkb2vMWFjSJ2Enqw5S0v7XcEI0liavYDr648Dtt9WGOQZXMaSOiWUGpt1t6TP2D6/ulPSpyk0TVWp6zTsCqiWXCAVLZTaSVqe1EtnVeBQYDfbbxaKMxpYxPavc7arSd3fDfygUCaAEbnuewSpivopANuTJJW6Vu8jDYDdhtSech5wiu1bC+XB9pS57fJ4qH1JY7N2s/2/hTI9k/OMII1D/B5wA/DZ9g4BdVNaQmJP0pRnfyRNjPxMoTh7AudJ2oqpcxSuQfoitEmhTED912nYFlBNI+l0UlXjL0jVepOBUa1BerafrTnS90kN6i2zkP6TzEka71NqJoJ5SP9pBVjSQrYfy9/oity52J5MmjDzb7ln4TbApZIOsH1kiUwAkj5F+sLzGvBfti8plSXnmZk0DurbpPFZm9m+u3Cm9wLfJfWYPQ5YtTrur4Q8SH9F0pedVnvTBGDXPHazdqWu07DrJNFUku6nUh/f2t3arns2gsHSHb9FaSLNBZ1nXi9w/llJY562AZYgjbU7zvYjhfJcQ7rjPRS4ov247eve8abeZ3qYNIj5f0gzgbRnqn1qIUmTSGONjgfeMQmr7cPqztREpa5T3EE1hO0lSmdoM291o1U4ZfNTWJ5gtF2RGQAknUT6pns+cIDtW0rkaDMJeAn4Qn5UmTSGpW7/l8+9cn5UmTJz3x3K1C+EnTonNIKkm22vWODURa5T3EE1THXuu7w9mtT9/ayaczSyO34lx82k/zAita0sCdxp+8Md39ibLG8xdUG5d/RSc/k1xsIgImmL/g6RZjMv/gWxLlFANYwasvR0nsniLFL7xTu649t+os48A5G0GrC77Z1LZ2kSSWOBk/Mciq3BldvY/k3BTD8lDWCuZvqu7VLT+CDpRNI6WdVMv7D91QJZ3gBOpu9hAl9o71FXp7qvUxRQDaPK+lSVfaVu62nrjn9r6e74nZS8Tk3VlC88A51f0nW2V2tYpiLXSdK1wPZ9VRVLesiFpj3L56/1OkUbVPO0L4c9loLLYTexOz6ApO9UNkcAqwGPForTZCMlyfmbqNLaULM0INOstl/LmWYn3ZmXNELSvLb/nTPNR7nPxz2BF/o5tnmNOfpS63WKAqp5vkkas3Jq3h5PKqTC21WrOd4kjT06s5/XDmd/A06V9Lu8vWveV9LJwEV5ei+AHYHS63n9ArgiD/cQqWPJf5UIYvsfHY5NrDNLH2q9TlHFFwY1SXO47BpHjZYHxe7K1KmNxgPH5LFbxeRZEaZksn1ByTwAkj7E1N6NFzdg8HBj2sXactV2naKAaoimzn3XVJI+QpoCai7bi0lamTSQcffC0cIgkquo+lVggPwUDWsXK3KdooqvOfqa+y7073+AT5EXn7R9o6T1iyZqkEo3/D61d8SpQ9uEo287RLnu+NcydbgCvH2QvIFaB8i3aVK7WJHrFAVUQ9ieUDrDYGP7odZUUFlTlupugqJztvWlZPfo/thesnSGDprULlbkOkUB1RCSNuXtk7NexdQZG75v+4xi4ZrpIUkfJc3HNzOwB2mxwAC40AKX00LSwkxdSubREhMj516NszuvMCxpDFN7OV5v+x3T+tTF9kmSJjK1vWeLUu1ipa5TtEE1hKR/klamfChv30BqRJ4TON527ev3NFmevPJw0kq/Ai4kNSiXmoG6USTtBMxn+9C8/Qip56OA79n+bYFM+wAz2z4wbz8IPE9aQuVE2z8rkOnnpGXUD8nb9wG3kGYnuc527TP3N7FdrNR1ijuo5pilVThll+UP22eUll0PFbafBr5cOkeD7UZa4r3lSdsLS5oNuACovYAiLQ++XmX7Gdur5m/nE4DaCyjSl8A1K9vP2f6cUt1xv929e6yJ7WJFrlMUUM3R6MlZm0LSfh0O2/ZPagvTbGq7mzwdwPareWBsEbYnVTYPz/smF8w0oq1q8Qc5k/MSLrVraLtYkes0olc/OEyzq5SWMH+bPDnr1QXyNNWkPh4AO1F2IcWmGV3dsP1TmDIu6r0lAgFz5fbCVqYTcqZZgVIT6s4iaUrnDdsX5kzzkKqvaidpZPVDX9IYSevnR6mOJkWuU7RBNcRgm5y1CfJ/mD1IhdNppEGMpZYzbxRJvwGebZ+AVdJBwHtt71Yg009Jqw9/ozW4OldfHwk8bnufApm+Q2rH3M32g3nf4sBRpEGotQ//aGi7WJHrFAVUwwymyVlLyY3I3yG1QZ0IHN4aKxKS/MF/DKnd4Ma8e2VgIrBzqzdWzZlGkrpJ7ww8QGpTWZQ04PpHJXrx5Vy7AT8kdUgSaUG+g20fVSjP9aSl1N9sbee2OgH/sL1uoVy1X6cooBooT2nyfuAV4H7bbxWO1BiSDgW2AMYBvy7xQTuYSFqKqV94brN9T8k8MGVy2A/kzbttv1IyT0urCqtk1/Kc40bbK1e2P1mpUnvH7PR1q/M6RQHVELkudyxpyfBZgCeB2YEFgSuB39i+pFzCZlBaHPA10gSxsTjgAHKb08pM/cJzS+lq0FydvU41EzCx5BcxSWuQehhWM40vcWcu6XZgrfYCIH9GXGV7+bozVTLUep2igGoISeOBk4C/tiaHrBxbHdgOuNn2sQXihUFG0tKkTiOfAO4CniK1YSwLvAz8jjT2qLZCQdKGwN7AfMD1pC9hrUxLA2eQ2hH7W2qiF5l2JK0gcB+pe3c10zqkD+B9W+0uNWVqYrtYkesU3cwbwvbGHY5dS8E1ocKgdBDpA21Xt30LzXcw25K+9NS5zMVngF36+hCTNBNpeqaNqXfZlDmAdfqrZpS0CrAMUFsBZfswSS8Dl+W2xOLtYhS6TnEH1WCSfmr7h6VzhBDKaEq7WClxB9UQkn7VvgvYrjUewva36k8VhhJJJ9n+SsHzizSbhEnVeRsBmwJ3AL8t0QYlaXNggu1nJc1PmqB1VeA24Lu2H647UyXblPYeSUXbxfoi6V+2l+3pOeIOqhkkPUSa7uVCpk5x8nNgLwDbpVccDYOIpHPadwEbAhdDmfXF8tisBUidgF4gjfE7B/gs8ITtPQpkus32h/LzU0kdkk4ntQF9uVPVew8zNbFdrLpUSuvzaQ5Se2bPOidFAdUQ+Vb+J6T/wHvZflTSvbZLrkcTBilJ15HuAo5h6rxupwBbQ5nlXSTdbHvFPJvE48BCtl/P7U/XucwaVXfaXi4/v9b26pVjRbp0SxoLHDdAe897bF9UY6ZfkWYn+V5r0gBJ9/V6WqaY6qghbL9oe09SFcPJkvYi/n7Cu7cG6dv3fwLP274UeMX2hBKFU/YmgO03gGtsv5633wRKdTG/VNKBeWzWpbnKr9Xj8PkSgWz/utPYMNs31Fk45XN+izR34imSvpWHL/T87iY+ABsm99jbiDTG4LLCccIgZfst278EdgT+U9KRlG9zfrzSpjplpnVJ7wNeL5TpG6TC8U5S+9iZuTprF1Ivx0aQ9K/SGfJn0yfy5gRqmKswqvgaTNImts8tnSMMfpI+S+om3Lheobkr9ZwNGEA8DzCTC68pVqq9Z1pIWghY1fb5vTxP3EE124GlA4ShwfZ5wP2lc7STtL/tSaULJwDbz9t+RtL+haMcT5o4ehnbc9ueG3gwPy9eOAHYfgzYrNfniQKq2TTwS0LoWu0zmHeh9t6EXSiaqVR7z7uwRq9PEAVUs+1aOkAYUpr4hScy9aFEe8+70PO73miDGgQkbWx7fOkcYXCTtEjJgad9kTSiabP1Ny1TXe09TRR3UINDTBAbplurcMoDQRuhVRBI2q90lpamZaqrvefdkjSuZz877qCaoY+R/1MOARvZnrPOPGHokvSg7cVK56iKTJ1Jus72agXPP19/h4AbbS/Si/OWHhcRploP+H9A+wJ8AtaqP04YzCTd1N8h0hpjtZPU3zIaIq19VrsmZupH6V6OTzF1FeSW1gwlC/TqpFFANceVwMt9jfKXdGeBPGFwWxD4FNA+saiAy+uPA8BzpKXMn2g/kOeiLOE5mpfpHaoDmwu5F/h4P0ul9Ow6RQHVELY/3eHY+nVmCUPCucBctm9oPyDp0trTJCcBiwPvKAyAP9acpaWJmfolaZztrxU49f8A89L3ek+H9Oqk0QbVQHnql7VIt9DX2H68cKQQQk1Ktfc0URRQDSNpZ2A/0rIIAj4GHGj7uKLBwqAlaTVgXdIXnn/avq5wJCTNS1qBdcr4Htt/L5eoOZkkTab/9p6Fbc9Sd6YWSbMBuzP139NlwFG2X+3J+aKAapbc3vTR1nxgkt4DXN5aEiCEaZG7Sn8R+HPetRlwuu2DCmbaGdgDWAS4ARgDXGF7o8gEku6iQ3uP7UXrzlQ5/2mk5ef/kHdtC4y2/cVenC/aoJrnGdI/gJYX874Q3o0vAyu3vuFKOpj0AVysgCIVBGsCV9reUNLywE8L5mlapv+hQHtPl1ZoLfCYXSLptl6dLAqohpD0nfz0buAqSWeTbqE3BfrrMhzCQB4lVVm1qmBmBR4pFweAV22/KglJs9q+Q1LpGoLGZLL96w7HjqgzSx+ukzTG9pUAktYGJvbqZFFANcfc+c978qPl7AJZwiAn6QjSF5zngVsljc/bGwNXl8wGPCxpNGnG7vGS/k1qcympcZnqbu8ZIMvNOcPMwOWSHszbiwN39Oy80QYVwtAjaftOx22fWFeWTiR9DJgH+Ftrhd3SmpKp7vaeAbIs3um47Z4U5lFANYSko4HDbd/Sx7E5gS8Br9k+ufZwIcwAkuay3T5TyjS/Zqhnqpz3trb2nj731ZSlyHWKyWKb49fAfpJul3S6pN9IOk7SP0gj/+cGzigbMQwWkv4q6XOSZu7j2FKSDpT01ZpjnS3pF5LWz1+6qnl2knQBUPeMCU3M1HKdpDGVTD1t7xlAkesUd1ANI2ku0kJgCwGvALfbjqmOwjTJg72/A2wJPEuaS202YElSR5wjbdfevinpM6SeheuQeqq9CdwJnAccW2JQetMytbX3LEfqzTelvafEHVTOVb1O8wFv0OPrFAVUCEOcpCWY+oXnX7ZfLpsodFKqvaeJooBqiMq3pnccAmx7pZojhSFA0pLAY5VxULMB77N9f8FMmwMX234+b48GNrB9VmRqdrtYPvfCpLu5Kb3AezXjRhRQDRHfmkIvSJpImpnk9bw9C2m6ozULZrrB9ipt+663vWqhSI3KJOki0mDqs4FrbU/K+5cCNgS2Ao62XXubtKT/JnXYug2YnHfb9ud7cb4YB9UQUQCFHpmp2lXa9uu5kCqpr85ZpT+LGpPJ9sdze8+uwDp58thqe8/2BSeQ3gxYzvZrdZys9D+K0Cb32jkC+CAwCzASmGR7VNFgYbB6StLnbZ8DIGlT4OnCmSZKOozUcxVgLHBtwTzQsEy2zwfOL3X+Du4ldd6opYCKKr6GyVUyWwOnk3rzfQVY1vY+RYOFQUnS0sDJwPtJ7ZkPAdvZvqfjG3ubaU5gX+ATedd44KBWVVZkmqrO9p4u85wJrAxcRKWQsv2tnpwvCqhmkTTR9hqSbmp1jChdPx8Gvzx8AdsvSVrT9jWlM4XO6m7v6TJTnzOU9Gpmkqjia56XcxvBjZIOAR4jBlSH6bcYsI2krUnz861RKoikS+ijx2rh5TYal4ma23u6UfcUWVFANc92pAJpLPBt0vo0WxZNFAalPP5pm/x4g1RVtEbJLubZXpXns5H+fb9ZKEtLEzPV2t7TDUnLAD8DPsTbF3ZcqhfniwKqIXLj9SKtqfYlTQAWIH2ru4I0+j+Erki6AhgF/AnY0vZdku5rQOGE7fbOB/+UVHSG9SZmAl4Gbsjdznve3tOl44EfA78kdXnfkR7W8EQB1RzfJ3WOaJkVWB2Yi/SPIubhC9PiCWBhYEFgfuAu+h4IXrvcbbplBOnf+TyF4gDNzASckx9NMrvtiyQpD43ZX9K1wH69OFkUUM0xi+2HKtuX2X4WeLY6OWMI3bC9maR5gC1IHyLLAKMlrWW79J1B9W7lTeA+YKdCWVoal6kpS6K0eU3SCOAuSd8gLX45V69OFr34GkLS3bY/0M+xe2wvXXemMHRIWoDUI2xrYDHbixbIsJjtvpYxL6aJmVrqbu/pMtOawO3AaOAnpLvMQ1or7M5o0TusOa6StEv7Tkm7Un4F1DDI2X7S9hG21yGt0FrCWa0neTxNE5zVetKgTC3HA0eR7ug2BE5i6uKFRdi+xvZLth+2vaPtLXpVOEFU8TXJt4GzJG0LXJf3rU5qi9qsVKgwOEk6nv7bnEyZ6itVnhe7C2jTxEwttbb3dCKpY1tYzMU3xNl+EviopI2AD+fd59m+uGCsMHid28e+RUlfhEbWnKXF/TwvqYmZWmpt7xnAR0izkJwCXMXbC/aeiTaoEIa4PAv2D4H1Sd2Dj61OIFtjjsnAJNKH2+ykbtQwdUmZ2uebbGKmSrZa23sGyDIS2Jg0pm4l0qS1p9i+tafnjQIqhKFJ0vLAj4BVgUOBP9guPfg0DHKSZiUVVIcCB9g+smfnigIqhKFH0umkNsxfAKcxdS43APIQhtBApdp7BpILps+SCqclSGO0jrP9SM/OGQVUCEOPpPuZ2qbS+rPVbuCSXZVDZ5KeokN7j+0JBTKdBKxAWgLkT7ZvqeW8UUCFEEJzlGrvGSDTW6S2Onh7Z5KettXFOKgQhjBJm+cZJVrboyVtVjBSGIDtybb/Znt7YAxpHs5Lc0++UplG2J47P0ZVHnP3siNJ3EGFMIRJusH2Km37Yn2xhivR3tNEMQ4qhKGtr1qS+H/fYG3tPQfU1d7TRHEHFcIQJuk44Dng13nXWGA+2zuUyhQ6K9Xe00RRQIUwhOWZ8PcFPpF3jQcOsj2p/3eF0AxRQIUQQmikqIsOYQiS9Fc6zC1XarBnCNMiCqgQhqaf97GvfcBuCI0WBVQIQ9NoYBHbvwaQdDVp6XcDPyiYK4SuxUDdEIam75PGzrTMAqwBbADsViJQCNMq7qBCGJpmsf1QZfsy288Az+SefSE0XtxBhTA0zVvdsF2dJmf+mrOE8K5EARXC0HSVpF3ad0raFbi6QJ4QplmMgwphCJK0AHAW8BpwXd69OjArsJntJwpFC6FrUUCFMIRJ2gj4cN681fbFJfOEMC2igAohhNBI0QYVQgihkaKACiGE0EhRQIUhTdJLNZ9vCUnbTufP2FPSHH3s/7Gkn7XtW0XS7dPwsz8vae8BXrO/pL362L+EpGG7NlGoXxRQIcwgkmYirX46XQUUsCfwjgIKOAX4Utu+rfP+AUmayfY5tg+evngh1CMKqDAsSNpA0gRJZ0u6V9LBkr4s6WpJN0taOr/uBEm/lTRR0r8kbZL3zybp+Pza6yVtmPfvIOkcSRcDFwEHA+tJukHSt/Ndxz8kXZcfH63kuVTSGZLukHSykm8B7wcukXRJ9Xew/S/g35LWruzeCjhF0i6SrpF0o6QzW3dgld/nKuCQnPfIfOxzkq7Kv8//SVqw8nNXlnSFpLv6GU81UtKh+Zw35fFVSFpI0t/z73+LpPWm/28vDFcx1VEYTlYGPgg8C9wLHGN7LUl7AN8k3blAugtaC1iaVFB8gLQSrW2vKGl54EJJy+bXrwasZPtZSRsAe9luFWxzABvbflXSMqS7nTXy+1YldQF/FPgnsI7tX0n6DrCh7af7+B1OId01XSVpDPCs7bskPWv76HzOg4CdgCPyexYBPmp7sqQdKj/rMmCMbUvamTR/33fzsZWAMcCcwPWSzmvLsRPwvO01Jc0K/FPShcAWwAW2/0vSSPq+EwyhK1FAheHkGtuPAUi6B7gw778Z2LDyutNsvwXcJeleYHlgXfIHvu07JD0AtAqo8baf7eecMwNHSloFmFx5D8DVth/OeW4gFYyXDfA7nApcLum7vL16b4VcMI0G5gIuqLzndNuT+/hZiwCnSlqINJnsfZVjZ9t+BXgl38mtBdxQOf5JYCVJX8jb8wDLANcAx0maGTjLdvU9IUyTqOILw8lrledvVbbf4u1f1toHBw40WLDT8unfBp4g3b2tQSoI+sozmS6+MOYJYO8DPgZsSSqwAE4AvmF7ReAAYLYu8h0BHJnfs2vbewa6BgK+aXuV/FjS9oW2/w6sDzwCnCDpKwP9TiH0JwqoEN7pi5JG5HappYA7gX8AXwbIVXuL5f3tXgTmrmzPAzyW78i2A0Z2cf72n9HuFOCXwL2tO7D8+sfyncuXuzhHK9sj+fn2bcc2ze1u7yEt0XFN2/ELgK/n8yFpWUlzSloceCJXNx5Dqv4M4V2JAiqEd3qQNKHq/wK72X4V+A0wQtLNpLuWHWy/1sd7bwIm584K387v217SjaSqwk53Wy3jgL+1d5KoOJ3UdlXtvbcvcBWpLeuOLs4BsD9wuqRrgfb2rpuAS4ArgZ/YfrTt+DHAbcB1uev570h3gBsAN0q6ntTj8PAus4TwDjHVUQgVkk4AzrV9RuksIQx3cQcVQgihkeIOKoQQQiPFHVQIIYRGigIqhBBCI0UBFUIIoZGigAohhNBIUUCFEEJopP8PHLAiptZA6S8AAAAASUVORK5CYII=
"
>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_mismatch" class="doc_header"><code>get_mismatch</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/model_training.py#L446" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_mismatch</code>(<strong><code>model</code></strong>, <strong><code>X_test</code></strong>, <strong><code>y_test</code></strong>, <strong><code>n</code></strong>=<em><code>10</code></em>)</p>
</blockquote>
<p>analyzes misclassifications of trained machine learning model</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">model (model object): trained machine learning model from train_ml_model</td>
</tr>
<tr>
<td style="text-align:left">X_test (dataframe): motif dataframe used for validating model</td>
</tr>
<tr>
<td style="text-align:left">y_test (list): list of labels</td>
</tr>
<tr>
<td style="text-align:left">n (int): number of returned misclassifications; default:10</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns tuples of misclassifications and their predicted probability</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_mismatch</span><span class="p">(</span><span class="n">model_ft</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_test</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;Man(a1-3)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, 0.4353863000869751),
 (&#39;GalNAc(b1-3)Gal&#39;, 0.8447189331054688),
 (&#39;Man(a1-3)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, 0.4353863000869751),
 (&#39;Fuc(a1-3)[GalNAc(b1-4)]GlcNAc(b1-2)Man(a1-3)[Man(a1-z)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;,
  0.3797847330570221),
 (&#39;Man(a1-6)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, 0.4353863000869751),
 (&#39;Gal(a1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;,
  0.9270392656326294),
 (&#39;Man(a1-6)[Xyl(b1-2)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, 0.13527607917785645),
 (&#39;Fuc2S4S(a1-3)Fuc2S4S(a1-3)Fuc2S4S&#39;, 0.6127856373786926),
 (&#39;Man(a1-3)[Gal(b1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, 0.544141948223114),
 (&#39;Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)GalNAc&#39;, 0.9209876656532288)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="models">models<a class="anchor-link" href="#models"> </a></h2><blockquote><p>describes some examples for machine learning architectures applicable to glycans. The main portal is prep_models which allows users to setup (trained) models by their string names</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(SweetNet)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(LectinOracle)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(LectinOracle_flex)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(NSequonPred)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(init_weights)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(prep_model)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="processing">processing<a class="anchor-link" href="#processing"> </a></h2><blockquote><p>contains helper functions to prepare glycan data for model training</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(dataset_to_graphs)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(dataset_to_dataloader)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(split_data_to_train)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="inference">inference<a class="anchor-link" href="#inference"> </a></h2><blockquote><p>can be used to analyze trained models, make predictions, or obtain glycan representations</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#show_doc(glycans_to_emb)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In order to run <code>get_esm1b_representations</code>, you first have to run this snippet:</p>
<p><code>!pip install fair-esm
import esm
model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="train_test_split">train_test_split<a class="anchor-link" href="#train_test_split"> </a></h2><blockquote><p>contains various data split functions to get appropriate training and test sets</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hierarchy_filter" class="doc_header"><code>hierarchy_filter</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L35" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hierarchy_filter</code>(<strong><code>df_in</code></strong>, <strong><code>rank</code></strong>=<em><code>'Domain'</code></em>, <strong><code>min_seq</code></strong>=<em><code>5</code></em>, <strong><code>wildcard_seed</code></strong>=<em><code>False</code></em>, <strong><code>wildcard_list</code></strong>=<em><code>None</code></em>, <strong><code>wildcard_name</code></strong>=<em><code>None</code></em>, <strong><code>r</code></strong>=<em><code>0.1</code></em>, <strong><code>col</code></strong>=<em><code>'target'</code></em>)</p>
</blockquote>
<p>stratified data split in train/test at the taxonomic level, removing duplicate glycans and infrequent classes</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df_in (dataframe): dataframe of glycan sequences and taxonomic labels</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which rank should be filtered; default:'domain'</td>
</tr>
<tr>
<td style="text-align:left">min_seq (int): how many glycans need to be present in class to keep it; default:5</td>
</tr>
<tr>
<td style="text-align:left">wildcard_seed (bool): set to True if you want to seed wildcard glycoletters; default:False</td>
</tr>
<tr>
<td style="text-align:left">wildcard_list (list): list which glycoletters a wildcard encompasses</td>
</tr>
<tr>
<td style="text-align:left">wildcard_name (string): how the wildcard should be named in the IUPAC-condensed nomenclature</td>
</tr>
<tr>
<td style="text-align:left">r (float): rate of replacement, default:0.1 or 10%</td>
</tr>
<tr>
<td style="text-align:left">col (string): column name for glycan sequences; default:target</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns train_x, val_x (lists of glycans (strings) after stratified shuffle split)</td>
</tr>
<tr>
<td style="text-align:left">train_y, val_y (lists of taxonomic labels (mapped integers))</td>
</tr>
<tr>
<td style="text-align:left">id_val (taxonomic labels in text form (strings))</td>
</tr>
<tr>
<td style="text-align:left">class_list (list of unique taxonomic classes (strings))</td>
</tr>
<tr>
<td style="text-align:left">class_converter (dictionary to map mapped integers back to text labels)</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_x</span><span class="p">,</span> <span class="n">val_x</span><span class="p">,</span> <span class="n">train_y</span><span class="p">,</span> <span class="n">val_y</span><span class="p">,</span> <span class="n">id_val</span><span class="p">,</span> <span class="n">class_list</span><span class="p">,</span> <span class="n">class_converter</span> <span class="o">=</span> <span class="n">hierarchy_filter</span><span class="p">(</span><span class="n">df_species</span><span class="p">,</span>
                                                                                       <span class="n">rank</span> <span class="o">=</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">train_x</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[&#39;Glc(b1-6)GlcNAc(a1-3)FucNAc(a1-3)GlcNAc(b1-2)Glc&#39;, &#39;Gal(b1-4)GlcNAc(b1-4)Man(a1-3)[Gal(b1-4)GlcNAc(b1-6)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;, &#39;D-Ara(a1-2)Gal(b1-3)Gal(b1-4)Man&#39;, &#39;GalfOAcOLac(b1-3)Glc(b1-6)GalfOAcOLac&#39;, &#39;Neu5Ac(a2-3)Gal(b1-4)GlcNAcOS(b1-6)[Gal(b1-3)]GalNAc&#39;, &#39;Gal(z1-z)GlcNAc(z1-z)Fuc&#39;, &#39;Gal4S(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc&#39;, &#39;GlcA(a1-2)[GlcN(a1-7)]LDManHep(a1-3)[GlcN(a1-4)[GalA6N(a1-6)]Glc(b1-4)]LDManHep(a1-5)KdoOPEtN&#39;, &#39;GalNAcAOAcN(a1-4)GalNFoAN(a1-3)QuiNAc(a1-3)Rha(a1-4)GalNAcAOAcN&#39;, &#39;Man(a1-2)Man(a1-3)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="general_split" class="doc_header"><code>general_split</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L108" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>general_split</code>(<strong><code>glycans</code></strong>, <strong><code>labels</code></strong>, <strong><code>test_size</code></strong>=<em><code>0.2</code></em>)</p>
</blockquote>
<p>splits glycans and labels into train / test sets</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr>
<td style="text-align:left">labels (list): list of labels used for prediction</td>
</tr>
<tr>
<td style="text-align:left">test_size (float): % size of test set; default:0.2 / 20%</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Returns X_train, X_test, y_train, y_test</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_x</span><span class="p">,</span> <span class="n">val_x</span><span class="p">,</span> <span class="n">train_y</span><span class="p">,</span> <span class="n">val_y</span> <span class="o">=</span> <span class="n">general_split</span><span class="p">(</span><span class="n">df_species</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                                              <span class="n">df_species</span><span class="o">.</span><span class="n">Species</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">train_x</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[&#39;Rha(a1-3)ManNAc(b1-4)Rha&#39;, &#39;Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)DDManHep(a1-6)Glc(b1-4)[LDManHep(a1-2)LDManHep(a1-3)]LDManHep(a1-5)Kdo&#39;, &#39;Rha(a1-3)Rha(b1-4)[Rha(a1-2)]Rha(b1-4)Rha(a1-3)Rha&#39;, &#39;Fuc(a1-2)Gal(b1-z)GlcNAc(b1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-6)]Gal(b1-z)GlcNAc(b1-3)Gal(b1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-6)]GalNAc&#39;, &#39;Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc&#39;, &#39;Glc6Ac(b1-2)Glc1Ole6Ac&#39;, &#39;Araf(a1-2)Xyl(a1-6)Glc(b1-4)[Araf(a1-2)Xyl(a1-6)]Glc(b1-4)Glc(b1-4)Glc-ol&#39;, &#39;Glc(b1-3)Gal(a1-3)[Gal(b1-4)]Rha2Ac(b1-4)Rha(a1-4)GlcA(b1-3)[Glc(a1-2)]Galf(b1-3)Glc&#39;, &#39;GlcNAcOS(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAcOS(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc&#39;, &#39;GlcNAc(b1-4)GlcA(b1-3)GlcNAc&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="prepare_multilabel" class="doc_header"><code>prepare_multilabel</code><a href="https://github.com/BojarLab/glycowork/tree/master/glycowork/ml/train_test_split.py#L122" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>prepare_multilabel</code>(<strong><code>df</code></strong>, <strong><code>rank</code></strong>=<em><code>'Species'</code></em>, <strong><code>glycan_col</code></strong>=<em><code>'target'</code></em>)</p>
</blockquote>
<p>converts a one row per glycan-species/tissue/disease association file to a format of one glycan - all associations</p>
<table>
<thead><tr>
<th style="text-align:left">Arguments:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">df (dataframe): dataframe where each row is one glycan - species association</td>
</tr>
<tr>
<td style="text-align:left">rank (string): which label column should be used; default:Species</td>
</tr>
<tr>
<td style="text-align:left">glycan_col (string): column name of where the glycan sequences are stored; default:target</td>
</tr>
</tbody>
</table>
<table>
<thead><tr>
<th style="text-align:left">Returns:</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">(1) list of unique glycans in df</td>
</tr>
<tr>
<td style="text-align:left">(2) list of lists, where each inner list are all the labels of a glycan</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

